// @atlaskit/adf-utils/dist/cjs/transforms/transform-media-single-width.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transformMediaSingleWidth = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _traverse = require("../traverse/traverse");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// when media single has a invalid width, we fallback to default width
// this the last resort when revert the pixel width, we couldnt determine the correct width value
var transformMediaSingleWidth = exports.transformMediaSingleWidth = function transformMediaSingleWidth(adf) {
  var isTransformed = false;
  var transformedAdf = (0, _traverse.traverse)(adf, {
    mediaSingle: function mediaSingle(node) {
      var _ref = node.attrs || {},
        width = _ref.width,
        widthType = _ref.widthType;

      // if width is not a number or greater than 100, and widthType is not set, we set width to 100
      if (typeof width === 'number' && width > 100 && !widthType) {
        isTransformed = true;
        var newAttrs = _objectSpread(_objectSpread({}, node.attrs), {}, {
          width: 100
        });
        return _objectSpread(_objectSpread({}, node), {}, {
          attrs: newAttrs
        });
      }
      return node;
    }
  });
  return {
    transformedAdf: transformedAdf,
    isTransformed: isTransformed
  };
};
