// @atlaskit/side-nav-items/dist/es2019/ui/menu-item/drag-and-drop/use-menu-item-drag-and-drop.js
/* use-menu-item-drag-and-drop.tsx generated by @compiled/babel-plugin v0.38.1 */
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { createPortal } from 'react-dom';
import { draggable, dropTargetForElements } from '@atlaskit/pragmatic-drag-and-drop/element/adapter';
import { pointerOutsideOfPreview } from '@atlaskit/pragmatic-drag-and-drop/element/pointer-outside-of-preview';
import { setCustomNativeDragPreview } from '@atlaskit/pragmatic-drag-and-drop/element/set-custom-native-drag-preview';
import { DragPreview } from './drag-preview';
import { DropIndicator } from './drop-indicator';
import { attachInstruction, extractInstruction } from './hitbox';
const idle = {
  type: 'idle'
};
/**
 * A convenience helper for setting up drag and drop for menu items
 *
 * - Don't include the `draggable` property if you don't want the menu item to be a draggable
 * - Don't include the `dropTarget` property if you don't want the menu item to be a drop target
 */
export function useMenuItemDragAndDrop({
  draggable: draggableArgs,
  dropTarget: dropTargetArgs
}) {
  const draggableAnchorRef = useRef(null);
  const draggableButtonRef = useRef(null);
  const dropTargetRef = useRef(null);
  const [internalState, setInternalState] = useState(idle);
  const getDraggableElement = useCallback(() => {
    var _ref, _draggableAnchorRef$c;
    return (_ref = (_draggableAnchorRef$c = draggableAnchorRef.current) !== null && _draggableAnchorRef$c !== void 0 ? _draggableAnchorRef$c : draggableButtonRef.current) !== null && _ref !== void 0 ? _ref : null;
  }, []);

  // Set up draggable
  useEffect(() => {
    // Don't set up a draggable if there are no draggable args
    if (!draggableArgs) {
      return;
    }
    const element = getDraggableElement();

    // Some elements may initially want to use DnD, then decide not to render.
    // In that case, we don't want to throw an error.
    if (!element) {
      return;
    }
    return draggable({
      element,
      getInitialData: draggableArgs.getInitialData,
      canDrag: draggableArgs.canDrag,
      onGenerateDragPreview({
        nativeSetDragImage,
        source,
        location
      }) {
        setCustomNativeDragPreview({
          nativeSetDragImage,
          getOffset: pointerOutsideOfPreview({
            x: "var(--ds-space-200, 16px)",
            y: "var(--ds-space-100, 8px)"
          }),
          render({
            container
          }) {
            const args = {
              dragHandle: source.dragHandle,
              element: source.element,
              input: location.current.input
            };
            setInternalState({
              type: 'preview',
              container,
              ui: draggableArgs.getDragPreviewPieces(args)
            });
          }
        });
      },
      onDragStart() {
        setInternalState({
          type: 'dragging'
        });
      },
      onDrop() {
        setInternalState(idle);
      }
    });
  }, [draggableArgs, getDraggableElement]);

  // Set up drop target
  useEffect(() => {
    if (!dropTargetArgs) {
      return;
    }

    // Don't need to provide a draggable element to have a drop target.
    // Using this element in our `canDrop` check
    const draggableElement = getDraggableElement();
    const dropTarget = dropTargetRef.current;

    // Some elements may initially want to use DnD, then decide not to render.
    // In that case, we don't want to throw an error
    if (!dropTarget) {
      return;
    }
    return dropTargetForElements({
      element: dropTarget,
      // cannot drop on self
      canDrop: args => {
        // cannot drop on self
        if (args.source.element === draggableElement) {
          return false;
        }
        if (dropTargetArgs.canDrop) {
          return dropTargetArgs.canDrop(args);
        }
        return true;
      },
      // menu items are always sticky, and the GroupDropIndicator should clear stickiness
      getIsSticky: () => true,
      getData(args) {
        var _dropTargetArgs$getDa, _dropTargetArgs$getDa2;
        const data = (_dropTargetArgs$getDa = (_dropTargetArgs$getDa2 = dropTargetArgs.getData) === null || _dropTargetArgs$getDa2 === void 0 ? void 0 : _dropTargetArgs$getDa2.call(dropTargetArgs, args)) !== null && _dropTargetArgs$getDa !== void 0 ? _dropTargetArgs$getDa : {};
        const operations = dropTargetArgs.getOperations(args);
        return attachInstruction(data, {
          input: args.input,
          element: args.element,
          operations
        });
      },
      onDragStart({
        self
      }) {
        const instruction = extractInstruction(self.data);
        setInternalState({
          type: 'is-over',
          instruction
        });
      },
      onDrag({
        self
      }) {
        const instruction = extractInstruction(self.data);
        setInternalState(current => {
          var _current$instruction, _current$instruction2;
          if (current.type === 'is-over' && (instruction === null || instruction === void 0 ? void 0 : instruction.operation) === ((_current$instruction = current.instruction) === null || _current$instruction === void 0 ? void 0 : _current$instruction.operation) && (instruction === null || instruction === void 0 ? void 0 : instruction.blocked) === ((_current$instruction2 = current.instruction) === null || _current$instruction2 === void 0 ? void 0 : _current$instruction2.blocked)) {
            return current;
          }
          return {
            type: 'is-over',
            instruction
          };
        });
      },
      onDragLeave() {
        setInternalState(idle);
      },
      onDrop() {
        setInternalState(idle);
      },
      onDragEnter({
        self
      }) {
        const instruction = extractInstruction(self.data);
        setInternalState({
          type: 'is-over',
          instruction
        });
      }
    });
  }, [dropTargetArgs, getDraggableElement]);
  const dragPreview = (() => {
    if (internalState.type !== 'preview') {
      return null;
    }
    return /*#__PURE__*/createPortal( /*#__PURE__*/React.createElement(DragPreview, {
      elemBefore: internalState.ui.elemBefore
    }, internalState.ui.content), internalState.container);
  })();
  const dropIndicator = internalState.type === 'is-over' && internalState.instruction && /*#__PURE__*/React.createElement(DropIndicator, {
    instruction: internalState.instruction
  });
  const state = useMemo(() => {
    if (internalState.type === 'preview') {
      return {
        type: 'preview'
      };
    }
    // returning a new object to avoid modification of our `internalState` object
    return {
      ...internalState
    };
  }, [internalState]);
  return {
    state,
    draggableButtonRef,
    dropTargetRef,
    draggableAnchorRef,
    dragPreview,
    dropIndicator
  };
}
