// @atlaskit/side-nav-items/dist/cjs/ui/skeleton/skeleton-menu-item.js
/* skeleton-menu-item.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SkeletonMenuItem = void 0;
require("./skeleton-menu-item.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _compiled = require("@atlaskit/primitives/compiled");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var containerStyles = {
  root: "_4t3i1tcg _bozg1b66 _y4ti1b66 _1q511b66 _85i51b66 _1e0c1txw _4cvr1h6o",
  text: "_bozg1b66 _y4ti1b66",
  item: "_11c8fhey _4t3iu7e1 _1e0c1txw _4cvr1h6o",
  hasDescription: "_4t3i1ylp",
  description: "_11c8wadc _4t3iu7e1 _1e0c1txw _4cvr1h6o",
  elemBefore: "_bozgv77o _y4tiv77o _1q51v77o _85i5v77o _1e0c1txw _4cvr1h6o _1bah1h6o"
};
var skeletonStyles = {
  item: "_2rko1rr0 _bfhkizff _1bsb53f4 _4t3ic3s8",
  elemBefore: "_2rko12b0 _bfhkizff _1bsbgktf _4t3igktf",
  description: "_2rko1rr0 _bfhkizff _1bsb1wqb _4t3ic3s8"
};
/**
 * __Skeleton menu item__
 *
 * A skeleton menu item is used to display a loading state for a side nav menu item.
 * This component can have an icon element skeleton, and a description skeleton included.
 */
var SkeletonMenuItem = exports.SkeletonMenuItem = function SkeletonMenuItem(props) {
  var _props$hasDescription = props.hasDescription,
    hasDescription = _props$hasDescription === void 0 ? false : _props$hasDescription,
    _props$hasElemBefore = props.hasElemBefore,
    hasElemBefore = _props$hasElemBefore === void 0 ? false : _props$hasElemBefore,
    testId = props.testId;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId,
    className: (0, _runtime.ax)([containerStyles.root, hasDescription && containerStyles.hasDescription])
  }, /*#__PURE__*/React.createElement(_compiled.Flex, {
    alignItems: "center"
  }, hasElemBefore && /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([containerStyles.elemBefore])
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([skeletonStyles.elemBefore])
  })), /*#__PURE__*/React.createElement(_compiled.Flex, {
    xcss: containerStyles.text,
    gap: "space.025",
    direction: "column"
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([containerStyles.item])
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([skeletonStyles.item])
  })), hasDescription && /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([containerStyles.description])
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([skeletonStyles.description])
  })))));
};
