// @atlaskit/forge-react-types/dist/es2019/components/__generated__/BoxProps.codegen.js
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 *
 * Extract component prop types from UIKit 2 components - BoxProps
 *
 * @codegen <<SignedSource::653f98d927e1ec1646db39eca8e994fb>>
 * @codegenCommand yarn workspace @atlaskit/forge-react-types codegen
 * @codegenDependency ../../../../forge-ui/src/components/UIKit/box/index.tsx <<SignedSource::8875ae64ad6c03e0827dd8f8a893ac04>>
 */
/* eslint @repo/internal/codegen/signed-source-integrity: "warn" */
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage/preview */

/**
 *
 * @param supportedXCSSProps - the list of css props to be supported for the intended component.
 *    If not provided, all the props will be supported. The props could be either standard css props
 *    or design token based props. If the prop is a design token based prop, the value of the prop
 *    will be validated against the design tokens map to ensure the value is a valid design token string.
 * @returns a function that takes a style object and returns a style object with only the supported props
 *    as specified in the supportedXCSSProps list. The props that are not supported will be removed from the
 *    returned style object and a warning will be logged in the console.
 */

/**
 * Generates SerializedStyles from xcss style object.
 *
 * Previously we used `&&` specificity hack to ensure Emotion xcss styles could
 * override Compiled styles when wrapping @atlaskit/primitives/compiled components.
 *
 * Now that components like Box and Pressable are reimplemented entirely in Emotion,
 * we no longer need the specificity increase - Emotion merges styles in array order (last wins).
 */

const borderRadiusTokens = ['radius.xsmall', 'radius.small', 'radius.medium', 'radius.large', 'radius.xlarge', 'radius.full', 'radius.tile'];
const borderRadiusSupportedValues = [...borderRadiusTokens, 'border.radius'];
const xcssValidator = makeXCSSValidator({
  // text related props
  textAlign: {
    allowCSS: true
  },
  // color related props
  color: true,
  boxShadow: true,
  opacity: true,
  backgroundColor: true,
  borderColor: true,
  borderBlockColor: true,
  borderBlockEndColor: true,
  borderBlockStartColor: true,
  borderBottomColor: true,
  borderInlineColor: true,
  borderInlineEndColor: true,
  borderInlineStartColor: true,
  borderLeftColor: true,
  borderRightColor: true,
  borderTopColor: true,
  overflow: {
    supportedValues: ['hidden', 'visible', 'scroll', 'auto']
  },
  overflowX: {
    supportedValues: ['hidden', 'visible', 'scroll', 'auto']
  },
  overflowY: {
    supportedValues: ['hidden', 'visible', 'scroll', 'auto']
  },
  // layout and space related props
  display: {
    supportedValues: ['block', 'inline-block', 'inline', 'none']
  },
  flexGrow: {
    allowCSS: true
  },
  width: {
    allowCSS: true
  },
  height: {
    allowCSS: true
  },
  minWidth: {
    allowCSS: true
  },
  maxWidth: {
    allowCSS: true
  },
  minHeight: {
    allowCSS: true
  },
  maxHeight: {
    allowCSS: true
  },
  margin: true,
  marginBlock: true,
  marginBlockEnd: true,
  marginBlockStart: true,
  marginBottom: true,
  marginInline: true,
  marginInlineEnd: true,
  marginInlineStart: true,
  marginLeft: true,
  marginRight: true,
  marginTop: true,
  padding: true,
  paddingBlock: true,
  paddingBlockEnd: true,
  paddingBlockStart: true,
  paddingBottom: true,
  paddingInline: true,
  paddingInlineEnd: true,
  paddingInlineStart: true,
  paddingLeft: true,
  paddingRight: true,
  paddingTop: true,
  // other box related props
  borderRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderBottomLeftRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderBottomRightRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderTopLeftRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderTopRightRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderEndEndRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderEndStartRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderStartEndRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderStartStartRadius: {
    supportedValues: borderRadiusSupportedValues
  },
  borderWidth: true,
  borderBlockWidth: true,
  borderBlockEndWidth: true,
  borderBlockStartWidth: true,
  borderBottomWidth: true,
  borderInlineWidth: true,
  borderInlineEndWidth: true,
  borderInlineStartWidth: true,
  borderLeftWidth: true,
  borderRightWidth: true,
  borderTopWidth: true,
  // other props not in tokens based props
  borderTopStyle: {
    supportedValues: ['dotted', 'dashed', 'solid', 'none', 'hidden']
  },
  borderBottomStyle: {
    supportedValues: ['dotted', 'dashed', 'solid', 'none', 'hidden']
  },
  borderRightStyle: {
    supportedValues: ['dotted', 'dashed', 'solid', 'none', 'hidden']
  },
  borderLeftStyle: {
    supportedValues: ['dotted', 'dashed', 'solid', 'none', 'hidden']
  },
  borderStyle: {
    supportedValues: ['dotted', 'dashed', 'solid', 'none', 'hidden']
  },
  position: {
    supportedValues: ['relative', 'static']
  }
});

/**
 * A box is a generic container that provides managed access to design tokens.
 *
 * @see [Box](https://developer.atlassian.com/platform/forge/ui-kit/components/box/) in UI Kit documentation for more information
 */
export {};
