// @atlaskit/adf-utils/dist/cjs/validator/specs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "alignment", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.alignment;
  }
});
Object.defineProperty(exports, "annotation", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.annotation;
  }
});
Object.defineProperty(exports, "backgroundColor", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.backgroundColor;
  }
});
Object.defineProperty(exports, "blockCard", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.blockCard;
  }
});
Object.defineProperty(exports, "blockRootOnly", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.blockRootOnly;
  }
});
Object.defineProperty(exports, "blockTaskItem", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.blockTaskItem;
  }
});
Object.defineProperty(exports, "block_content", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.block_content;
  }
});
Object.defineProperty(exports, "blockquote", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.blockquote;
  }
});
Object.defineProperty(exports, "bodiedExtension", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.bodiedExtension;
  }
});
Object.defineProperty(exports, "bodiedExtension_with_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.bodiedExtension_with_marks;
  }
});
Object.defineProperty(exports, "bodiedSyncBlock", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.bodiedSyncBlock;
  }
});
Object.defineProperty(exports, "border", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.border;
  }
});
Object.defineProperty(exports, "breakout", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.breakout;
  }
});
Object.defineProperty(exports, "bulletList", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.bulletList;
  }
});
Object.defineProperty(exports, "caption", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.caption;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.code;
  }
});
Object.defineProperty(exports, "codeBlock", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.codeBlock;
  }
});
Object.defineProperty(exports, "codeBlock_root_only", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.codeBlock_root_only;
  }
});
Object.defineProperty(exports, "confluenceInlineComment", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.confluenceInlineComment;
  }
});
Object.defineProperty(exports, "dataConsumer", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.dataConsumer;
  }
});
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.date;
  }
});
Object.defineProperty(exports, "decisionItem", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.decisionItem;
  }
});
Object.defineProperty(exports, "decisionList", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.decisionList;
  }
});
Object.defineProperty(exports, "doc", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.doc;
  }
});
Object.defineProperty(exports, "em", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.em;
  }
});
Object.defineProperty(exports, "embedCard", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.embedCard;
  }
});
Object.defineProperty(exports, "emoji", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.emoji;
  }
});
Object.defineProperty(exports, "expand", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.expand;
  }
});
Object.defineProperty(exports, "expand_root_only", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.expand_root_only;
  }
});
Object.defineProperty(exports, "extension", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.extension;
  }
});
Object.defineProperty(exports, "extensionFrame", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.extensionFrame;
  }
});
Object.defineProperty(exports, "extension_with_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.extension_with_marks;
  }
});
Object.defineProperty(exports, "fragment", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.fragment;
  }
});
Object.defineProperty(exports, "hardBreak", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.hardBreak;
  }
});
Object.defineProperty(exports, "heading", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.heading;
  }
});
Object.defineProperty(exports, "heading_with_alignment", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.heading_with_alignment;
  }
});
Object.defineProperty(exports, "heading_with_indentation", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.heading_with_indentation;
  }
});
Object.defineProperty(exports, "heading_with_no_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.heading_with_no_marks;
  }
});
Object.defineProperty(exports, "indentation", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.indentation;
  }
});
Object.defineProperty(exports, "inlineCard", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.inlineCard;
  }
});
Object.defineProperty(exports, "inlineExtension", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.inlineExtension;
  }
});
Object.defineProperty(exports, "inlineExtension_with_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.inlineExtension_with_marks;
  }
});
Object.defineProperty(exports, "inline_content", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.inline_content;
  }
});
Object.defineProperty(exports, "layoutColumn", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.layoutColumn;
  }
});
Object.defineProperty(exports, "layoutSection", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.layoutSection;
  }
});
Object.defineProperty(exports, "layoutSection_full", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.layoutSection_full;
  }
});
Object.defineProperty(exports, "layoutSection_with_single_column", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.layoutSection_with_single_column;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.link;
  }
});
Object.defineProperty(exports, "listItem", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.listItem;
  }
});
Object.defineProperty(exports, "media", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.media;
  }
});
Object.defineProperty(exports, "mediaGroup", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.mediaGroup;
  }
});
Object.defineProperty(exports, "mediaInline", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.mediaInline;
  }
});
Object.defineProperty(exports, "mediaSingle", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.mediaSingle;
  }
});
Object.defineProperty(exports, "mediaSingle_caption", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.mediaSingle_caption;
  }
});
Object.defineProperty(exports, "mediaSingle_full", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.mediaSingle_full;
  }
});
Object.defineProperty(exports, "mention", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.mention;
  }
});
Object.defineProperty(exports, "multiBodiedExtension", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.multiBodiedExtension;
  }
});
Object.defineProperty(exports, "nestedExpand", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.nestedExpand;
  }
});
Object.defineProperty(exports, "nestedExpand_content", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.nestedExpand_content;
  }
});
Object.defineProperty(exports, "nestedExpand_with_no_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.nestedExpand_with_no_marks;
  }
});
Object.defineProperty(exports, "non_nestable_block_content", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.non_nestable_block_content;
  }
});
Object.defineProperty(exports, "orderedList", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.orderedList;
  }
});
Object.defineProperty(exports, "panel", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.panel;
  }
});
Object.defineProperty(exports, "paragraph", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.paragraph;
  }
});
Object.defineProperty(exports, "paragraph_with_alignment", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.paragraph_with_alignment;
  }
});
Object.defineProperty(exports, "paragraph_with_indentation", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.paragraph_with_indentation;
  }
});
Object.defineProperty(exports, "paragraph_with_no_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.paragraph_with_no_marks;
  }
});
Object.defineProperty(exports, "placeholder", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.placeholder;
  }
});
Object.defineProperty(exports, "rule", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.rule;
  }
});
Object.defineProperty(exports, "status", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.status;
  }
});
Object.defineProperty(exports, "strike", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.strike;
  }
});
Object.defineProperty(exports, "strong", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.strong;
  }
});
Object.defineProperty(exports, "subsup", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.subsup;
  }
});
Object.defineProperty(exports, "syncBlock", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.syncBlock;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.table;
  }
});
Object.defineProperty(exports, "tableCell", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.tableCell;
  }
});
Object.defineProperty(exports, "tableHeader", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.tableHeader;
  }
});
Object.defineProperty(exports, "tableRow", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.tableRow;
  }
});
Object.defineProperty(exports, "taskItem", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.taskItem;
  }
});
Object.defineProperty(exports, "taskList", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.taskList;
  }
});
Object.defineProperty(exports, "text", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.text;
  }
});
Object.defineProperty(exports, "textColor", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.textColor;
  }
});
Object.defineProperty(exports, "text_code_inline", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.text_code_inline;
  }
});
Object.defineProperty(exports, "text_formatted", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.text_formatted;
  }
});
Object.defineProperty(exports, "text_link_inline", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.text_link_inline;
  }
});
Object.defineProperty(exports, "text_with_no_marks", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.text_with_no_marks;
  }
});
Object.defineProperty(exports, "underline", {
  enumerable: true,
  get: function get() {
    return _schemaValidator.underline;
  }
});
var _schemaValidator = require("@atlaskit/adf-schema/schema-validator");
