// @atlaskit/feature-gate-js-client/dist/cjs/index.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CLIENT_VERSION", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.CLIENT_VERSION;
  }
});
Object.defineProperty(exports, "DynamicConfig", {
  enumerable: true,
  get: function get() {
    return _DynamicConfig.DynamicConfig;
  }
});
Object.defineProperty(exports, "EvaluationReason", {
  enumerable: true,
  get: function get() {
    return _types.EvaluationReason;
  }
});
Object.defineProperty(exports, "FeatureGateEnvironment", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.FeatureGateEnvironment;
  }
});
Object.defineProperty(exports, "Layer", {
  enumerable: true,
  get: function get() {
    return _Layer.Layer;
  }
});
Object.defineProperty(exports, "PerimeterType", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.PerimeterType;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.default;
  }
});
var _FeatureGates = _interopRequireWildcard(require("./client/FeatureGates"));
var _DynamicConfig = require("./client/compat/DynamicConfig");
var _Layer = require("./client/compat/Layer");
var _types = require("./client/compat/types");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
