// @atlaskit/motion/dist/esm/entering/animate-in.js
import React from 'react';
import KeyframesMotion from './keyframes-motion';

/**
 * Props for controlling the behavior of the AnimateIn animation
 */

/**
 * __AnimateIn__
 *
 * Useful for Animating in one or more elements.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/motion/docs/entering-motions)
 */
var AnimateIn = function AnimateIn(_ref) {
  var children = _ref.children,
    _ref$duration = _ref.duration,
    duration = _ref$duration === void 0 ? 'large' : _ref$duration,
    enteringAnimation = _ref.enteringAnimation,
    exitingAnimation = _ref.exitingAnimation,
    onFinish = _ref.onFinish,
    isPaused = _ref.isPaused,
    animationTimingFunction = _ref.animationTimingFunction;
  return /*#__PURE__*/React.createElement(KeyframesMotion, {
    duration: duration,
    enteringAnimation: enteringAnimation,
    exitingAnimation: exitingAnimation,
    animationTimingFunction: animationTimingFunction,
    onFinish: onFinish,
    isPaused: isPaused
  }, children);
};
export default AnimateIn;
