// @atlaskit/navigation-system/dist/esm/components/badge-container.js
/* badge-container.tsx generated by @compiled/babel-plugin v0.38.1 */
import "./badge-container.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
var styles = {
  root: "_kqswh2mm _1e0c1txw",
  badgeContainer: "_kqswstnw _o5721q9c _152tr5cr _1e02utpp _lcxvglyw"
};
/**
 * __Badge container__
 *
 * Adds a badge to the top right corner of the child element.
 * Example of usage is showing the number of notifications in the top navigation bar.
 */
export var BadgeContainer = function BadgeContainer(_ref) {
  var children = _ref.children,
    badgeId = _ref.id,
    Badge = _ref.badge,
    _ref$isListItem = _ref.isListItem,
    isListItem = _ref$isListItem === void 0 ? true : _ref$isListItem;
  return /*#__PURE__*/React.createElement("div", {
    role: isListItem ? 'listitem' : undefined,
    className: ax([styles.root])
  }, children, /*#__PURE__*/React.createElement("div", {
    id: badgeId,
    "aria-hidden": true,
    className: ax([styles.badgeContainer])
  }, /*#__PURE__*/React.createElement(Badge, null)));
};
