// @atlaskit/select/dist/cjs/select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SelectWithoutAnalytics = void 0;
var _analyticsNext = require("@atlaskit/analytics-next");
var _async = _interopRequireDefault(require("@atlaskit/react-select/async"));
var _createSelect = _interopRequireDefault(require("./create-select"));
/* eslint-disable @repo/internal/react/require-jsdoc */

var packageName = "@atlaskit/select";
var packageVersion = "0.0.0-development";
var SelectWithoutAnalytics = exports.SelectWithoutAnalytics = (0, _createSelect.default)(_async.default);
var createAndFireEventOnAtlaskit = (0, _analyticsNext.createAndFireEvent)('atlaskit');
var Select = (0, _analyticsNext.withAnalyticsContext)({
  componentName: 'select',
  packageName: packageName,
  packageVersion: packageVersion
})((0, _analyticsNext.withAnalyticsEvents)({
  onChange: createAndFireEventOnAtlaskit({
    action: 'changed',
    actionSubject: 'option',
    attributes: {
      componentName: 'select',
      packageName: packageName,
      packageVersion: packageVersion
    }
  })
})(SelectWithoutAnalytics));
var _default = exports.default = Select;
