// @atlaskit/navigation-system/dist/esm/ui/page-layout/constants.js
/**
 * These variables are shared across the different page layout slots.
 * This violates the [UI styling standard](https://atlassian.design/components/eslint-plugin-ui-styling-standard/no-imported-style-values/usage).
 * We have chosen to ignore this rule to prevent duplicating them across the different slots, and to provide a high level view of them and how they relate.
 */

// These are "compressed" to discourage usage.
export var sideNavVar = '--n_sNvw';
export var asideVar = '--n_asDw';
export var panelVar = '--n_pnlW';
export var bannerMountedVar = '--n_bnrM';
export var topNavMountedVar = '--n_tNvM';
export var ribbonVar = '--n_rbnW';

/**
 * Captures the current width of the side navigation, at all times, including during resizing.
 *
 * The standard `sideNavVar` only captures the 'committed' width, not the resizing width.
 */
export var sideNavLiveWidthVar = '--n_sNvlw';
export var sideNavPanelSplitterId = Symbol('SideNav PanelSplitter');
export var asidePanelSplitterId = Symbol('Aside PanelSplitter');
export var panelPanelSplitterId = Symbol('Panel PanelSplitter');

// We aren't using template literals here because Compiled can't compiled them in platform ATM.
export var contentHeightWhenFixed = "calc(100vh - var(--n_bnrM, 0px) - var(--n_tNvM, 0px))";
export var contentInsetBlockStart = "calc(var(--n_bnrM, 0px) + var(--n_tNvM, 0px))";

// We define variables as they can change their size depending on the viewport width. That isn't needed for
// general grid item slots because, well, they just take up the size of the grid item! In this instance since
// the UNSAFE absolutely positioned sibling isn't on the grid it needs extra layout information.
// Note: THESE ARE THE SAME VALUES AS THE LEGACY PAGE LAYOUT.
// WHEN WE ELIMINATE USAGE OF THOSE VARIABLES THESE CAN BE RENAMED TO BE HASHED.
export var UNSAFE_topNavVar = '--topNavigationHeight';
export var UNSAFE_bannerVar = '--bannerHeight';
export var UNSAFE_sideNavLayoutVar = '--leftSidebarWidth';
export var UNSAFE_ribbonVar = '--leftPanelWidth';
export var UNSAFE_asideLayoutVar = '--rightSidebarWidth';
export var UNSAFE_panelLayoutVar = '--rightPanelWidth';

// The following UNSAFE variables are used to absolutely position elements that aren't a child of page layout.
// Known use cases: Legacy pages rendered inside Confluence and Jira.
export var UNSAFE_MAIN_BLOCK_START_FOR_LEGACY_PAGES_ONLY = "calc(var(".concat(UNSAFE_bannerVar, ", 0px) + var(").concat(UNSAFE_topNavVar, ", 0px))");
export var UNSAFE_MAIN_INLINE_START_FOR_LEGACY_PAGES_ONLY = "calc(var(".concat(UNSAFE_ribbonVar, ", 0px) + var(").concat(UNSAFE_sideNavLayoutVar, ", 0px))");
export var UNSAFE_MAIN_INLINE_END_FOR_LEGACY_PAGES_ONLY = "calc(var(".concat(UNSAFE_asideLayoutVar, ", 0px) + var(").concat(UNSAFE_panelLayoutVar, ", 0px))");

/**
 * We define the z-indexes here so each page slot can be locally layered against each other.
 * For globally defined values such as flag, modal, etc, we can continue to
 * rely on accessing them through global means.
 */
export var localSlotLayers = {
  ribbon: 4,
  // The side nav panel splitter is layered above the top nav when FHS and 'platform-dst-side-nav-layering-fixes' is enabled.
  // It has the same z-index value, but is rendered after the top nav in the DOM so is stacked above.
  sideNavPanelSplitterFHS: 4,
  topBar: 4,
  banner: 4,
  // When FHS and 'platform-dst-side-nav-layering-fixes' is enabled, the side nav is layered below the top nav,
  // but above the panel
  bannerFHS: 3,
  topNavFHS: 3,
  sideNav: 2,
  panelSmallViewports: 1
};
export var openLayerObserverSideNavNamespace = 'side-nav';
// We have separate OpenLayerObserver namespaces for each top nav layout area, instead of a single namespace for the entire top nav,
// as we need to know if there are open layers in TopNavStart specifically to prevent the side nav from peeking.
// The OpenLayerObserver doesn't support nested namespaces. We could update it, but for now it's simple enough to just have separate
// namespaces.
export var openLayerObserverTopNavStartNamespace = 'top-nav-start';
export var openLayerObserverTopNavMiddleNamespace = 'top-nav-middle';
export var openLayerObserverTopNavEndNamespace = 'top-nav-end';

/**
 * CSS scroll timeline variable for the side nav content scroll indicator.
 * The scroll timeline is created in SideNavContent, and then used by TopNavStart to apply the scroll indicator line.
 */
export var sideNavContentScrollTimelineVar = '--sNcst';
