// @atlaskit/tile/dist/cjs/skeleton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TileSkeleton;
var _react = _interopRequireDefault(require("react"));
var _skeleton = _interopRequireDefault(require("@atlaskit/skeleton"));
var _tile = _interopRequireDefault(require("./tile"));
/**
 * __Tile Skeleton__
 *
 * A tile skeleton is the loading state for the tile component.
 *
 */
function TileSkeleton(_ref) {
  var color = _ref.color,
    isShimmering = _ref.isShimmering,
    shimmeringEndColor = _ref.shimmeringEndColor,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    testId = _ref.testId;
  return /*#__PURE__*/_react.default.createElement(_tile.default, {
    label: "",
    size: size,
    testId: testId,
    backgroundColor: "transparent"
  }, /*#__PURE__*/_react.default.createElement(_skeleton.default, {
    color: color,
    isShimmering: isShimmering,
    ShimmeringEndColor: shimmeringEndColor,
    width: '100%',
    height: '100%',
    testId: testId ? "".concat(testId, "--skeleton") : undefined
  }));
}
