// @atlaskit/adf-schema/dist/cjs/schema/nodes/bodied-extension.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bodiedExtension = void 0;
var _extensions = require("../../utils/extensions");
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
/**
 * @name bodiedExtension_node
 */

/**
 * @name bodiedExtension_with_marks_node
 */

var bodiedExtension = exports.bodiedExtension = (0, _nodeTypes.bodiedExtension)({
  parseDOM: [{
    context: 'bodiedExtension//',
    tag: '[data-node-type="bodied-extension"]',
    skip: true
  }, {
    tag: '[data-node-type="bodied-extension"]',
    getAttrs: function getAttrs(domNode) {
      return (0, _extensions.getExtensionAttrs)(domNode);
    }
  }],
  toDOM: function toDOM(node) {
    var attrs = {
      'data-node-type': 'bodied-extension',
      'data-extension-type': node.attrs.extensionType,
      'data-extension-key': node.attrs.extensionKey,
      'data-text': node.attrs.text,
      'data-parameters': JSON.stringify(node.attrs.parameters),
      'data-layout': node.attrs.layout,
      'data-local-id:': node.attrs.localId
    };
    return ['div', attrs, 0];
  }
});
