// @atlaskit/adf-schema/dist/cjs/schema/nodes/confluence-jira-issue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.confluenceJiraIssue = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var name = 'confluenceJiraIssue';
var confluenceJiraIssue = exports.confluenceJiraIssue = (0, _nodeTypes.confluenceJiraIssue)({
  parseDOM: [{
    tag: "span[data-node-type=\"".concat(name, "\"]"),
    getAttrs: function getAttrs(domNode) {
      var dom = domNode;
      return {
        issueKey: dom.textContent,
        macroId: dom.dataset && dom.dataset.macroId,
        schemaVersion: dom.dataset && dom.dataset.schemaVersion,
        server: dom.dataset && dom.dataset.server,
        serverId: dom.dataset && dom.dataset.serverId
      };
    }
  }],
  toDOM: function toDOM(node) {
    var attrs = {
      'data-node-type': name,
      'data-macro-id': node.attrs.macroId,
      'data-schema-version': node.attrs.schemaVersion,
      'data-server': node.attrs.server,
      'data-server-id': node.attrs.serverId,
      'data-jira-issue': node.attrs.issueKey
    };
    return ['span', attrs, node.attrs.issueKey];
  }
});
