// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/bodiedSyncBlock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bodiedSyncBlock = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _breakout = require("../marks/breakout");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _unsupportedBlock = require("../nodes/unsupportedBlock");
var _blockCard = require("./blockCard");
var _blockquote = require("./blockquote");
var _codeBlock = require("./codeBlock");
var _confluenceUnsupportedBlock = require("./confluenceUnsupportedBlock");
var _decisionList = require("./decisionList");
var _embedCard = require("./embedCard");
var _expand = require("./expand");
var _list = require("./list");
var _mediaGroup = require("./mediaGroup");
var _mediaSingle = require("./mediaSingle");
var _panel = require("./panel");
var _paragraph = require("./paragraph");
var _rule = require("./rule");
var _heading = require("./heading");
var _tableNodes = require("./tableNodes");
var _task = require("./task");
var _layoutSection = require("./layoutSection");
var bodiedSyncBlock = exports.bodiedSyncBlock = (0, _adfSchemaGenerator.adfNode)('bodiedSyncBlock').define({
  selectable: true,
  isolating: true,
  marks: [_breakout.breakout, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    resourceId: {
      type: 'string',
      default: ''
    },
    localId: {
      type: 'string',
      default: ''
    }
  },
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(_paragraph.paragraph, _paragraph.paragraph.use('with_alignment'), _paragraph.paragraph.use('with_indentation'), _paragraph.paragraph.use('with_no_marks'), _blockCard.blockCard, _blockquote.blockquote, _blockquote.blockquote.use('legacy'), _list.bulletList, _codeBlock.codeBlock, _confluenceUnsupportedBlock.confluenceUnsupportedBlock, _decisionList.decisionList, _embedCard.embedCard, _expand.expand, _heading.heading, _heading.heading.use('with_alignment'), _heading.heading.use('with_indentation'), _heading.heading.use('with_no_marks'), _layoutSection.layoutSection, _layoutSection.layoutSection.use('with_single_column'), _layoutSection.layoutSection.use('full'), _mediaGroup.mediaGroup, _mediaSingle.mediaSingle, _mediaSingle.mediaSingle.use('caption'), _mediaSingle.mediaSingle.use('full'), _mediaSingle.mediaSingle.use('width_type'), _list.orderedList, _panel.panel, _rule.rule, _tableNodes.table,
  // @ts-expect-error - types don't deal well with circular references for the variant
  _tableNodes.table.use('with_nested_table'), _task.taskList, _unsupportedBlock.unsupportedBlock))]
});
