// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/heading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.headingWithMarks = exports.heading = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _inlineContentGroup = require("../groups/inlineContentGroup");
var _inlineGroup = require("../groups/inlineGroup");
var _alignmentAndIndentation = require("../marks/alignmentAndIndentation");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var heading = exports.heading = (0, _adfSchemaGenerator.adfNode)('heading').define({
  defining: true,
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  hasEmptyMarks: true,
  attrs: {
    level: {
      type: 'number',
      default: 1,
      minimum: 1,
      maximum: 6
    },
    localId: {
      type: 'string',
      optional: true,
      default: null
    }
  },
  content: [(0, _adfSchemaGenerator.$zeroPlus)((0, _adfSchemaGenerator.$or)(_inlineGroup.inlineGroup, _inlineContentGroup.inlineContentGroup))]
}).variant('with_alignment', {
  marks: [_alignmentAndIndentation.alignment, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  content: [],
  ignore: []
}).variant('with_indentation', {
  marks: [_alignmentAndIndentation.indentation, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  content: [],
  ignore: []
}).variant('with_no_marks', {
  content: [],
  marks: [],
  noMarks: true,
  ignore: []
});
var headingWithMarks = exports.headingWithMarks = heading;
