// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/mention.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mention = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _annotation = require("../marks/annotation");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var mention = exports.mention = (0, _adfSchemaGenerator.adfNode)('mention').define({
  inline: true,
  selectable: true,
  marks: [_unsupportedNodeAttribute.unsupportedNodeAttribute, _unsupportedMark.unsupportedMark],
  attrs: {
    id: {
      type: 'string',
      default: ''
    },
    localId: {
      type: 'string',
      optional: true,
      default: null
    },
    text: {
      type: 'string',
      default: '',
      optional: true
    },
    accessLevel: {
      type: 'string',
      default: '',
      optional: true
    },
    userType: {
      type: 'enum',
      values: ['DEFAULT', 'SPECIAL', 'APP'],
      default: null,
      optional: true
    }
  },
  stage0: {
    marks: [_unsupportedNodeAttribute.unsupportedNodeAttribute, _unsupportedMark.unsupportedMark, _annotation.annotation]
  }
});
