// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/task.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.taskList = exports.taskItem = exports.blockTaskItem = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _inlineContentGroup = require("../groups/inlineContentGroup");
var _inlineGroup = require("../groups/inlineGroup");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _unsupportedBlock = require("./unsupportedBlock");
var _paragraph = require("./paragraph");
var _extension = require("./extension");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var taskItem = exports.taskItem = (0, _adfSchemaGenerator.adfNode)('taskItem');
var taskList = exports.taskList = (0, _adfSchemaGenerator.adfNode)('taskList');
var blockTaskItem = exports.blockTaskItem = (0, _adfSchemaGenerator.adfNode)('blockTaskItem');
var commonTaskItemProps = {
  defining: true,
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  allowAnyChildMark: true,
  attrs: {
    localId: {
      type: 'string',
      default: ''
    },
    state: {
      type: 'enum',
      values: ['TODO', 'DONE'],
      default: 'TODO'
    }
  }
};
taskItem.define(_objectSpread(_objectSpread({}, commonTaskItemProps), {}, {
  content: [(0, _adfSchemaGenerator.$zeroPlus)((0, _adfSchemaGenerator.$or)(_inlineGroup.inlineGroup, _inlineContentGroup.inlineContentGroup))]
}));
blockTaskItem.define(_objectSpread(_objectSpread({}, commonTaskItemProps), {}, {
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  allowAnyChildMark: false,
  // Override the commonTaskItemProps setting
  contentMinItems: 1,
  content: [(0, _adfSchemaGenerator.$or)(_paragraph.paragraph.use('with_no_marks'), _extension.extension.use('with_marks')), (0, _adfSchemaGenerator.$zeroPlus)((0, _adfSchemaGenerator.$or)(_paragraph.paragraph.use('with_no_marks'), _extension.extension.use('with_marks')))]
}));
taskList.define({
  defining: true,
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    localId: {
      type: 'string',
      default: ''
    }
  },
  contentMinItems: 1,
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(taskItem, _unsupportedBlock.unsupportedBlock, blockTaskItem)), (0, _adfSchemaGenerator.$zeroPlus)((0, _adfSchemaGenerator.$or)(taskItem, taskList, _unsupportedBlock.unsupportedBlock, blockTaskItem))]
});
taskList.variant('flexible_first_child', {
  contentMinItems: 1,
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(taskItem, taskList, _unsupportedBlock.unsupportedBlock, blockTaskItem))],
  stage0: true
});
