// @atlaskit/adf-schema/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AnnotationMarkStates", {
  enumerable: true,
  get: function get() {
    return _schema.AnnotationMarkStates;
  }
});
Object.defineProperty(exports, "AnnotationTypes", {
  enumerable: true,
  get: function get() {
    return _schema.AnnotationTypes;
  }
});
Object.defineProperty(exports, "B100", {
  enumerable: true,
  get: function get() {
    return _utils.B100;
  }
});
Object.defineProperty(exports, "B400", {
  enumerable: true,
  get: function get() {
    return _utils.B400;
  }
});
Object.defineProperty(exports, "B50", {
  enumerable: true,
  get: function get() {
    return _utils.B50;
  }
});
Object.defineProperty(exports, "B500", {
  enumerable: true,
  get: function get() {
    return _utils.B500;
  }
});
Object.defineProperty(exports, "B75", {
  enumerable: true,
  get: function get() {
    return _utils.B75;
  }
});
Object.defineProperty(exports, "G200", {
  enumerable: true,
  get: function get() {
    return _utils.G200;
  }
});
Object.defineProperty(exports, "G300", {
  enumerable: true,
  get: function get() {
    return _utils.G300;
  }
});
Object.defineProperty(exports, "G400", {
  enumerable: true,
  get: function get() {
    return _utils.G400;
  }
});
Object.defineProperty(exports, "G50", {
  enumerable: true,
  get: function get() {
    return _utils.G50;
  }
});
Object.defineProperty(exports, "G500", {
  enumerable: true,
  get: function get() {
    return _utils.G500;
  }
});
Object.defineProperty(exports, "G75", {
  enumerable: true,
  get: function get() {
    return _utils.G75;
  }
});
Object.defineProperty(exports, "N0", {
  enumerable: true,
  get: function get() {
    return _utils.N0;
  }
});
Object.defineProperty(exports, "N20", {
  enumerable: true,
  get: function get() {
    return _utils.N20;
  }
});
Object.defineProperty(exports, "N200", {
  enumerable: true,
  get: function get() {
    return _utils.N200;
  }
});
Object.defineProperty(exports, "N30", {
  enumerable: true,
  get: function get() {
    return _utils.N30;
  }
});
Object.defineProperty(exports, "N300", {
  enumerable: true,
  get: function get() {
    return _utils.N300;
  }
});
Object.defineProperty(exports, "N40", {
  enumerable: true,
  get: function get() {
    return _utils.N40;
  }
});
Object.defineProperty(exports, "N50", {
  enumerable: true,
  get: function get() {
    return _utils.N50;
  }
});
Object.defineProperty(exports, "N500", {
  enumerable: true,
  get: function get() {
    return _utils.N500;
  }
});
Object.defineProperty(exports, "N60", {
  enumerable: true,
  get: function get() {
    return _utils.N60;
  }
});
Object.defineProperty(exports, "N80", {
  enumerable: true,
  get: function get() {
    return _utils.N80;
  }
});
Object.defineProperty(exports, "N800", {
  enumerable: true,
  get: function get() {
    return _utils.N800;
  }
});
Object.defineProperty(exports, "N90", {
  enumerable: true,
  get: function get() {
    return _utils.N90;
  }
});
Object.defineProperty(exports, "P100", {
  enumerable: true,
  get: function get() {
    return _utils.P100;
  }
});
Object.defineProperty(exports, "P300", {
  enumerable: true,
  get: function get() {
    return _utils.P300;
  }
});
Object.defineProperty(exports, "P400", {
  enumerable: true,
  get: function get() {
    return _utils.P400;
  }
});
Object.defineProperty(exports, "P50", {
  enumerable: true,
  get: function get() {
    return _utils.P50;
  }
});
Object.defineProperty(exports, "P500", {
  enumerable: true,
  get: function get() {
    return _utils.P500;
  }
});
Object.defineProperty(exports, "P75", {
  enumerable: true,
  get: function get() {
    return _utils.P75;
  }
});
Object.defineProperty(exports, "PanelType", {
  enumerable: true,
  get: function get() {
    return _schema.PanelType;
  }
});
Object.defineProperty(exports, "R100", {
  enumerable: true,
  get: function get() {
    return _utils.R100;
  }
});
Object.defineProperty(exports, "R300", {
  enumerable: true,
  get: function get() {
    return _utils.R300;
  }
});
Object.defineProperty(exports, "R400", {
  enumerable: true,
  get: function get() {
    return _utils.R400;
  }
});
Object.defineProperty(exports, "R50", {
  enumerable: true,
  get: function get() {
    return _utils.R50;
  }
});
Object.defineProperty(exports, "R500", {
  enumerable: true,
  get: function get() {
    return _utils.R500;
  }
});
Object.defineProperty(exports, "R75", {
  enumerable: true,
  get: function get() {
    return _utils.R75;
  }
});
Object.defineProperty(exports, "T100", {
  enumerable: true,
  get: function get() {
    return _utils.T100;
  }
});
Object.defineProperty(exports, "T300", {
  enumerable: true,
  get: function get() {
    return _utils.T300;
  }
});
Object.defineProperty(exports, "T50", {
  enumerable: true,
  get: function get() {
    return _utils.T50;
  }
});
Object.defineProperty(exports, "T500", {
  enumerable: true,
  get: function get() {
    return _utils.T500;
  }
});
Object.defineProperty(exports, "T75", {
  enumerable: true,
  get: function get() {
    return _utils.T75;
  }
});
Object.defineProperty(exports, "Y200", {
  enumerable: true,
  get: function get() {
    return _utils.Y200;
  }
});
Object.defineProperty(exports, "Y400", {
  enumerable: true,
  get: function get() {
    return _utils.Y400;
  }
});
Object.defineProperty(exports, "Y50", {
  enumerable: true,
  get: function get() {
    return _utils.Y50;
  }
});
Object.defineProperty(exports, "Y500", {
  enumerable: true,
  get: function get() {
    return _utils.Y500;
  }
});
Object.defineProperty(exports, "Y75", {
  enumerable: true,
  get: function get() {
    return _utils.Y75;
  }
});
Object.defineProperty(exports, "acNameToEmoji", {
  enumerable: true,
  get: function get() {
    return _utils.acNameToEmoji;
  }
});
Object.defineProperty(exports, "acShortcutToEmoji", {
  enumerable: true,
  get: function get() {
    return _utils.acShortcutToEmoji;
  }
});
Object.defineProperty(exports, "alignment", {
  enumerable: true,
  get: function get() {
    return _schema.alignment;
  }
});
Object.defineProperty(exports, "alignmentPositionMap", {
  enumerable: true,
  get: function get() {
    return _schema.alignmentPositionMap;
  }
});
Object.defineProperty(exports, "annotation", {
  enumerable: true,
  get: function get() {
    return _schema.annotation;
  }
});
Object.defineProperty(exports, "backgroundColor", {
  enumerable: true,
  get: function get() {
    return _schema.backgroundColor;
  }
});
Object.defineProperty(exports, "backgroundColorPalette", {
  enumerable: true,
  get: function get() {
    return _schema.backgroundColorPalette;
  }
});
Object.defineProperty(exports, "blockCard", {
  enumerable: true,
  get: function get() {
    return _schema.blockCard;
  }
});
Object.defineProperty(exports, "blockquote", {
  enumerable: true,
  get: function get() {
    return _schema.blockquote;
  }
});
Object.defineProperty(exports, "blockquoteWithList", {
  enumerable: true,
  get: function get() {
    return _schema.blockquoteWithList;
  }
});
Object.defineProperty(exports, "blockquoteWithNestedCodeblockOrMedia", {
  enumerable: true,
  get: function get() {
    return _schema.blockquoteWithNestedCodeblockOrMedia;
  }
});
Object.defineProperty(exports, "blockquoteWithoutNonBodiedMacros", {
  enumerable: true,
  get: function get() {
    return _schema.blockquoteWithoutNonBodiedMacros;
  }
});
Object.defineProperty(exports, "bodiedExtension", {
  enumerable: true,
  get: function get() {
    return _schema.bodiedExtension;
  }
});
Object.defineProperty(exports, "border", {
  enumerable: true,
  get: function get() {
    return _schema.border;
  }
});
Object.defineProperty(exports, "borderColorPalette", {
  enumerable: true,
  get: function get() {
    return _schema.borderColorPalette;
  }
});
Object.defineProperty(exports, "breakout", {
  enumerable: true,
  get: function get() {
    return _schema.breakout;
  }
});
Object.defineProperty(exports, "buildAnnotationMarkDataAttributes", {
  enumerable: true,
  get: function get() {
    return _schema.buildAnnotationMarkDataAttributes;
  }
});
Object.defineProperty(exports, "bulletList", {
  enumerable: true,
  get: function get() {
    return _schema.bulletList;
  }
});
Object.defineProperty(exports, "bulletListSelector", {
  enumerable: true,
  get: function get() {
    return _schema.bulletListSelector;
  }
});
Object.defineProperty(exports, "caption", {
  enumerable: true,
  get: function get() {
    return _schema.caption;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function get() {
    return _schema.code;
  }
});
Object.defineProperty(exports, "codeBlock", {
  enumerable: true,
  get: function get() {
    return _schema.codeBlock;
  }
});
Object.defineProperty(exports, "codeBlockToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.codeBlockToJSON;
  }
});
Object.defineProperty(exports, "colorPalette", {
  enumerable: true,
  get: function get() {
    return _schema.colorPalette;
  }
});
Object.defineProperty(exports, "colorPaletteExtended", {
  enumerable: true,
  get: function get() {
    return _schema.colorPaletteExtended;
  }
});
Object.defineProperty(exports, "confluenceInlineComment", {
  enumerable: true,
  get: function get() {
    return _schema.confluenceInlineComment;
  }
});
Object.defineProperty(exports, "confluenceJiraIssue", {
  enumerable: true,
  get: function get() {
    return _schema.confluenceJiraIssue;
  }
});
Object.defineProperty(exports, "confluenceUnsupportedBlock", {
  enumerable: true,
  get: function get() {
    return _schema.confluenceUnsupportedBlock;
  }
});
Object.defineProperty(exports, "confluenceUnsupportedInline", {
  enumerable: true,
  get: function get() {
    return _schema.confluenceUnsupportedInline;
  }
});
Object.defineProperty(exports, "copyPrivateMediaAttributes", {
  enumerable: true,
  get: function get() {
    return _schema.copyPrivateMediaAttributes;
  }
});
Object.defineProperty(exports, "createPMMarkSpecFactory", {
  enumerable: true,
  get: function get() {
    return _schema.createPMMarkSpecFactory;
  }
});
Object.defineProperty(exports, "createPMNodeSpecFactory", {
  enumerable: true,
  get: function get() {
    return _schema.createPMNodeSpecFactory;
  }
});
Object.defineProperty(exports, "createSchema", {
  enumerable: true,
  get: function get() {
    return _schema.createSchema;
  }
});
Object.defineProperty(exports, "dataConsumer", {
  enumerable: true,
  get: function get() {
    return _schema.dataConsumer;
  }
});
Object.defineProperty(exports, "dataConsumerToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.dataConsumerToJSON;
  }
});
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function get() {
    return _schema.date;
  }
});
Object.defineProperty(exports, "decisionItem", {
  enumerable: true,
  get: function get() {
    return _schema.decisionItem;
  }
});
Object.defineProperty(exports, "decisionList", {
  enumerable: true,
  get: function get() {
    return _schema.decisionList;
  }
});
Object.defineProperty(exports, "decisionListSelector", {
  enumerable: true,
  get: function get() {
    return _schema.decisionListSelector;
  }
});
Object.defineProperty(exports, "doc", {
  enumerable: true,
  get: function get() {
    return _schema.doc;
  }
});
Object.defineProperty(exports, "em", {
  enumerable: true,
  get: function get() {
    return _schema.em;
  }
});
Object.defineProperty(exports, "embedCard", {
  enumerable: true,
  get: function get() {
    return _schema.embedCard;
  }
});
Object.defineProperty(exports, "emoji", {
  enumerable: true,
  get: function get() {
    return _schema.emoji;
  }
});
Object.defineProperty(exports, "emojiIdToAcName", {
  enumerable: true,
  get: function get() {
    return _utils.emojiIdToAcName;
  }
});
Object.defineProperty(exports, "expandToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.expandToJSON;
  }
});
Object.defineProperty(exports, "expandWithNestedExpand", {
  enumerable: true,
  get: function get() {
    return _schema.expandWithNestedExpand;
  }
});
Object.defineProperty(exports, "extendedBlockquote", {
  enumerable: true,
  get: function get() {
    return _schema.extendedBlockquote;
  }
});
Object.defineProperty(exports, "extendedPanel", {
  enumerable: true,
  get: function get() {
    return _schema.extendedPanel;
  }
});
Object.defineProperty(exports, "extension", {
  enumerable: true,
  get: function get() {
    return _schema.extension;
  }
});
Object.defineProperty(exports, "extensionFrame", {
  enumerable: true,
  get: function get() {
    return _schema.extensionFrame;
  }
});
Object.defineProperty(exports, "fragment", {
  enumerable: true,
  get: function get() {
    return _schema.fragment;
  }
});
Object.defineProperty(exports, "fragmentToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.fragmentToJSON;
  }
});
Object.defineProperty(exports, "generateUuid", {
  enumerable: true,
  get: function get() {
    return _utils.generateUuid;
  }
});
Object.defineProperty(exports, "getCellAttrs", {
  enumerable: true,
  get: function get() {
    return _schema.getCellAttrs;
  }
});
Object.defineProperty(exports, "getCellDomAttrs", {
  enumerable: true,
  get: function get() {
    return _schema.getCellDomAttrs;
  }
});
Object.defineProperty(exports, "getDarkModeLCHColor", {
  enumerable: true,
  get: function get() {
    return _utils.getDarkModeLCHColor;
  }
});
Object.defineProperty(exports, "getEmojiAcName", {
  enumerable: true,
  get: function get() {
    return _utils.getEmojiAcName;
  }
});
Object.defineProperty(exports, "getLinkMatch", {
  enumerable: true,
  get: function get() {
    return _utils.getLinkMatch;
  }
});
Object.defineProperty(exports, "hardBreak", {
  enumerable: true,
  get: function get() {
    return _schema.hardBreak;
  }
});
Object.defineProperty(exports, "heading", {
  enumerable: true,
  get: function get() {
    return _schema.heading;
  }
});
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _utils.hexToRgb;
  }
});
Object.defineProperty(exports, "hexToRgba", {
  enumerable: true,
  get: function get() {
    return _utils.hexToRgba;
  }
});
Object.defineProperty(exports, "image", {
  enumerable: true,
  get: function get() {
    return _schema.image;
  }
});
Object.defineProperty(exports, "indentation", {
  enumerable: true,
  get: function get() {
    return _schema.indentation;
  }
});
Object.defineProperty(exports, "inlineCard", {
  enumerable: true,
  get: function get() {
    return _schema.inlineCard;
  }
});
Object.defineProperty(exports, "inlineExtension", {
  enumerable: true,
  get: function get() {
    return _schema.inlineExtension;
  }
});
Object.defineProperty(exports, "inlineNodes", {
  enumerable: true,
  get: function get() {
    return _schema.inlineNodes;
  }
});
Object.defineProperty(exports, "isHex", {
  enumerable: true,
  get: function get() {
    return _utils.isHex;
  }
});
Object.defineProperty(exports, "isRgb", {
  enumerable: true,
  get: function get() {
    return _utils.isRgb;
  }
});
Object.defineProperty(exports, "isSafeUrl", {
  enumerable: true,
  get: function get() {
    return _utils.isSafeUrl;
  }
});
Object.defineProperty(exports, "layoutColumn", {
  enumerable: true,
  get: function get() {
    return _schema.layoutColumn;
  }
});
Object.defineProperty(exports, "layoutSection", {
  enumerable: true,
  get: function get() {
    return _schema.layoutSection;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function get() {
    return _schema.link;
  }
});
Object.defineProperty(exports, "linkToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.linkToJSON;
  }
});
Object.defineProperty(exports, "linkify", {
  enumerable: true,
  get: function get() {
    return _utils.linkify;
  }
});
Object.defineProperty(exports, "linkifyMatch", {
  enumerable: true,
  get: function get() {
    return _utils.linkifyMatch;
  }
});
Object.defineProperty(exports, "listItem", {
  enumerable: true,
  get: function get() {
    return _schema.listItem;
  }
});
Object.defineProperty(exports, "listItemWithDecisionStage0", {
  enumerable: true,
  get: function get() {
    return _schema.listItemWithDecisionStage0;
  }
});
Object.defineProperty(exports, "listItemWithTask", {
  enumerable: true,
  get: function get() {
    return _schema.listItemWithTask;
  }
});
Object.defineProperty(exports, "listItemWithoutNonBodiedMacros", {
  enumerable: true,
  get: function get() {
    return _schema.listItemWithoutNonBodiedMacros;
  }
});
Object.defineProperty(exports, "media", {
  enumerable: true,
  get: function get() {
    return _schema.media;
  }
});
Object.defineProperty(exports, "mediaGroup", {
  enumerable: true,
  get: function get() {
    return _schema.mediaGroup;
  }
});
Object.defineProperty(exports, "mediaInline", {
  enumerable: true,
  get: function get() {
    return _schema.mediaInline;
  }
});
Object.defineProperty(exports, "mediaSingle", {
  enumerable: true,
  get: function get() {
    return _schema.mediaSingle;
  }
});
Object.defineProperty(exports, "mediaSingleFull", {
  enumerable: true,
  get: function get() {
    return _schema.mediaSingleFull;
  }
});
Object.defineProperty(exports, "mediaSingleSpec", {
  enumerable: true,
  get: function get() {
    return _schema.mediaSingleSpec;
  }
});
Object.defineProperty(exports, "mediaSingleToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.mediaSingleToJSON;
  }
});
Object.defineProperty(exports, "mediaSingleWithCaption", {
  enumerable: true,
  get: function get() {
    return _schema.mediaSingleWithCaption;
  }
});
Object.defineProperty(exports, "mediaSingleWithWidthType", {
  enumerable: true,
  get: function get() {
    return _schema.mediaSingleWithWidthType;
  }
});
Object.defineProperty(exports, "mediaToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.mediaToJSON;
  }
});
Object.defineProperty(exports, "mention", {
  enumerable: true,
  get: function get() {
    return _schema.mention;
  }
});
Object.defineProperty(exports, "mentionToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.mentionToJSON;
  }
});
Object.defineProperty(exports, "multiBodiedExtension", {
  enumerable: true,
  get: function get() {
    return _schema.multiBodiedExtension;
  }
});
Object.defineProperty(exports, "nestedExpand", {
  enumerable: true,
  get: function get() {
    return _schema.nestedExpand;
  }
});
Object.defineProperty(exports, "nestedExpandWithoutNonBodiedMacros", {
  enumerable: true,
  get: function get() {
    return _schema.nestedExpandWithoutNonBodiedMacros;
  }
});
Object.defineProperty(exports, "normalizeHexColor", {
  enumerable: true,
  get: function get() {
    return _utils.normalizeHexColor;
  }
});
Object.defineProperty(exports, "normalizeUrl", {
  enumerable: true,
  get: function get() {
    return _utils.normalizeUrl;
  }
});
Object.defineProperty(exports, "orderedList", {
  enumerable: true,
  get: function get() {
    return _schema.orderedList;
  }
});
Object.defineProperty(exports, "orderedListSelector", {
  enumerable: true,
  get: function get() {
    return _schema.orderedListSelector;
  }
});
Object.defineProperty(exports, "orderedListWithOrder", {
  enumerable: true,
  get: function get() {
    return _schema.orderedListWithOrder;
  }
});
Object.defineProperty(exports, "panelWithoutNestedNonBodiedMacros", {
  enumerable: true,
  get: function get() {
    return _schema.panelWithoutNestedNonBodiedMacros;
  }
});
Object.defineProperty(exports, "paragraph", {
  enumerable: true,
  get: function get() {
    return _schema.paragraph;
  }
});
Object.defineProperty(exports, "placeholder", {
  enumerable: true,
  get: function get() {
    return _schema.placeholder;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _utils.rgbToHex;
  }
});
Object.defineProperty(exports, "rule", {
  enumerable: true,
  get: function get() {
    return _schema.rule;
  }
});
Object.defineProperty(exports, "sanitizeNodes", {
  enumerable: true,
  get: function get() {
    return _schema.sanitizeNodes;
  }
});
Object.defineProperty(exports, "status", {
  enumerable: true,
  get: function get() {
    return _schema.status;
  }
});
Object.defineProperty(exports, "strike", {
  enumerable: true,
  get: function get() {
    return _schema.strike;
  }
});
Object.defineProperty(exports, "strong", {
  enumerable: true,
  get: function get() {
    return _schema.strong;
  }
});
Object.defineProperty(exports, "subsup", {
  enumerable: true,
  get: function get() {
    return _schema.subsup;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function get() {
    return _schema.table;
  }
});
Object.defineProperty(exports, "tableBackgroundBorderColor", {
  enumerable: true,
  get: function get() {
    return _schema.tableBackgroundBorderColor;
  }
});
Object.defineProperty(exports, "tableBackgroundColorNames", {
  enumerable: true,
  get: function get() {
    return _schema.tableBackgroundColorNames;
  }
});
Object.defineProperty(exports, "tableBackgroundColorPalette", {
  enumerable: true,
  get: function get() {
    return _schema.tableBackgroundColorPalette;
  }
});
Object.defineProperty(exports, "tableCell", {
  enumerable: true,
  get: function get() {
    return _schema.tableCell;
  }
});
Object.defineProperty(exports, "tableCellContentDomSelector", {
  enumerable: true,
  get: function get() {
    return _schema.tableCellContentDomSelector;
  }
});
Object.defineProperty(exports, "tableCellContentWrapperSelector", {
  enumerable: true,
  get: function get() {
    return _schema.tableCellContentWrapperSelector;
  }
});
Object.defineProperty(exports, "tableCellSelector", {
  enumerable: true,
  get: function get() {
    return _schema.tableCellSelector;
  }
});
Object.defineProperty(exports, "tableCellWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _schema.tableCellWithNestedTable;
  }
});
Object.defineProperty(exports, "tableHeader", {
  enumerable: true,
  get: function get() {
    return _schema.tableHeader;
  }
});
Object.defineProperty(exports, "tableHeaderSelector", {
  enumerable: true,
  get: function get() {
    return _schema.tableHeaderSelector;
  }
});
Object.defineProperty(exports, "tableHeaderWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _schema.tableHeaderWithNestedTable;
  }
});
Object.defineProperty(exports, "tablePrefixSelector", {
  enumerable: true,
  get: function get() {
    return _schema.tablePrefixSelector;
  }
});
Object.defineProperty(exports, "tableRow", {
  enumerable: true,
  get: function get() {
    return _schema.tableRow;
  }
});
Object.defineProperty(exports, "tableRowWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _schema.tableRowWithNestedTable;
  }
});
Object.defineProperty(exports, "tableStage0", {
  enumerable: true,
  get: function get() {
    return _schema.tableStage0;
  }
});
Object.defineProperty(exports, "tableToJSON", {
  enumerable: true,
  get: function get() {
    return _schema.tableToJSON;
  }
});
Object.defineProperty(exports, "tableWithCustomWidth", {
  enumerable: true,
  get: function get() {
    return _schema.tableWithCustomWidth;
  }
});
Object.defineProperty(exports, "tableWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _schema.tableWithNestedTable;
  }
});
Object.defineProperty(exports, "taskItem", {
  enumerable: true,
  get: function get() {
    return _schema.taskItem;
  }
});
Object.defineProperty(exports, "taskList", {
  enumerable: true,
  get: function get() {
    return _schema.taskList;
  }
});
Object.defineProperty(exports, "taskListSelector", {
  enumerable: true,
  get: function get() {
    return _schema.taskListSelector;
  }
});
Object.defineProperty(exports, "text", {
  enumerable: true,
  get: function get() {
    return _schema.text;
  }
});
Object.defineProperty(exports, "textColor", {
  enumerable: true,
  get: function get() {
    return _schema.textColor;
  }
});
Object.defineProperty(exports, "toJSONTableCell", {
  enumerable: true,
  get: function get() {
    return _schema.toJSONTableCell;
  }
});
Object.defineProperty(exports, "toJSONTableHeader", {
  enumerable: true,
  get: function get() {
    return _schema.toJSONTableHeader;
  }
});
Object.defineProperty(exports, "typeAheadQuery", {
  enumerable: true,
  get: function get() {
    return _schema.typeAheadQuery;
  }
});
Object.defineProperty(exports, "underline", {
  enumerable: true,
  get: function get() {
    return _schema.underline;
  }
});
Object.defineProperty(exports, "unknownBlock", {
  enumerable: true,
  get: function get() {
    return _schema.unknownBlock;
  }
});
Object.defineProperty(exports, "unsupportedBlock", {
  enumerable: true,
  get: function get() {
    return _schema.unsupportedBlock;
  }
});
Object.defineProperty(exports, "unsupportedInline", {
  enumerable: true,
  get: function get() {
    return _schema.unsupportedInline;
  }
});
Object.defineProperty(exports, "unsupportedMark", {
  enumerable: true,
  get: function get() {
    return _schema.unsupportedMark;
  }
});
Object.defineProperty(exports, "unsupportedNodeAttribute", {
  enumerable: true,
  get: function get() {
    return _schema.unsupportedNodeAttribute;
  }
});
Object.defineProperty(exports, "unsupportedNodeTypesForMediaCards", {
  enumerable: true,
  get: function get() {
    return _schema.unsupportedNodeTypesForMediaCards;
  }
});
Object.defineProperty(exports, "uuid", {
  enumerable: true,
  get: function get() {
    return _utils.uuid;
  }
});
var _schema = require("./schema");
var _utils = require("./utils");
