// @atlaskit/navigation-system/dist/cjs/ui/top-nav-items/chat-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChatButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _aiChat = _interopRequireDefault(require("@atlaskit/icon/core/ai-chat"));
var _listItem = require("../../components/list-item");
var _migration = require("./themed/migration");
/**
 * __Chat button__
 *
 * The chat button for the top navigation.
 * TODO: check if this is still needed and remove. Jira doesn't use it - there's a separate button for ConversationAssistant
 */
var ChatButton = exports.ChatButton = function ChatButton(_ref) {
  var children = _ref.children,
    onClick = _ref.onClick,
    isSelected = _ref.isSelected,
    interactionName = _ref.interactionName;
  return /*#__PURE__*/_react.default.createElement(_listItem.ListItem, null, /*#__PURE__*/_react.default.createElement(_migration.Button, {
    appearance: "default",
    iconBefore: _aiChat.default,
    onClick: onClick,
    isSelected: isSelected,
    interactionName: interactionName
  }, children));
};
