// @atlaskit/react-ufo/dist/cjs/vc/vc-observer/attachAbortListeners.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.attachAbortListeners = attachAbortListeners;
var _bindEventListener = require("bind-event-listener");
var _types = require("../../common/vc/types");
var _getViewport = require("./getViewport");
function attachAbortListeners(window, initialViewPort, callback) {
  var unbindWheel = (0, _bindEventListener.bind)(window, {
    type: _types.AbortEvent.wheel,
    listener: function listener(e) {
      callback(_types.AbortEvent.wheel, e.timeStamp);
    },
    options: {
      passive: true,
      once: true
    }
  });
  var unbindKeydown = (0, _bindEventListener.bind)(window, {
    type: _types.AbortEvent.keydown,
    listener: function listener(e) {
      callback(_types.AbortEvent.keydown, e.timeStamp);
    },
    options: {
      once: true
    }
  });
  var unbindResize = (0, _bindEventListener.bind)(window, {
    type: _types.AbortEvent.resize,
    listener: function listener(e) {
      if ((0, _getViewport.getViewportWidth)() !== initialViewPort.w || (0, _getViewport.getViewportHeight)() !== initialViewPort.h) {
        callback(_types.AbortEvent.resize, e.timeStamp);
        unbindResize();
      }
    }
  });
  return [unbindWheel, unbindKeydown, unbindResize];
}
