// @atlaskit/navigation-system/dist/cjs/ui/top-nav-items/nav-logo/logo-renderer.js
/* logo-renderer.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogoRenderer = LogoRenderer;
exports.themedLogoText = exports.themedLogoIcon = void 0;
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _hasCustomThemeContext = require("../themed/has-custom-theme-context");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var themedLogoIcon = exports.themedLogoIcon = '--ds-top-bar-logo-icon';
var themedLogoText = exports.themedLogoText = '--ds-top-bar-logo-text';
function LogoRenderer(_ref) {
  var LogoOrIcon = _ref.logoOrIcon,
    shouldUseNewLogoDesign = _ref.shouldUseNewLogoDesign;
  var hasCustomTheme = (0, _hasCustomThemeContext.useHasCustomTheme)();
  if (hasCustomTheme) {
    return /*#__PURE__*/React.createElement(LogoOrIcon, {
      label: "",
      size: "small",
      shouldUseNewLogoDesign: shouldUseNewLogoDesign,
      iconColor: "var(".concat(themedLogoIcon, ")"),
      textColor: "var(".concat(themedLogoText, ")")
    });
  }
  return /*#__PURE__*/React.createElement(LogoOrIcon, {
    size: "small",
    shouldUseNewLogoDesign: shouldUseNewLogoDesign,
    label: "",
    appearance: "brand"
  });
}
