// @atlaskit/adf-schema-generator/dist/cjs/transforms/adfToPm/buildPmSpec.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildNodeSpec = exports.buildMarkSpec = exports.buildMarkExcludes = exports.buildAttrs = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _merge = _interopRequireDefault(require("lodash/merge"));
var _adfMark = require("../../adfMark");
var _ADFMarkSpec = require("../../types/ADFMarkSpec");
var _buildContentExpression = require("./buildContentExpression");
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// @DSLCompatibilityException
var excludesLinkMarksNodes = ['doc', 'layoutColumn'];

// @DSLCompatibilityException
var excludesIndentationNodes = ['tableHeader', 'tableCell'];

// @DSLCompatibilityException
var NODES_MARKS_OVERRIDES = {
  bodiedExtension: ['dataConsumer', 'fragment', 'unsupportedMark', 'unsupportedNodeAttribute'],
  codeBlock: ['unsupportedMark', 'unsupportedNodeAttribute'],
  extensionFrame: ['dataConsumer', 'fragment', 'unsupportedMark', 'unsupportedNodeAttribute'],
  multiBodiedExtension: ['unsupportedNodeAttribute', 'unsupportedMark']
};
var buildAttrs = exports.buildAttrs = function buildAttrs(attrs) {
  var attributes = attrs;
  if (!attrs) {
    return undefined;
  }
  if (attrs['anyOf']) {
    attributes = _merge.default.apply(void 0, [{}].concat((0, _toConsumableArray2.default)(attrs['anyOf'])));
  }
  // @ts-expect-error
  return Object.entries(attributes).reduce(function (acc, _ref) {
    var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
      key = _ref2[0],
      value = _ref2[1];
    if (value.default === undefined && value.optional) {
      return acc;
    }
    return _objectSpread(_objectSpread({}, acc), {}, (0, _defineProperty2.default)({}, key, _objectSpread(_objectSpread({}, acc[key]), 'default' in value && {
      default: value.default
    })));
  }, {});
};
var buildMarkExcludes = exports.buildMarkExcludes = function buildMarkExcludes(mark) {
  var originalMarkExcludes = mark.getSpec().excludes;
  if (originalMarkExcludes === undefined) {
    return undefined;
  }
  if (originalMarkExcludes === _ADFMarkSpec.MarkExcludesAll || originalMarkExcludes === _ADFMarkSpec.MarkExcludesNone) {
    return originalMarkExcludes;
  }
  return originalMarkExcludes.filter(function (markOrGroup) {
    if (markOrGroup) {
      return true;
    }
    // eslint-disable-next-line no-console
    console.warn('Excludes markOrGroup is undefined for', "\"".concat(mark.getType(), "\"."), 'Likely caused by a circular dependency.');
    return false;
  }).map(function (markOrGroup) {
    if (markOrGroup instanceof _adfMark.ADFMark) {
      return markOrGroup.getType();
    } else {
      return markOrGroup.group;
    }
  }).join(' ');
};
var filterMarks = function filterMarks(nodeType) {
  return function (mark) {
    if (mark === 'link') {
      return !excludesLinkMarksNodes.includes(nodeType);
    }
    if (mark === 'indentation') {
      return !excludesIndentationNodes.includes(nodeType);
    }
    return true;
  };
};
var buildMarkSpec = exports.buildMarkSpec = function buildMarkSpec(mark) {
  var pmMarkSpec = {};
  pmMarkSpec.attrs = buildAttrs(mark.getSpec().attrs);
  pmMarkSpec.inclusive = mark.getSpec().inclusive;
  pmMarkSpec.excludes = buildMarkExcludes(mark);
  pmMarkSpec.group = mark.getGroup();
  pmMarkSpec.spanning = mark.getSpec().spanning;
  return Object.entries(pmMarkSpec).reduce(function (acc, _ref3) {
    var _ref4 = (0, _slicedToArray2.default)(_ref3, 2),
      key = _ref4[0],
      value = _ref4[1];
    if (value !== undefined && value !== null) {
      acc[key] = value;
    }
    return acc;
  }, {});
};
var buildNodeSpec = exports.buildNodeSpec = function buildNodeSpec(nodeType, nodeSpec, nodeGroups, content) {
  var _nodeSpec$DANGEROUS_M;
  var pmNodeSpec = {};
  var marks = [];
  var contents = [];
  var _iterator = _createForOfIteratorHelper(content),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var _marks;
      var child = _step.value;
      contents.push((0, _buildContentExpression.buildContentExpression)(child.expr));
      (_marks = marks).push.apply(_marks, (0, _toConsumableArray2.default)(child.marks));
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  if (nodeType in NODES_MARKS_OVERRIDES) {
    marks = NODES_MARKS_OVERRIDES[nodeType];
  } else if (nodeSpec.allowAnyChildMark) {
    marks = ['_'];
  } else if (nodeSpec.allowNoChildMark) {
    marks = [''];
  } else {
    marks = Array.apply(void 0, (0, _toConsumableArray2.default)(new Set(marks))).filter(filterMarks(nodeType));
  }
  if (content.length) {
    pmNodeSpec.content = Array.apply(void 0, (0, _toConsumableArray2.default)(new Set(contents))).join(' ');
  }
  if (marks.length) {
    pmNodeSpec.marks = marks.join(' ');
  }
  if (nodeGroups.length > 0) {
    pmNodeSpec.group = nodeGroups.join(' ');
  }
  pmNodeSpec.inline = nodeSpec.inline;
  pmNodeSpec.atom = nodeSpec.atom;
  pmNodeSpec.attrs = buildAttrs(nodeSpec.attrs);
  pmNodeSpec.selectable = nodeSpec.selectable;
  pmNodeSpec.draggable = nodeSpec.draggable;
  pmNodeSpec.code = nodeSpec.code;
  pmNodeSpec.whitespace = nodeSpec.whitespace;
  pmNodeSpec.definingAsContext = nodeSpec.definingAsContext;
  pmNodeSpec.definingForContent = nodeSpec.definingForContent;
  pmNodeSpec.defining = nodeSpec.defining;
  pmNodeSpec.isolating = nodeSpec.isolating;
  pmNodeSpec.tableRole = nodeSpec.tableRole;
  pmNodeSpec.linebreakReplacement = nodeSpec.linebreakReplacement;
  var overrides = ((_nodeSpec$DANGEROUS_M = nodeSpec.DANGEROUS_MANUAL_OVERRIDE) === null || _nodeSpec$DANGEROUS_M === void 0 ? void 0 : _nodeSpec$DANGEROUS_M['pm-spec']) || {};
  for (var _i = 0, _Object$entries = Object.entries(overrides); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = (0, _slicedToArray2.default)(_Object$entries[_i], 2),
      key = _Object$entries$_i[0],
      override = _Object$entries$_i[1];
    pmNodeSpec[key] = override.value;
  }
  return Object.entries(pmNodeSpec).reduce(function (acc, _ref5) {
    var _ref6 = (0, _slicedToArray2.default)(_ref5, 2),
      key = _ref6[0],
      value = _ref6[1];
    if (value !== undefined && value !== null) {
      acc[key] = value;
    }
    return acc;
  }, {});
};
