// @atlaskit/side-nav-items/dist/cjs/ui/menu-item/flyout-menu-item/flyout-header.js
/* flyout-header.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutHeader = void 0;
require("./flyout-header.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _compiled = require("@atlaskit/primitives/compiled");
var _closeButton = require("./close-button");
var _flyoutMenuItemContext = require("./flyout-menu-item-context");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var headerStyles = {
  root: "_zulp12x7 _bozg1b66 _85i512x7 _1e0c1txw _2lx21bp4",
  flex: "_zulppxbi _1bah1yb4 _2lx21sbv _4cvr1h6o _1bsb1osq _bozgv77o"
};
/**
 * __FlyoutHeader__
 *
 * The header section of a flyout menu. This component displays the title, and
 * close button, as well as any other provided actions relevant to the menu.
 * This component should be placed first within the FlyoutMenuItemContent.
 */
var FlyoutHeader = exports.FlyoutHeader = function FlyoutHeader(props) {
  var children = props.children,
    testId = props.testId,
    title = props.title,
    closeButtonLabel = props.closeButtonLabel;
  var id = (0, _flyoutMenuItemContext.useTitleId)();
  var onCloseRef = (0, _react.useContext)(_flyoutMenuItemContext.OnCloseContext);
  var handleClose = (0, _react.useCallback)(function (event) {
    var _onCloseRef$current;
    (_onCloseRef$current = onCloseRef.current) === null || _onCloseRef$current === void 0 || _onCloseRef$current.call(onCloseRef, event, 'close-button');
  }, [onCloseRef]);
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": testId,
    className: (0, _runtime.ax)([headerStyles.root])
  }, /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    xcss: headerStyles.flex
  }, /*#__PURE__*/_react.default.createElement(_closeButton.CloseButton, {
    label: closeButtonLabel,
    onClick: handleClose,
    testId: testId && "".concat(testId, "--close-button")
  }), /*#__PURE__*/_react.default.createElement(_heading.default, {
    size: "xsmall",
    as: "h2",
    id: id
  }, title)), children);
};
