// @atlaskit/side-nav-items/dist/cjs/ui/menu-item/flyout-menu-item/flyout-body.js
/* flyout-body.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutBody = void 0;
require("./flyout-body.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var bodyStyles = {
  root: "_1e0c1txw _2lx21bp4 _18m91wug _4t3i1osq _1bahv2br _ahbqx0bf _12l2x0bf _6rthx0bf _bozgu2gc _y4tiu2gc _85i5u2gc"
};
/**
 * __Flyout menu item body__
 *
 * The main section of a flyout menu. This component is used to render the
 * primary contents of the flyout menu. This component should be placed between
 * FlyoutHeader and FlyoutFooter (if present), as is scrollable if the content
 * exceeds the available space.
 */
var FlyoutBody = exports.FlyoutBody = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var children = props.children,
    testId = props.testId,
    onKeyDown = props.onKeyDown,
    onKeyUp = props.onKeyUp,
    onBlurCapture = props.onBlurCapture;
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    "data-testid": testId,
    onKeyDown: onKeyDown,
    onKeyUp: onKeyUp,
    onBlurCapture: onBlurCapture,
    role: "presentation",
    className: (0, _runtime.ax)([bodyStyles.root])
  }, children);
});
