// @atlaskit/tooltip/dist/es2019/utilities.js
export function getMousePosition(mouseCoordinates) {
  const safeMouse = mouseCoordinates || {
    top: 0,
    left: 0
  };
  const getBoundingClientRect = () => ({
    top: safeMouse.top,
    left: safeMouse.left,
    bottom: safeMouse.top,
    right: safeMouse.left,
    width: 0,
    height: 0
  });
  return {
    getBoundingClientRect,
    clientWidth: 0,
    clientHeight: 0
  };
}
export function getVirtualElementFromMousePos(mousePos, {
  targetElement,
  tooltipPosition
}) {
  if (tooltipPosition === 'mouse') {
    return {
      getBoundingClientRect: () => DOMRect.fromRect({
        x: mousePos.clientX,
        y: mousePos.clientY,
        width: 0,
        height: 0
      })
    };
  }
  const targetRect = targetElement.getBoundingClientRect();
  if (tooltipPosition === 'mouse-x') {
    return {
      getBoundingClientRect: () => DOMRect.fromRect({
        x: mousePos.clientX,
        y: targetRect.top,
        width: 0,
        height: targetRect.height
      })
    };
  }
  if (tooltipPosition === 'mouse-y') {
    return {
      getBoundingClientRect: () => DOMRect.fromRect({
        x: targetRect.left,
        y: mousePos.clientY,
        width: targetRect.width,
        height: 0
      })
    };
  }
  throw new Error(`Invalid tooltip position for virtual element: ${tooltipPosition}`);
}
