// @atlaskit/react-ufo/dist/esm/segment/ssr-render-profiler.js
var _process;
import React, { createContext, useContext, useMemo } from 'react';
import { getConfig } from '../config';
import { getActiveInteraction } from '../interaction-metrics';
// These are stored outside react context to be resilient to concurrent mode
// restarting the start marker from a suspense and losing the initial render
var startTimes = new Map();
var spanStates = new Map();
export var clearState = function clearState() {
  startTimes.clear();
  spanStates.clear();
};
// Keep track of the last interaction id and reset the start timers if it ever changes.
// This is to allow multi-step ssr to track the render timers from different "interaction"s seperately
var lastActiveInteraction;
function checkActiveInteractionAndResetStartMarksIfSet() {
  var _getActiveInteraction;
  var activeInteractionId = (_getActiveInteraction = getActiveInteraction()) === null || _getActiveInteraction === void 0 ? void 0 : _getActiveInteraction.id;
  if (!!lastActiveInteraction && lastActiveInteraction !== activeInteractionId) {
    clearState();
  }
  lastActiveInteraction = activeInteractionId;
}
var getIsNativeTracingEnabled = function getIsNativeTracingEnabled() {
  var _getConfig$ssr$enable, _getConfig;
  return (_getConfig$ssr$enable = (_getConfig = getConfig()) === null || _getConfig === void 0 || (_getConfig = _getConfig.ssr) === null || _getConfig === void 0 ? void 0 : _getConfig.enableNativeTracing) !== null && _getConfig$ssr$enable !== void 0 ? _getConfig$ssr$enable : false;
};
var onStartRender = function onStartRender(id, currentSegmentName, parentContext) {
  var _startTimes$get, _vm_internals__, _spanStates$get;
  if (!startTimes.has(id)) {
    startTimes.set(id, []);
  }
  (_startTimes$get = startTimes.get(id)) === null || _startTimes$get === void 0 || _startTimes$get.push(performance.now());
  if (!getIsNativeTracingEnabled()) {
    return null;
  }
  var startSpan = (_vm_internals__ = globalThis.__vm_internals__) === null || _vm_internals__ === void 0 || (_vm_internals__ = _vm_internals__.telemetry) === null || _vm_internals__ === void 0 ? void 0 : _vm_internals__.startSpan;
  var spanState = (_spanStates$get = spanStates.get(id)) !== null && _spanStates$get !== void 0 ? _spanStates$get : null;
  if (!spanState && startSpan) {
    var span = startSpan(currentSegmentName, {
      parentSpanId: parentContext === null || parentContext === void 0 ? void 0 : parentContext.spanId
    });
    spanState = {
      span: span
    };
    spanStates.set(id, spanState);
  }
  return spanState;
};
var isInSSR = typeof __SERVER__ !== 'undefined' && __SERVER__ || typeof process !== 'undefined' && Boolean(((_process = process) === null || _process === void 0 || (_process = _process.env) === null || _process === void 0 ? void 0 : _process.REACT_SSR) || false);
var ProfilerMarker = function ProfilerMarker(_ref) {
  var onRender = _ref.onRender;
  onRender === null || onRender === void 0 || onRender();
  return null;
};
var ParentSpanContext = /*#__PURE__*/createContext(null);

// For profiler spans in SSR, we 'end' any with their latest end times set, we need to do this at UFO-interaction-end since until then the component could rerender/remount again and spans fundamentally don't use the last 'end' but the first
export var flushSsrRenderProfilerTraces = function flushSsrRenderProfilerTraces() {
  spanStates.forEach(function (spanState) {
    if (spanState.latestEndTime != null) {
      spanState.span.end(spanState.latestEndTime);
    }
  });
  spanStates.clear();
};
export var SsrRenderProfilerInner = function SsrRenderProfilerInner(_ref2) {
  var _currentSpanState$spa;
  var children = _ref2.children,
    labelStack = _ref2.labelStack,
    _onRender = _ref2.onRender;
  var reactProfilerId = useMemo(function () {
    return labelStack.map(function (l) {
      return l.name;
    }).join('/');
  }, [labelStack]);
  checkActiveInteractionAndResetStartMarksIfSet();
  var parentSpan = useContext(ParentSpanContext);
  var currentSpanState = onStartRender(reactProfilerId, labelStack[labelStack.length - 1].name, parentSpan);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ParentSpanContext.Provider, {
    value: (_currentSpanState$spa = currentSpanState === null || currentSpanState === void 0 ? void 0 : currentSpanState.span.getSpanContext()) !== null && _currentSpanState$spa !== void 0 ? _currentSpanState$spa : null
  }, children, /*#__PURE__*/React.createElement(ProfilerMarker, {
    onRender: function onRender() {
      var startTimesForId = startTimes.get(reactProfilerId);
      if (startTimesForId !== null && startTimesForId !== void 0 && startTimesForId.length) {
        var _endTime = performance.now();
        var firstStartTime = startTimesForId[0];
        var lastStartTime = startTimesForId[startTimesForId.length - 1];
        var baseDuration = _endTime - lastStartTime;
        var actualDuration = _endTime - firstStartTime;
        _onRender('mount',
        // this is incorrect, but on the server there is no mount phase
        actualDuration, baseDuration, firstStartTime, _endTime);
      }
      var spanState = spanStates.get(reactProfilerId);
      if (spanState) {
        spanState.latestEndTime = performance.now();
      }
    }
  })));
};
var SsrRenderProfiler = function SsrRenderProfiler(props) {
  if (isInSSR) {
    return /*#__PURE__*/React.createElement(SsrRenderProfilerInner, props);
  }

  // ensure structure similar to SSR implementation
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ProfilerMarker, null), props.children);
};
export default SsrRenderProfiler;
