// @atlaskit/dynamic-table/dist/es2019/styled/dynamic-table.js
/* dynamic-table.tsx generated by @compiled/babel-plugin v0.38.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./dynamic-table.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @repo/internal/react/require-jsdoc */

import { forwardRef } from 'react';
const fixedSizeTableStyles = null;
const tableStyles = null;
const bodyBorder = null;
export const Table = /*#__PURE__*/forwardRef(({
  isFixedSize,
  hasDataRow,
  children,
  testId,
  isLoading,
  ...rest
}, ref) => {
  return /*#__PURE__*/React.createElement("table", _extends({
    inert: isLoading ? '' : undefined,
    style: {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      '--local-dynamic-table-hover-bg': "var(--ds-background-neutral-subtle-hovered, #0515240F)",
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      '--local-dynamic-table-highlighted-bg': "var(--ds-background-selected, #E9F2FE)",
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      '--local-dynamic-table-hover-highlighted-bg': "var(--ds-background-selected-hovered, #CFE1FD)",
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      '--local-dynamic-table-row-focus-outline': "var(--ds-border-focused, #4688EC)"
    },
    ref: ref
  }, rest, {
    "data-testid": testId && `${testId}--table`,
    className: ax(["_1bsb1osq _yq5hus1c _btyzidpf _ect4ttxp", isFixedSize && "_1kqm1n9t", hasDataRow && "_179r1uh4"])
  }), children);
});
const captionStyles = null;
export const Caption = ({
  children
}) => /*#__PURE__*/React.createElement("caption", {
  className: ax(["_11c81af2 _6rthu2gc _1pfh1ejb _m6k41e03"])
}, children);
const paginationWrapperStyles = null;
export const PaginationWrapper = ({
  children,
  testId
}) => /*#__PURE__*/React.createElement("div", {
  "data-testid": testId && `${testId}--pagination--container`,
  className: ax(["_1e0c1txw _1bah1h6o"])
}, children);
