// @atlaskit/object/dist/cjs/components/object-tile/components/database.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DatabaseObjectTile;
var _react = _interopRequireDefault(require("react"));
var _database = _interopRequireDefault(require("@atlaskit/icon/core/database"));
var _objectTileBase = _interopRequireDefault(require("../object-tile-base"));
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::a5c890b63fcdc09226ded606bd7c322e>>
 * @codegenCommand yarn build-glyphs
 */

function DatabaseObjectTile(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Database' : _ref$label,
    size = _ref.size,
    testId = _ref.testId,
    isBold = _ref.isBold;
  return /*#__PURE__*/_react.default.createElement(_objectTileBase.default, {
    label: label,
    size: size,
    testId: testId,
    isBold: isBold,
    icon: _database.default,
    color: isBold ? "var(--ds-icon, #292A2E)" : "var(--ds-icon-accent-purple, #AF59E1)",
    backgroundColor: isBold ? 'color.background.accent.purple.subtle' : undefined
  });
}
