// @atlaskit/object/dist/cjs/components/object-tile/components/incident.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = IncidentObjectTile;
var _react = _interopRequireDefault(require("react"));
var _incident = _interopRequireDefault(require("@atlaskit/icon/core/incident"));
var _objectTileBase = _interopRequireDefault(require("../object-tile-base"));
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::7b52a9f186ad8e116ad7e1e0109b235a>>
 * @codegenCommand yarn build-glyphs
 */

function IncidentObjectTile(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Incident' : _ref$label,
    size = _ref.size,
    testId = _ref.testId,
    isBold = _ref.isBold;
  return /*#__PURE__*/_react.default.createElement(_objectTileBase.default, {
    label: label,
    size: size,
    testId: testId,
    isBold: isBold,
    icon: _incident.default,
    color: isBold ? "var(--ds-icon, #292A2E)" : "var(--ds-icon-accent-red, #C9372C)",
    backgroundColor: isBold ? 'color.background.accent.red.subtle' : undefined
  });
}
