// @atlaskit/object/dist/cjs/components/object-tile/components/new-feature.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NewFeatureObjectTile;
var _react = _interopRequireDefault(require("react"));
var _add = _interopRequireDefault(require("@atlaskit/icon/core/add"));
var _objectTileBase = _interopRequireDefault(require("../object-tile-base"));
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::6f2158d44f4432eda77e2e6957e015a7>>
 * @codegenCommand yarn build-glyphs
 */

function NewFeatureObjectTile(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'New Feature' : _ref$label,
    size = _ref.size,
    testId = _ref.testId,
    isBold = _ref.isBold;
  return /*#__PURE__*/_react.default.createElement(_objectTileBase.default, {
    label: label,
    size: size,
    testId: testId,
    isBold: isBold,
    icon: _add.default,
    color: isBold ? "var(--ds-icon, #292A2E)" : "var(--ds-icon-accent-green, #22A06B)",
    backgroundColor: isBold ? 'color.background.accent.green.subtle' : undefined
  });
}
