// @atlaskit/object/dist/cjs/components/object-tile/components/page-live-doc.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PageLiveDocObjectTile;
var _react = _interopRequireDefault(require("react"));
var _pageLiveDoc = _interopRequireDefault(require("@atlaskit/icon-lab/core/page-live-doc"));
var _objectTileBase = _interopRequireDefault(require("../object-tile-base"));
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::a95f50e5543417b936016ac441fe40a9>>
 * @codegenCommand yarn build-glyphs
 */

function PageLiveDocObjectTile(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Page Live Doc' : _ref$label,
    size = _ref.size,
    testId = _ref.testId,
    isBold = _ref.isBold;
  return /*#__PURE__*/_react.default.createElement(_objectTileBase.default, {
    label: label,
    size: size,
    testId: testId,
    isBold: isBold,
    icon: _pageLiveDoc.default,
    color: isBold ? "var(--ds-icon, #292A2E)" : "var(--ds-icon-accent-magenta, #CD519D)",
    backgroundColor: isBold ? 'color.background.accent.magenta.subtle' : undefined
  });
}
