// @atlaskit/object/dist/cjs/components/object-tile/components/question.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = QuestionObjectTile;
var _react = _interopRequireDefault(require("react"));
var _questionCircle = _interopRequireDefault(require("@atlaskit/icon/core/question-circle"));
var _objectTileBase = _interopRequireDefault(require("../object-tile-base"));
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::4724c4007b146c29264b0c47a7639898>>
 * @codegenCommand yarn build-glyphs
 */

function QuestionObjectTile(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Question' : _ref$label,
    size = _ref.size,
    testId = _ref.testId,
    isBold = _ref.isBold;
  return /*#__PURE__*/_react.default.createElement(_objectTileBase.default, {
    label: label,
    size: size,
    testId: testId,
    isBold: isBold,
    icon: _questionCircle.default,
    color: isBold ? "var(--ds-icon, #292A2E)" : "var(--ds-icon-accent-purple, #AF59E1)",
    backgroundColor: isBold ? 'color.background.accent.purple.subtle' : undefined
  });
}
