// @atlaskit/object/dist/cjs/components/object-tile/components/whiteboard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = WhiteboardObjectTile;
var _react = _interopRequireDefault(require("react"));
var _whiteboard = _interopRequireDefault(require("@atlaskit/icon/core/whiteboard"));
var _objectTileBase = _interopRequireDefault(require("../object-tile-base"));
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::e2e669e2245e139e9284b5e3e471ca9f>>
 * @codegenCommand yarn build-glyphs
 */

function WhiteboardObjectTile(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Whiteboard' : _ref$label,
    size = _ref.size,
    testId = _ref.testId,
    isBold = _ref.isBold;
  return /*#__PURE__*/_react.default.createElement(_objectTileBase.default, {
    label: label,
    size: size,
    testId: testId,
    isBold: isBold,
    icon: _whiteboard.default,
    color: isBold ? "var(--ds-icon, #292A2E)" : "var(--ds-icon-accent-teal, #2898BD)",
    backgroundColor: isBold ? 'color.background.accent.teal.subtle' : undefined
  });
}
