// @atlaskit/tooltip/dist/cjs/utilities.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMousePosition = getMousePosition;
exports.getVirtualElementFromMousePos = getVirtualElementFromMousePos;
function getMousePosition(mouseCoordinates) {
  var safeMouse = mouseCoordinates || {
    top: 0,
    left: 0
  };
  var getBoundingClientRect = function getBoundingClientRect() {
    return {
      top: safeMouse.top,
      left: safeMouse.left,
      bottom: safeMouse.top,
      right: safeMouse.left,
      width: 0,
      height: 0
    };
  };
  return {
    getBoundingClientRect: getBoundingClientRect,
    clientWidth: 0,
    clientHeight: 0
  };
}
function getVirtualElementFromMousePos(mousePos, _ref) {
  var targetElement = _ref.targetElement,
    tooltipPosition = _ref.tooltipPosition;
  if (tooltipPosition === 'mouse') {
    return {
      getBoundingClientRect: function getBoundingClientRect() {
        return DOMRect.fromRect({
          x: mousePos.clientX,
          y: mousePos.clientY,
          width: 0,
          height: 0
        });
      }
    };
  }
  var targetRect = targetElement.getBoundingClientRect();
  if (tooltipPosition === 'mouse-x') {
    return {
      getBoundingClientRect: function getBoundingClientRect() {
        return DOMRect.fromRect({
          x: mousePos.clientX,
          y: targetRect.top,
          width: 0,
          height: targetRect.height
        });
      }
    };
  }
  if (tooltipPosition === 'mouse-y') {
    return {
      getBoundingClientRect: function getBoundingClientRect() {
        return DOMRect.fromRect({
          x: targetRect.left,
          y: mousePos.clientY,
          width: targetRect.width,
          height: 0
        });
      }
    };
  }
  throw new Error("Invalid tooltip position for virtual element: ".concat(tooltipPosition));
}
