// @atlaskit/adf-schema/dist/cjs/next-schema/generated/nodeTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unsupportedInline = exports.unsupportedBlock = exports.textWithNoMarks = exports.textFormatted = exports.textCodeInline = exports.text = exports.taskList = exports.taskItem = exports.tableWithNestedTable = exports.tableRowWithNestedTable = exports.tableRow = exports.tableHeaderWithNestedTable = exports.tableHeader = exports.tableCellWithNestedTable = exports.tableCell = exports.table = exports.statusStage0 = exports.status = exports.rule = exports.placeholder = exports.paragraphWithNoMarks = exports.paragraphWithIndentation = exports.paragraphWithAlignment = exports.paragraph = exports.panelWithoutNestedNonBodiedMacros = exports.panel = exports.orderedList = exports.nestedExpandWithoutNonBodiedMacros = exports.nestedExpandWithNoMarks = exports.nestedExpand = exports.multiBodiedExtensionStage0 = exports.mentionStage0 = exports.mention = exports.mediaSingleWidthType = exports.mediaSingleFull = exports.mediaSingleCaption = exports.mediaSingle = exports.mediaInline = exports.mediaGroup = exports.media = exports.listItemWithoutNonBodiedMacros = exports.listItemWithNestedDecisionStage0 = exports.listItem = exports.layoutSectionWithSingleColumnStage0 = exports.layoutSectionFull = exports.layoutSection = exports.layoutColumn = exports.inlineExtensionWithMarks = exports.inlineExtension = exports.inlineCardStage0 = exports.inlineCard = exports.image = exports.headingWithNoMarks = exports.headingWithIndentation = exports.headingWithAlignment = exports.heading = exports.hardBreak = exports.extensionWithMarks = exports.extensionFrameStage0 = exports.extension = exports.expandRootOnly = exports.expand = exports.emojiStage0 = exports.emoji = exports.embedCard = exports.doc = exports.decisionList = exports.decisionItem = exports.dateStage0 = exports.date = exports.confluenceUnsupportedInline = exports.confluenceUnsupportedBlock = exports.confluenceJiraIssue = exports.codeBlockRootOnly = exports.codeBlock = exports.caption = exports.bulletList = exports.bodiedExtensionWithMarks = exports.bodiedExtension = exports.blockquoteWithoutNonBodiedMacros = exports.blockquoteWithoutNestedCodeblockOrMedia = exports.blockquoteLegacy = exports.blockquote = exports.blockCard = void 0;
var _createPMSpecFactory = require("../../schema/createPMSpecFactory");
/**
 * This file was automatically generated by @atlaskit/adf-schema-generator
 * DO NOT MODIFY IT BY HAND. Instead, modify the source files in "packages/adf-schema/src/next-schema" ,
 * and run "yarn workspace @atlaskit/adf-schema build:schema:all" to regenerate this file.
 */

var blockCard = exports.blockCard = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    url: {
      default: null
    },
    datasource: {
      default: null
    },
    width: {
      default: null
    },
    layout: {
      default: null
    },
    data: {
      default: null
    }
  },
  selectable: true,
  draggable: true
});
var blockquote = exports.blockquote = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | orderedList | bulletList | unsupportedBlock | codeBlock | mediaSingle | mediaGroup | extension)+',
  marks: 'unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  group: 'block',
  selectable: true,
  defining: true
});
var blockquoteLegacy = exports.blockquoteLegacy = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | unsupportedBlock)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  selectable: true,
  defining: true
});
var blockquoteWithoutNestedCodeblockOrMedia = exports.blockquoteWithoutNestedCodeblockOrMedia = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | orderedList | bulletList | unsupportedBlock)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  selectable: true,
  defining: true
});
var blockquoteWithoutNonBodiedMacros = exports.blockquoteWithoutNonBodiedMacros = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | orderedList | bulletList | unsupportedBlock | codeBlock | mediaSingle | mediaGroup)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  selectable: true,
  defining: true
});
var bodiedExtension = exports.bodiedExtension = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaGroup | mediaSingle | decisionList | taskList | table | blockCard | embedCard | extension | unsupportedBlock)+',
  marks: 'dataConsumer fragment unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: null
    }
  },
  selectable: true,
  defining: true,
  isolating: true
});
var bodiedExtensionWithMarks = exports.bodiedExtensionWithMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  marks: 'dataConsumer fragment unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: null
    }
  },
  selectable: true,
  defining: true,
  isolating: true
});
var bulletList = exports.bulletList = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'listItem+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  selectable: false
});
var caption = exports.caption = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(hardBreak | mention | emoji | date | placeholder | inlineCard | status | text | unsupportedInline)*',
  marks: '_',
  selectable: false,
  isolating: true
});
var codeBlock = exports.codeBlock = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(text | unsupportedInline)*',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    language: {
      default: null
    },
    uniqueId: {
      default: null
    }
  },
  code: true,
  defining: true
});
var codeBlockRootOnly = exports.codeBlockRootOnly = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(text | unsupportedInline)*',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  attrs: {
    language: {
      default: null
    },
    uniqueId: {
      default: null
    }
  },
  code: true,
  defining: true
});
var confluenceJiraIssue = exports.confluenceJiraIssue = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  atom: true,
  attrs: {
    issueKey: {
      default: ''
    },
    macroId: {
      default: null
    },
    schemaVersion: {
      default: null
    },
    server: {
      default: null
    },
    serverId: {
      default: null
    }
  }
});
var confluenceUnsupportedBlock = exports.confluenceUnsupportedBlock = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    cxhtml: {
      default: null
    }
  }
});
var confluenceUnsupportedInline = exports.confluenceUnsupportedInline = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  atom: true,
  attrs: {
    cxhtml: {
      default: null
    }
  }
});
var date = exports.date = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    timestamp: {
      default: ''
    }
  },
  selectable: true
});
var dateStage0 = exports.dateStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    timestamp: {
      default: ''
    }
  },
  selectable: true
});
var decisionItem = exports.decisionItem = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'inline*',
  marks: '_',
  attrs: {
    localId: {
      default: ''
    },
    state: {
      default: 'DECIDED'
    }
  },
  defining: true
});
var decisionList = exports.decisionList = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(decisionItem | unsupportedBlock)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    localId: {
      default: ''
    }
  },
  selectable: false,
  defining: true
});
var doc = exports.doc = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(block | codeBlock | layoutSection | blockRootOnly | expand)+',
  marks: 'unsupportedMark unsupportedNodeAttribute alignment indentation dataConsumer fragment breakout'
});
var embedCard = exports.embedCard = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    url: {
      default: ''
    },
    layout: {
      default: 'center'
    },
    width: {
      default: 100
    },
    originalHeight: {
      default: null
    },
    originalWidth: {
      default: null
    }
  },
  selectable: true
});
var emoji = exports.emoji = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    shortName: {
      default: ''
    },
    id: {
      default: ''
    },
    text: {
      default: ''
    }
  },
  selectable: true
});
var emojiStage0 = exports.emojiStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    shortName: {
      default: ''
    },
    id: {
      default: ''
    },
    text: {
      default: ''
    }
  },
  selectable: true
});
var expand = exports.expand = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaGroup | mediaSingle | decisionList | taskList | table | blockCard | embedCard | extension | unsupportedBlock | nestedExpand)+',
  marks: 'unsupportedMark unsupportedNodeAttribute fragment dataConsumer',
  group: 'block',
  attrs: {
    title: {
      default: ''
    },
    __expanded: {
      default: true
    }
  },
  selectable: true,
  isolating: true
});
var expandRootOnly = exports.expandRootOnly = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaGroup | mediaSingle | decisionList | taskList | table | blockCard | embedCard | extension | unsupportedBlock | nestedExpand)+',
  marks: 'unsupportedMark unsupportedNodeAttribute fragment dataConsumer',
  attrs: {
    title: {
      default: ''
    },
    __expanded: {
      default: true
    }
  },
  selectable: true,
  isolating: true
});
var extension = exports.extension = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  atom: true,
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: null
    }
  },
  selectable: true
});
var extensionWithMarks = exports.extensionWithMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  atom: true,
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: null
    }
  },
  selectable: true
});
var extensionFrameStage0 = exports.extensionFrameStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaGroup | mediaSingle | decisionList | taskList | table | extension | bodiedExtension | unsupportedBlock | blockCard | embedCard)+',
  marks: 'dataConsumer fragment unsupportedMark unsupportedNodeAttribute',
  attrs: {},
  selectable: false,
  definingAsContext: false,
  definingForContent: true,
  isolating: true
});
var hardBreak = exports.hardBreak = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    text: {
      default: '\n'
    }
  },
  selectable: false,
  linebreakReplacement: true
});
var heading = exports.heading = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'inline*',
  marks: 'link em strong strike subsup underline textColor annotation backgroundColor typeAheadQuery confluenceInlineComment unsupportedNodeAttribute unsupportedMark code dataConsumer fragment border',
  group: 'block',
  attrs: {
    level: {
      default: 1
    },
    localId: {
      default: null
    }
  },
  selectable: false,
  defining: true
});
var headingWithAlignment = exports.headingWithAlignment = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    level: {
      default: 1
    },
    localId: {
      default: null
    }
  },
  selectable: false,
  defining: true
});
var headingWithIndentation = exports.headingWithIndentation = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    level: {
      default: 1
    },
    localId: {
      default: null
    }
  },
  selectable: false,
  defining: true
});
var headingWithNoMarks = exports.headingWithNoMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    level: {
      default: 1
    },
    localId: {
      default: null
    }
  },
  selectable: false,
  defining: true
});
var image = exports.image = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    src: {
      default: ''
    },
    alt: {
      default: ''
    },
    title: {
      default: null
    }
  },
  draggable: true
});
var inlineCard = exports.inlineCard = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    url: {
      default: null
    },
    data: {
      default: null
    }
  },
  selectable: true,
  draggable: true
});
var inlineCardStage0 = exports.inlineCardStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    url: {
      default: null
    },
    data: {
      default: null
    }
  },
  selectable: true,
  draggable: true
});
var inlineExtension = exports.inlineExtension = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    localId: {
      default: null
    }
  },
  selectable: true
});
var inlineExtensionWithMarks = exports.inlineExtensionWithMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    localId: {
      default: null
    }
  },
  selectable: true
});
var layoutColumn = exports.layoutColumn = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(block | unsupportedBlock)+',
  marks: 'unsupportedMark unsupportedNodeAttribute alignment indentation dataConsumer fragment',
  attrs: {
    width: {
      default: undefined
    }
  },
  selectable: false,
  isolating: true
});
var layoutSection = exports.layoutSection = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(layoutColumn | unsupportedBlock){1,3} unsupportedBlock*',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  isolating: true
});
var layoutSectionFull = exports.layoutSectionFull = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(layoutColumn | unsupportedBlock){2,3}',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  isolating: true
});
var layoutSectionWithSingleColumnStage0 = exports.layoutSectionWithSingleColumnStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(layoutColumn | unsupportedBlock){1,5} unsupportedBlock*',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  attrs: {
    columnRuleStyle: {
      default: null
    }
  },
  isolating: true
});
var listItem = exports.listItem = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | mediaSingle | codeBlock | unsupportedBlock | extension) (paragraph | bulletList | orderedList | taskList | mediaSingle | codeBlock | unsupportedBlock | extension)*',
  marks: 'dataConsumer fragment unsupportedMark unsupportedNodeAttribute',
  selectable: false,
  defining: true
});
var listItemWithNestedDecisionStage0 = exports.listItemWithNestedDecisionStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | mediaSingle | codeBlock | unsupportedBlock | decisionList | extension) (paragraph | bulletList | orderedList | taskList | mediaSingle | codeBlock | unsupportedBlock | decisionList | extension)*',
  marks: 'unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  selectable: false,
  defining: true
});
var listItemWithoutNonBodiedMacros = exports.listItemWithoutNonBodiedMacros = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | mediaSingle | codeBlock | unsupportedBlock) (paragraph | bulletList | orderedList | taskList | mediaSingle | codeBlock | unsupportedBlock)*',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  selectable: false,
  defining: true
});
var media = exports.media = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  attrs: {
    type: {
      default: 'file'
    },
    id: {
      default: ''
    },
    alt: {
      default: ''
    },
    collection: {
      default: ''
    },
    height: {
      default: null
    },
    occurrenceKey: {
      default: null
    },
    width: {
      default: null
    },
    __contextId: {
      default: null
    },
    __displayType: {
      default: null
    },
    __external: {
      default: false
    },
    __fileMimeType: {
      default: null
    },
    __fileName: {
      default: null
    },
    __fileSize: {
      default: null
    },
    __mediaTraceId: {
      default: null
    },
    url: {
      default: null
    }
  },
  selectable: true
});
var mediaGroup = exports.mediaGroup = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(media | unsupportedBlock)+',
  marks: 'link annotation border unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {},
  selectable: false
});
var mediaInline = exports.mediaInline = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    type: {
      default: 'file'
    },
    url: {
      default: null
    },
    id: {
      default: ''
    },
    alt: {
      default: ''
    },
    collection: {
      default: ''
    },
    occurrenceKey: {
      default: null
    },
    width: {
      default: null
    },
    height: {
      default: null
    },
    __fileName: {
      default: null
    },
    __fileSize: {
      default: null
    },
    __fileMimeType: {
      default: null
    },
    __displayType: {
      default: null
    },
    __contextId: {
      default: null
    },
    __mediaTraceId: {
      default: null
    },
    __external: {
      default: false
    }
  },
  selectable: true
});
var mediaSingle = exports.mediaSingle = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  atom: true,
  attrs: {
    layout: {
      default: 'center'
    },
    width: {
      default: null
    }
  },
  selectable: true,
  content: 'media|unsupportedBlock+|media unsupportedBlock+',
  marks: 'annotation border link unsupportedMark unsupportedNodeAttribute'
});
var mediaSingleCaption = exports.mediaSingleCaption = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'media|unsupportedBlock+|media (caption|unsupportedBlock) unsupportedBlock*',
  marks: 'link annotation border unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  atom: false,
  attrs: {
    layout: {
      default: 'center'
    },
    width: {
      default: null
    }
  },
  selectable: true
});
var mediaSingleFull = exports.mediaSingleFull = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'media|unsupportedBlock+|media (caption|unsupportedBlock) unsupportedBlock*',
  marks: 'link annotation border unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  atom: false,
  attrs: {
    width: {
      default: null
    },
    layout: {
      default: 'center'
    },
    widthType: {
      default: null
    }
  },
  selectable: true
});
var mediaSingleWidthType = exports.mediaSingleWidthType = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'media|unsupportedBlock+|media unsupportedBlock+',
  marks: 'link annotation border unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  atom: true,
  attrs: {
    width: {
      default: null
    },
    layout: {
      default: 'center'
    },
    widthType: {
      default: null
    }
  },
  selectable: true
});
var mention = exports.mention = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    id: {
      default: ''
    },
    localId: {
      default: null
    },
    text: {
      default: ''
    },
    accessLevel: {
      default: ''
    },
    userType: {
      default: null
    }
  },
  selectable: true
});
var mentionStage0 = exports.mentionStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    id: {
      default: ''
    },
    localId: {
      default: null
    },
    text: {
      default: ''
    },
    accessLevel: {
      default: ''
    },
    userType: {
      default: null
    }
  },
  selectable: true
});
var multiBodiedExtensionStage0 = exports.multiBodiedExtensionStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'extensionFrame+',
  marks: 'unsupportedNodeAttribute unsupportedMark',
  group: 'blockRootOnly',
  attrs: {
    extensionKey: {
      default: ''
    },
    extensionType: {
      default: ''
    },
    parameters: {
      default: null
    },
    text: {
      default: null
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: null
    }
  },
  selectable: true,
  definingAsContext: true
});
var nestedExpand = exports.nestedExpand = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | heading | mediaSingle | mediaGroup | codeBlock | bulletList | orderedList | taskList | decisionList | rule | panel | blockquote | unsupportedBlock | extension)+',
  marks: 'unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  attrs: {
    title: {
      default: ''
    },
    __expanded: {
      default: true
    }
  },
  selectable: true,
  isolating: true
});
var nestedExpandWithNoMarks = exports.nestedExpandWithNoMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  attrs: {
    title: {
      default: ''
    },
    __expanded: {
      default: true
    }
  },
  selectable: true,
  isolating: true
});
var nestedExpandWithoutNonBodiedMacros = exports.nestedExpandWithoutNonBodiedMacros = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | heading | mediaSingle | mediaGroup | codeBlock | bulletList | orderedList | taskList | decisionList | rule | panel | blockquote | unsupportedBlock)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  attrs: {
    title: {
      default: ''
    },
    __expanded: {
      default: true
    }
  },
  selectable: true,
  isolating: true
});
var orderedList = exports.orderedList = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'listItem+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    order: {
      default: 1
    }
  },
  selectable: false
});
var panel = exports.panel = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | heading | bulletList | orderedList | blockCard | mediaGroup | mediaSingle | codeBlock | taskList | rule | decisionList | unsupportedBlock | extension)+',
  marks: 'unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  group: 'block',
  attrs: {
    panelType: {
      default: 'info'
    },
    panelIcon: {
      default: null
    },
    panelIconId: {
      default: null
    },
    panelIconText: {
      default: null
    },
    panelColor: {
      default: null
    }
  },
  selectable: true
});
var panelWithoutNestedNonBodiedMacros = exports.panelWithoutNestedNonBodiedMacros = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | heading | bulletList | orderedList | blockCard | mediaGroup | mediaSingle | codeBlock | taskList | rule | decisionList | unsupportedBlock)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    panelType: {
      default: 'info'
    },
    panelIcon: {
      default: null
    },
    panelIconId: {
      default: null
    },
    panelIconText: {
      default: null
    },
    panelColor: {
      default: null
    }
  },
  selectable: true
});
var paragraph = exports.paragraph = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'inline*',
  marks: 'link em strong strike subsup underline textColor annotation backgroundColor typeAheadQuery confluenceInlineComment unsupportedNodeAttribute unsupportedMark code dataConsumer fragment border',
  group: 'block',
  attrs: {
    localId: {
      default: null
    }
  },
  selectable: false
});
var paragraphWithAlignment = exports.paragraphWithAlignment = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    localId: {
      default: null
    }
  },
  selectable: false
});
var paragraphWithIndentation = exports.paragraphWithIndentation = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    localId: {
      default: null
    }
  },
  selectable: false
});
var paragraphWithNoMarks = exports.paragraphWithNoMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  attrs: {
    localId: {
      default: null
    }
  },
  selectable: false
});
var placeholder = exports.placeholder = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  marks: '',
  group: 'inline',
  inline: true,
  attrs: {
    text: {
      default: ''
    }
  },
  selectable: false
});
var rule = exports.rule = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block'
});
var status = exports.status = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    text: {
      default: ''
    },
    color: {
      default: ''
    },
    localId: {
      default: ''
    },
    style: {
      default: ''
    }
  },
  selectable: true
});
var statusStage0 = exports.statusStage0 = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    text: {
      default: ''
    },
    color: {
      default: ''
    },
    localId: {
      default: ''
    },
    style: {
      default: ''
    }
  },
  selectable: true
});
var table = exports.table = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'tableRow+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    displayMode: {
      default: null
    },
    isNumberColumnEnabled: {
      default: false
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: ''
    },
    width: {
      default: null
    },
    __autoSize: {
      default: false
    }
  },
  selectable: true,
  isolating: true,
  tableRole: 'table'
});
var tableWithNestedTable = exports.tableWithNestedTable = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'tableRow+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    displayMode: {
      default: null
    },
    isNumberColumnEnabled: {
      default: false
    },
    layout: {
      default: 'default'
    },
    localId: {
      default: ''
    },
    width: {
      default: null
    },
    __autoSize: {
      default: false
    }
  },
  selectable: true,
  isolating: true,
  tableRole: 'table'
});
var tableCell = exports.tableCell = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaSingle | mediaGroup | decisionList | taskList | blockCard | embedCard | extension | nestedExpand | unsupportedBlock)+',
  marks: 'alignment unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  attrs: {
    colspan: {
      default: 1
    },
    rowspan: {
      default: 1
    },
    colwidth: {
      default: null
    },
    background: {
      default: null
    }
  },
  selectable: false,
  isolating: true,
  tableRole: 'cell'
});
var tableCellWithNestedTable = exports.tableCellWithNestedTable = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaSingle | mediaGroup | decisionList | taskList | blockCard | embedCard | extension | nestedExpand | unsupportedBlock | table)+',
  marks: 'alignment unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  attrs: {
    colspan: {
      default: 1
    },
    rowspan: {
      default: 1
    },
    colwidth: {
      default: null
    },
    background: {
      default: null
    }
  },
  selectable: false,
  isolating: true,
  tableRole: 'cell'
});
var tableHeader = exports.tableHeader = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaSingle | mediaGroup | decisionList | taskList | blockCard | embedCard | extension | nestedExpand)+',
  marks: 'alignment unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  attrs: {
    colspan: {
      default: 1
    },
    rowspan: {
      default: 1
    },
    colwidth: {
      default: null
    },
    background: {
      default: null
    }
  },
  selectable: false,
  isolating: true,
  tableRole: 'header_cell'
});
var tableHeaderWithNestedTable = exports.tableHeaderWithNestedTable = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaSingle | mediaGroup | decisionList | taskList | blockCard | embedCard | extension | nestedExpand | table)+',
  marks: 'alignment unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  attrs: {
    colspan: {
      default: 1
    },
    rowspan: {
      default: 1
    },
    colwidth: {
      default: null
    },
    background: {
      default: null
    }
  },
  selectable: false,
  isolating: true,
  tableRole: 'header_cell'
});
var tableRow = exports.tableRow = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(tableCell | tableHeader)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  selectable: false,
  tableRole: 'row'
});
var tableRowWithNestedTable = exports.tableRowWithNestedTable = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(tableCell | tableHeader)+',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  selectable: false,
  tableRole: 'row'
});
var taskItem = exports.taskItem = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: 'inline*',
  marks: '_',
  attrs: {
    localId: {
      default: ''
    },
    state: {
      default: 'TODO'
    }
  },
  selectable: false,
  defining: true
});
var taskList = exports.taskList = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  content: '(taskItem | unsupportedBlock)+ (taskItem | taskList | unsupportedBlock)*',
  marks: 'unsupportedMark unsupportedNodeAttribute',
  group: 'block',
  attrs: {
    localId: {
      default: ''
    }
  },
  selectable: false,
  defining: true
});
var text = exports.text = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline'
});
var textCodeInline = exports.textCodeInline = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline'
});
var textFormatted = exports.textFormatted = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline'
});
var textWithNoMarks = exports.textWithNoMarks = (0, _createPMSpecFactory.createPMNodeSpecFactory)({});
var unsupportedBlock = exports.unsupportedBlock = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'block',
  atom: true,
  attrs: {
    originalValue: {
      default: {}
    }
  },
  selectable: true
});
var unsupportedInline = exports.unsupportedInline = (0, _createPMSpecFactory.createPMNodeSpecFactory)({
  group: 'inline',
  inline: true,
  attrs: {
    originalValue: {
      default: {}
    }
  },
  selectable: true
});
