// @atlaskit/adf-schema/dist/cjs/schema/marks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AnnotationMarkStates", {
  enumerable: true,
  get: function get() {
    return _annotation.AnnotationMarkStates;
  }
});
Object.defineProperty(exports, "AnnotationTypes", {
  enumerable: true,
  get: function get() {
    return _annotation.AnnotationTypes;
  }
});
Object.defineProperty(exports, "alignment", {
  enumerable: true,
  get: function get() {
    return _alignment.alignment;
  }
});
Object.defineProperty(exports, "alignmentPositionMap", {
  enumerable: true,
  get: function get() {
    return _alignment.alignmentPositionMap;
  }
});
Object.defineProperty(exports, "annotation", {
  enumerable: true,
  get: function get() {
    return _annotation.annotation;
  }
});
Object.defineProperty(exports, "backgroundColor", {
  enumerable: true,
  get: function get() {
    return _backgroundColor.backgroundColor;
  }
});
Object.defineProperty(exports, "backgroundColorPalette", {
  enumerable: true,
  get: function get() {
    return _backgroundColor.backgroundColorPalette;
  }
});
Object.defineProperty(exports, "border", {
  enumerable: true,
  get: function get() {
    return _border.border;
  }
});
Object.defineProperty(exports, "borderColorPalette", {
  enumerable: true,
  get: function get() {
    return _border.borderColorPalette;
  }
});
Object.defineProperty(exports, "breakout", {
  enumerable: true,
  get: function get() {
    return _breakout.breakout;
  }
});
Object.defineProperty(exports, "buildAnnotationMarkDataAttributes", {
  enumerable: true,
  get: function get() {
    return _annotation.buildDataAttributes;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function get() {
    return _code.code;
  }
});
Object.defineProperty(exports, "colorPalette", {
  enumerable: true,
  get: function get() {
    return _textColor.colorPalette;
  }
});
Object.defineProperty(exports, "colorPaletteExtended", {
  enumerable: true,
  get: function get() {
    return _textColor.colorPaletteExtended;
  }
});
Object.defineProperty(exports, "confluenceInlineComment", {
  enumerable: true,
  get: function get() {
    return _confluenceInlineComment.confluenceInlineComment;
  }
});
Object.defineProperty(exports, "dataConsumer", {
  enumerable: true,
  get: function get() {
    return _dataConsumer.dataConsumer;
  }
});
Object.defineProperty(exports, "dataConsumerToJSON", {
  enumerable: true,
  get: function get() {
    return _dataConsumer.toJSON;
  }
});
Object.defineProperty(exports, "em", {
  enumerable: true,
  get: function get() {
    return _em.em;
  }
});
Object.defineProperty(exports, "fragment", {
  enumerable: true,
  get: function get() {
    return _fragment.fragment;
  }
});
Object.defineProperty(exports, "fragmentToJSON", {
  enumerable: true,
  get: function get() {
    return _fragment.toJSON;
  }
});
Object.defineProperty(exports, "indentation", {
  enumerable: true,
  get: function get() {
    return _indentation.indentation;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function get() {
    return _link.link;
  }
});
Object.defineProperty(exports, "linkToJSON", {
  enumerable: true,
  get: function get() {
    return _link.toJSON;
  }
});
Object.defineProperty(exports, "strike", {
  enumerable: true,
  get: function get() {
    return _strike.strike;
  }
});
Object.defineProperty(exports, "strong", {
  enumerable: true,
  get: function get() {
    return _strong.strong;
  }
});
Object.defineProperty(exports, "subsup", {
  enumerable: true,
  get: function get() {
    return _subsup.subsup;
  }
});
Object.defineProperty(exports, "textColor", {
  enumerable: true,
  get: function get() {
    return _textColor.textColor;
  }
});
Object.defineProperty(exports, "typeAheadQuery", {
  enumerable: true,
  get: function get() {
    return _typeAheadQuery.typeAheadQuery;
  }
});
Object.defineProperty(exports, "underline", {
  enumerable: true,
  get: function get() {
    return _underline.underline;
  }
});
Object.defineProperty(exports, "unsupportedMark", {
  enumerable: true,
  get: function get() {
    return _unsupportedMark.unsupportedMark;
  }
});
Object.defineProperty(exports, "unsupportedNodeAttribute", {
  enumerable: true,
  get: function get() {
    return _unsupportedNodeAttributes.unsupportedNodeAttribute;
  }
});
var _em = require("./em");
var _code = require("./code");
var _strike = require("./strike");
var _strong = require("./strong");
var _underline = require("./underline");
var _link = require("./link");
var _typeAheadQuery = require("./type-ahead-query");
var _subsup = require("./subsup");
var _textColor = require("./text-color");
var _confluenceInlineComment = require("./confluence-inline-comment");
var _breakout = require("./breakout");
var _alignment = require("./alignment");
var _indentation = require("./indentation");
var _annotation = require("./annotation");
var _unsupportedMark = require("./unsupported-mark");
var _unsupportedNodeAttributes = require("./unsupported-node-attributes");
var _dataConsumer = require("./data-consumer");
var _fragment = require("./fragment");
var _border = require("./border");
var _backgroundColor = require("./background-color");
