// @atlaskit/navigation-system/dist/cjs/components/badge-container.js
/* badge-container.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BadgeContainer = void 0;
require("./badge-container.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var styles = {
  root: "_kqswh2mm _1e0c1txw",
  badgeContainer: "_kqswstnw _o5721q9c _152tr5cr _1e02utpp _lcxvglyw"
};
/**
 * __Badge container__
 *
 * Adds a badge to the top right corner of the child element.
 * Example of usage is showing the number of notifications in the top navigation bar.
 */
var BadgeContainer = exports.BadgeContainer = function BadgeContainer(_ref) {
  var children = _ref.children,
    badgeId = _ref.id,
    Badge = _ref.badge,
    _ref$isListItem = _ref.isListItem,
    isListItem = _ref$isListItem === void 0 ? true : _ref$isListItem;
  return /*#__PURE__*/React.createElement("div", {
    role: isListItem ? 'listitem' : undefined,
    className: (0, _runtime.ax)([styles.root])
  }, children, /*#__PURE__*/React.createElement("div", {
    id: badgeId,
    "aria-hidden": true,
    className: (0, _runtime.ax)([styles.badgeContainer])
  }, /*#__PURE__*/React.createElement(Badge, null)));
};
