// @forge/react/out/types/forge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderingLocations = exports.CompassContextTypes = exports.isDashboardGadgetExtensionContext = exports.isContentActionExtensionContext = exports.isContextMenuExtensionContext = exports.isCustomFieldExtensionContext = exports.isCustomFieldContextConfigExtensionContext = exports.isIssuePanelExtensionContext = exports.isJiraContext = exports.EcosystemLicenseMode = exports.isForgeElement = void 0;
const isForgeElement = (auxNode) => {
    return auxNode !== null && typeof auxNode !== 'boolean' && auxNode !== undefined;
};
exports.isForgeElement = isForgeElement;
var EcosystemLicenseMode;
(function (EcosystemLicenseMode) {
    EcosystemLicenseMode["USER_ACCESS"] = "USER_ACCESS";
})(EcosystemLicenseMode = exports.EcosystemLicenseMode || (exports.EcosystemLicenseMode = {}));
const isJiraContext = (context) => context.type === 'jira';
exports.isJiraContext = isJiraContext;
const isIssuePanelExtensionContext = (extensionContext) => extensionContext.type === 'issuePanel';
exports.isIssuePanelExtensionContext = isIssuePanelExtensionContext;
const isCustomFieldContextConfigExtensionContext = (extensionContext) => extensionContext.type === 'contextConfig';
exports.isCustomFieldContextConfigExtensionContext = isCustomFieldContextConfigExtensionContext;
const isCustomFieldExtensionContext = (extensionContext) => extensionContext.type === 'customField' || extensionContext.type === 'customFieldType';
exports.isCustomFieldExtensionContext = isCustomFieldExtensionContext;
const isContextMenuExtensionContext = (extensionContext) => extensionContext.type === 'contextMenu';
exports.isContextMenuExtensionContext = isContextMenuExtensionContext;
const isContentActionExtensionContext = (extensionContext) => extensionContext.type === 'contentAction';
exports.isContentActionExtensionContext = isContentActionExtensionContext;
const isDashboardGadgetExtensionContext = (extensionContext) => extensionContext.type === 'dashboardGadget';
exports.isDashboardGadgetExtensionContext = isDashboardGadgetExtensionContext;
var CompassContextTypes;
(function (CompassContextTypes) {
    CompassContextTypes["AdminPage"] = "compass:adminPage";
    CompassContextTypes["ComponentPage"] = "compass:componentPage";
    CompassContextTypes["TeamPage"] = "compass:teamPage";
    CompassContextTypes["ComponentImporter"] = "compass:componentImporter";
})(CompassContextTypes = exports.CompassContextTypes || (exports.CompassContextTypes = {}));
var RenderingLocations;
(function (RenderingLocations) {
    RenderingLocations["AppConfigurationPage"] = "appConfigurationPage";
    RenderingLocations["OnboardingFlow"] = "onboardingFlow";
})(RenderingLocations = exports.RenderingLocations || (exports.RenderingLocations = {}));
