// @atlaskit/adf-schema/dist/esm/next-schema/nodes/task.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { $onePlus, $or, $zeroPlus, adfNode } from '@atlaskit/adf-schema-generator';
import { inlineContentGroup } from '../groups/inlineContentGroup';
import { inlineGroup } from '../groups/inlineGroup';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
import { unsupportedBlock } from './unsupportedBlock';
import { paragraph } from './paragraph';
import { extension } from './extension';
export var taskItem = adfNode('taskItem');
export var taskList = adfNode('taskList');
export var blockTaskItem = adfNode('blockTaskItem');
var commonTaskItemProps = {
  defining: true,
  selectable: false,
  marks: [unsupportedMark, unsupportedNodeAttribute],
  allowAnyChildMark: true,
  attrs: {
    localId: {
      type: 'string',
      default: ''
    },
    state: {
      type: 'enum',
      values: ['TODO', 'DONE'],
      default: 'TODO'
    }
  }
};
taskItem.define(_objectSpread(_objectSpread({}, commonTaskItemProps), {}, {
  content: [$zeroPlus($or(inlineGroup, inlineContentGroup))]
}));
blockTaskItem.define(_objectSpread(_objectSpread({}, commonTaskItemProps), {}, {
  marks: [unsupportedMark, unsupportedNodeAttribute],
  allowAnyChildMark: false,
  // Override the commonTaskItemProps setting
  contentMinItems: 1,
  content: [$or(paragraph.use('with_no_marks'), extension.use('with_marks')), $zeroPlus($or(paragraph.use('with_no_marks'), extension.use('with_marks')))]
}));
taskList.define({
  defining: true,
  selectable: false,
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    localId: {
      type: 'string',
      default: ''
    }
  },
  contentMinItems: 1,
  content: [$onePlus($or(taskItem, unsupportedBlock, blockTaskItem)), $zeroPlus($or(taskItem, taskList, unsupportedBlock, blockTaskItem))]
});
taskList.variant('flexible_first_child', {
  contentMinItems: 1,
  content: [$onePlus($or(taskItem, taskList, unsupportedBlock, blockTaskItem))],
  stage0: true
});
