// @atlaskit/datetime-picker/dist/es2019/internal/single-value.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { components } from '@atlaskit/select';

/**
 * This creates a functional component that `react-select` will use to make the
 * SingleValue part of the different pickers.
 */
export const makeSingleValue =
// TODO: Make ID required
({
  id,
  lang
}) => ({
  children,
  className,
  clearValue,
  cx,
  data,
  getStyles,
  getValue,
  hasValue,
  isDisabled,
  isMulti,
  isRtl,
  options,
  selectOption,
  selectProps,
  setValue,
  ...rest
}) => {
  return /*#__PURE__*/React.createElement(components.SingleValue, _extends({}, rest, {
    // eslint-disable-next-line @atlaskit/design-system/no-unsafe-style-overrides, @atlaskit/ui-styling-standard/no-classname-prop
    className: className,
    clearValue: clearValue,
    cx: cx,
    data: data,
    getStyles: getStyles,
    getValue: getValue,
    hasValue: hasValue,
    innerProps: {
      id,
      lang
    },
    isDisabled: isDisabled,
    isMulti: isMulti,
    isRtl: isRtl,
    options: options,
    selectOption: selectOption,
    selectProps: selectProps,
    setValue: setValue
  }), children);
};
