// @atlaskit/lozenge/dist/cjs/new/lozenge.js
/* lozenge.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _lozengeBase = _interopRequireDefault(require("./lozenge-base"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
/**
 * __Lozenge__
 *
 * A lozenge is a visual indicator used to highlight an item's status for quick recognition.
 *
 * - [Examples](https://atlassian.design/components/lozenge/examples)
 * - [Code](https://atlassian.design/components/lozenge/code)
 * - [Usage](https://atlassian.design/components/lozenge/usage)
 */
var Lozenge = function Lozenge(_ref) {
  var _ref$appearance = _ref.appearance,
    appearance = _ref$appearance === void 0 ? 'neutral' : _ref$appearance,
    _ref$spacing = _ref.spacing,
    spacing = _ref$spacing === void 0 ? 'default' : _ref$spacing,
    _ref$maxWidth = _ref.maxWidth,
    maxWidth = _ref$maxWidth === void 0 ? 200 : _ref$maxWidth,
    style = _ref.style,
    testId = _ref.testId,
    children = _ref.children,
    iconBefore = _ref.iconBefore,
    trailingMetric = _ref.trailingMetric,
    trailingMetricAppearance = _ref.trailingMetricAppearance;
  var baseProps = {
    appearance: appearance,
    spacing: spacing,
    iconBefore: iconBefore,
    trailingMetric: trailingMetric,
    trailingMetricAppearance: trailingMetricAppearance,
    maxWidth: maxWidth,
    style: style,
    testId: testId,
    children: children
  };
  return /*#__PURE__*/React.createElement(_lozengeBase.default, baseProps, children);
};
Lozenge.displayName = 'Lozenge';
var _default = exports.default = Lozenge;
