// @atlaskit/editor-prosemirror/dist/cjs/history.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "closeHistory", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.closeHistory;
  }
});
Object.defineProperty(exports, "history", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.history;
  }
});
Object.defineProperty(exports, "isHistoryTransaction", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.isHistoryTransaction;
  }
});
Object.defineProperty(exports, "redo", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.redo;
  }
});
Object.defineProperty(exports, "redoDepth", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.redoDepth;
  }
});
Object.defineProperty(exports, "redoNoScroll", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.redoNoScroll;
  }
});
Object.defineProperty(exports, "undo", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.undo;
  }
});
Object.defineProperty(exports, "undoDepth", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.undoDepth;
  }
});
Object.defineProperty(exports, "undoNoScroll", {
  enumerable: true,
  get: function get() {
    return _prosemirrorHistory.undoNoScroll;
  }
});
var _prosemirrorHistory = require("prosemirror-history");
