// @atlaskit/editor-prosemirror/dist/cjs/model.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ContentMatch", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.ContentMatch;
  }
});
Object.defineProperty(exports, "DOMParser", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.DOMParser;
  }
});
Object.defineProperty(exports, "DOMSerializer", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.DOMSerializer;
  }
});
Object.defineProperty(exports, "Fragment", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.Fragment;
  }
});
Object.defineProperty(exports, "Mark", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.Mark;
  }
});
Object.defineProperty(exports, "MarkType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.MarkType;
  }
});
Object.defineProperty(exports, "Node", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.Node;
  }
});
Object.defineProperty(exports, "NodeRange", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.NodeRange;
  }
});
Object.defineProperty(exports, "NodeType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.NodeType;
  }
});
Object.defineProperty(exports, "ReplaceError", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.ReplaceError;
  }
});
Object.defineProperty(exports, "ResolvedPos", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.ResolvedPos;
  }
});
Object.defineProperty(exports, "Schema", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.Schema;
  }
});
Object.defineProperty(exports, "Slice", {
  enumerable: true,
  get: function get() {
    return _prosemirrorModel.Slice;
  }
});
var _prosemirrorModel = require("prosemirror-model");
