// @atlaskit/editor-prosemirror/dist/cjs/state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AllSelection", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.AllSelection;
  }
});
Object.defineProperty(exports, "EditorState", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.EditorState;
  }
});
Object.defineProperty(exports, "NodeSelection", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.NodeSelection;
  }
});
Object.defineProperty(exports, "Plugin", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.Plugin;
  }
});
Object.defineProperty(exports, "PluginKey", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.PluginKey;
  }
});
Object.defineProperty(exports, "Selection", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.Selection;
  }
});
Object.defineProperty(exports, "SelectionRange", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.SelectionRange;
  }
});
Object.defineProperty(exports, "TextSelection", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.TextSelection;
  }
});
Object.defineProperty(exports, "Transaction", {
  enumerable: true,
  get: function get() {
    return _prosemirrorState.Transaction;
  }
});
var _prosemirrorState = require("prosemirror-state");
