// @atlaskit/editor-prosemirror/dist/cjs/transform.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AddMarkStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.AddMarkStep;
  }
});
Object.defineProperty(exports, "AddNodeMarkStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.AddNodeMarkStep;
  }
});
Object.defineProperty(exports, "AttrStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.AttrStep;
  }
});
Object.defineProperty(exports, "DocAttrStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.DocAttrStep;
  }
});
Object.defineProperty(exports, "MapResult", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.MapResult;
  }
});
Object.defineProperty(exports, "Mapping", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.Mapping;
  }
});
Object.defineProperty(exports, "RemoveMarkStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.RemoveMarkStep;
  }
});
Object.defineProperty(exports, "RemoveNodeMarkStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.RemoveNodeMarkStep;
  }
});
Object.defineProperty(exports, "ReplaceAroundStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.ReplaceAroundStep;
  }
});
Object.defineProperty(exports, "ReplaceStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.ReplaceStep;
  }
});
Object.defineProperty(exports, "Step", {
  enumerable: true,
  get: function get() {
    return _transformOverride.Step;
  }
});
Object.defineProperty(exports, "StepMap", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.StepMap;
  }
});
Object.defineProperty(exports, "StepResult", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.StepResult;
  }
});
Object.defineProperty(exports, "Transform", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.Transform;
  }
});
Object.defineProperty(exports, "canJoin", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.canJoin;
  }
});
Object.defineProperty(exports, "canSplit", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.canSplit;
  }
});
Object.defineProperty(exports, "dropPoint", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.dropPoint;
  }
});
Object.defineProperty(exports, "findWrapping", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.findWrapping;
  }
});
Object.defineProperty(exports, "insertPoint", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.insertPoint;
  }
});
Object.defineProperty(exports, "joinPoint", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.joinPoint;
  }
});
Object.defineProperty(exports, "liftTarget", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.liftTarget;
  }
});
Object.defineProperty(exports, "replaceStep", {
  enumerable: true,
  get: function get() {
    return _prosemirrorTransform.replaceStep;
  }
});
var _transformOverride = require("./transform-override");
var _prosemirrorTransform = require("prosemirror-transform");
