// @atlaskit/editor-prosemirror/dist/cjs/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "canInsert", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.canInsert;
  }
});
Object.defineProperty(exports, "contains", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.contains;
  }
});
Object.defineProperty(exports, "findBlockNodes", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findBlockNodes;
  }
});
Object.defineProperty(exports, "findChildren", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findChildren;
  }
});
Object.defineProperty(exports, "findChildrenByAttr", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findChildrenByAttr;
  }
});
Object.defineProperty(exports, "findChildrenByMark", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findChildrenByMark;
  }
});
Object.defineProperty(exports, "findChildrenByType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findChildrenByType;
  }
});
Object.defineProperty(exports, "findDomRefAtPos", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findDomRefAtPos;
  }
});
Object.defineProperty(exports, "findInlineNodes", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findInlineNodes;
  }
});
Object.defineProperty(exports, "findParentDomRef", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findParentDomRef;
  }
});
Object.defineProperty(exports, "findParentDomRefOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findParentDomRefOfType;
  }
});
Object.defineProperty(exports, "findParentNode", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findParentNode;
  }
});
Object.defineProperty(exports, "findParentNodeClosestToPos", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findParentNodeClosestToPos;
  }
});
Object.defineProperty(exports, "findParentNodeOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findParentNodeOfType;
  }
});
Object.defineProperty(exports, "findParentNodeOfTypeClosestToPos", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findParentNodeOfTypeClosestToPos;
  }
});
Object.defineProperty(exports, "findPositionOfNodeBefore", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findPositionOfNodeBefore;
  }
});
Object.defineProperty(exports, "findSelectedNodeOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findSelectedNodeOfType;
  }
});
Object.defineProperty(exports, "findTextNodes", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.findTextNodes;
  }
});
Object.defineProperty(exports, "flatten", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.flatten;
  }
});
Object.defineProperty(exports, "hasParentNode", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.hasParentNode;
  }
});
Object.defineProperty(exports, "hasParentNodeOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.hasParentNodeOfType;
  }
});
Object.defineProperty(exports, "isNodeSelection", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.isNodeSelection;
  }
});
Object.defineProperty(exports, "removeNodeBefore", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.removeNodeBefore;
  }
});
Object.defineProperty(exports, "removeParentNodeOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.removeParentNodeOfType;
  }
});
Object.defineProperty(exports, "removeSelectedNode", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.removeSelectedNode;
  }
});
Object.defineProperty(exports, "replaceParentNodeOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.replaceParentNodeOfType;
  }
});
Object.defineProperty(exports, "replaceSelectedNode", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.replaceSelectedNode;
  }
});
Object.defineProperty(exports, "safeInsert", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.safeInsert;
  }
});
Object.defineProperty(exports, "selectParentNodeOfType", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.selectParentNodeOfType;
  }
});
Object.defineProperty(exports, "setParentNodeMarkup", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.setParentNodeMarkup;
  }
});
Object.defineProperty(exports, "setTextSelection", {
  enumerable: true,
  get: function get() {
    return _prosemirrorUtils.setTextSelection;
  }
});
var _prosemirrorUtils = require("prosemirror-utils");
