// @atlaskit/react-ufo/dist/es2019/create-payload/critical-metrics-payload/root-metrics.js
import { getLighthouseMetrics } from '../../additional-payload';
import { getConfig } from '../../config';
import { getPageVisibilityState } from '../../hidden-timing';
import { sanitizeUfoName } from '../common/utils';
import getBrowserMetadata from '../utils/get-browser-metadata';
import { getFMP } from '../utils/get-fmp';
import getInteractionStatus from '../utils/get-interaction-status';
import getNavigationMetrics from '../utils/get-navigation-metrics';
import getPageVisibilityUpToTTAI from '../utils/get-page-visibility-up-to-ttai';
import getPaintMetrics from '../utils/get-paint-metrics';
import { LATEST_REACT_UFO_PAYLOAD_VERSION } from '../utils/get-react-ufo-payload-version';
import getSSRSuccess from '../utils/get-ssr-success';
import getTTAI from '../utils/get-ttai';
import { getTTI } from '../utils/get-tti';
import getVCMetrics from '../utils/get-vc-metrics';

// Re-export types for convenience

// Local utility functions
function getPageVisibilityUpToTTI(interaction) {
  var _interaction$apdex$0$, _interaction$apdex, _interaction$apdex$;
  const {
    start
  } = interaction;
  const bm3EndTimeOrInteractionEndTime = (_interaction$apdex$0$ = (_interaction$apdex = interaction.apdex) === null || _interaction$apdex === void 0 ? void 0 : (_interaction$apdex$ = _interaction$apdex[0]) === null || _interaction$apdex$ === void 0 ? void 0 : _interaction$apdex$.stopTime) !== null && _interaction$apdex$0$ !== void 0 ? _interaction$apdex$0$ : interaction.end;
  return getPageVisibilityState(start, bm3EndTimeOrInteractionEndTime);
}

// TODO Write tests for this function
export async function createRootCriticalMetricsPayload(interactionId, interaction, vcMetrics) {
  var _finalVCMetrics$ufoV, _interaction$cohortin, _window$location;
  const config = getConfig();
  if (!config) {
    throw Error('UFO Configuration not provided');
  }
  const {
    end,
    start,
    ufoName,
    rate,
    type,
    abortReason,
    routeName,
    previousInteractionName,
    isPreviousInteractionAborted,
    abortedByInteractionName,
    holdInfo,
    responsiveness
  } = interaction;
  const pageVisibilityAtTTI = getPageVisibilityUpToTTI(interaction);
  const pageVisibilityAtTTAI = getPageVisibilityUpToTTAI(interaction);
  const interactionStatus = getInteractionStatus(interaction);
  if (interactionStatus.originalInteractionStatus !== 'SUCCEEDED') {
    // To reduce payload sent from the client, we don't send critical perf metrics for non-success interactions
    return null;
  }
  const newUFOName = sanitizeUfoName(ufoName);

  // Get performance metrics
  const ttai = getTTAI(interaction);
  const paintMetrics = await getPaintMetrics(type, end);
  const navigationMetrics = getNavigationMetrics(type);
  const ssrSuccess = getSSRSuccess(type);

  // Calculate BM3 metrics (TTI and FMP) directly
  const tti = getTTI(interaction, newUFOName);
  const fmp = getFMP(interaction, newUFOName);

  // Get browser metadata (using compact nested format)
  const browserMetadata = getBrowserMetadata();
  const lighthouseMetrics = getLighthouseMetrics({
    start,
    stop: end
  });

  // Use provided vcMetrics or calculate if not provided
  const finalVCMetrics = vcMetrics || (await getVCMetrics(interaction));
  const ttvc = (_finalVCMetrics$ufoV = finalVCMetrics['ufo:vc:rev']) === null || _finalVCMetrics$ufoV === void 0 ? void 0 : _finalVCMetrics$ufoV.map(revision => {
    if (revision['metric:vc90'] === null || revision.clean !== true) {
      return null;
    }
    return {
      revision: revision.revision,
      vc90: revision['metric:vc90']
    };
  }).filter(revision => revision != null);

  // find earliest hold
  const earliestHold = holdInfo === null || holdInfo === void 0 ? void 0 : holdInfo.reduce((a, b) => {
    if (a && a.start < b.start) {
      return a;
    }
    return b;
  }, null);

  // Process cohorting custom data
  const cohortingCustomData = (_interaction$cohortin = interaction.cohortingCustomData) !== null && _interaction$cohortin !== void 0 && _interaction$cohortin.size ? Object.fromEntries(interaction.cohortingCustomData) : undefined;
  const properties = {
    // Basic metadata
    'event:hostname': ((_window$location = window.location) === null || _window$location === void 0 ? void 0 : _window$location.hostname) || 'unknown',
    'event:product': config.product,
    'event:schema': '1.0.0',
    'event:region': config.region || 'unknown',
    'event:source': {
      name: 'react-ufo/web',
      version: LATEST_REACT_UFO_PAYLOAD_VERSION
    },
    'experience:key': 'custom.ufo.critical-metrics',
    'experience:name': newUFOName,
    // Browser metadata (compact nested format)
    browser: browserMetadata.browser,
    device: browserMetadata.device,
    network: browserMetadata.network,
    time: browserMetadata.time,
    metrics: {
      fp: paintMetrics.fp,
      fcp: paintMetrics.fcp,
      lcp: paintMetrics.lcp,
      ttai,
      tti,
      fmp,
      tbt: lighthouseMetrics['metric:tbt'],
      tbtObserved: lighthouseMetrics['metric:tbt:observed'],
      cls: lighthouseMetrics['metric:cls'],
      ttvc: ttvc !== null && ttvc !== void 0 ? ttvc : undefined,
      earliestHoldStart: earliestHold !== null && earliestHold !== void 0 && earliestHold.start ? Math.round(earliestHold.start - start) : undefined,
      // for interaction response
      inputDelay: responsiveness !== null && responsiveness !== void 0 && responsiveness.inputDelay ? Math.round(responsiveness.inputDelay) : undefined,
      inp: responsiveness !== null && responsiveness !== void 0 && responsiveness.experimentalInputToNextPaint ? Math.round(responsiveness.experimentalInputToNextPaint) : undefined,
      ...(navigationMetrics && {
        navigation: navigationMetrics
      })
    },
    ...(ssrSuccess !== undefined && {
      ssrSuccess
    }),
    interactionId,
    type,
    rate,
    routeName: routeName !== null && routeName !== void 0 ? routeName : undefined,
    // Performance timings
    start: Math.round(start),
    end: Math.round(end),
    // Status and outcome
    status: interactionStatus.originalInteractionStatus,
    abortReason,
    previousInteractionName,
    isPreviousInteractionAborted,
    abortedByInteractionName,
    pageVisibilityAtTTI: pageVisibilityAtTTI !== null && pageVisibilityAtTTI !== void 0 ? pageVisibilityAtTTI : undefined,
    pageVisibilityAtTTAI,
    // Basic error count (not detailed error count)
    errorCount: interaction.errors.length,
    // Cohorting custom data
    ...(Object.keys(cohortingCustomData || {}).length > 0 && {
      cohortingCustomData
    })
  };
  const payload = {
    actionSubject: 'experience',
    action: 'measured',
    eventType: 'operational',
    source: 'measured',
    tags: ['observability'],
    attributes: {
      properties: properties
    }
  };
  return payload;
}
