// @atlaskit/react-select/dist/es2019/components/internal/scroll-manager.js
import React, { Fragment } from 'react';
import useScrollCapture from './use-scroll-capture';
import useScrollLock from './use-scroll-lock';
const blurSelectInput = event => {
  const element = event.target;
  return element.ownerDocument.activeElement && element.ownerDocument.activeElement.blur();
};
export default function ScrollManager({
  children,
  lockEnabled,
  captureEnabled = true,
  onBottomArrive,
  onBottomLeave,
  onTopArrive,
  onTopLeave
}) {
  const setScrollCaptureTarget = useScrollCapture({
    isEnabled: captureEnabled,
    onBottomArrive,
    onBottomLeave,
    onTopArrive,
    onTopLeave
  });
  const setScrollLockTarget = useScrollLock({
    isEnabled: lockEnabled
  });
  const targetRef = element => {
    setScrollCaptureTarget(element);
    setScrollLockTarget(element);
  };
  return /*#__PURE__*/React.createElement(Fragment, null, lockEnabled &&
  /*#__PURE__*/
  // eslint-disable-next-line @atlassian/a11y/click-events-have-key-events, @atlassian/a11y/interactive-element-not-keyboard-focusable, @atlassian/a11y/no-static-element-interactions
  React.createElement("div", {
    onClick: blurSelectInput,
    style: {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      position: 'fixed',
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      insetBlockEnd: 0,
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      insetBlockStart: 0,
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      insetInlineEnd: 0,
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      insetInlineStart: 0
    }
  }), children(targetRef));
}
