// @compiled/react/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsx = exports.cx = exports.styled = exports.keyframes = exports.cssMap = exports.css = exports.createStrictAPI = exports.ClassNames = void 0;
const react_1 = require("react");
var class_names_1 = require("./class-names");
Object.defineProperty(exports, "ClassNames", { enumerable: true, get: function () { return class_names_1.ClassNames; } });
var create_strict_api_1 = require("./create-strict-api");
Object.defineProperty(exports, "createStrictAPI", { enumerable: true, get: function () { return create_strict_api_1.createStrictAPI; } });
var css_1 = require("./css");
Object.defineProperty(exports, "css", { enumerable: true, get: function () { return __importDefault(css_1).default; } });
var css_map_1 = require("./css-map");
Object.defineProperty(exports, "cssMap", { enumerable: true, get: function () { return __importDefault(css_map_1).default; } });
var keyframes_1 = require("./keyframes");
Object.defineProperty(exports, "keyframes", { enumerable: true, get: function () { return keyframes_1.keyframes; } });
var styled_1 = require("./styled");
Object.defineProperty(exports, "styled", { enumerable: true, get: function () { return styled_1.styled; } });
var xcss_prop_1 = require("./xcss-prop");
Object.defineProperty(exports, "cx", { enumerable: true, get: function () { return xcss_prop_1.cx; } });
// Pass through the (classic) jsx runtime.
// Compiled currently doesn't define its own and uses this purely to enable a local jsx namespace.
exports.jsx = react_1.createElement;
