// @atlaskit/pagination/dist/cjs/internal/components/render-ellipsis.js
/* render-ellipsis.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = renderEllipsis;
require("./render-ellipsis.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _compiled = require("@atlaskit/primitives/compiled");
var _visuallyHidden = _interopRequireDefault(require("@atlaskit/visually-hidden"));
var styles = {
  container: "_1e0c1txw _kqswh2mm"
};
function renderEllipsis(_ref) {
  var key = _ref.key,
    testId = _ref.testId,
    from = _ref.from,
    to = _ref.to;
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    as: "li",
    testId: testId,
    key: key,
    xcss: styles.container,
    paddingInline: "space.100"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    testId: testId && "".concat(testId, "-text")
  }, /*#__PURE__*/_react.default.createElement(_visuallyHidden.default, null, "Skipped pages from ", from, " to ", to), "\u2026"));
}
