// @atlaskit/calendar/dist/cjs/internal/components/week-header.js
/* week-header.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./week-header.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _compiled = require("@atlaskit/primitives/compiled");
var _weekDayGrid = _interopRequireDefault(require("./week-day-grid"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var styles = {
  columnHeader: "_1ul98pqr _o5721q9c _1e0c1txw _1bah1h6o"
};
var WeekHeader = /*#__PURE__*/(0, _react.memo)(function WeekHeader(_ref) {
  var daysShort = _ref.daysShort,
    testId = _ref.testId;
  return /*#__PURE__*/React.createElement(_weekDayGrid.default, {
    isHidden: true,
    testId: testId && "".concat(testId, "--column-headers")
  }, daysShort.map(function (shortDay) {
    return /*#__PURE__*/React.createElement(_compiled.Box, {
      padding: "space.100",
      xcss: styles.columnHeader,
      key: shortDay,
      role: "columnheader",
      testId: testId && "".concat(testId, "--column-header")
    }, /*#__PURE__*/React.createElement(_compiled.Text, {
      weight: "bold",
      size: "small",
      align: "center",
      color: "color.text.subtle"
    }, shortDay));
  }));
});
WeekHeader.displayName = 'WeekHeader';

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = WeekHeader;
