// @atlaskit/navigation-system/dist/esm/ui/page-layout/panel-splitter/side-nav-panel-splitter.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React, { useContext, useEffect, useState } from 'react';
import invariant from 'tiny-invariant';
import { useOpenLayerObserver } from '@atlaskit/layering/experimental/open-layer-observer';
import { fg } from '@atlaskit/platform-feature-flags';
import { useIsFhsEnabled } from '../../fhs-rollout/use-is-fhs-enabled';
import { openLayerObserverSideNavNamespace, openLayerObserverTopNavEndNamespace, openLayerObserverTopNavMiddleNamespace, openLayerObserverTopNavStartNamespace, sideNavPanelSplitterId } from '../constants';
import { useToggleSideNav } from '../side-nav/use-toggle-side-nav';
import { OnDoubleClickContext, PanelSplitterContext } from './context';
import { PanelSplitter } from './panel-splitter';

/**
 * Namespaces to check for open layers that would interfere with the panel splitter.
 *
 * Placed outside the component for stability, as the list is used as an effect dependency.
 */
var openLayerNamespacesToCheck = [
// The side nav panel splitter is layered above the side nav, so needs to be disabled when there are open popups in the side nav.
openLayerObserverSideNavNamespace,
// The side nav panel splitter is layered above the top nav, so needs to be disabled when there are open popups in the top nav.
openLayerObserverTopNavStartNamespace, openLayerObserverTopNavMiddleNamespace, openLayerObserverTopNavEndNamespace];

/**
 * Returns whether there are any open popups in the side nav or top nav
 */
function useHasOpenPopupsInSideNavOrTopNav() {
  var isFhsEnabled = useIsFhsEnabled();
  var openLayerObserver = useOpenLayerObserver();
  // Setting the initial state to false, as it is unlikely that there would be any open popups when the app starts.
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    hasOpenPopups = _useState2[0],
    setHasOpenPopups = _useState2[1];
  useEffect(function () {
    if (!openLayerObserver || !isFhsEnabled || !fg('platform-dst-side-nav-layering-fixes')) {
      return;
    }
    function updateState() {
      if (!openLayerObserver) {
        return;
      }
      var hasAnyOpenLayers = openLayerNamespacesToCheck.some(function (namespace) {
        return openLayerObserver.getCount({
          namespace: namespace,
          type: 'popup'
        }) > 0;
      });
      setHasOpenPopups(hasAnyOpenLayers);
    }

    // Initial check. We do this _in case_ a popup is already open when the component mounts.
    updateState();

    // Subscribe to each namespace
    var cleanups = openLayerNamespacesToCheck.map(function (namespace) {
      return openLayerObserver.onChange(updateState, {
        namespace: namespace
      });
    });
    return function cleanupHook() {
      cleanups.forEach(function (cleanup) {
        return cleanup();
      });
    };
  }, [isFhsEnabled, openLayerObserver]);
  return hasOpenPopups;
}
/**
 * _SideNavPanelSplitter_
 *
 * A component that allows the user to resize or collapse the side nav.
 * It must be used within the `SideNav` layout area.
 *
 * Example usage:
 * ```tsx
 * <SideNav>
 *   <SideNavPanelSplitter label="Double click to collapse"  />
 * </SideNav>
 * ```
 */
export var SideNavPanelSplitter = function SideNavPanelSplitter(_ref) {
  var label = _ref.label,
    onResizeStart = _ref.onResizeStart,
    onResizeEnd = _ref.onResizeEnd,
    testId = _ref.testId,
    _ref$shouldCollapseOn = _ref.shouldCollapseOnDoubleClick,
    shouldCollapseOnDoubleClick = _ref$shouldCollapseOn === void 0 ? true : _ref$shouldCollapseOn,
    tooltipContent = _ref.tooltipContent;
  var context = useContext(PanelSplitterContext);
  invariant((context === null || context === void 0 ? void 0 : context.panelId) === sideNavPanelSplitterId, 'SideNavPanelSplitter must be rendered as a child of <SideNav />.');
  var toggleSideNav = useToggleSideNav({
    trigger: 'double-click'
  });
  var isFhsEnabled = useIsFhsEnabled();

  // The logic and state for disabling the panel splitter when there are open popups
  // in the side nav or top nav is being placed here, instead of in `SideNav`, to prevent
  // re-rendering the side nav anytime the number of open popups changes.
  var hasOpenLayersInSideNavOrTopNav = useHasOpenPopupsInSideNavOrTopNav();
  if (hasOpenLayersInSideNavOrTopNav && isFhsEnabled && fg('platform-dst-side-nav-layering-fixes')) {
    return null;
  }
  return /*#__PURE__*/React.createElement(OnDoubleClickContext.Provider, {
    value: shouldCollapseOnDoubleClick ? toggleSideNav : undefined
  }, /*#__PURE__*/React.createElement(PanelSplitter, {
    label: label,
    onResizeStart: onResizeStart,
    onResizeEnd: onResizeEnd,
    testId: testId,
    tooltipContent:
    // Only displaying the tooltip if double click to collapse is enabled.
    // Otherwise, we would be displaying a tooltip with "Double click to collapse", when it actually won't.
    shouldCollapseOnDoubleClick ? tooltipContent : undefined
  }));
};
