// @atlaskit/adf-schema/dist/cjs/next-schema/marks/alignmentAndIndentation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.indentation = exports.alignment = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
// These marks defined together because they form a cycle within the excludes.
var alignment = exports.alignment = (0, _adfSchemaGenerator.adfMark)('alignment');
var indentation = exports.indentation = (0, _adfSchemaGenerator.adfMark)('indentation');
var alignementMarkExclusionGroup = (0, _adfSchemaGenerator.adfMarkGroup)('alignment', [alignment]);
var indentationMarkExclusionGroup = (0, _adfSchemaGenerator.adfMarkGroup)('indentation', [indentation]);
alignment.define({
  excludes: [alignementMarkExclusionGroup, indentationMarkExclusionGroup],
  attrs: {
    align: {
      type: 'enum',
      values: ['center', 'end']
    }
  }
});
indentation.define({
  excludes: [indentationMarkExclusionGroup, alignementMarkExclusionGroup],
  attrs: {
    level: {
      type: 'number',
      minimum: 1,
      maximum: 6
    }
  }
});
