// @atlaskit/calendar/dist/es2019/internal/components/week-header.js
/* week-header.tsx generated by @compiled/babel-plugin v0.38.1 */
import "./week-header.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { memo } from 'react';
import { Box, Text } from '@atlaskit/primitives/compiled';
import WeekDayGrid from './week-day-grid';
const styles = {
  columnHeader: "_1ul98pqr _o5721q9c _1e0c1txw _1bah1h6o"
};
const WeekHeader = /*#__PURE__*/memo(function WeekHeader({
  daysShort,
  testId
}) {
  return /*#__PURE__*/React.createElement(WeekDayGrid, {
    isHidden: true,
    testId: testId && `${testId}--column-headers`
  }, daysShort.map(shortDay => /*#__PURE__*/React.createElement(Box, {
    padding: "space.100",
    xcss: styles.columnHeader,
    key: shortDay,
    role: "columnheader",
    testId: testId && `${testId}--column-header`
  }, /*#__PURE__*/React.createElement(Text, {
    weight: "bold",
    size: "small",
    align: "center",
    color: "color.text.subtle"
  }, shortDay))));
});
WeekHeader.displayName = 'WeekHeader';

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export default WeekHeader;
