// @atlaskit/react-ufo/dist/cjs/vc/no-op-vc-observer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VCObserverNOOP = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
// This class should be used on scenarios where VC can not be calculate
// such as: SSR environment and legacy browsers
var VCObserverNOOP = exports.VCObserverNOOP = /*#__PURE__*/function () {
  function VCObserverNOOP() {
    (0, _classCallCheck2.default)(this, VCObserverNOOP);
  }
  return (0, _createClass2.default)(VCObserverNOOP, [{
    key: "start",
    value: function start(_) {}
  }, {
    key: "stop",
    value: function stop() {}
  }, {
    key: "getVCRawData",
    value: function getVCRawData() {
      return null;
    }
  }, {
    key: "getVCResult",
    value: function getVCResult(_) {
      return Promise.resolve({
        'ufo:vc:noop': true
      });
    }
  }, {
    key: "setSSRElement",
    value: function setSSRElement(_) {}
  }, {
    key: "setReactRootRenderStart",
    value: function setReactRootRenderStart(_) {}
  }, {
    key: "setReactRootRenderStop",
    value: function setReactRootRenderStop(_) {}
  }]);
}();
