// @atlaskit/side-nav-items/dist/es2019/ui/menu-item/flyout-menu-item/flyout-header.js
/* flyout-header.tsx generated by @compiled/babel-plugin v0.38.1 */
import "./flyout-header.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { useCallback, useContext } from 'react';
import Heading from '@atlaskit/heading';
import { Flex } from '@atlaskit/primitives/compiled';
import { CloseButton } from './close-button';
import { OnCloseContext, useTitleId } from './flyout-menu-item-context';
const headerStyles = {
  root: "_zulp12x7 _bozg1b66 _85i512x7 _1e0c1txw _2lx21bp4",
  flex: "_zulppxbi _1bah1yb4 _2lx21sbv _4cvr1h6o _1bsb1osq _bozgv77o"
};
/**
 * __FlyoutHeader__
 *
 * The header section of a flyout menu. This component displays the title, and
 * close button, as well as any other provided actions relevant to the menu.
 * This component should be placed first within the FlyoutMenuItemContent.
 */
export const FlyoutHeader = props => {
  const {
    children,
    testId,
    title,
    closeButtonLabel
  } = props;
  const id = useTitleId();
  const onCloseRef = useContext(OnCloseContext);
  const handleClose = useCallback(event => {
    var _onCloseRef$current;
    (_onCloseRef$current = onCloseRef.current) === null || _onCloseRef$current === void 0 ? void 0 : _onCloseRef$current.call(onCloseRef, event, 'close-button');
  }, [onCloseRef]);
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId,
    className: ax([headerStyles.root])
  }, /*#__PURE__*/React.createElement(Flex, {
    xcss: headerStyles.flex
  }, /*#__PURE__*/React.createElement(CloseButton, {
    label: closeButtonLabel,
    onClick: handleClose,
    testId: testId && `${testId}--close-button`
  }), /*#__PURE__*/React.createElement(Heading, {
    size: "xsmall",
    as: "h2",
    id: id
  }, title)), children);
};
