// @atlaskit/side-nav-items/dist/es2019/ui/menu-item/flyout-menu-item/flyout-menu-item-trigger.js
/* flyout-menu-item-trigger.tsx generated by @compiled/babel-plugin v0.38.1 */
import "./flyout-menu-item-trigger.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef, useCallback, useContext } from 'react';
import mergeRefs from '@atlaskit/ds-lib/merge-refs';
import ChevronRightIcon from '@atlaskit/icon/core/chevron-right';
import { fg } from '@atlaskit/platform-feature-flags';
import { PopupTrigger } from '@atlaskit/popup/experimental';
import { MenuItemBase } from '../menu-item';
import { IsOpenContext, OnCloseContext } from './flyout-menu-item-context';
const elemAfterStyles = {
  root: "_18zr12x7 _1tz3r0mg"
};
/**
 * __FlyoutMenuItemTrigger__
 *
 * The button that toggles the flyout menu.
 */
export const FlyoutMenuItemTrigger = /*#__PURE__*/forwardRef(({
  children,
  elemBefore,
  onClick,
  isSelected,
  interactionName,
  testId,
  isContentTooltipDisabled,
  visualContentRef,
  isDragging,
  hasDragIndicator,
  dropIndicator
}, forwardedRef) => {
  const isOpen = useContext(IsOpenContext);
  const onCloseRef = useContext(OnCloseContext);
  const handleClick = useCallback((event, analyticsEvent) => {
    // If the flyout is open and the trigger is clicked, close the flyout and call the onClick
    // handler with the source information set to 'outside-click'.
    if (fg('platform_dst_nav4_flyout_menu_slots_close_button')) {
      if (isOpen && onCloseRef.current) {
        onCloseRef.current(event, 'outside-click');
      }
    }
    onClick === null || onClick === void 0 ? void 0 : onClick(event, analyticsEvent);
  }, [isOpen, onCloseRef, onClick]);
  return /*#__PURE__*/React.createElement(PopupTrigger, null, ({
    ref,
    'aria-controls': ariaControls,
    'aria-expanded': ariaExpanded,
    'aria-haspopup': ariaHasPopup
  }) => /*#__PURE__*/React.createElement(MenuItemBase, {
    testId: testId,
    ref: mergeRefs([ref, forwardedRef]),
    visualContentRef: visualContentRef,
    elemBefore: elemBefore,
    elemAfter: /*#__PURE__*/React.createElement("div", {
      className: ax([elemAfterStyles.root])
    }, /*#__PURE__*/React.createElement(ChevronRightIcon, {
      label: "",
      color: "currentColor",
      size: "small"
    })),
    onClick: handleClick,
    ariaControls: ariaControls,
    ariaExpanded: ariaExpanded,
    ariaHasPopup: ariaHasPopup,
    interactionName: interactionName,
    isContentTooltipDisabled: isContentTooltipDisabled,
    isSelected: isSelected,
    isDragging: isDragging,
    hasDragIndicator: hasDragIndicator,
    dropIndicator: dropIndicator
  }, children));
});
