// @forge/manifest/out/validators/compatibility-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompatibilityValidator = void 0;
const text_1 = require("../text");
const text_2 = require("../text");
const utils_1 = require("../utils");
class CompatibilityValidator {
    async validate(manifest) {
        if (!manifest || !manifest.typedContent || !manifest.filePath) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        const validationErrors = [];
        const modules = manifest.typedContent.modules || {};
        const compatibility = manifest.typedContent.app?.compatibility || {};
        const isJiraPrimary = compatibility.jira?.required === true;
        const isConfluencePrimary = compatibility.confluence?.required === true;
        const moduleKeys = Object.keys(modules);
        if (isJiraPrimary && moduleKeys.some((key) => key === 'confluence:fullPage')) {
            validationErrors.push({
                message: text_1.errors.app.compatibility.invalidConfluenceModuleInJira(),
                reference: text_2.References.SchemaError,
                level: 'error',
                ...(0, utils_1.findPosition)('confluence:fullPage:', manifest.yamlContentByLine)
            });
        }
        if (isConfluencePrimary && moduleKeys.some((key) => key === 'jira:fullPage')) {
            validationErrors.push({
                message: text_1.errors.app.compatibility.invalidJiraModuleInConfluence(),
                reference: text_2.References.SchemaError,
                level: 'error',
                ...(0, utils_1.findPosition)('jira:fullPage', manifest.yamlContentByLine)
            });
        }
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.CompatibilityValidator = CompatibilityValidator;
