// @forge/manifest/out/validators/connect-remote-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectRemoteValidator = void 0;
const text_1 = require("../text");
const utils_1 = require("../utils");
class ConnectRemoteValidator {
    async validate(manifest) {
        if (!manifest?.typedContent?.app) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        const app = manifest.typedContent.app;
        const validationErrors = [];
        if (manifest.typedContent.connectModules && !app.connect?.remote) {
            validationErrors.push({
                message: text_1.errors.app.missingAppConnectRemote(),
                reference: text_1.References.App,
                level: 'error',
                ...(0, utils_1.findPosition)('connect', manifest.yamlContentByLine)
            });
        }
        if (manifest.typedContent.connectModules &&
            app?.connect?.remote &&
            !manifest.typedContent.remotes?.find((remote) => remote.key === app?.connect?.remote)) {
            validationErrors.push({
                message: text_1.errors.app.missingRemoteForConnect(app?.connect?.remote || 'undefined'),
                reference: text_1.References.App,
                level: 'error',
                ...(0, utils_1.findPosition)('connect', manifest.yamlContentByLine)
            });
        }
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.ConnectRemoteValidator = ConnectRemoteValidator;
