// @forge/manifest/out/validators/connect-authentication-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectAuthenticationValidator = void 0;
const text_1 = require("../text");
const utils_1 = require("../utils");
class ConnectAuthenticationValidator {
    async validate(manifest) {
        if (!manifest?.typedContent?.app) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        const app = manifest.typedContent.app;
        const validationErrors = [];
        if (!manifest.typedContent.connectModules && app?.connect?.authentication) {
            validationErrors.push({
                message: text_1.errors.app.missingConnectModules(),
                reference: text_1.References.App,
                level: 'error',
                ...(0, utils_1.findPosition)('connect', manifest.yamlContentByLine)
            });
        }
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.ConnectAuthenticationValidator = ConnectAuthenticationValidator;
