// @forge/manifest/out/validators/data-classification-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataClassificationValidator = void 0;
const text_1 = require("../text");
const utils_1 = require("../utils");
class DataClassificationValidator {
    async validate(manifest) {
        if (!manifest?.typedContent?.remotes) {
            return {
                success: true,
                manifestObject: manifest
            };
        }
        const validationErrors = [];
        manifest.typedContent.remotes.forEach((remote) => {
            if (remote.operations && remote.operations.includes('storage') && remote.storage?.inScopeEUD === undefined) {
                validationErrors.push({
                    message: text_1.errors.app.remotes.missingRemotesStorageInScopeEUD(remote.key),
                    reference: text_1.References.App,
                    level: 'error',
                    ...(0, utils_1.findPosition)(remote.key, manifest.yamlContentByLine)
                });
            }
        });
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.DataClassificationValidator = DataClassificationValidator;
