// @forge/manifest/out/validators/display-conditions-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisplayConditionsValidator = void 0;
const tslib_1 = require("tslib");
const ajv_1 = tslib_1.__importDefault(require("ajv"));
const text_1 = require("../text");
const isPlainObject = (obj) => Object.prototype.toString.call(obj) === '[object Object]';
const errors = text_1.errors.modules.displayConditions;
class DisplayConditionsValidator {
    schema;
    validateSchema;
    moduleSchemas;
    supportedOperations;
    constructor(schema) {
        this.schema = schema;
        const ajv = new ajv_1.default({ allErrors: true, verbose: true, strict: false });
        this.validateSchema = ajv.compile(this.schema);
        const properties = this.schema.properties;
        this.moduleSchemas = properties.modules?.properties;
        this.supportedOperations = properties.operations?.items?.enum;
    }
    flattenDisplayConditions(moduleKey, displayConditions, nesting = 0) {
        if (nesting >= 10) {
            throw Error(errors.overcomplicated(moduleKey));
        }
        if (Array.isArray(displayConditions)) {
            return [displayConditions];
        }
        const result = [];
        const obj = {};
        const keys = Object.keys(displayConditions);
        for (let i = 0; i < keys.length; i += 1) {
            if (this.supportedOperations.includes(keys[i])) {
                result.push(...this.flattenDisplayConditions(moduleKey, displayConditions[keys[i]], nesting + 1));
            }
            else {
                obj[keys[i]] = displayConditions[keys[i]];
            }
        }
        return [...result, obj].filter((entry) => Object.keys(entry).length > 0);
    }
    toValidationError(message) {
        return {
            message,
            reference: text_1.References.SchemaError,
            level: 'error',
            line: 0,
            column: 0
        };
    }
    async validate(manifest) {
        if (!manifest ||
            !manifest.typedContent ||
            (!manifest.typedContent.modules && !manifest.typedContent.connectModules)) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        const { typedContent: { modules } } = manifest;
        if (!modules) {
            return {
                success: true,
                manifestObject: manifest
            };
        }
        const validationErrors = new Map();
        const reportError = (moduleKey, instancePath, message) => {
            const key = `${moduleKey}__${instancePath}`;
            if (!validationErrors.has(key)) {
                validationErrors.set(key, this.toValidationError(message));
            }
        };
        Object.keys(this.moduleSchemas).forEach((m) => {
            modules[m]?.forEach((module) => {
                if (!module.displayConditions) {
                    return;
                }
                if (!isPlainObject(module.displayConditions)) {
                    reportError(module.key, 'n/a', errors.invalid(module.key));
                    return;
                }
                let displayConditions = [];
                try {
                    displayConditions = this.flattenDisplayConditions(module.key, module.displayConditions);
                }
                catch (e) {
                    reportError(module.key, 'n/a', e.message);
                    return;
                }
                displayConditions.forEach((result) => {
                    void this.validateSchema({ modules: { [m]: result } });
                    this.validateSchema.errors?.forEach((error) => {
                        const { keyword, instancePath, params, message } = error;
                        const split = instancePath.split('/');
                        const path = String(split.pop());
                        if (keyword === 'type' && message) {
                            if (path === String(Number(path))) {
                                reportError(module.key, instancePath, errors.typeMismatch(module.key, `${split.pop()}[${path}]`, message));
                            }
                            else {
                                reportError(module.key, instancePath, errors.typeMismatch(module.key, path, message));
                            }
                        }
                        else if (keyword === 'enum' && params) {
                            reportError(module.key, instancePath, `${path}; ${text_1.errors.schema.enumValues(params.allowedValues)}`);
                        }
                        else if (keyword === 'additionalProperties' && params) {
                            reportError(module.key, instancePath, errors.propertyNotAllowed(module.key, String(params.additionalProperty)));
                        }
                        else {
                            reportError(module.key, instancePath, errors.invalid(module.key));
                        }
                    });
                });
            });
        });
        return {
            success: validationErrors.size === 0,
            manifestObject: manifest,
            errors: Array.from(validationErrors.values())
        };
    }
}
exports.DisplayConditionsValidator = DisplayConditionsValidator;
