// @forge/manifest/out/validators/entity-property-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityPropertyValidator = void 0;
const types_1 = require("../types");
const text_1 = require("../text");
const utils_1 = require("../utils");
const isManifestWithModules = (manifest) => {
    return !!manifest?.typedContent?.['modules'];
};
class EntityPropertyValidator {
    async validate(manifest) {
        const validationErrors = [];
        if (manifest?.typedContent && isManifestWithModules(manifest)) {
            const { yamlContentByLine, typedContent: { modules } } = manifest;
            const appValuesLength = modules[types_1.AllModuleTypes.JiraEntityProperty]?.reduce((acc, module) => acc + module.values.length, 0) || 0;
            if (appValuesLength > 30) {
                validationErrors.push({
                    message: text_1.errors.modules.jiraEntityProperty.maxExtractionsPerApp(30),
                    reference: text_1.References.Modules,
                    level: 'error',
                    ...(0, utils_1.findPosition)('app', yamlContentByLine)
                });
            }
        }
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.EntityPropertyValidator = EntityPropertyValidator;
