// @forge/manifest/out/validators/file-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileValidator = void 0;
const tslib_1 = require("tslib");
const types_1 = require("../types");
const fs_1 = tslib_1.__importDefault(require("fs"));
const text_1 = require("../text");
class FileValidator {
    async validate(manifest) {
        const file = manifest || types_1.MANIFEST_FILE;
        if (!fs_1.default.existsSync(file)) {
            return {
                success: false,
                errors: [
                    {
                        message: text_1.errors.missingManifest(),
                        reference: text_1.References.MissingManifest,
                        level: 'error',
                        line: 0,
                        column: 0
                    }
                ]
            };
        }
        return {
            success: true,
            manifestObject: {
                filePath: file
            }
        };
    }
}
exports.FileValidator = FileValidator;
