// @forge/manifest/out/validators/package-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageValidator = void 0;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const text_1 = require("../text");
const glob_1 = require("glob");
class PackageValidator {
    async fileExistsForGlob(pattern) {
        return (await (0, glob_1.glob)(pattern)).some((file) => fs_1.default.lstatSync(file).isFile());
    }
    async validate(manifest) {
        const extraFiles = manifest?.typedContent?.app?.package?.extraFiles ?? [];
        const validationErrors = [];
        for (const glob of extraFiles) {
            if (!(await this.fileExistsForGlob(glob))) {
                validationErrors.push({
                    message: text_1.errors.app.package.extraFiles.missing(glob),
                    reference: text_1.References.App,
                    level: 'error'
                });
            }
        }
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.PackageValidator = PackageValidator;
