// @forge/manifest/out/validators/product-trigger-scopes-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductTriggerScopesValidator = exports.validateEventScopes = void 0;
const text_1 = require("../text");
const utils_1 = require("../utils");
const scopes_1 = require("../scopes");
async function validateEventScopes(event, scopes, config) {
    const emptyRequiredScopes = { current: [] };
    const requiredScopes = config.mapping.find((entry) => entry.productEvent === event)?.oAuthScopes || emptyRequiredScopes;
    const missingScopes = await (0, scopes_1.getMissingScopes)(scopes, requiredScopes);
    return { requiredScopes, missingScopes };
}
exports.validateEventScopes = validateEventScopes;
class ProductTriggerScopesValidator {
    config;
    constructor(config) {
        this.config = config;
    }
    async validate(manifest) {
        if (!manifest || !manifest.typedContent) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        const validationErrors = [];
        const addValidationError = (scope, event) => {
            validationErrors.push({
                message: text_1.errors.permissions.missingPermissionFromScope(scope, event),
                reference: text_1.References.MissingScopes,
                level: 'error',
                metadata: {
                    missingPermission: scope
                },
                ...(0, utils_1.findPosition)('scopes', manifest.yamlContentByLine)
            });
        };
        if (!manifest.typedContent.modules || !manifest.typedContent.modules.trigger) {
            return {
                success: true,
                manifestObject: manifest
            };
        }
        const manifestScopes = manifest.typedContent.permissions?.scopes || [];
        for (const element of manifest.typedContent.modules.trigger) {
            for (const event of element.events) {
                const { missingScopes } = await validateEventScopes(event, manifestScopes, this.config);
                missingScopes.forEach((scope) => addValidationError(scope, event));
            }
        }
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.ProductTriggerScopesValidator = ProductTriggerScopesValidator;
