// @forge/manifest/out/validators/yaml-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YamlValidator = void 0;
const tslib_1 = require("tslib");
const yaml_1 = require("yaml");
const fs_1 = tslib_1.__importDefault(require("fs"));
const text_1 = require("../text");
const utils_1 = require("../utils");
class YamlValidator {
    async validate(manifest) {
        if (!manifest || !manifest.filePath) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        try {
            const content = fs_1.default.readFileSync(manifest.filePath, 'utf8');
            const manifestContent = new utils_1.ManifestParserBuilder()
                .withInterpolators()
                .build()
                .parseManifest({ content, filePath: manifest.filePath });
            return {
                success: true,
                manifestObject: {
                    filePath: manifest.filePath,
                    yamlContent: manifestContent,
                    yamlContentByLine: content.split('\n')
                }
            };
        }
        catch (e) {
            if (e instanceof yaml_1.YAMLError) {
                const pos = e.linePos?.[0];
                return {
                    success: false,
                    errors: [
                        {
                            message: text_1.errors.invalidManifest(e.message.replace(/(at line).+/gms, '').trim()),
                            reference: text_1.References.InvalidManifest,
                            level: 'error',
                            line: pos?.line ?? 0,
                            column: pos?.col ?? 0
                        }
                    ]
                };
            }
            return {
                success: false,
                errors: [
                    {
                        message: text_1.errors.invalidManifest(e.message),
                        reference: text_1.References.InvalidManifest,
                        level: 'error',
                        line: 0,
                        column: 0
                    }
                ]
            };
        }
    }
}
exports.YamlValidator = YamlValidator;
