// @atlaskit/navigation-system/dist/es2019/ui/top-nav-items/app-switcher.js
import React, { forwardRef } from 'react';
import AppSwitcherIcon from '@atlaskit/icon/core/app-switcher';
import { IconButton } from './themed/migration';
const toggleButtonTooltipOptions = {
  // We're disabling pointer events on the tooltip to prevent it from blocking mouse events, so that the side nav flyout stays open
  // when moving the mouse from the top bar to the side nav.
  ignoreTooltipPointerEvents: true
};

/**
 * __App switcher__
 *
 * The trigger button for the app switcher. Allows users to switch between Atlassian products.
 */
export const AppSwitcher = /*#__PURE__*/forwardRef(({
  label,
  onClick,
  onMouseEnter,
  interactionName,
  isSelected,
  'aria-controls': ariaControls,
  'aria-expanded': ariaExpanded,
  'aria-haspopup': ariaHasPopup
}, ref) => /*#__PURE__*/React.createElement(IconButton, {
  ref: ref,
  "aria-controls": ariaControls,
  "aria-expanded": ariaExpanded,
  "aria-haspopup": ariaHasPopup,
  icon: AppSwitcherIcon,
  label: label,
  appearance: "subtle",
  onClick: onClick,
  onMouseEnter: onMouseEnter,
  interactionName: interactionName,
  isTooltipDisabled: false,
  tooltip: toggleButtonTooltipOptions,
  isSelected: isSelected
}));
