// @atlaskit/react-ufo/dist/cjs/additional-payload/utils/lighthouse-metrics/utils/buffer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EntriesBuffer = exports.BufferWithMaxLength = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _const = require("../../const");
var BufferWithMaxLength = exports.BufferWithMaxLength = /*#__PURE__*/function () {
  function BufferWithMaxLength() {
    var maxLength = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1000;
    (0, _classCallCheck2.default)(this, BufferWithMaxLength);
    (0, _defineProperty2.default)(this, "buffer", []);
    (0, _defineProperty2.default)(this, "full", false);
    this.maxLength = maxLength;
  }
  return (0, _createClass2.default)(BufferWithMaxLength, [{
    key: "push",
    value: function push(item) {
      if (this.full || this.maxLength === this.buffer.length) {
        this.full = true;
        this.buffer.shift();
      }
      this.buffer.push(item);
    }
  }, {
    key: "getAll",
    value: function getAll() {
      return this.buffer;
    }
  }]);
}();
var EntriesBuffer = exports.EntriesBuffer = (0, _defineProperty2.default)((0, _defineProperty2.default)({}, _const.PerformanceObserverEntryTypes.LongTask, new BufferWithMaxLength()), _const.PerformanceObserverEntryTypes.LayoutShift, new BufferWithMaxLength());
