// @atlaskit/side-nav-items/dist/esm/ui/menu-item/flyout-menu-item/flyout-menu-item.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React, { forwardRef, useEffect, useRef } from 'react';
import { useAnalyticsEvents } from '@atlaskit/analytics-next';
import useControlled from '@atlaskit/ds-lib/use-controlled';
import usePreviousValue from '@atlaskit/ds-lib/use-previous-value';
import { fg } from '@atlaskit/platform-feature-flags';
import { Popup } from '@atlaskit/popup/experimental';
import { MenuListItem } from '../menu-list-item';
import { IsOpenContext, OnCloseContext, SetIsOpenContext } from './flyout-menu-item-context';
/**
 * __FlyoutMenuItem__
 *
 * Displays content in a flyout menu, triggered by a button.
 *
 * The top-level component that contains the trigger and content of a flyout menu.
 *
 * Usage example:
 * ```tsx
 * <FlyoutMenuItem>
 *   <FlyoutMenuItemTrigger>Trigger</FlyoutMenuItemTrigger>
 *   <FlyoutMenuItemContent>
 *     <MenuList>
 *       <ButtonMenuItem>Item 1</ButtonMenuItem>
 *       <ButtonMenuItem>Item 2</ButtonMenuItem>
 *     </MenuList>
 *   </FlyoutMenuItemContent>
 * </FlyoutMenuItem>
 * ```
 */
export var FlyoutMenuItem = /*#__PURE__*/forwardRef(function (_ref, forwardedRef) {
  var children = _ref.children,
    id = _ref.id,
    isOpenControlled = _ref.isOpen,
    _ref$isDefaultOpen = _ref.isDefaultOpen,
    isDefaultOpen = _ref$isDefaultOpen === void 0 ? false : _ref$isDefaultOpen,
    onOpenChange = _ref.onOpenChange;
  var _useControlled = useControlled(isOpenControlled, function () {
      return isDefaultOpen;
    }),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    isOpen = _useControlled2[0],
    setIsOpen = _useControlled2[1];
  var previousIsOpen = usePreviousValue(isOpen);
  var onCloseRef = useRef(null);
  var _useAnalyticsEvents = useAnalyticsEvents(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  useEffect(function () {
    if (previousIsOpen === undefined || previousIsOpen === isOpen) {
      /**
       * The previous value is `undefined` on initialization, so if it is `undefined` then the value hasn't changed.
       *
       * The previous value can be equal to the current one if the component re-renders due to something else changing.
       *
       * In both cases the value hasn't changed and we don't want to notify consumers.
       */
      return;
    }

    // When flyout menu is opened, fire analytics event
    if (isOpen && previousIsOpen === false) {
      if (fg('platform_dst_nav4_flyout_menu_slots_close_button')) {
        var navigationAnalyticsEvent = createAnalyticsEvent({
          source: 'sideNav',
          actionSubject: 'flyoutMenu',
          action: 'opened'
        });
        navigationAnalyticsEvent.fire('navigation');
      }
    }
    onOpenChange === null || onOpenChange === void 0 || onOpenChange(isOpen);
  }, [isOpen, onOpenChange, previousIsOpen, createAnalyticsEvent]);
  return /*#__PURE__*/React.createElement(IsOpenContext.Provider, {
    value: isOpen
  }, /*#__PURE__*/React.createElement(SetIsOpenContext.Provider, {
    value: setIsOpen
  }, /*#__PURE__*/React.createElement(OnCloseContext.Provider, {
    value: onCloseRef
  }, /*#__PURE__*/React.createElement(MenuListItem, {
    ref: forwardedRef
  }, /*#__PURE__*/React.createElement(Popup, {
    id: id,
    isOpen: isOpen,
    role: fg('platform_dst_nav4_flyout_menu_slots_close_button') ? 'dialog' : undefined
  }, children)))));
});
