// @atlaskit/adf-schema/dist/cjs/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "B100", {
  enumerable: true,
  get: function get() {
    return _colors.B100;
  }
});
Object.defineProperty(exports, "B400", {
  enumerable: true,
  get: function get() {
    return _colors.B400;
  }
});
Object.defineProperty(exports, "B50", {
  enumerable: true,
  get: function get() {
    return _colors.B50;
  }
});
Object.defineProperty(exports, "B500", {
  enumerable: true,
  get: function get() {
    return _colors.B500;
  }
});
Object.defineProperty(exports, "B75", {
  enumerable: true,
  get: function get() {
    return _colors.B75;
  }
});
Object.defineProperty(exports, "G200", {
  enumerable: true,
  get: function get() {
    return _colors.G200;
  }
});
Object.defineProperty(exports, "G300", {
  enumerable: true,
  get: function get() {
    return _colors.G300;
  }
});
Object.defineProperty(exports, "G400", {
  enumerable: true,
  get: function get() {
    return _colors.G400;
  }
});
Object.defineProperty(exports, "G50", {
  enumerable: true,
  get: function get() {
    return _colors.G50;
  }
});
Object.defineProperty(exports, "G500", {
  enumerable: true,
  get: function get() {
    return _colors.G500;
  }
});
Object.defineProperty(exports, "G75", {
  enumerable: true,
  get: function get() {
    return _colors.G75;
  }
});
Object.defineProperty(exports, "N0", {
  enumerable: true,
  get: function get() {
    return _colors.N0;
  }
});
Object.defineProperty(exports, "N20", {
  enumerable: true,
  get: function get() {
    return _colors.N20;
  }
});
Object.defineProperty(exports, "N200", {
  enumerable: true,
  get: function get() {
    return _colors.N200;
  }
});
Object.defineProperty(exports, "N30", {
  enumerable: true,
  get: function get() {
    return _colors.N30;
  }
});
Object.defineProperty(exports, "N300", {
  enumerable: true,
  get: function get() {
    return _colors.N300;
  }
});
Object.defineProperty(exports, "N40", {
  enumerable: true,
  get: function get() {
    return _colors.N40;
  }
});
Object.defineProperty(exports, "N50", {
  enumerable: true,
  get: function get() {
    return _colors.N50;
  }
});
Object.defineProperty(exports, "N500", {
  enumerable: true,
  get: function get() {
    return _colors.N500;
  }
});
Object.defineProperty(exports, "N60", {
  enumerable: true,
  get: function get() {
    return _colors.N60;
  }
});
Object.defineProperty(exports, "N80", {
  enumerable: true,
  get: function get() {
    return _colors.N80;
  }
});
Object.defineProperty(exports, "N800", {
  enumerable: true,
  get: function get() {
    return _colors.N800;
  }
});
Object.defineProperty(exports, "N90", {
  enumerable: true,
  get: function get() {
    return _colors.N90;
  }
});
Object.defineProperty(exports, "P100", {
  enumerable: true,
  get: function get() {
    return _colors.P100;
  }
});
Object.defineProperty(exports, "P300", {
  enumerable: true,
  get: function get() {
    return _colors.P300;
  }
});
Object.defineProperty(exports, "P400", {
  enumerable: true,
  get: function get() {
    return _colors.P400;
  }
});
Object.defineProperty(exports, "P50", {
  enumerable: true,
  get: function get() {
    return _colors.P50;
  }
});
Object.defineProperty(exports, "P500", {
  enumerable: true,
  get: function get() {
    return _colors.P500;
  }
});
Object.defineProperty(exports, "P75", {
  enumerable: true,
  get: function get() {
    return _colors.P75;
  }
});
Object.defineProperty(exports, "R100", {
  enumerable: true,
  get: function get() {
    return _colors.R100;
  }
});
Object.defineProperty(exports, "R300", {
  enumerable: true,
  get: function get() {
    return _colors.R300;
  }
});
Object.defineProperty(exports, "R400", {
  enumerable: true,
  get: function get() {
    return _colors.R400;
  }
});
Object.defineProperty(exports, "R50", {
  enumerable: true,
  get: function get() {
    return _colors.R50;
  }
});
Object.defineProperty(exports, "R500", {
  enumerable: true,
  get: function get() {
    return _colors.R500;
  }
});
Object.defineProperty(exports, "R75", {
  enumerable: true,
  get: function get() {
    return _colors.R75;
  }
});
Object.defineProperty(exports, "T100", {
  enumerable: true,
  get: function get() {
    return _colors.T100;
  }
});
Object.defineProperty(exports, "T300", {
  enumerable: true,
  get: function get() {
    return _colors.T300;
  }
});
Object.defineProperty(exports, "T50", {
  enumerable: true,
  get: function get() {
    return _colors.T50;
  }
});
Object.defineProperty(exports, "T500", {
  enumerable: true,
  get: function get() {
    return _colors.T500;
  }
});
Object.defineProperty(exports, "T75", {
  enumerable: true,
  get: function get() {
    return _colors.T75;
  }
});
Object.defineProperty(exports, "Y200", {
  enumerable: true,
  get: function get() {
    return _colors.Y200;
  }
});
Object.defineProperty(exports, "Y400", {
  enumerable: true,
  get: function get() {
    return _colors.Y400;
  }
});
Object.defineProperty(exports, "Y50", {
  enumerable: true,
  get: function get() {
    return _colors.Y50;
  }
});
Object.defineProperty(exports, "Y500", {
  enumerable: true,
  get: function get() {
    return _colors.Y500;
  }
});
Object.defineProperty(exports, "Y75", {
  enumerable: true,
  get: function get() {
    return _colors.Y75;
  }
});
Object.defineProperty(exports, "acNameToEmoji", {
  enumerable: true,
  get: function get() {
    return _emoji.acNameToEmoji;
  }
});
Object.defineProperty(exports, "acShortcutToEmoji", {
  enumerable: true,
  get: function get() {
    return _emoji.acShortcutToEmoji;
  }
});
Object.defineProperty(exports, "emojiIdToAcName", {
  enumerable: true,
  get: function get() {
    return _emoji.emojiIdToAcName;
  }
});
Object.defineProperty(exports, "generateUuid", {
  enumerable: true,
  get: function get() {
    return _uuid.generateUuid;
  }
});
Object.defineProperty(exports, "getDarkModeLCHColor", {
  enumerable: true,
  get: function get() {
    return _lchColorInversion.getDarkModeLCHColor;
  }
});
Object.defineProperty(exports, "getEmojiAcName", {
  enumerable: true,
  get: function get() {
    return _emoji.getEmojiAcName;
  }
});
Object.defineProperty(exports, "getLinkMatch", {
  enumerable: true,
  get: function get() {
    return _url.getLinkMatch;
  }
});
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _colors.hexToRgb;
  }
});
Object.defineProperty(exports, "hexToRgba", {
  enumerable: true,
  get: function get() {
    return _colors.hexToRgba;
  }
});
Object.defineProperty(exports, "isHex", {
  enumerable: true,
  get: function get() {
    return _colors.isHex;
  }
});
Object.defineProperty(exports, "isRgb", {
  enumerable: true,
  get: function get() {
    return _colors.isRgb;
  }
});
Object.defineProperty(exports, "isSafeUrl", {
  enumerable: true,
  get: function get() {
    return _url.isSafeUrl;
  }
});
Object.defineProperty(exports, "linkify", {
  enumerable: true,
  get: function get() {
    return _url.linkify;
  }
});
Object.defineProperty(exports, "linkifyMatch", {
  enumerable: true,
  get: function get() {
    return _url.linkifyMatch;
  }
});
Object.defineProperty(exports, "normalizeHexColor", {
  enumerable: true,
  get: function get() {
    return _colors.normalizeHexColor;
  }
});
Object.defineProperty(exports, "normalizeUrl", {
  enumerable: true,
  get: function get() {
    return _url.normalizeUrl;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _colors.rgbToHex;
  }
});
Object.defineProperty(exports, "uuid", {
  enumerable: true,
  get: function get() {
    return _uuid.uuid;
  }
});
var _emoji = require("./confluence/emoji");
var _uuid = require("./uuid");
var _colors = require("./colors");
var _url = require("./url");
var _lchColorInversion = require("./lch-color-inversion");
