// @atlaskit/adf-schema-generator/dist/es2019/adfMark.js
function _classPrivateFieldInitSpec(e, t, a) { _checkPrivateRedeclaration(e, t), t.set(e, a); }
function _checkPrivateRedeclaration(e, t) { if (t.has(e)) throw new TypeError("Cannot initialize the same private elements twice on an object"); }
function _classPrivateFieldGet(s, a) { return s.get(_assertClassBrand(s, a)); }
function _classPrivateFieldSet(s, a, r) { return s.set(_assertClassBrand(s, a), r), r; }
function _assertClassBrand(e, t, n) { if ("function" == typeof e ? e === t : e.has(t)) return arguments.length < 3 ? t : n; throw new TypeError("Private element is not present on this object"); }
var _type = /*#__PURE__*/new WeakMap();
var _spec = /*#__PURE__*/new WeakMap();
var _group = /*#__PURE__*/new WeakMap();
export class ADFMark {
  constructor(type) {
    _classPrivateFieldInitSpec(this, _type, void 0);
    _classPrivateFieldInitSpec(this, _spec, null);
    _classPrivateFieldInitSpec(this, _group, null);
    this.type = type;
    _classPrivateFieldSet(_type, this, type);
  }

  /**
   * Returns true if the mark has been defined, e.g. mark.define() method was called.
   */
  isDefined() {
    return !!_classPrivateFieldGet(_spec, this);
  }

  /**
   * Checks whether the mark is ignored by a given transformer.
   */
  isIgnored(transformerName) {
    var _classPrivateFieldGet2;
    if (!_classPrivateFieldGet(_spec, this)) {
      throw new Error('Mark is not defined');
    }
    return (_classPrivateFieldGet2 = _classPrivateFieldGet(_spec, this).ignore) === null || _classPrivateFieldGet2 === void 0 ? void 0 : _classPrivateFieldGet2.includes(transformerName);
  }

  /**
   * Define a mark.
   *
   * Assigns MarkSpec to the mark.
   */
  define(spec) {
    if (_classPrivateFieldGet(_spec, this)) {
      throw new Error('Cannot re-define a mark');
    }
    _classPrivateFieldSet(_spec, this, spec);
    return this;
  }
  getSpec() {
    // @ts-expect-error
    return _classPrivateFieldGet(_spec, this);
  }
  getType() {
    return _classPrivateFieldGet(_type, this);
  }

  /**
   * Returns a full name of the mark.
   */
  getName() {
    return `${_classPrivateFieldGet(_type, this)}_mark`;
  }
  setGroup(group) {
    if (_classPrivateFieldGet(_group, this)) {
      throw new Error('Cannot re-define a group');
    }
    _classPrivateFieldSet(_group, this, group);
    return this;
  }
  getGroup() {
    return _classPrivateFieldGet(_group, this);
  }
}
export function adfMark(type) {
  return new ADFMark(type);
}
