// @atlaskit/atlassian-context/dist/cjs/common/constants/domains.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.globalDomains = exports.fullDomainOverride = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _index = require("./index");
var globalDomains = exports.globalDomains = {
  confluence: 'confluence.atlassian.com',
  // Atlassian design system documentation
  design: 'design.atlassian.com',
  // Atlassian design system documentation
  extranet: 'extranet.atlassian.com',
  // Old staff intranet (now hello.atlassian.net)
  go: 'go.atlassian.com',
  // Go links
  my: 'my.atlassian.com',
  surveys: 'surveys.atlassian.com',
  'wac-cdn': 'wac-cdn.atlassian.com',
  // WAC CDN links
  integrations: 'integrations.atlassian.com',
  // Integrations links
  permalink: 'permalink.atlassian.com',
  support: 'support.atlassian.com'
};

/**
 * Contains overrides for domains where the pattern of subdomain only doesn't work.
 * Priority is exact match > fallback to commercial production
 *
 * IF YOU ADD SOMETHING HERE YOU MUST EXPLICITLY ADD EVERY PERIMETER AND EVERY ENVIRONMENT
 * ANYTHING NOT SET EXPLICITLY WILL DEFAULT TO COMMERCIAL/PRODUCTION
 */
var fullDomainOverride = exports.fullDomainOverride = {
  id: (0, _defineProperty2.default)((0, _defineProperty2.default)({}, _index.COMMERCIAL, (0, _defineProperty2.default)((0, _defineProperty2.default)((0, _defineProperty2.default)({}, _index.PRODUCTION, 'id.atlassian.com'), _index.STAGING, 'id.stg.internal.atlassian.com'), _index.DEV, 'id.dev.internal.atlassian.com')), _index.FEDRAMP_MODERATE, (0, _defineProperty2.default)((0, _defineProperty2.default)({}, _index.PRODUCTION, 'id.atlassian-us-gov-mod.com'), _index.STAGING, 'id.stg.atlassian-us-gov-mod.com'))
};
