// @atlaskit/adf-schema-generator/dist/cjs/transforms/adfToJson/nodeBuilder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildNode = void 0;
exports.buildNodeMarks = buildNodeMarks;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _transformerNames = require("../transformerNames");
var _attrBuilder = require("./attrBuilder");
var _contentBuilder = require("./contentBuilder");
var _nameResolver = require("./nameResolver");
var _requiredBuilder = require("./requiredBuilder");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var buildNode = exports.buildNode = function buildNode(node, content, fullSchema
// eslint-disable-next-line @typescript-eslint/no-explicit-any
) {
  var _nodeSpec$marks, _content$, _content$2;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  var json = {};
  var isStage0 = !fullSchema && node.hasStage0();
  var nodeSpec = node.getSpec(isStage0);
  var jsonContent = (0, _contentBuilder.buildContent)(content, node.getName(), nodeSpec, fullSchema);
  var nodeMarks = ((_nodeSpec$marks = nodeSpec.marks) === null || _nodeSpec$marks === void 0 ? void 0 : _nodeSpec$marks.map(function (mark) {
    if (mark.isIgnored(_transformerNames.JSONSchemaTransformerName)) {
      return;
    }
    return mark.getType();
  }).filter(Boolean)) || [];
  var marks = buildNodeMarks(nodeMarks, {
    hasNoMarks: !!nodeSpec.noMarks,
    hasEmptyMarks: !!nodeSpec.hasEmptyMarks,
    marksMaxItems: nodeSpec.marksMaxItems
  });
  var attrs = (0, _attrBuilder.buildAttrs)(node.getSpec().attrs);
  var version = nodeSpec.version;
  var jsonVersion = version ? {
    version: {
      enum: [version]
    }
  } : {};
  var required = (0, _requiredBuilder.buildRequired)(nodeSpec.attrs, ((_content$ = content[0]) === null || _content$ === void 0 ? void 0 : _content$.minItems) === 1 || Boolean((_content$2 = content[0]) === null || _content$2 === void 0 || (_content$2 = _content$2.range) === null || _content$2 === void 0 ? void 0 : _content$2.type), node.getName());
  if (node.getVariant() === 'base' || nodeSpec.noExtend || node.hasAttributeOverride()) {
    json.type = 'object';
    json.properties = _objectSpread(_objectSpread(_objectSpread(_objectSpread({
      type: {
        enum: [node.getType()]
      }
    }, marks), attrs), jsonContent), jsonVersion);
    json.additionalProperties = false;
    json.required = required;

    // Text is a special node in PM and it has a text property
    if (node.getName() === 'text') {
      json.properties.text = {
        minLength: 1,
        type: 'string'
      };
      json.required.push('text');
    }
  } else {
    var _required = (0, _requiredBuilder.buildVariantRequired)(content);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    var allOfItem = _objectSpread(_objectSpread({
      type: 'object',
      properties: _objectSpread(_objectSpread({}, marks), jsonContent)
    }, _required), {}, {
      additionalProperties: true
    });
    json.allOf = [{
      // @ts-expect-error
      $ref: "#/definitions/".concat((0, _nameResolver.resolveName)(node.getBase().getName()))
    }, allOfItem];
  }
  return json;
};
function buildNodeMarks(nodeMarks, opts) {
  var hasNoMarks = opts.hasNoMarks,
    hasEmptyMarks = opts.hasEmptyMarks,
    marksMaxItems = opts.marksMaxItems;
  if (hasNoMarks) {
    return {
      marks: {
        type: 'array',
        maxItems: marksMaxItems !== null && marksMaxItems !== void 0 ? marksMaxItems : 0
      }
    };
  }
  if (hasEmptyMarks) {
    return {
      marks: {
        type: 'array',
        maxItems: marksMaxItems
      }
    };
  }
  if (nodeMarks.length === 0) {
    return {};
  }
  if (nodeMarks.length === 1) {
    return {
      marks: {
        type: 'array',
        items: {
          $ref: "#/definitions/".concat(nodeMarks[0], "_mark")
        },
        maxItems: marksMaxItems
      }
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  var markDefinitions = [];
  nodeMarks.forEach(function (markName) {
    markDefinitions.push({
      $ref: "#/definitions/".concat(markName, "_mark")
    });
  });
  var jsonMarks = {
    marks: {
      type: 'array',
      items: {
        anyOf: markDefinitions
      },
      maxItems: marksMaxItems
    }
  };
  return jsonMarks;
}
