// @atlaskit/modal-dialog/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CloseButton", {
  enumerable: true,
  get: function get() {
    return _closeButton.CloseButton;
  }
});
Object.defineProperty(exports, "ModalBody", {
  enumerable: true,
  get: function get() {
    return _modalBody.default;
  }
});
Object.defineProperty(exports, "ModalFooter", {
  enumerable: true,
  get: function get() {
    return _modalFooter.default;
  }
});
Object.defineProperty(exports, "ModalHeader", {
  enumerable: true,
  get: function get() {
    return _modalHeader.default;
  }
});
Object.defineProperty(exports, "ModalTitle", {
  enumerable: true,
  get: function get() {
    return _modalTitle.default;
  }
});
Object.defineProperty(exports, "ModalTransition", {
  enumerable: true,
  get: function get() {
    return _modalTransition.default;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _modalDialog.default;
  }
});
Object.defineProperty(exports, "useModal", {
  enumerable: true,
  get: function get() {
    return _hooks.useModal;
  }
});
var _modalDialog = _interopRequireDefault(require("./modal-dialog"));
var _modalTransition = _interopRequireDefault(require("./modal-transition"));
var _modalHeader = _interopRequireDefault(require("./modal-header"));
var _modalTitle = _interopRequireDefault(require("./modal-title"));
var _modalBody = _interopRequireDefault(require("./modal-body"));
var _modalFooter = _interopRequireDefault(require("./modal-footer"));
var _closeButton = require("./close-button");
var _hooks = require("./hooks");
