// @atlaskit/react-ufo/dist/cjs/placeholder/loosely-lazy/lazy-suspense.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LazySuspense = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var isNodeEnvironment = function isNodeEnvironment() {
  return typeof window === 'undefined' || window.name === 'nodejs';
};
var LazySuspenseContext = /*#__PURE__*/(0, _react.createContext)({
  fallback: /*#__PURE__*/_react.default.createElement(_react.Fragment, null),
  setFallback: function setFallback() {
    // eslint-disable-next-line no-console
    console.warn('Missing <LooselySuspense /> boundary');
  },
  name: ''
});
var DynamicFallback = function DynamicFallback(_ref) {
  var children = _ref.children,
    outsideSuspense = _ref.outsideSuspense;
  var _React$useContext = _react.default.useContext(LazySuspenseContext),
    fallback = _React$useContext.fallback,
    setFallback = _React$useContext.setFallback;
  (0, _react.useLayoutEffect)(function () {
    return function () {
      if (!outsideSuspense) {
        setFallback(null);
      }
    };
  }, [outsideSuspense, setFallback]);
  return children(outsideSuspense ? null : fallback);
};
DynamicFallback.displayName = 'DynamicFallback';
var LazySuspense = exports.LazySuspense = function LazySuspense(_ref2) {
  var fallback = _ref2.fallback,
    children = _ref2.children,
    name = _ref2.name;
  var _useState = (0, _react.useState)(null),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    hydrationFallback = _useState2[0],
    setHydrationFallback = _useState2[1];
  var mountedRef = (0, _react.useRef)(false);
  (0, _react.useLayoutEffect)(function () {
    mountedRef.current = true;
    return function () {
      mountedRef.current = false;
    };
  }, []);
  var contextValue = (0, _react.useMemo)(function () {
    return {
      fallback: fallback,
      setFallback: function setFallback(newFallback) {
        if (hydrationFallback === newFallback) {
          return;
        }
        setHydrationFallback(newFallback);
      },
      name: name
    };
  }, [fallback, hydrationFallback, name]);
  var renderFallback = function renderFallback(outsideSuspense) {
    return /*#__PURE__*/_react.default.createElement(DynamicFallback, {
      outsideSuspense: outsideSuspense
    }, function (resolvedFallback) {
      return outsideSuspense && hydrationFallback ? hydrationFallback : resolvedFallback;
    });
  };
  return /*#__PURE__*/_react.default.createElement(LazySuspenseContext.Provider, {
    value: contextValue
  }, isNodeEnvironment() ? children : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: renderFallback(false)
  }, children), (!mountedRef.current || hydrationFallback) && renderFallback(true)));
};
LazySuspense.displayName = 'LazySuspense';
