// @atlaskit/navigation-system/dist/esm/context/skip-links/skip-links-context.js
import { createContext, useContext, useEffect } from 'react';
import noop from '@atlaskit/ds-lib/noop';
/**
 * Provides a way to register and unregister skip links
 */
export var SkipLinksContext = /*#__PURE__*/createContext({
  registerSkipLink: noop,
  unregisterSkipLink: noop
});
var useSkipLinks = function useSkipLinks() {
  return useContext(SkipLinksContext);
};

/**
 * Internal-only hook for registering skip links.
 *
 * `useSkipLink` is the public API wrapper of this.
 *
 * This private version exists for us to support `onBeforeNavigate` for the side nav use case,
 * where we might need to expand it before moving focus, without having to support `onBeforeNavigate` publicly.
 */
export var useSkipLinkInternal = function useSkipLinkInternal(_ref) {
  var id = _ref.id,
    label = _ref.label,
    listIndex = _ref.listIndex,
    onBeforeNavigate = _ref.onBeforeNavigate,
    isHidden = _ref.isHidden;
  var _useSkipLinks = useSkipLinks(),
    registerSkipLink = _useSkipLinks.registerSkipLink,
    unregisterSkipLink = _useSkipLinks.unregisterSkipLink;
  useEffect(function () {
    if (isHidden) {
      /**
       * Skip links are hidden for slots with 0 height or width.
       * They should not be registered.
       */
      return;
    }
    registerSkipLink({
      id: id,
      label: label,
      listIndex: listIndex,
      onBeforeNavigate: onBeforeNavigate,
      isHidden: isHidden
    });
    return function () {
      unregisterSkipLink(id);
    };
  }, [id, isHidden, label, listIndex, onBeforeNavigate, registerSkipLink, unregisterSkipLink]);
};

/**
 * Call `useSkipLink` to register a skip link for important elements on the page.
 */
export var useSkipLink = function useSkipLink(id, label, listIndex) {
  useSkipLinkInternal({
    id: id,
    label: label,
    listIndex: listIndex
  });
};
