// @atlaskit/feature-gate-js-client/dist/cjs/index.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CLIENT_VERSION", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.CLIENT_VERSION;
  }
});
Object.defineProperty(exports, "DynamicConfig", {
  enumerable: true,
  get: function get() {
    return _DynamicConfig.DynamicConfig;
  }
});
Object.defineProperty(exports, "EvaluationReason", {
  enumerable: true,
  get: function get() {
    return _types.EvaluationReason;
  }
});
Object.defineProperty(exports, "FeatureGateEnvironment", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.FeatureGateEnvironment;
  }
});
Object.defineProperty(exports, "Layer", {
  enumerable: true,
  get: function get() {
    return _Layer.Layer;
  }
});
Object.defineProperty(exports, "PerimeterType", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.PerimeterType;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _FeatureGates.default;
  }
});
var _FeatureGates = _interopRequireWildcard(require("./client/FeatureGates"));
var _DynamicConfig = require("./client/compat/DynamicConfig");
var _Layer = require("./client/compat/Layer");
var _types = require("./client/compat/types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
