// @atlaskit/react-select/dist/cjs/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultStyles = void 0;
exports.mergeStyles = mergeStyles;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _containers = require("./components/containers");
var _control = require("./components/control");
var _group = require("./components/group");
var _indicators = require("./components/indicators");
var _input = require("./components/input");
var _menu = require("./components/menu");
var _multiValue = require("./components/multi-value");
var _option = require("./components/option");
var _placeholder = require("./components/placeholder");
var _singleValue = require("./components/single-value");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var defaultStyles = exports.defaultStyles = {
  clearIndicator: _indicators.clearIndicatorCSS,
  container: _containers.containerCSS,
  control: _control.css,
  dropdownIndicator: _indicators.dropdownIndicatorCSS,
  group: _group.groupCSS,
  groupHeading: _group.groupHeadingCSS,
  indicatorsContainer: _containers.indicatorsContainerCSS,
  input: _input.inputCSS,
  loadingIndicator: _indicators.loadingIndicatorCSS,
  loadingMessage: _menu.loadingMessageCSS,
  menu: _menu.menuCSS,
  menuList: _menu.menuListCSS,
  menuPortal: _menu.menuPortalCSS,
  multiValue: _multiValue.multiValueCSS,
  multiValueLabel: _multiValue.multiValueLabelCSS,
  multiValueRemove: _multiValue.multiValueRemoveCSS,
  noOptionsMessage: _menu.noOptionsMessageCSS,
  option: _option.optionCSS,
  placeholder: _placeholder.placeholderCSS,
  singleValue: _singleValue.css,
  valueContainer: _containers.valueContainerCSS
};
// Merge Utility
// Allows consumers to extend a base Select with additional styles
function mergeStyles(source) {
  var target = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  // initialize with source styles
  var styles = _objectSpread({}, source);

  // massage in target styles
  Object.keys(target).forEach(function (keyAsString) {
    var key = keyAsString;
    if (source[key]) {
      styles[key] = function (rsCss, props) {
        return target[key](source[key](rsCss, props), props);
      };
    } else {
      styles[key] = target[key];
    }
  });
  return styles;
}
