// @atlaskit/layering/dist/cjs/classes/layer-node.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LayerNode = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var LayerNode = exports.LayerNode = /*#__PURE__*/function () {
  function LayerNode(id, parent) {
    (0, _classCallCheck2.default)(this, LayerNode);
    this.id = id;
    this.childNodes = [];
    this.parent = parent;
  }
  return (0, _createClass2.default)(LayerNode, [{
    key: "addChild",
    value: function addChild(node) {
      this.childNodes.push(node);
    }
  }, {
    key: "removeChild",
    value: function removeChild(node) {
      this.childNodes = this.childNodes.filter(function (child) {
        return child.id !== node.id;
      });
    }
  }, {
    key: "getLevel",
    value: function getLevel() {
      if (!this.parent) {
        return 1;
      }
      return this.parent.getLevel() + 1;
    }
  }, {
    key: "getHeight",
    value: function getHeight() {
      if (this.childNodes.length === 0) {
        return 1;
      }
      return Math.max.apply(Math, (0, _toConsumableArray2.default)(this.childNodes.map(function (child) {
        return child.getHeight();
      }))) + 1;
    }
  }]);
}();
