// @atlaskit/primitives/dist/es2019/xcss/style-maps.partial.js
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css } from '@emotion/react';
import { CURRENT_SURFACE_CSS_VAR } from '@atlaskit/tokens';
/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::df710bcd99c0a8704ca36bddb12c2154>>
 * @codegenId dimensions
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenParams ["dimensions"]
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/dimensions.tsx <<SignedSource::cc9b3f12104c6ede803da6a42daac0b0>>
 */
export const dimensionMap = {
  '100%': '100%',
  'size.100': '1rem',
  'size.200': '1.5rem',
  'size.300': '2rem',
  'size.400': '2.5rem',
  'size.500': '3rem',
  'size.600': '6rem',
  'size.1000': '12rem'
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::c9b68acdd57654884efd2f5087890624>>
 * @codegenId spacing
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenDependency ../../../tokens/src/artifacts/tokens-raw/atlassian-spacing.tsx <<SignedSource::535518e7add48ef24f526d0904f70060>>
 */
export const positiveSpaceMap = {
  'space.0': "var(--ds-space-0, 0px)",
  'space.025': "var(--ds-space-025, 2px)",
  'space.050': "var(--ds-space-050, 4px)",
  'space.075': "var(--ds-space-075, 6px)",
  'space.100': "var(--ds-space-100, 8px)",
  'space.150': "var(--ds-space-150, 12px)",
  'space.200': "var(--ds-space-200, 16px)",
  'space.250': "var(--ds-space-250, 20px)",
  'space.300': "var(--ds-space-300, 24px)",
  'space.400': "var(--ds-space-400, 32px)",
  'space.500': "var(--ds-space-500, 40px)",
  'space.600': "var(--ds-space-600, 48px)",
  'space.800': "var(--ds-space-800, 64px)",
  'space.1000': "var(--ds-space-1000, 80px)"
};
export const negativeSpaceMap = {
  'space.negative.025': "var(--ds-space-negative-025, -2px)",
  'space.negative.050': "var(--ds-space-negative-050, -4px)",
  'space.negative.075': "var(--ds-space-negative-075, -6px)",
  'space.negative.100': "var(--ds-space-negative-100, -8px)",
  'space.negative.150': "var(--ds-space-negative-150, -12px)",
  'space.negative.200': "var(--ds-space-negative-200, -16px)",
  'space.negative.250': "var(--ds-space-negative-250, -20px)",
  'space.negative.300': "var(--ds-space-negative-300, -24px)",
  'space.negative.400': "var(--ds-space-negative-400, -32px)"
};
export const allSpaceMap = {
  ...positiveSpaceMap,
  ...negativeSpaceMap
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::cefa2c59e15e6a3749ef38fa86c6adea>>
 * @codegenId inverse-colors
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenDependency ../../../tokens/src/artifacts/tokens-raw/atlassian-light.tsx <<SignedSource::38a7d4716f6999a6bdda9e4fe2bca6a1>>
 */
export const inverseColorMap = {
  'color.background.neutral.bold': 'color.text.inverse',
  'color.background.neutral.bold.hovered': 'color.text.inverse',
  'color.background.neutral.bold.pressed': 'color.text.inverse',
  'color.background.selected.bold': 'color.text.inverse',
  'color.background.selected.bold.hovered': 'color.text.inverse',
  'color.background.selected.bold.pressed': 'color.text.inverse',
  'color.background.brand.bold': 'color.text.inverse',
  'color.background.brand.bold.hovered': 'color.text.inverse',
  'color.background.brand.bold.pressed': 'color.text.inverse',
  'color.background.brand.boldest': 'color.text.inverse',
  'color.background.brand.boldest.hovered': 'color.text.inverse',
  'color.background.brand.boldest.pressed': 'color.text.inverse',
  'color.background.danger.bold': 'color.text.inverse',
  'color.background.danger.bold.hovered': 'color.text.inverse',
  'color.background.danger.bold.pressed': 'color.text.inverse',
  'color.background.warning.bold': 'color.text.warning.inverse',
  'color.background.warning.bold.hovered': 'color.text.warning.inverse',
  'color.background.warning.bold.pressed': 'color.text.warning.inverse',
  'color.background.success.bold': 'color.text.inverse',
  'color.background.success.bold.hovered': 'color.text.inverse',
  'color.background.success.bold.pressed': 'color.text.inverse',
  'color.background.discovery.bold': 'color.text.inverse',
  'color.background.discovery.bold.hovered': 'color.text.inverse',
  'color.background.discovery.bold.pressed': 'color.text.inverse',
  'color.background.information.bold': 'color.text.inverse',
  'color.background.information.bold.hovered': 'color.text.inverse',
  'color.background.information.bold.pressed': 'color.text.inverse'
};

/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::c673d53a0a31524e65f1c7051579eae5>>
 * @codegenId elevation
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenParams ["opacity", "shadow", "surface"]
 * @codegenDependency ../../../tokens/src/artifacts/tokens-raw/atlassian-light.tsx <<SignedSource::38a7d4716f6999a6bdda9e4fe2bca6a1>>
 */
export const opacityMap = {
  'opacity.disabled': "var(--ds-opacity-disabled, 0.4)",
  'opacity.loading': "var(--ds-opacity-loading, 0.2)"
};
export const shadowMap = {
  'elevation.shadow.overflow': "var(--ds-shadow-overflow, 0px 0px 8px #1E1F21, 0px 0px 1px #1E1F21)",
  'elevation.shadow.overflow.perimeter': "var(--ds-shadow-overflow-perimeter, #1E1F211f)",
  'elevation.shadow.overflow.spread': "var(--ds-shadow-overflow-spread, #1E1F2129)",
  'elevation.shadow.overlay': "var(--ds-shadow-overlay, 0px 8px 12px #1E1F21, 0px 0px 1px #1E1F21)",
  'elevation.shadow.raised': "var(--ds-shadow-raised, 0px 1px 1px #1E1F21, 0px 0px 1px #1E1F21)"
};
export const surfaceColorMap = {
  'elevation.surface': "var(--ds-surface, #FFFFFF)",
  'elevation.surface.hovered': "var(--ds-surface-hovered, #F0F1F2)",
  'elevation.surface.pressed': "var(--ds-surface-pressed, #DDDEE1)",
  'elevation.surface.overlay': "var(--ds-surface-overlay, #FFFFFF)",
  'elevation.surface.overlay.hovered': "var(--ds-surface-overlay-hovered, #F0F1F2)",
  'elevation.surface.overlay.pressed': "var(--ds-surface-overlay-pressed, #DDDEE1)",
  'elevation.surface.raised': "var(--ds-surface-raised, #FFFFFF)",
  'elevation.surface.raised.hovered': "var(--ds-surface-raised-hovered, #F0F1F2)",
  'elevation.surface.raised.pressed': "var(--ds-surface-raised-pressed, #DDDEE1)",
  'elevation.surface.sunken': "var(--ds-surface-sunken, #F8F8F8)"
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::fa88ed1ec5a817750b0348ca36dbb5f1>>
 * @codegenId colors
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenParams ["border", "background", "text", "fill"]
 * @codegenDependency ../../../tokens/src/artifacts/tokens-raw/atlassian-light.tsx <<SignedSource::38a7d4716f6999a6bdda9e4fe2bca6a1>>
 */
export const borderColorMap = {
  'color.border': "var(--ds-border, #0B120E24)",
  'color.border.accent.lime': "var(--ds-border-accent-lime, #6A9A23)",
  'color.border.accent.red': "var(--ds-border-accent-red, #E2483D)",
  'color.border.accent.orange': "var(--ds-border-accent-orange, #E06C00)",
  'color.border.accent.yellow': "var(--ds-border-accent-yellow, #B38600)",
  'color.border.accent.green': "var(--ds-border-accent-green, #22A06B)",
  'color.border.accent.teal': "var(--ds-border-accent-teal, #2898BD)",
  'color.border.accent.blue': "var(--ds-border-accent-blue, #357DE8)",
  'color.border.accent.purple': "var(--ds-border-accent-purple, #AF59E1)",
  'color.border.accent.magenta': "var(--ds-border-accent-magenta, #CD519D)",
  'color.border.accent.gray': "var(--ds-border-accent-gray, #7D818A)",
  'color.border.disabled': "var(--ds-border-disabled, #0515240F)",
  'color.border.focused': "var(--ds-border-focused, #4688EC)",
  'color.border.input': "var(--ds-border-input, #8C8F97)",
  'color.border.inverse': "var(--ds-border-inverse, #FFFFFF)",
  'color.border.selected': "var(--ds-border-selected, #1868DB)",
  'color.border.brand': "var(--ds-border-brand, #1868DB)",
  'color.border.danger': "var(--ds-border-danger, #E2483D)",
  'color.border.warning': "var(--ds-border-warning, #E06C00)",
  'color.border.success': "var(--ds-border-success, #6A9A23)",
  'color.border.discovery': "var(--ds-border-discovery, #AF59E1)",
  'color.border.information': "var(--ds-border-information, #357DE8)",
  'color.border.bold': "var(--ds-border-bold, #7D818A)"
};
export const backgroundColorMap = {
  'color.background.accent.lime.subtlest': "var(--ds-background-accent-lime-subtlest, #EFFFD6)",
  'color.background.accent.lime.subtlest.hovered': "var(--ds-background-accent-lime-subtlest-hovered, #D3F1A7)",
  'color.background.accent.lime.subtlest.pressed': "var(--ds-background-accent-lime-subtlest-pressed, #BDE97C)",
  'color.background.accent.lime.subtler': "var(--ds-background-accent-lime-subtler, #D3F1A7)",
  'color.background.accent.lime.subtler.hovered': "var(--ds-background-accent-lime-subtler-hovered, #BDE97C)",
  'color.background.accent.lime.subtler.pressed': "var(--ds-background-accent-lime-subtler-pressed, #B3DF72)",
  'color.background.accent.lime.subtle': "var(--ds-background-accent-lime-subtle, #94C748)",
  'color.background.accent.lime.subtle.hovered': "var(--ds-background-accent-lime-subtle-hovered, #B3DF72)",
  'color.background.accent.lime.subtle.pressed': "var(--ds-background-accent-lime-subtle-pressed, #BDE97C)",
  'color.background.accent.lime.bolder': "var(--ds-background-accent-lime-bolder, #5B7F24)",
  'color.background.accent.lime.bolder.hovered': "var(--ds-background-accent-lime-bolder-hovered, #4C6B1F)",
  'color.background.accent.lime.bolder.pressed': "var(--ds-background-accent-lime-bolder-pressed, #3F5224)",
  'color.background.accent.red.subtlest': "var(--ds-background-accent-red-subtlest, #FFECEB)",
  'color.background.accent.red.subtlest.hovered': "var(--ds-background-accent-red-subtlest-hovered, #FFD5D2)",
  'color.background.accent.red.subtlest.pressed': "var(--ds-background-accent-red-subtlest-pressed, #FFB8B2)",
  'color.background.accent.red.subtler': "var(--ds-background-accent-red-subtler, #FFD5D2)",
  'color.background.accent.red.subtler.hovered': "var(--ds-background-accent-red-subtler-hovered, #FFB8B2)",
  'color.background.accent.red.subtler.pressed': "var(--ds-background-accent-red-subtler-pressed, #FD9891)",
  'color.background.accent.red.subtle': "var(--ds-background-accent-red-subtle, #F87168)",
  'color.background.accent.red.subtle.hovered': "var(--ds-background-accent-red-subtle-hovered, #FD9891)",
  'color.background.accent.red.subtle.pressed': "var(--ds-background-accent-red-subtle-pressed, #FFB8B2)",
  'color.background.accent.red.bolder': "var(--ds-background-accent-red-bolder, #C9372C)",
  'color.background.accent.red.bolder.hovered': "var(--ds-background-accent-red-bolder-hovered, #AE2E24)",
  'color.background.accent.red.bolder.pressed': "var(--ds-background-accent-red-bolder-pressed, #872821)",
  'color.background.accent.orange.subtlest': "var(--ds-background-accent-orange-subtlest, #FFF5DB)",
  'color.background.accent.orange.subtlest.hovered': "var(--ds-background-accent-orange-subtlest-hovered, #FCE4A6)",
  'color.background.accent.orange.subtlest.pressed': "var(--ds-background-accent-orange-subtlest-pressed, #FBD779)",
  'color.background.accent.orange.subtler': "var(--ds-background-accent-orange-subtler, #FCE4A6)",
  'color.background.accent.orange.subtler.hovered': "var(--ds-background-accent-orange-subtler-hovered, #FBD779)",
  'color.background.accent.orange.subtler.pressed': "var(--ds-background-accent-orange-subtler-pressed, #FBC828)",
  'color.background.accent.orange.subtle': "var(--ds-background-accent-orange-subtle, #FCA700)",
  'color.background.accent.orange.subtle.hovered': "var(--ds-background-accent-orange-subtle-hovered, #FBC828)",
  'color.background.accent.orange.subtle.pressed': "var(--ds-background-accent-orange-subtle-pressed, #FBD779)",
  'color.background.accent.orange.bolder': "var(--ds-background-accent-orange-bolder, #BD5B00)",
  'color.background.accent.orange.bolder.hovered': "var(--ds-background-accent-orange-bolder-hovered, #9E4C00)",
  'color.background.accent.orange.bolder.pressed': "var(--ds-background-accent-orange-bolder-pressed, #7A3B00)",
  'color.background.accent.yellow.subtlest': "var(--ds-background-accent-yellow-subtlest, #FEF7C8)",
  'color.background.accent.yellow.subtlest.hovered': "var(--ds-background-accent-yellow-subtlest-hovered, #F5E989)",
  'color.background.accent.yellow.subtlest.pressed': "var(--ds-background-accent-yellow-subtlest-pressed, #EFDD4E)",
  'color.background.accent.yellow.subtler': "var(--ds-background-accent-yellow-subtler, #F5E989)",
  'color.background.accent.yellow.subtler.hovered': "var(--ds-background-accent-yellow-subtler-hovered, #EFDD4E)",
  'color.background.accent.yellow.subtler.pressed': "var(--ds-background-accent-yellow-subtler-pressed, #EED12B)",
  'color.background.accent.yellow.subtle': "var(--ds-background-accent-yellow-subtle, #EED12B)",
  'color.background.accent.yellow.subtle.hovered': "var(--ds-background-accent-yellow-subtle-hovered, #DDB30E)",
  'color.background.accent.yellow.subtle.pressed': "var(--ds-background-accent-yellow-subtle-pressed, #EFDD4E)",
  'color.background.accent.yellow.bolder': "var(--ds-background-accent-yellow-bolder, #946F00)",
  'color.background.accent.yellow.bolder.hovered': "var(--ds-background-accent-yellow-bolder-hovered, #7F5F01)",
  'color.background.accent.yellow.bolder.pressed': "var(--ds-background-accent-yellow-bolder-pressed, #614A05)",
  'color.background.accent.green.subtlest': "var(--ds-background-accent-green-subtlest, #DCFFF1)",
  'color.background.accent.green.subtlest.hovered': "var(--ds-background-accent-green-subtlest-hovered, #BAF3DB)",
  'color.background.accent.green.subtlest.pressed': "var(--ds-background-accent-green-subtlest-pressed, #97EDC9)",
  'color.background.accent.green.subtler': "var(--ds-background-accent-green-subtler, #BAF3DB)",
  'color.background.accent.green.subtler.hovered': "var(--ds-background-accent-green-subtler-hovered, #97EDC9)",
  'color.background.accent.green.subtler.pressed': "var(--ds-background-accent-green-subtler-pressed, #7EE2B8)",
  'color.background.accent.green.subtle': "var(--ds-background-accent-green-subtle, #4BCE97)",
  'color.background.accent.green.subtle.hovered': "var(--ds-background-accent-green-subtle-hovered, #7EE2B8)",
  'color.background.accent.green.subtle.pressed': "var(--ds-background-accent-green-subtle-pressed, #97EDC9)",
  'color.background.accent.green.bolder': "var(--ds-background-accent-green-bolder, #1F845A)",
  'color.background.accent.green.bolder.hovered': "var(--ds-background-accent-green-bolder-hovered, #216E4E)",
  'color.background.accent.green.bolder.pressed': "var(--ds-background-accent-green-bolder-pressed, #19573D)",
  'color.background.accent.teal.subtlest': "var(--ds-background-accent-teal-subtlest, #E7F9FF)",
  'color.background.accent.teal.subtlest.hovered': "var(--ds-background-accent-teal-subtlest-hovered, #C6EDFB)",
  'color.background.accent.teal.subtlest.pressed': "var(--ds-background-accent-teal-subtlest-pressed, #B1E4F7)",
  'color.background.accent.teal.subtler': "var(--ds-background-accent-teal-subtler, #C6EDFB)",
  'color.background.accent.teal.subtler.hovered': "var(--ds-background-accent-teal-subtler-hovered, #B1E4F7)",
  'color.background.accent.teal.subtler.pressed': "var(--ds-background-accent-teal-subtler-pressed, #9DD9EE)",
  'color.background.accent.teal.subtle': "var(--ds-background-accent-teal-subtle, #6CC3E0)",
  'color.background.accent.teal.subtle.hovered': "var(--ds-background-accent-teal-subtle-hovered, #9DD9EE)",
  'color.background.accent.teal.subtle.pressed': "var(--ds-background-accent-teal-subtle-pressed, #B1E4F7)",
  'color.background.accent.teal.bolder': "var(--ds-background-accent-teal-bolder, #227D9B)",
  'color.background.accent.teal.bolder.hovered': "var(--ds-background-accent-teal-bolder-hovered, #206A83)",
  'color.background.accent.teal.bolder.pressed': "var(--ds-background-accent-teal-bolder-pressed, #1A5265)",
  'color.background.accent.blue.subtlest': "var(--ds-background-accent-blue-subtlest, #E9F2FE)",
  'color.background.accent.blue.subtlest.hovered': "var(--ds-background-accent-blue-subtlest-hovered, #CFE1FD)",
  'color.background.accent.blue.subtlest.pressed': "var(--ds-background-accent-blue-subtlest-pressed, #ADCBFB)",
  'color.background.accent.blue.subtler': "var(--ds-background-accent-blue-subtler, #CFE1FD)",
  'color.background.accent.blue.subtler.hovered': "var(--ds-background-accent-blue-subtler-hovered, #ADCBFB)",
  'color.background.accent.blue.subtler.pressed': "var(--ds-background-accent-blue-subtler-pressed, #8FB8F6)",
  'color.background.accent.blue.subtle': "var(--ds-background-accent-blue-subtle, #669DF1)",
  'color.background.accent.blue.subtle.hovered': "var(--ds-background-accent-blue-subtle-hovered, #8FB8F6)",
  'color.background.accent.blue.subtle.pressed': "var(--ds-background-accent-blue-subtle-pressed, #ADCBFB)",
  'color.background.accent.blue.bolder': "var(--ds-background-accent-blue-bolder, #1868DB)",
  'color.background.accent.blue.bolder.hovered': "var(--ds-background-accent-blue-bolder-hovered, #1558BC)",
  'color.background.accent.blue.bolder.pressed': "var(--ds-background-accent-blue-bolder-pressed, #144794)",
  'color.background.accent.purple.subtlest': "var(--ds-background-accent-purple-subtlest, #F8EEFE)",
  'color.background.accent.purple.subtlest.hovered': "var(--ds-background-accent-purple-subtlest-hovered, #EED7FC)",
  'color.background.accent.purple.subtlest.pressed': "var(--ds-background-accent-purple-subtlest-pressed, #E3BDFA)",
  'color.background.accent.purple.subtler': "var(--ds-background-accent-purple-subtler, #EED7FC)",
  'color.background.accent.purple.subtler.hovered': "var(--ds-background-accent-purple-subtler-hovered, #E3BDFA)",
  'color.background.accent.purple.subtler.pressed': "var(--ds-background-accent-purple-subtler-pressed, #D8A0F7)",
  'color.background.accent.purple.subtle': "var(--ds-background-accent-purple-subtle, #C97CF4)",
  'color.background.accent.purple.subtle.hovered': "var(--ds-background-accent-purple-subtle-hovered, #D8A0F7)",
  'color.background.accent.purple.subtle.pressed': "var(--ds-background-accent-purple-subtle-pressed, #E3BDFA)",
  'color.background.accent.purple.bolder': "var(--ds-background-accent-purple-bolder, #964AC0)",
  'color.background.accent.purple.bolder.hovered': "var(--ds-background-accent-purple-bolder-hovered, #803FA5)",
  'color.background.accent.purple.bolder.pressed': "var(--ds-background-accent-purple-bolder-pressed, #673286)",
  'color.background.accent.magenta.subtlest': "var(--ds-background-accent-magenta-subtlest, #FFECF8)",
  'color.background.accent.magenta.subtlest.hovered': "var(--ds-background-accent-magenta-subtlest-hovered, #FDD0EC)",
  'color.background.accent.magenta.subtlest.pressed': "var(--ds-background-accent-magenta-subtlest-pressed, #FCB6E1)",
  'color.background.accent.magenta.subtler': "var(--ds-background-accent-magenta-subtler, #FDD0EC)",
  'color.background.accent.magenta.subtler.hovered': "var(--ds-background-accent-magenta-subtler-hovered, #FCB6E1)",
  'color.background.accent.magenta.subtler.pressed': "var(--ds-background-accent-magenta-subtler-pressed, #F797D2)",
  'color.background.accent.magenta.subtle': "var(--ds-background-accent-magenta-subtle, #E774BB)",
  'color.background.accent.magenta.subtle.hovered': "var(--ds-background-accent-magenta-subtle-hovered, #F797D2)",
  'color.background.accent.magenta.subtle.pressed': "var(--ds-background-accent-magenta-subtle-pressed, #FCB6E1)",
  'color.background.accent.magenta.bolder': "var(--ds-background-accent-magenta-bolder, #AE4787)",
  'color.background.accent.magenta.bolder.hovered': "var(--ds-background-accent-magenta-bolder-hovered, #943D73)",
  'color.background.accent.magenta.bolder.pressed': "var(--ds-background-accent-magenta-bolder-pressed, #77325B)",
  'color.background.accent.gray.subtlest': "var(--ds-background-accent-gray-subtlest, #F0F1F2)",
  'color.background.accent.gray.subtlest.hovered': "var(--ds-background-accent-gray-subtlest-hovered, #DDDEE1)",
  'color.background.accent.gray.subtlest.pressed': "var(--ds-background-accent-gray-subtlest-pressed, #B7B9BE)",
  'color.background.accent.gray.subtler': "var(--ds-background-accent-gray-subtler, #DDDEE1)",
  'color.background.accent.gray.subtler.hovered': "var(--ds-background-accent-gray-subtler-hovered, #B7B9BE)",
  'color.background.accent.gray.subtler.pressed': "var(--ds-background-accent-gray-subtler-pressed, #8C8F97)",
  'color.background.accent.gray.subtle': "var(--ds-background-accent-gray-subtle, #8C8F97)",
  'color.background.accent.gray.subtle.hovered': "var(--ds-background-accent-gray-subtle-hovered, #B7B9BE)",
  'color.background.accent.gray.subtle.pressed': "var(--ds-background-accent-gray-subtle-pressed, #DDDEE1)",
  'color.background.accent.gray.bolder': "var(--ds-background-accent-gray-bolder, #6B6E76)",
  'color.background.accent.gray.bolder.hovered': "var(--ds-background-accent-gray-bolder-hovered, #505258)",
  'color.background.accent.gray.bolder.pressed': "var(--ds-background-accent-gray-bolder-pressed, #3B3D42)",
  'color.background.disabled': "var(--ds-background-disabled, #17171708)",
  'color.background.input': "var(--ds-background-input, #FFFFFF)",
  'color.background.input.hovered': "var(--ds-background-input-hovered, #F8F8F8)",
  'color.background.input.pressed': "var(--ds-background-input-pressed, #FFFFFF)",
  'color.background.inverse.subtle': "var(--ds-background-inverse-subtle, #00000029)",
  'color.background.inverse.subtle.hovered': "var(--ds-background-inverse-subtle-hovered, #0000003D)",
  'color.background.inverse.subtle.pressed': "var(--ds-background-inverse-subtle-pressed, #00000052)",
  'color.background.neutral': "var(--ds-background-neutral, #0515240F)",
  'color.background.neutral.hovered': "var(--ds-background-neutral-hovered, #0B120E24)",
  'color.background.neutral.pressed': "var(--ds-background-neutral-pressed, #080F214A)",
  'color.background.neutral.subtle': "var(--ds-background-neutral-subtle, #00000000)",
  'color.background.neutral.subtle.hovered': "var(--ds-background-neutral-subtle-hovered, #0515240F)",
  'color.background.neutral.subtle.pressed': "var(--ds-background-neutral-subtle-pressed, #0B120E24)",
  'color.background.neutral.bold': "var(--ds-background-neutral-bold, #292A2E)",
  'color.background.neutral.bold.hovered': "var(--ds-background-neutral-bold-hovered, #3B3D42)",
  'color.background.neutral.bold.pressed': "var(--ds-background-neutral-bold-pressed, #505258)",
  'color.background.selected': "var(--ds-background-selected, #E9F2FE)",
  'color.background.selected.hovered': "var(--ds-background-selected-hovered, #CFE1FD)",
  'color.background.selected.pressed': "var(--ds-background-selected-pressed, #8FB8F6)",
  'color.background.selected.bold': "var(--ds-background-selected-bold, #1868DB)",
  'color.background.selected.bold.hovered': "var(--ds-background-selected-bold-hovered, #1558BC)",
  'color.background.selected.bold.pressed': "var(--ds-background-selected-bold-pressed, #123263)",
  'color.background.brand.subtlest': "var(--ds-background-brand-subtlest, #E9F2FE)",
  'color.background.brand.subtlest.hovered': "var(--ds-background-brand-subtlest-hovered, #CFE1FD)",
  'color.background.brand.subtlest.pressed': "var(--ds-background-brand-subtlest-pressed, #ADCBFB)",
  'color.background.brand.bold': "var(--ds-background-brand-bold, #1868DB)",
  'color.background.brand.bold.hovered': "var(--ds-background-brand-bold-hovered, #1558BC)",
  'color.background.brand.bold.pressed': "var(--ds-background-brand-bold-pressed, #144794)",
  'color.background.brand.boldest': "var(--ds-background-brand-boldest, #1C2B42)",
  'color.background.brand.boldest.hovered': "var(--ds-background-brand-boldest-hovered, #123263)",
  'color.background.brand.boldest.pressed': "var(--ds-background-brand-boldest-pressed, #144794)",
  'color.background.danger': "var(--ds-background-danger, #FFECEB)",
  'color.background.danger.hovered': "var(--ds-background-danger-hovered, #FFD5D2)",
  'color.background.danger.pressed': "var(--ds-background-danger-pressed, #FFB8B2)",
  'color.background.danger.subtler': "var(--ds-background-danger-subtler, #FFD5D2)",
  'color.background.danger.subtler.hovered': "var(--ds-background-danger-subtler-hovered, #FFB8B2)",
  'color.background.danger.subtler.pressed': "var(--ds-background-danger-subtler-pressed, #FD9891)",
  'color.background.danger.bold': "var(--ds-background-danger-bold, #C9372C)",
  'color.background.danger.bold.hovered': "var(--ds-background-danger-bold-hovered, #AE2E24)",
  'color.background.danger.bold.pressed': "var(--ds-background-danger-bold-pressed, #872821)",
  'color.background.warning': "var(--ds-background-warning, #FFF5DB)",
  'color.background.warning.hovered': "var(--ds-background-warning-hovered, #FCE4A6)",
  'color.background.warning.pressed': "var(--ds-background-warning-pressed, #FBD779)",
  'color.background.warning.subtler': "var(--ds-background-warning-subtler, #FCE4A6)",
  'color.background.warning.subtler.hovered': "var(--ds-background-warning-subtler-hovered, #FBD779)",
  'color.background.warning.subtler.pressed': "var(--ds-background-warning-subtler-pressed, #FBC828)",
  'color.background.warning.bold': "var(--ds-background-warning-bold, #FBC828)",
  'color.background.warning.bold.hovered': "var(--ds-background-warning-bold-hovered, #FCA700)",
  'color.background.warning.bold.pressed': "var(--ds-background-warning-bold-pressed, #F68909)",
  'color.background.success': "var(--ds-background-success, #EFFFD6)",
  'color.background.success.hovered': "var(--ds-background-success-hovered, #D3F1A7)",
  'color.background.success.pressed': "var(--ds-background-success-pressed, #BDE97C)",
  'color.background.success.subtler': "var(--ds-background-success-subtler, #D3F1A7)",
  'color.background.success.subtler.hovered': "var(--ds-background-success-subtler-hovered, #BDE97C)",
  'color.background.success.subtler.pressed': "var(--ds-background-success-subtler-pressed, #B3DF72)",
  'color.background.success.bold': "var(--ds-background-success-bold, #5B7F24)",
  'color.background.success.bold.hovered': "var(--ds-background-success-bold-hovered, #4C6B1F)",
  'color.background.success.bold.pressed': "var(--ds-background-success-bold-pressed, #3F5224)",
  'color.background.discovery': "var(--ds-background-discovery, #F8EEFE)",
  'color.background.discovery.hovered': "var(--ds-background-discovery-hovered, #EED7FC)",
  'color.background.discovery.pressed': "var(--ds-background-discovery-pressed, #E3BDFA)",
  'color.background.discovery.subtler': "var(--ds-background-discovery-subtler, #EED7FC)",
  'color.background.discovery.subtler.hovered': "var(--ds-background-discovery-subtler-hovered, #E3BDFA)",
  'color.background.discovery.subtler.pressed': "var(--ds-background-discovery-subtler-pressed, #D8A0F7)",
  'color.background.discovery.bold': "var(--ds-background-discovery-bold, #964AC0)",
  'color.background.discovery.bold.hovered': "var(--ds-background-discovery-bold-hovered, #803FA5)",
  'color.background.discovery.bold.pressed': "var(--ds-background-discovery-bold-pressed, #673286)",
  'color.background.information': "var(--ds-background-information, #E9F2FE)",
  'color.background.information.hovered': "var(--ds-background-information-hovered, #CFE1FD)",
  'color.background.information.pressed': "var(--ds-background-information-pressed, #ADCBFB)",
  'color.background.information.subtler': "var(--ds-background-information-subtler, #CFE1FD)",
  'color.background.information.subtler.hovered': "var(--ds-background-information-subtler-hovered, #ADCBFB)",
  'color.background.information.subtler.pressed': "var(--ds-background-information-subtler-pressed, #8FB8F6)",
  'color.background.information.bold': "var(--ds-background-information-bold, #1868DB)",
  'color.background.information.bold.hovered': "var(--ds-background-information-bold-hovered, #1558BC)",
  'color.background.information.bold.pressed': "var(--ds-background-information-bold-pressed, #144794)",
  'color.blanket': "var(--ds-blanket, #050C1F75)",
  'color.blanket.selected': "var(--ds-blanket-selected, #388BFF14)",
  'color.blanket.danger': "var(--ds-blanket-danger, #EF5C4814)",
  'color.skeleton': "var(--ds-skeleton, #0515240F)",
  'color.skeleton.subtle': "var(--ds-skeleton-subtle, #17171708)",
  'elevation.surface': "var(--ds-surface, #FFFFFF)",
  'elevation.surface.hovered': "var(--ds-surface-hovered, #F0F1F2)",
  'elevation.surface.pressed': "var(--ds-surface-pressed, #DDDEE1)",
  'elevation.surface.overlay': "var(--ds-surface-overlay, #FFFFFF)",
  'elevation.surface.overlay.hovered': "var(--ds-surface-overlay-hovered, #F0F1F2)",
  'elevation.surface.overlay.pressed': "var(--ds-surface-overlay-pressed, #DDDEE1)",
  'elevation.surface.raised': "var(--ds-surface-raised, #FFFFFF)",
  'elevation.surface.raised.hovered': "var(--ds-surface-raised-hovered, #F0F1F2)",
  'elevation.surface.raised.pressed': "var(--ds-surface-raised-pressed, #DDDEE1)",
  'elevation.surface.sunken': "var(--ds-surface-sunken, #F8F8F8)",
  'utility.elevation.surface.current': "var(--ds-elevation-surface-current, #FFFFFF)"
};
export const textColorMap = {
  'color.text': "var(--ds-text, #292A2E)",
  'color.text.accent.lime': "var(--ds-text-accent-lime, #4C6B1F)",
  'color.text.accent.lime.bolder': "var(--ds-text-accent-lime-bolder, #37471F)",
  'color.text.accent.red': "var(--ds-text-accent-red, #AE2E24)",
  'color.text.accent.red.bolder': "var(--ds-text-accent-red-bolder, #5D1F1A)",
  'color.text.accent.orange': "var(--ds-text-accent-orange, #9E4C00)",
  'color.text.accent.orange.bolder': "var(--ds-text-accent-orange-bolder, #693200)",
  'color.text.accent.yellow': "var(--ds-text-accent-yellow, #7F5F01)",
  'color.text.accent.yellow.bolder': "var(--ds-text-accent-yellow-bolder, #533F04)",
  'color.text.accent.green': "var(--ds-text-accent-green, #216E4E)",
  'color.text.accent.green.bolder': "var(--ds-text-accent-green-bolder, #164B35)",
  'color.text.accent.teal': "var(--ds-text-accent-teal, #206A83)",
  'color.text.accent.teal.bolder': "var(--ds-text-accent-teal-bolder, #164555)",
  'color.text.accent.blue': "var(--ds-text-accent-blue, #1558BC)",
  'color.text.accent.blue.bolder': "var(--ds-text-accent-blue-bolder, #123263)",
  'color.text.accent.purple': "var(--ds-text-accent-purple, #803FA5)",
  'color.text.accent.purple.bolder': "var(--ds-text-accent-purple-bolder, #48245D)",
  'color.text.accent.magenta': "var(--ds-text-accent-magenta, #943D73)",
  'color.text.accent.magenta.bolder': "var(--ds-text-accent-magenta-bolder, #50253F)",
  'color.text.accent.gray': "var(--ds-text-accent-gray, #505258)",
  'color.text.accent.gray.bolder': "var(--ds-text-accent-gray-bolder, #1E1F21)",
  'color.text.disabled': "var(--ds-text-disabled, #080F214A)",
  'color.text.inverse': "var(--ds-text-inverse, #FFFFFF)",
  'color.text.selected': "var(--ds-text-selected, #1868DB)",
  'color.text.brand': "var(--ds-text-brand, #1868DB)",
  'color.text.danger': "var(--ds-text-danger, #AE2E24)",
  'color.text.danger.bolder': "var(--ds-text-danger-bolder, #5D1F1A)",
  'color.text.warning': "var(--ds-text-warning, #9E4C00)",
  'color.text.warning.inverse': "var(--ds-text-warning-inverse, #292A2E)",
  'color.text.warning.bolder': "var(--ds-text-warning-bolder, #693200)",
  'color.text.success': "var(--ds-text-success, #4C6B1F)",
  'color.text.success.bolder': "var(--ds-text-success-bolder, #37471F)",
  'color.text.discovery': "var(--ds-text-discovery, #803FA5)",
  'color.text.discovery.bolder': "var(--ds-text-discovery-bolder, #48245D)",
  'color.text.information': "var(--ds-text-information, #1558BC)",
  'color.text.information.bolder': "var(--ds-text-information-bolder, #123263)",
  'color.text.subtlest': "var(--ds-text-subtlest, #6B6E76)",
  'color.text.subtle': "var(--ds-text-subtle, #505258)",
  'color.link': "var(--ds-link, #1868DB)",
  'color.link.pressed': "var(--ds-link-pressed, #1558BC)",
  'color.link.visited': "var(--ds-link-visited, #803FA5)",
  'color.link.visited.pressed': "var(--ds-link-visited-pressed, #48245D)"
};
export const fillMap = {
  'color.icon': "var(--ds-icon, #292A2E)",
  'color.icon.accent.lime': "var(--ds-icon-accent-lime, #6A9A23)",
  'color.icon.accent.red': "var(--ds-icon-accent-red, #C9372C)",
  'color.icon.accent.orange': "var(--ds-icon-accent-orange, #E06C00)",
  'color.icon.accent.yellow': "var(--ds-icon-accent-yellow, #B38600)",
  'color.icon.accent.green': "var(--ds-icon-accent-green, #22A06B)",
  'color.icon.accent.teal': "var(--ds-icon-accent-teal, #2898BD)",
  'color.icon.accent.blue': "var(--ds-icon-accent-blue, #357DE8)",
  'color.icon.accent.purple': "var(--ds-icon-accent-purple, #AF59E1)",
  'color.icon.accent.magenta': "var(--ds-icon-accent-magenta, #CD519D)",
  'color.icon.accent.gray': "var(--ds-icon-accent-gray, #7D818A)",
  'color.icon.disabled': "var(--ds-icon-disabled, #080F214A)",
  'color.icon.inverse': "var(--ds-icon-inverse, #FFFFFF)",
  'color.icon.selected': "var(--ds-icon-selected, #1868DB)",
  'color.icon.brand': "var(--ds-icon-brand, #1868DB)",
  'color.icon.danger': "var(--ds-icon-danger, #C9372C)",
  'color.icon.warning': "var(--ds-icon-warning, #E06C00)",
  'color.icon.warning.inverse': "var(--ds-icon-warning-inverse, #292A2E)",
  'color.icon.success': "var(--ds-icon-success, #6A9A23)",
  'color.icon.discovery': "var(--ds-icon-discovery, #AF59E1)",
  'color.icon.information': "var(--ds-icon-information, #357DE8)",
  'color.icon.subtlest': "var(--ds-icon-subtlest, #6B6E76)",
  'color.icon.subtle': "var(--ds-icon-subtle, #505258)"
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::042cbfe8041c09e3817ae74154994f32>>
 * @codegenId misc
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenParams ["layer"]
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/dimensions.tsx <<SignedSource::cc9b3f12104c6ede803da6a42daac0b0>>
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/layer.tsx <<SignedSource::92793ca02dbfdad66e53ffbe9f0baa0a>>
 */
export const layerMap = {
  '1': 1,
  card: 100,
  navigation: 200,
  dialog: 300,
  layer: 400,
  blanket: 500,
  modal: 510,
  flag: 600,
  spotlight: 700,
  tooltip: 800
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::6ab2b394dd5a830d6966e333e20b1473>>
 * @codegenId border
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenParams ["width", "radius"]
 * @codegenDependency ../../../tokens/src/artifacts/tokens-raw/atlassian-shape.tsx <<SignedSource::8817f4073995e5dc9c2bb766316632d6>>
 */
export const borderWidthMap = {
  'border.width': "var(--ds-border-width, 1px)",
  'border.width.selected': "var(--ds-border-width-selected, 2px)",
  'border.width.focused': "var(--ds-border-width-focused, 2px)"
};
export const borderRadiusMap = {
  'radius.xsmall': "var(--ds-radius-xsmall, 2px)",
  'radius.small': "var(--ds-radius-small, 3px)",
  'radius.medium': "var(--ds-radius-medium, 6px)",
  'radius.large': "var(--ds-radius-large, 8px)",
  'radius.xlarge': "var(--ds-radius-xlarge, 12px)",
  'radius.xxlarge': "var(--ds-radius-xxlarge, 16px)",
  'radius.full': "var(--ds-radius-full, 9999px)",
  'radius.tile': "var(--ds-radius-tile, 25%)"
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::64331f0b8b5e2f090a3aad153d92deb6>>
 * @codegenId typography
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenParams ["fontSize", "fontWeight", "fontFamily", "lineHeight", "body", "ui"]
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/dimensions.tsx <<SignedSource::cc9b3f12104c6ede803da6a42daac0b0>>
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/layer.tsx <<SignedSource::92793ca02dbfdad66e53ffbe9f0baa0a>>
 */
export const fontMap = {
  'font.body': "var(--ds-font-body, normal 400 14px/20px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  // @deprecated
  'font.body.UNSAFE_small': "var(--ds-font-body-UNSAFE_small, normal 400 12px/16px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.body.large': "var(--ds-font-body-large, normal 400 16px/24px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.body.small': "var(--ds-font-body-small, normal 400 12px/16px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.code': "var(--ds-font-code, normal 400 0.875em/1 \"Atlassian Mono\", ui-monospace, Menlo, \"Segoe UI Mono\", \"Ubuntu Mono\", monospace)",
  'font.heading.large': "var(--ds-font-heading-large, normal 653 24px/28px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.heading.medium': "var(--ds-font-heading-medium, normal 653 20px/24px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.heading.small': "var(--ds-font-heading-small, normal 653 16px/20px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.heading.xlarge': "var(--ds-font-heading-xlarge, normal 653 28px/32px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.heading.xsmall': "var(--ds-font-heading-xsmall, normal 653 14px/20px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.heading.xxlarge': "var(--ds-font-heading-xxlarge, normal 653 32px/36px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.heading.xxsmall': "var(--ds-font-heading-xxsmall, normal 653 12px/16px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.metric.large': "var(--ds-font-metric-large, normal 653 28px/32px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.metric.medium': "var(--ds-font-metric-medium, normal 653 24px/28px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.metric.small': "var(--ds-font-metric-small, normal 653 16px/20px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
};
export const fontWeightMap = {
  'font.weight.bold': "var(--ds-font-weight-bold, 653)",
  'font.weight.medium': "var(--ds-font-weight-medium, 500)",
  'font.weight.regular': "var(--ds-font-weight-regular, 400)",
  'font.weight.semibold': "var(--ds-font-weight-semibold, 600)"
};
export const fontFamilyMap = {
  'font.family.body': "var(--ds-font-family-body, \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.family.brand.body': "var(--ds-font-family-brand-body, \"Charlie Text\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.family.brand.heading': "var(--ds-font-family-brand-heading, \"Charlie Display\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  'font.family.code': "var(--ds-font-family-code, \"Atlassian Mono\", ui-monospace, Menlo, \"Segoe UI Mono\", \"Ubuntu Mono\", monospace)",
  'font.family.heading': "var(--ds-font-family-heading, \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
};
/**
 * @codegenEnd
 */

/**
 * THIS SECTION WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::aeebc24b95e2189021f1d98f1715a6a0>>
 * @codegenId text
 * @codegenCommand yarn workspace @atlaskit/primitives codegen-styles
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/dimensions.tsx <<SignedSource::cc9b3f12104c6ede803da6a42daac0b0>>
 * @codegenDependency ../../../primitives/scripts/codegen-file-templates/layer.tsx <<SignedSource::92793ca02dbfdad66e53ffbe9f0baa0a>>
 */
export const textSizeMap = {
  medium: "var(--ds-font-body, normal 400 14px/20px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  UNSAFE_small: "var(--ds-font-body-UNSAFE_small, normal 400 12px/16px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  large: "var(--ds-font-body-large, normal 400 16px/24px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  small: "var(--ds-font-body-small, normal 400 12px/16px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
};
export const textWeightMap = {
  bold: "var(--ds-font-weight-bold, 653)",
  medium: "var(--ds-font-weight-medium, 500)",
  regular: "var(--ds-font-weight-regular, 400)",
  semibold: "var(--ds-font-weight-semibold, 600)"
};
export const metricTextSizeMap = {
  large: "var(--ds-font-metric-large, normal 653 28px/32px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  medium: "var(--ds-font-metric-medium, normal 653 24px/28px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  small: "var(--ds-font-metric-small, normal 653 16px/20px \"Atlassian Sans\", ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
};

/**
 * @codegenEnd
 */

// Margin needs some bespoke types: https://atlassian.slack.com/archives/CKRHB23K8/p1712623192772909

/**
 * Token functions are currently allowed for borderRadius in xcss to ease migration.
 * Because 'radius.small' defaults to 3px, we need a mechanism to allow for 4px values to be provided:
 * This is done by calling the token function - token('radius.small')
 *
 * TODO: Remove once shape them is rolled out and fg platform-dst-shape-theme-default removed
 */

// Generate maps for Box props. backgroundColor, padding, etc
const spacingProperties = [
// Used by Box
'padding', 'paddingBlock', 'paddingBlockStart', 'paddingBlockEnd', 'paddingInline', 'paddingInlineStart', 'paddingInlineEnd',
// Used by Inline and Stack
'gap', 'rowGap', 'columnGap'];
export const getSerializedStylesMap = (cssProperty, tokenMap) => {
  return Object.keys(tokenMap).reduce((emotionSpacingMap, token) => {
    // eslint-disable-next-line @repo/internal/styles/no-exported-styles
    emotionSpacingMap[token] = css({
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
      [cssProperty]: tokenMap[token]
    });
    return emotionSpacingMap;
  }, {});
};
export const paddingStylesMap = spacingProperties.reduce((styleMap, spacingProperty) => {
  styleMap[spacingProperty] = getSerializedStylesMap(spacingProperty, positiveSpaceMap);
  return styleMap;
}, {});
export const spaceStylesMap = spacingProperties.reduce((styleMap, spacingProperty) => {
  styleMap[spacingProperty] = getSerializedStylesMap(spacingProperty, positiveSpaceMap);
  return styleMap;
}, {});
export const backgroundColorStylesMap = getSerializedStylesMap('backgroundColor', backgroundColorMap);
export const textColorStylesMap = getSerializedStylesMap('color', textColorMap);
export const fontWeightStylesMap = getSerializedStylesMap('fontWeight', fontWeightMap);
export const fontFamilyStylesMap = getSerializedStylesMap('fontFamily', fontFamilyMap);
export const fontStylesMap = getSerializedStylesMap('font', fontMap);
export const textSizeStylesMap = getSerializedStylesMap('font', textSizeMap);
export const textWeightStylesMap = getSerializedStylesMap('fontWeight', textWeightMap);
export const surfaceColorStylesMap = getSerializedStylesMap(CURRENT_SURFACE_CSS_VAR, surfaceColorMap);
export const isSurfaceColorToken = color => surfaceColorMap[color] !== undefined;
export const metricTextSizeStylesMap = getSerializedStylesMap('font', metricTextSizeMap);
