// @atlaskit/adf-schema/dist/es2019/next-schema/nodes/task.js
import { $onePlus, $or, $zeroPlus, adfNode } from '@atlaskit/adf-schema-generator';
import { inlineContentGroup } from '../groups/inlineContentGroup';
import { inlineGroup } from '../groups/inlineGroup';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
import { unsupportedBlock } from './unsupportedBlock';
export const taskItem = adfNode('taskItem');
export const taskList = adfNode('taskList');
taskItem.define({
  defining: true,
  selectable: false,
  marks: [unsupportedMark, unsupportedNodeAttribute],
  allowAnyChildMark: true,
  attrs: {
    localId: {
      type: 'string',
      default: ''
    },
    state: {
      type: 'enum',
      values: ['TODO', 'DONE'],
      default: 'TODO'
    }
  },
  content: [$zeroPlus($or(inlineGroup, inlineContentGroup))]
});
taskList.define({
  defining: true,
  selectable: false,
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    localId: {
      type: 'string',
      default: ''
    }
  },
  contentMinItems: 1,
  content: [$onePlus($or(taskItem, unsupportedBlock)), $zeroPlus($or(taskItem, taskList, unsupportedBlock))]
});
