// @atlaskit/react-ufo/dist/es2019/segment/ssr-render-profiler.js
var _process, _process$env;
import React, { createContext, useContext, useMemo } from 'react';
import { getConfig } from '../config';
import { getActiveInteraction } from '../interaction-metrics';
// These are stored outside react context to be resilient to concurrent mode
// restarting the start marker from a suspense and losing the initial render
const startTimes = new Map();
const spanStates = new Map();
export const clearState = () => {
  startTimes.clear();
  spanStates.clear();
};
// Keep track of the last interaction id and reset the start timers if it ever changes.
// This is to allow multi-step ssr to track the render timers from different "interaction"s seperately
let lastActiveInteraction;
function checkActiveInteractionAndResetStartMarksIfSet() {
  var _getActiveInteraction;
  const activeInteractionId = (_getActiveInteraction = getActiveInteraction()) === null || _getActiveInteraction === void 0 ? void 0 : _getActiveInteraction.id;
  if (!!lastActiveInteraction && lastActiveInteraction !== activeInteractionId) {
    clearState();
  }
  lastActiveInteraction = activeInteractionId;
}
const getIsNativeTracingEnabled = () => {
  var _getConfig$ssr$enable, _getConfig, _getConfig$ssr;
  return (_getConfig$ssr$enable = (_getConfig = getConfig()) === null || _getConfig === void 0 ? void 0 : (_getConfig$ssr = _getConfig.ssr) === null || _getConfig$ssr === void 0 ? void 0 : _getConfig$ssr.enableNativeTracing) !== null && _getConfig$ssr$enable !== void 0 ? _getConfig$ssr$enable : false;
};
const onStartRender = (id, currentSegmentName, parentContext) => {
  var _startTimes$get, _vm_internals__, _vm_internals__$telem, _spanStates$get;
  if (!startTimes.has(id)) {
    startTimes.set(id, []);
  }
  (_startTimes$get = startTimes.get(id)) === null || _startTimes$get === void 0 ? void 0 : _startTimes$get.push(performance.now());
  if (!getIsNativeTracingEnabled()) {
    return null;
  }
  const startSpan = (_vm_internals__ = globalThis.__vm_internals__) === null || _vm_internals__ === void 0 ? void 0 : (_vm_internals__$telem = _vm_internals__.telemetry) === null || _vm_internals__$telem === void 0 ? void 0 : _vm_internals__$telem.startSpan;
  let spanState = (_spanStates$get = spanStates.get(id)) !== null && _spanStates$get !== void 0 ? _spanStates$get : null;
  if (!spanState && startSpan) {
    const span = startSpan(currentSegmentName, {
      parentSpanId: parentContext === null || parentContext === void 0 ? void 0 : parentContext.spanId
    });
    spanState = {
      span
    };
    spanStates.set(id, spanState);
  }
  return spanState;
};
const isInSSR = typeof __SERVER__ !== 'undefined' && __SERVER__ || typeof process !== 'undefined' && Boolean(((_process = process) === null || _process === void 0 ? void 0 : (_process$env = _process.env) === null || _process$env === void 0 ? void 0 : _process$env.REACT_SSR) || false);
const ProfilerMarker = ({
  onRender
}) => {
  onRender === null || onRender === void 0 ? void 0 : onRender();
  return null;
};
const ParentSpanContext = /*#__PURE__*/createContext(null);

// For profiler spans in SSR, we 'end' any with their latest end times set, we need to do this at UFO-interaction-end since until then the component could rerender/remount again and spans fundamentally don't use the last 'end' but the first
export const flushSsrRenderProfilerTraces = () => {
  spanStates.forEach(spanState => {
    if (spanState.latestEndTime != null) {
      spanState.span.end(spanState.latestEndTime);
    }
  });
  spanStates.clear();
};
export const SsrRenderProfilerInner = ({
  children,
  labelStack,
  onRender
}) => {
  var _currentSpanState$spa;
  const reactProfilerId = useMemo(() => labelStack.map(l => l.name).join('/'), [labelStack]);
  checkActiveInteractionAndResetStartMarksIfSet();
  const parentSpan = useContext(ParentSpanContext);
  const currentSpanState = onStartRender(reactProfilerId, labelStack[labelStack.length - 1].name, parentSpan);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ParentSpanContext.Provider, {
    value: (_currentSpanState$spa = currentSpanState === null || currentSpanState === void 0 ? void 0 : currentSpanState.span.getSpanContext()) !== null && _currentSpanState$spa !== void 0 ? _currentSpanState$spa : null
  }, children, /*#__PURE__*/React.createElement(ProfilerMarker, {
    onRender: () => {
      const startTimesForId = startTimes.get(reactProfilerId);
      if (startTimesForId !== null && startTimesForId !== void 0 && startTimesForId.length) {
        const endTime = performance.now();
        const firstStartTime = startTimesForId[0];
        const lastStartTime = startTimesForId[startTimesForId.length - 1];
        const baseDuration = endTime - lastStartTime;
        const actualDuration = endTime - firstStartTime;
        onRender('mount',
        // this is incorrect, but on the server there is no mount phase
        actualDuration, baseDuration, firstStartTime, endTime);
      }
      const spanState = spanStates.get(reactProfilerId);
      if (spanState) {
        spanState.latestEndTime = performance.now();
      }
    }
  })));
};
const SsrRenderProfiler = props => {
  if (isInSSR) {
    return /*#__PURE__*/React.createElement(SsrRenderProfilerInner, props);
  }

  // ensure structure similar to SSR implementation
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ProfilerMarker, null), props.children);
};
export default SsrRenderProfiler;
