// @forge/bridge/out/featureFlags/dataAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForgeDataAdapter = void 0;
const initFeatureFlags_1 = require("./initFeatureFlags");
class ForgeDataAdapter {
    constructor() {
        this.options = null;
        this.environment = undefined;
        this.cache = null;
    }
    async getDataAsync(current, user, options) {
        var _a;
        if (current) {
            return current;
        }
        this.environment = ((_a = this.options) === null || _a === void 0 ? void 0 : _a.environment) || { tier: 'development' };
        const initUser = {
            ...user,
            statsigEnvironment: this.environment
        };
        const result = await (0, initFeatureFlags_1.initFeatureFlags)({ user: initUser });
        const data = {
            source: 'Network',
            data: JSON.stringify(result),
            receivedAt: Date.now(),
            stableID: null,
            fullUserHash: null
        };
        this.cache = data;
        return data;
    }
    getDataSync(_user) {
        return this.cache;
    }
    async attach(_sdkKey, options, _network) {
        this.options = options;
    }
    async prefetchData() { }
    async setData(_data) { }
    async setDataLegacy(_data) { }
    async shutdown() {
        this.options = null;
        this.cache = null;
        this.environment = undefined;
    }
}
exports.ForgeDataAdapter = ForgeDataAdapter;
