// @atlaskit/side-nav-items/dist/esm/ui/skeleton/skeleton-menu-item.js
/* skeleton-menu-item.tsx generated by @compiled/babel-plugin v0.38.1 */
import "./skeleton-menu-item.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Flex } from '@atlaskit/primitives/compiled';
var containerStyles = {
  root: "_4t3i1tcg _bozg1b66 _y4ti1b66 _1q511b66 _85i51b66 _1e0c1txw _4cvr1h6o",
  text: "_bozg1b66 _y4ti1b66",
  item: "_11c8fhey _4t3iu7e1 _1e0c1txw _4cvr1h6o",
  hasDescription: "_4t3i1ylp",
  description: "_11c8wadc _4t3iu7e1 _1e0c1txw _4cvr1h6o",
  elemBefore: "_bozgv77o _y4tiv77o _1q51v77o _85i5v77o _1e0c1txw _4cvr1h6o _1bah1h6o"
};
var skeletonStyles = {
  item: "_2rko1rr0 _bfhkizff _1bsb53f4 _4t3ic3s8",
  elemBefore: "_2rko12b0 _bfhkizff _1bsbgktf _4t3igktf",
  description: "_2rko1rr0 _bfhkizff _1bsb1wqb _4t3ic3s8"
};
/**
 * __Skeleton menu item__
 *
 * A skeleton menu item is used to display a loading state for a side nav menu item.
 * This component can have an icon element skeleton, and a description skeleton included.
 */
export var SkeletonMenuItem = function SkeletonMenuItem(props) {
  var _props$hasDescription = props.hasDescription,
    hasDescription = _props$hasDescription === void 0 ? false : _props$hasDescription,
    _props$hasElemBefore = props.hasElemBefore,
    hasElemBefore = _props$hasElemBefore === void 0 ? false : _props$hasElemBefore,
    testId = props.testId;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId,
    className: ax([containerStyles.root, hasDescription && containerStyles.hasDescription])
  }, /*#__PURE__*/React.createElement(Flex, {
    alignItems: "center"
  }, hasElemBefore && /*#__PURE__*/React.createElement("div", {
    className: ax([containerStyles.elemBefore])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([skeletonStyles.elemBefore])
  })), /*#__PURE__*/React.createElement(Flex, {
    xcss: containerStyles.text,
    gap: "space.025",
    direction: "column"
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([containerStyles.item])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([skeletonStyles.item])
  })), hasDescription && /*#__PURE__*/React.createElement("div", {
    className: ax([containerStyles.description])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([skeletonStyles.description])
  })))));
};
