// @atlaskit/button/dist/cjs/new-button/containers/split-button/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function get() {
    return _splitButton.Divider;
  }
});
Object.defineProperty(exports, "SplitButton", {
  enumerable: true,
  get: function get() {
    return _splitButton.SplitButton;
  }
});
Object.defineProperty(exports, "SplitButtonContainer", {
  enumerable: true,
  get: function get() {
    return _splitButton.SplitButtonContainer;
  }
});
Object.defineProperty(exports, "SplitButtonContext", {
  enumerable: true,
  get: function get() {
    return _splitButtonContext.SplitButtonContext;
  }
});
Object.defineProperty(exports, "SplitButtonWithSlots", {
  enumerable: true,
  get: function get() {
    return _splitButton.SplitButtonWithSlots;
  }
});
Object.defineProperty(exports, "getActions", {
  enumerable: true,
  get: function get() {
    return _utils.getActions;
  }
});
Object.defineProperty(exports, "useSplitButtonContext", {
  enumerable: true,
  get: function get() {
    return _splitButtonContext.useSplitButtonContext;
  }
});
var _splitButton = require("./split-button");
var _utils = require("./utils");
var _splitButtonContext = require("./split-button-context");
