// @atlaskit/button/dist/cjs/new-button/containers/split-button/split-button-context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSplitButtonContext = exports.SplitButtonContext = void 0;
var _react = require("react");
/**
 * TODO: Add jsdoc
 */
var SplitButtonContext = exports.SplitButtonContext = /*#__PURE__*/(0, _react.createContext)(undefined);
var useSplitButtonContext = exports.useSplitButtonContext = function useSplitButtonContext() {
  var context = (0, _react.useContext)(SplitButtonContext);
  if (!context) {
    return undefined;
  }
  if (context.appearance === 'navigation') {
    var newContext = {
      spacing: 'default',
      appearance: 'subtle',
      isDisabled: false,
      isActiveOverSelected: false,
      isNavigationSplitButton: true,
      isHighlighted: context.isHighlighted
    };
    return newContext;
  }
  var splitButtonContext = {
    spacing: context.spacing,
    appearance: context.appearance,
    isDisabled: context.isDisabled,
    isActiveOverSelected: true,
    isNavigationSplitButton: false,
    isHighlighted: false
  };
  return splitButtonContext;
};
