// @forge/manifest/out/types/display-condition-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EntityType;
(function (EntityType) {
    EntityType["Space"] = "space";
    EntityType["Content"] = "content";
    EntityType["Issue"] = "issue";
    EntityType["IssueType"] = "issueType";
    EntityType["Project"] = "project";
    EntityType["App"] = "app";
    EntityType["User"] = "user";
})(EntityType || (EntityType = {}));
var AppEditionType;
(function (AppEditionType) {
    AppEditionType["Free"] = "free";
    AppEditionType["Standard"] = "standard";
    AppEditionType["Advanced"] = "advanced";
})(AppEditionType || (AppEditionType = {}));
var ConfluenceContentPermission;
(function (ConfluenceContentPermission) {
    ConfluenceContentPermission["View"] = "VIEW";
    ConfluenceContentPermission["Edit"] = "EDIT";
    ConfluenceContentPermission["SetPermissions"] = "SET_PERMISSIONS";
    ConfluenceContentPermission["Remove"] = "REMOVE";
    ConfluenceContentPermission["Export"] = "EXPORT";
    ConfluenceContentPermission["Administer"] = "ADMINISTER";
})(ConfluenceContentPermission || (ConfluenceContentPermission = {}));
