// @forge/manifest/out/types/module-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUPPORTED_PRODUCTS_ONLY = exports.SUPPORTED_MODULES = exports.AllModuleTypes = void 0;
var AllModuleTypes;
(function (AllModuleTypes) {
    AllModuleTypes["CoreFunction"] = "core:function";
    AllModuleTypes["CoreWorkflow"] = "core:workflow";
    AllModuleTypes["CoreTrigger"] = "core:trigger";
    AllModuleTypes["CoreWebTrigger"] = "core:webtrigger";
    AllModuleTypes["CoreAPIRoute"] = "core:apiRoute";
    AllModuleTypes["CoreEvent"] = "core:event";
    AllModuleTypes["CoreConsumer"] = "core:consumer";
    AllModuleTypes["CoreFifoConsumer"] = "core:fifoConsumer";
    AllModuleTypes["CoreScheduledTrigger"] = "core:scheduledTrigger";
    AllModuleTypes["CoreEndpoint"] = "core:endpoint";
    AllModuleTypes["CoreMigration"] = "core:migration";
    AllModuleTypes["CoreRemote"] = "core:remote";
    AllModuleTypes["CoreSql"] = "core:sql";
    AllModuleTypes["CoreOs"] = "core:objectStore";
    AllModuleTypes["CoreLlm"] = "core:llm";
    AllModuleTypes["XenMacro"] = "xen:macro";
    AllModuleTypes["DataResidencyMigration"] = "migration:dataResidency";
    AllModuleTypes["DashboardsWidget"] = "dashboards:widget";
    AllModuleTypes["DashboardsBackgroundScript"] = "dashboards:backgroundScript";
    AllModuleTypes["ConfluenceContentAction"] = "confluence:contentAction";
    AllModuleTypes["ConfluenceContentBylineItem"] = "confluence:contentBylineItem";
    AllModuleTypes["ConfluenceContextMenu"] = "confluence:contextMenu";
    AllModuleTypes["ConfluenceCustomContent"] = "confluence:customContent";
    AllModuleTypes["ConfluenceGlobalSettings"] = "confluence:globalSettings";
    AllModuleTypes["ConfluenceGlobalPage"] = "confluence:globalPage";
    AllModuleTypes["ConfluenceHomepageFeed"] = "confluence:homepageFeed";
    AllModuleTypes["ConfluenceSpacePage"] = "confluence:spacePage";
    AllModuleTypes["ConfluenceSpaceSettings"] = "confluence:spaceSettings";
    AllModuleTypes["ConfluenceBackgroundScript"] = "confluence:backgroundScript";
    AllModuleTypes["ConfluencePageBanner"] = "confluence:pageBanner";
    AllModuleTypes["ConfluenceFullPage"] = "confluence:fullPage";
    AllModuleTypes["ConfluenceContentProperty"] = "confluence:contentProperty";
    AllModuleTypes["JiraEntityProperty"] = "jira:entityProperty";
    AllModuleTypes["JiraWorkflowValidator"] = "jira:workflowValidator";
    AllModuleTypes["JiraWorkflowCondition"] = "jira:workflowCondition";
    AllModuleTypes["JiraWorkflowPostFunction"] = "jira:workflowPostFunction";
    AllModuleTypes["JiraActionValidator"] = "jira:actionValidator";
    AllModuleTypes["JiraCustomField"] = "jira:customField";
    AllModuleTypes["JiraCustomFieldType"] = "jira:customFieldType";
    AllModuleTypes["JiraIssueAction"] = "jira:issueAction";
    AllModuleTypes["JiraIssueActivity"] = "jira:issueActivity";
    AllModuleTypes["JiraUiModifications"] = "jira:uiModifications";
    AllModuleTypes["JiraIssueGlance"] = "jira:issueGlance";
    AllModuleTypes["JiraIssueContext"] = "jira:issueContext";
    AllModuleTypes["JiraIssuePanel"] = "jira:issuePanel";
    AllModuleTypes["JiraIssueNavigatorAction"] = "jira:issueNavigatorAction";
    AllModuleTypes["JiraAdminPage"] = "jira:adminPage";
    AllModuleTypes["JiraProjectPage"] = "jira:projectPage";
    AllModuleTypes["JiraProjectSettingsPage"] = "jira:projectSettingsPage";
    AllModuleTypes["JiraGlobalPage"] = "jira:globalPage";
    AllModuleTypes["JiraDashboardGadget"] = "jira:dashboardGadget";
    AllModuleTypes["JiraDashboardBackgroundScript"] = "jira:dashboardBackgroundScript";
    AllModuleTypes["JiraViewIssueBackgroundScript"] = "jira:issueViewBackgroundScript";
    AllModuleTypes["JiraGlobalPermission"] = "jira:globalPermission";
    AllModuleTypes["JiraProjectPermission"] = "jira:projectPermission";
    AllModuleTypes["JiraJqlFunction"] = "jira:jqlFunction";
    AllModuleTypes["JiraTimeTrackingProvider"] = "jira:timeTrackingProvider";
    AllModuleTypes["JiraBacklogAction"] = "jira:backlogAction";
    AllModuleTypes["JiraPersonalSettingsPage"] = "jira:personalSettingsPage";
    AllModuleTypes["JiraBoardAction"] = "jira:boardAction";
    AllModuleTypes["JiraSprintAction"] = "jira:sprintAction";
    AllModuleTypes["JiraCommand"] = "jira:command";
    AllModuleTypes["JiraGlobalBackgroundScript"] = "jira:globalBackgroundScript";
    AllModuleTypes["JiraFullPage"] = "jira:fullPage";
    AllModuleTypes["CompassAdminPage"] = "compass:adminPage";
    AllModuleTypes["CompassComponentPage"] = "compass:componentPage";
    AllModuleTypes["CompassGlobalPage"] = "compass:globalPage";
    AllModuleTypes["CompassTeamPage"] = "compass:teamPage";
    AllModuleTypes["CompassDataProvider"] = "compass:dataProvider";
    AllModuleTypes["JiraServiceManagementAssetsImportType"] = "jiraServiceManagement:assetsImportType";
    AllModuleTypes["JiraServiceManagementAssetsSchemaPage"] = "jiraServiceManagement:assetsSchemaPage";
    AllModuleTypes["JiraServiceManagementQueuePage"] = "jiraServiceManagement:queuePage";
    AllModuleTypes["JiraServiceManagementPortalRequestDetail"] = "jiraServiceManagement:portalRequestDetail";
    AllModuleTypes["JiraServiceManagementPortalRequestDetailPanel"] = "jiraServiceManagement:portalRequestDetailPanel";
    AllModuleTypes["JiraServiceManagementOrganizationPanel"] = "jiraServiceManagement:organizationPanel";
    AllModuleTypes["JiraServiceManagementPortalHeader"] = "jiraServiceManagement:portalHeader";
    AllModuleTypes["JiraServiceManagementPortalSubheader"] = "jiraServiceManagement:portalSubheader";
    AllModuleTypes["JiraServiceManagementPortalFooter"] = "jiraServiceManagement:portalFooter";
    AllModuleTypes["JiraServiceManagementPortalProfilePanel"] = "jiraServiceManagement:portalProfilePanel";
    AllModuleTypes["JiraServiceManagementPortalRequestViewAction"] = "jiraServiceManagement:portalRequestViewAction";
    AllModuleTypes["JiraServiceManagementPortalUserMenuAction"] = "jiraServiceManagement:portalUserMenuAction";
    AllModuleTypes["JiraServiceManagementPortalRequestCreatePropertyPanel"] = "jiraServiceManagement:portalRequestCreatePropertyPanel";
    AllModuleTypes["BitbucketRepoCodeFileViewer"] = "bitbucket:repoCodeFileViewer";
    AllModuleTypes["BitbucketRepoCodeOverviewCard"] = "bitbucket:repoCodeOverviewCard";
    AllModuleTypes["BitbucketRepoCodeOverviewAction"] = "bitbucket:repoCodeOverviewAction";
    AllModuleTypes["BitbucketRepoCodeOverviewPanel"] = "bitbucket:repoCodeOverviewPanel";
    AllModuleTypes["BitbucketRepoMainMenuPage"] = "bitbucket:repoMainMenuPage";
    AllModuleTypes["BitbucketRepoPullRequestCard"] = "bitbucket:repoPullRequestCard";
    AllModuleTypes["BitbucketRepoPullRequestAction"] = "bitbucket:repoPullRequestAction";
    AllModuleTypes["BitbucketRepoPullRequestOverviewPanel"] = "bitbucket:repoPullRequestOverviewPanel";
    AllModuleTypes["BitbucketRepoSettingsMenuPage"] = "bitbucket:repoSettingsMenuPage";
    AllModuleTypes["BitbucketProjectSettingsMenuPage"] = "bitbucket:projectSettingsMenuPage";
    AllModuleTypes["BitbucketWorkspaceGlobalPage"] = "bitbucket:workspaceGlobalPage";
    AllModuleTypes["BitbucketWorkspacePersonalSettingsPage"] = "bitbucket:workspacePersonalSettingsPage";
    AllModuleTypes["BitbucketWorkspaceSettingsMenuPage"] = "bitbucket:workspaceSettingsMenuPage";
    AllModuleTypes["BitbucketMergeCheck"] = "bitbucket:mergeCheck";
    AllModuleTypes["BitbucketDynamicPipelinesProvider"] = "bitbucket:dynamicPipelinesProvider";
    AllModuleTypes["ConnectJiraAdminPages"] = "connect-jira:adminPages";
    AllModuleTypes["ConnectJiraJiraBackgroundScripts"] = "connect-jira:jiraBackgroundScripts";
    AllModuleTypes["ConnectJiraGeneralPages"] = "connect-jira:generalPages";
    AllModuleTypes["ConnectJiraJiraDashboardItems"] = "connect-jira:jiraDashboardItems";
    AllModuleTypes["ConnectJiraJiraGlobalPermissions"] = "connect-jira:jiraGlobalPermissions";
    AllModuleTypes["ConnectJiraJiraIssueFields"] = "connect-jira:jiraIssueFields";
    AllModuleTypes["ConnectJiraJiraIssueGlances"] = "connect-jira:jiraIssueGlances";
    AllModuleTypes["ConnectJiraJiraProjectAdminTabPanels"] = "connect-jira:jiraProjectAdminTabPanels";
    AllModuleTypes["ConnectJiraJiraProjectTabPanels"] = "connect-jira:jiraProjectTabPanels";
    AllModuleTypes["ConnectJiraJiraWorkflowConditions"] = "connect-jira:jiraWorkflowConditions";
    AllModuleTypes["ConnectJiraJiraWorkflowPostFunctions"] = "connect-jira:jiraWorkflowPostFunctions";
    AllModuleTypes["ConnectJiraKeyboardShortcuts"] = "connect-jira:keyboardShortcuts";
    AllModuleTypes["ConnectJiraWebItems"] = "connect-jira:webItems";
    AllModuleTypes["ConnectJiraLifecycle"] = "connect-jira:lifecycle";
    AllModuleTypes["ConnectJiraAppMigrations"] = "connect-jira:cloudAppMigration";
    AllModuleTypes["ConnectJiraTranslations"] = "connect-jira:translations";
    AllModuleTypes["ConnectConfluenceAdminPages"] = "connect-confluence:adminPages";
    AllModuleTypes["ConnectConfluenceBlueprints"] = "connect-confluence:blueprints";
    AllModuleTypes["ConnectConfluenceConfigurePage"] = "connect-confluence:configurePage";
    AllModuleTypes["ConnectConfluenceConfluenceContentProperties"] = "connect-confluence:confluenceContentProperties";
    AllModuleTypes["ConnectConfluenceConfluenceThemes"] = "connect-confluence:confluenceThemes";
    AllModuleTypes["ConnectConfluenceContentBylineItems"] = "connect-confluence:contentBylineItems";
    AllModuleTypes["ConnectConfluenceCustomContent"] = "connect-confluence:customContent";
    AllModuleTypes["ConnectConfluenceDialogs"] = "connect-confluence:dialogs";
    AllModuleTypes["ConnectConfluenceDynamicContentMacros"] = "connect-confluence:dynamicContentMacros";
    AllModuleTypes["ConnectConfluenceGeneralPages"] = "connect-confluence:generalPages";
    AllModuleTypes["ConnectConfluenceKeyboardShortcuts"] = "connect-confluence:keyboardShortcuts";
    AllModuleTypes["ConnectConfluencePostInstallPage"] = "connect-confluence:postInstallPage";
    AllModuleTypes["ConnectConfluenceSpaceToolsTabs"] = "connect-confluence:spaceToolsTabs";
    AllModuleTypes["ConnectConfluenceStaticContentMacros"] = "connect-confluence:staticContentMacros";
    AllModuleTypes["ConnectConfluenceWebItems"] = "connect-confluence:webItems";
    AllModuleTypes["ConnectConfluenceWebPanels"] = "connect-confluence:webPanels";
    AllModuleTypes["ConnectConfluenceWebSections"] = "connect-confluence:webSections";
    AllModuleTypes["ConnectConfluenceWebhooks"] = "connect-confluence:webhooks";
    AllModuleTypes["ConnectConfluencePublishConditions"] = "connect-confluence:publishConditions";
    AllModuleTypes["ConnectJiraConfigurePage"] = "connect-jira:configurePage";
    AllModuleTypes["ConnectJiraDialogs"] = "connect-jira:dialogs";
    AllModuleTypes["ConnectJiraJiraBuildInfoProvider"] = "connect-jira:jiraBuildInfoProvider";
    AllModuleTypes["ConnectJiraJiraDeploymentInfoProvider"] = "connect-jira:jiraDeploymentInfoProvider";
    AllModuleTypes["ConnectJiraJiraDevelopmentTool"] = "connect-jira:jiraDevelopmentTool";
    AllModuleTypes["ConnectJiraJiraEntityProperties"] = "connect-jira:jiraEntityProperties";
    AllModuleTypes["ConnectJiraJiraFeatureFlagInfoProvider"] = "connect-jira:jiraFeatureFlagInfoProvider";
    AllModuleTypes["ConnectJiraJiraIssueContents"] = "connect-jira:jiraIssueContents";
    AllModuleTypes["ConnectJiraJiraIssueContexts"] = "connect-jira:jiraIssueContexts";
    AllModuleTypes["ConnectJiraJiraRemoteLinkInfoProvider"] = "connect-jira:jiraRemoteLinkInfoProvider";
    AllModuleTypes["ConnectJiraJiraIssueTabPanels"] = "connect-jira:jiraIssueTabPanels";
    AllModuleTypes["ConnectJiraJiraIssueTypes"] = "connect-jira:jiraIssueTypes";
    AllModuleTypes["ConnectJiraJiraProjectPages"] = "connect-jira:jiraProjectPages";
    AllModuleTypes["ConnectJiraJiraProjectPermissions"] = "connect-jira:jiraProjectPermissions";
    AllModuleTypes["ConnectJiraJiraReports"] = "connect-jira:jiraReports";
    AllModuleTypes["ConnectJiraJiraSearchRequestViews"] = "connect-jira:jiraSearchRequestViews";
    AllModuleTypes["ConnectJiraJiraTimeTrackingProviders"] = "connect-jira:jiraTimeTrackingProviders";
    AllModuleTypes["ConnectJiraJiraWorkflowValidators"] = "connect-jira:jiraWorkflowValidators";
    AllModuleTypes["ConnectJiraPostInstallPage"] = "connect-jira:postInstallPage";
    AllModuleTypes["ConnectJiraWebPanels"] = "connect-jira:webPanels";
    AllModuleTypes["ConnectJiraWebSections"] = "connect-jira:webSections";
    AllModuleTypes["ConnectJiraWebhooks"] = "connect-jira:webhooks";
    AllModuleTypes["ConnectConfluenceLifecycle"] = "connect-confluence:lifecycle";
    AllModuleTypes["ConnectConfluenceAppMigrations"] = "connect-confluence:cloudAppMigration";
    AllModuleTypes["ConnectConfluenceTranslations"] = "connect-confluence:translations";
    AllModuleTypes["ConnectJiraJiraSecurityInfoProvider"] = "connect-jira:jiraSecurityInfoProvider";
    AllModuleTypes["ConnectJiraServiceDeskPortalHeaders"] = "connect-jira:serviceDeskPortalHeaders";
    AllModuleTypes["ConnectJiraServiceDeskPortalSubHeaders"] = "connect-jira:serviceDeskPortalSubHeaders";
    AllModuleTypes["ConnectJiraServiceDeskPortalFooters"] = "connect-jira:serviceDeskPortalFooters";
    AllModuleTypes["ConnectJiraServiceDeskPortalRequestCreatePropertyPanels"] = "connect-jira:serviceDeskPortalRequestCreatePropertyPanels";
    AllModuleTypes["ConnectJiraServiceDeskPortalRequestViewDetailsPanels"] = "connect-jira:serviceDeskPortalRequestViewDetailsPanels";
    AllModuleTypes["ConnectJiraServiceDeskPortalRequestViewPanels"] = "connect-jira:serviceDeskPortalRequestViewPanels";
    AllModuleTypes["ConnectJiraServiceDeskPortalProfilePanels"] = "connect-jira:serviceDeskPortalProfilePanels";
    AllModuleTypes["ConnectJiraServiceDeskQueues"] = "connect-jira:serviceDeskQueues";
    AllModuleTypes["ConnectJiraServiceDeskReports"] = "connect-jira:serviceDeskReports";
    AllModuleTypes["ConnectJiraServiceDeskOrganizationPanels"] = "connect-jira:serviceDeskOrganizationPanels";
    AllModuleTypes["ConnectJiraServiceDeskReportGroups"] = "connect-jira:serviceDeskReportGroups";
    AllModuleTypes["ConnectJiraServiceDeskQueueGroups"] = "connect-jira:serviceDeskQueueGroups";
    AllModuleTypes["ConnectJiraServiceDeskPortalUserMenuActions"] = "connect-jira:serviceDeskPortalUserMenuActions";
    AllModuleTypes["ConnectJiraServiceDeskPortalProfileActions"] = "connect-jira:serviceDeskPortalProfileActions";
    AllModuleTypes["ConnectJiraServiceDeskPortalRequestViewActions"] = "connect-jira:serviceDeskPortalRequestViewActions";
    AllModuleTypes["ConnectJiraServiceDeskOrganizationActions"] = "connect-jira:serviceDeskOrganizationActions";
    AllModuleTypes["ConnectJiraAutomationActions"] = "connect-jira:automationActions";
    AllModuleTypes["ConnectJiraJiraDeploymentGating"] = "connect-jira:jiraDeploymentGating";
    AllModuleTypes["ConnectJiraAssetPanels"] = "connect-jira:assetPanels";
    AllModuleTypes["ConnectJiraJiraJqlFunctions"] = "connect-jira:jiraJqlFunctions";
    AllModuleTypes["ConnectJiraJiraOperationsInfoProvider"] = "connect-jira:jiraOperationsInfoProvider";
    AllModuleTypes["ConnectJiraJiraDevOpsComponentProvider"] = "connect-jira:jiraDevOpsComponentProvider";
    AllModuleTypes["AutomationForConfluenceTrigger"] = "automation:trigger";
    AllModuleTypes["AutomationForConfluenceAction"] = "automation:action";
    AllModuleTypes["AutomationForConfluenceBranch"] = "automation:branch";
    AllModuleTypes["AutomationForConfluenceCondition"] = "automation:condition";
    AllModuleTypes["AutomationForConfluenceAttributeResolver"] = "automation:attributeResolver";
    AllModuleTypes["AutomationForConfluenceTemplate"] = "automation:template";
    AllModuleTypes["AutomationForConfluenceI18n"] = "automation:i18n";
    AllModuleTypes["AutomationActionProvider"] = "automation:actionProvider";
    AllModuleTypes["RovoAgent"] = "rovo:agent";
    AllModuleTypes["RovoAgentConnector"] = "rovo:agentConnector";
    AllModuleTypes["CoreAction"] = "core:action";
    AllModuleTypes["GraphEntityProvider"] = "graph:entityProvider";
    AllModuleTypes["GraphConnector"] = "graph:connector";
    AllModuleTypes["SmartLinks"] = "graph:smartLink";
    AllModuleTypes["DevopsDevelopmentInfoProvider"] = "devops:developmentInfoProvider";
    AllModuleTypes["DevopsFeatureFlagInfoProvider"] = "devops:featureFlagInfoProvider";
    AllModuleTypes["DevopsDeploymentInfoProvider"] = "devops:deploymentInfoProvider";
    AllModuleTypes["DevopsBuildInfoProvider"] = "devops:buildInfoProvider";
    AllModuleTypes["DevopsRemoteLinkInfoProvider"] = "devops:remoteLinkInfoProvider";
})(AllModuleTypes = exports.AllModuleTypes || (exports.AllModuleTypes = {}));
exports.SUPPORTED_MODULES = Object.values(AllModuleTypes);
exports.SUPPORTED_PRODUCTS_ONLY = [...new Set(exports.SUPPORTED_MODULES.map((module) => module.split(':')[0]))];
