// @atlaskit/adf-schema-generator/dist/cjs/transforms/adfToValidatorSpec/buildMarks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildMarks = buildMarks;
var _transformerNames = require("../transformerNames");
function buildMarks(marks) {
  var noMarks = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var hasEmptyMarks = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
  if (noMarks) {
    return {
      type: 'array',
      maxItems: 0,
      items: [],
      optional: true
    };
  }
  if (hasEmptyMarks) {
    return {
      type: 'array',
      items: [],
      optional: true
    };
  }
  var filteredMarks = filterIgnoredMarks(marks);
  if (!filteredMarks || filteredMarks.length === 0) {
    // @ts-expect-error
    return;
  }
  var returnMarks = {
    type: 'array',
    optional: true,
    items: []
  };
  if (filteredMarks.length > 1) {
    returnMarks.items = [filteredMarks.map(function (mark) {
      return mark.getType();
    })];
  } else {
    returnMarks.items = filteredMarks.map(function (mark) {
      return mark.getType();
    });
  }
  return returnMarks;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function filterIgnoredMarks(marks) {
  return (marks !== null && marks !== void 0 ? marks : []).filter(function (mark) {
    return !mark.isIgnored(_transformerNames.ValidatorSpecTransformerName);
  });
}
