// @atlaskit/adf-schema-generator/dist/cjs/transforms/adfToValidatorSpec/buildContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildContent = buildContent;
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function buildContent(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
node, content) {
  var isTupleLike = content.length > 1;
  if (isTupleLike) {
    return buildTupleLikeContent(node, content);
  }
  return buildRegularContent(content[0]);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function buildTupleLikeContent(node, content) {
  var nodeSpec = node.getSpec();
  var hasUnsupportedInline = content.some(function (item) {
    return item.hasUnsupportedInline;
  });
  var hasUnsupportedBlock = content.some(function (item) {
    return item.hasUnsupportedBlock;
  });
  var returnContent = {
    type: 'array',
    isTupleLike: true
  };
  var items = [];
  var _iterator = _createForOfIteratorHelper(content),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var child = _step.value;
      items.push(buildItems(child.contentTypes));
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  returnContent.items = items;
  if ('contentMinItems' in nodeSpec) {
    returnContent.minItems = nodeSpec.contentMinItems;
  }
  if ('contentMaxItems' in nodeSpec) {
    returnContent.maxItems = nodeSpec.contentMaxItems;
  }
  if (hasUnsupportedInline) {
    returnContent.allowUnsupportedInline = true;
  }
  if (hasUnsupportedBlock) {
    returnContent.allowUnsupportedBlock = true;
  }
  return returnContent;
}
function buildRegularContent(content) {
  var returnContent = {
    type: 'array'
  };
  returnContent.items = [buildItems(content.contentTypes)];
  if (content !== null && content !== void 0 && content.minItems) {
    returnContent.minItems = content === null || content === void 0 ? void 0 : content.minItems;
  }
  if (content !== null && content !== void 0 && content.maxItems) {
    returnContent.maxItems = content === null || content === void 0 ? void 0 : content.maxItems;
  }
  if (content !== null && content !== void 0 && content.optional) {
    returnContent.optional = content === null || content === void 0 ? void 0 : content.optional;
  }
  if (content.hasUnsupportedInline) {
    returnContent.allowUnsupportedInline = true;
  }
  if (content.hasUnsupportedBlock) {
    returnContent.allowUnsupportedBlock = true;
  }
  return returnContent;
}

/**
 * If contentTypes only has one item it will be a single array [],
 * however if it has more than one item it will be an array within an array [[]].
 */
function buildItems(contentTypes) {
  if (contentTypes.length === 1) {
    return contentTypes[0];
  } else {
    return contentTypes;
  }
}
