// @atlaskit/adf-schema-generator/dist/esm/transforms/adfToValidatorSpec/buildAttrs.js
import _typeof from "@babel/runtime/helpers/typeof";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { isAnyOf } from '../../utils/isAnyOf';
export var buildAttrs = function buildAttrs(attrs) {
  if (isAnyOf(attrs)) {
    return attrs.anyOf.map(function (attr) {
      return buildAttributesMap(attr);
    });
  }
  return buildAttributesMap(attrs);
};
function buildAttributesMap(attrs) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  var tmpAttrs = cleanObject(attrs);
  Object.entries(tmpAttrs).forEach(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
      key = _ref2[0],
      attr = _ref2[1];
    if (attr && attr.type === 'array') {
      tmpAttrs[key].type = 'array';
      tmpAttrs[key].items = attr.items.properties ? [buildAttrs(attr.items.properties)] : [attr.items];
    }
    if (attr && attr.type === 'object' && attr.properties) {
      tmpAttrs[key] = buildAttrs(attr.properties);
    }
  });
  var newAttrs = {
    props: tmpAttrs
  };
  if (Object.keys(attrs).every(function (attr) {
    var _attrs$attr;
    return ((_attrs$attr = attrs[attr]) === null || _attrs$attr === void 0 ? void 0 : _attrs$attr.optional) === true;
  })) {
    newAttrs.optional = true;
  }
  return newAttrs;
}

//recursive function to clean an object of any values that could be
//null, undefined, empty string, or empty object as those aren't
//necessary to include in the ValidatorSpec
function cleanObject(obj) {
  var isObject = function isObject(val) {
    return val && _typeof(val) === 'object' && !Array.isArray(val);
  };
  return Object.entries(obj).reduce(function (acc, _ref3) {
    var _ref4 = _slicedToArray(_ref3, 2),
      key = _ref4[0],
      value = _ref4[1];
    if (key.startsWith('__')) {
      return acc;
    }
    if (isObject(value)) {
      var nextValue = cleanObject(value);
      if (Object.keys(nextValue).length > 0) {
        // @ts-expect-error
        acc[key] = nextValue;
      }
    } else if (value !== null && value !== undefined && value !== '' && key !== 'default') {
      // @ts-expect-error
      acc[key] = value;
    }
    return acc;
  }, {});
}
