// @atlaskit/navigation-system/dist/cjs/entry-points/side-nav-items/flyout-menu-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "COLLAPSE_ELEM_BEFORE", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.COLLAPSE_ELEM_BEFORE;
  }
});
Object.defineProperty(exports, "FlyoutBody", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.FlyoutBody;
  }
});
Object.defineProperty(exports, "FlyoutFooter", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.FlyoutFooter;
  }
});
Object.defineProperty(exports, "FlyoutHeader", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.FlyoutHeader;
  }
});
Object.defineProperty(exports, "FlyoutMenuItem", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.FlyoutMenuItem;
  }
});
Object.defineProperty(exports, "FlyoutMenuItemContent", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.FlyoutMenuItemContent;
  }
});
Object.defineProperty(exports, "FlyoutMenuItemTrigger", {
  enumerable: true,
  get: function get() {
    return _flyoutMenuItem.FlyoutMenuItemTrigger;
  }
});
var _flyoutMenuItem = require("@atlaskit/side-nav-items/flyout-menu-item");
