// @atlaskit/adf-schema/dist/cjs/schema/nodes/blockquote.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.extendedBlockquoteWithLocalId = exports.extendedBlockquote = exports.blockquote = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var _utils = require("../../utils");
/**
 * @name blockquote_node
 */

var nodeSpecOptions = {
  parseDOM: [{
    tag: 'blockquote'
  }],
  toDOM: function toDOM() {
    return ['blockquote', 0];
  }
};
var blockquote = exports.blockquote = (0, _nodeTypes.blockquoteLegacy)(nodeSpecOptions);

/**
 * @name extentedBlockquote
 * @description the block quote node with nested code block, media, and extension
 */
var extendedBlockquote = exports.extendedBlockquote = (0, _nodeTypes.blockquote)(nodeSpecOptions);
var nodeSpecOptionsWithLocalId = {
  parseDOM: [{
    tag: 'blockquote',
    getAttrs: function getAttrs() {
      return {
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    return ['blockquote', {
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    }, 0];
  }
};
var extendedBlockquoteWithLocalId = exports.extendedBlockquoteWithLocalId = (0, _nodeTypes.blockquote)(nodeSpecOptionsWithLocalId);
