// @atlaskit/adf-schema/dist/cjs/schema/nodes/bullet-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bulletListWithLocalId = exports.bulletListSelector = exports.bulletList = void 0;
var _utils = require("../../utils");
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var bulletListSelector = exports.bulletListSelector = '.ak-ul';
var bulletList = exports.bulletList = (0, _nodeTypes.bulletList)({
  parseDOM: [{
    tag: 'ul'
  }],
  toDOM: function toDOM() {
    var attrs = {
      class: bulletListSelector.substr(1)
    };
    return ['ul', attrs, 0];
  }
});
var bulletListWithLocalId = exports.bulletListWithLocalId = (0, _nodeTypes.bulletList)({
  parseDOM: [{
    tag: 'ul',
    getAttrs: function getAttrs() {
      return {
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    var attrs = {
      class: bulletListSelector.substr(1),
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    };
    return ['ul', attrs, 0];
  }
});
