// @atlaskit/adf-schema/dist/cjs/schema/nodes/date.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dateWithLocalId = exports.date = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var _utils = require("../../utils");
/**
 * @name date_node
 */

var date = exports.date = (0, _nodeTypes.date)({
  parseDOM: [{
    tag: 'span[data-node-type="date"]',
    getAttrs: function getAttrs(dom) {
      return {
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        timestamp: dom.getAttribute('data-timestamp')
      };
    }
  }],
  toDOM: function toDOM(node) {
    var attrs = {
      'data-node-type': 'date',
      'data-timestamp': node.attrs.timestamp
    };
    return ['span', attrs];
  }
});
var dateWithLocalId = exports.dateWithLocalId = (0, _nodeTypes.date)({
  parseDOM: [{
    tag: 'span[data-node-type="date"]',
    getAttrs: function getAttrs(dom) {
      return {
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        timestamp: dom.getAttribute('data-timestamp'),
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    var attrs = {
      'data-node-type': 'date',
      'data-timestamp': node.attrs.timestamp,
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    };
    return ['span', attrs];
  }
});
