// @atlaskit/adf-schema/dist/cjs/schema/nodes/embed-card.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.embedCardWithLocalId = exports.embedCard = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var _uuid = require("../../utils/uuid");
/**
 * @name embedCard_node
 */

var embedCard = exports.embedCard = (0, _nodeTypes.embedCard)({
  parseDOM: [{
    tag: 'div[data-embed-card]',
    getAttrs: function getAttrs(dom) {
      return {
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        url: dom.getAttribute('data-card-url'),
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        layout: dom.getAttribute('data-layout') || 'center',
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        width: Number(dom.getAttribute('data-width')) || null,
        originalWidth:
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        Number(dom.getAttribute('data-card-original-width')) || null,
        originalHeight:
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        Number(dom.getAttribute('data-card-original-height')) || null
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs = node.attrs,
      url = _node$attrs.url,
      layout = _node$attrs.layout,
      width = _node$attrs.width,
      originalWidth = _node$attrs.originalWidth,
      originalHeight = _node$attrs.originalHeight;
    var attrs = {
      'data-embed-card': '',
      'data-card-url': url,
      'data-layout': layout,
      'data-width': width,
      'data-original-width': originalWidth,
      'data-original-height': originalHeight
    };
    return ['div', attrs];
  }
});
var embedCardWithLocalId = exports.embedCardWithLocalId = (0, _nodeTypes.embedCard)({
  parseDOM: [{
    tag: 'div[data-embed-card]',
    getAttrs: function getAttrs(dom) {
      return {
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        url: dom.getAttribute('data-card-url'),
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        layout: dom.getAttribute('data-layout') || 'center',
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        width: Number(dom.getAttribute('data-width')) || null,
        originalWidth:
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        Number(dom.getAttribute('data-card-original-width')) || null,
        originalHeight:
        // eslint-disable-next-line @atlaskit/editor/no-as-casting
        Number(dom.getAttribute('data-card-original-height')) || null,
        localId: _uuid.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs3;
    var _node$attrs2 = node.attrs,
      url = _node$attrs2.url,
      layout = _node$attrs2.layout,
      width = _node$attrs2.width,
      originalWidth = _node$attrs2.originalWidth,
      originalHeight = _node$attrs2.originalHeight;
    var attrs = {
      'data-embed-card': '',
      'data-card-url': url,
      'data-layout': layout,
      'data-width': width,
      'data-original-width': originalWidth,
      'data-original-height': originalHeight,
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs3 = node.attrs) === null || _node$attrs3 === void 0 ? void 0 : _node$attrs3.localId) || undefined
    };
    return ['div', attrs];
  }
});
