// @atlaskit/adf-schema/dist/cjs/schema/nodes/list-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listItemWithNestedDecisionAndLocalIdStage0 = exports.listItemWithLocalId = exports.listItemWithDecisionStage0 = exports.listItem = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _utils = require("../../utils");
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * @name list_item
 * @description this node allows task-list to be nested inside list-item
 */
var listItem = exports.listItem = (0, _nodeTypes.listItem)({
  parseDOM: [{
    tag: 'li'
  }],
  toDOM: function toDOM() {
    return ['li', 0];
  }
});
var listItemWithLocalId = exports.listItemWithLocalId = (0, _nodeTypes.listItem)({
  parseDOM: [{
    tag: 'li',
    getAttrs: function getAttrs() {
      return {
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    return ['li', {
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    }, 0];
  }
});

/**
 * @name list_item_with_decision_stage0
 * @description this node allows decisions to be nested inside list-item
 */
var listItemWithDecisionStage0Spec = (0, _nodeTypes.listItemWithNestedDecisionStage0)({
  parseDOM: [{
    tag: 'li'
  }],
  toDOM: function toDOM() {
    return ['li', 0];
  }
});

// Allow list-first content while preserving decisionList support in stage0.
var listItemStage0Content = '((paragraph | mediaSingle | codeBlock | unsupportedBlock | decisionList | extension) (paragraph | bulletList | orderedList | taskList | mediaSingle | codeBlock | unsupportedBlock | decisionList | extension)*) | ((paragraph | bulletList | orderedList | taskList | mediaSingle | codeBlock | unsupportedBlock | extension)+)';
var listItemWithDecisionStage0 = exports.listItemWithDecisionStage0 = _objectSpread(_objectSpread({}, listItemWithDecisionStage0Spec), {}, {
  content: listItemStage0Content
});
var listItemWithNestedDecisionAndLocalIdStage0 = exports.listItemWithNestedDecisionAndLocalIdStage0 = (0, _nodeTypes.listItemWithNestedDecisionStage0)({
  parseDOM: [{
    tag: 'li',
    getAttrs: function getAttrs() {
      return {
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs2;
    return ['li', {
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs2 = node.attrs) === null || _node$attrs2 === void 0 ? void 0 : _node$attrs2.localId) || undefined
    }, 0];
  }
});
