// @atlaskit/adf-schema/dist/cjs/schema/nodes/placeholder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.placeholderWithLocalId = exports.placeholder = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var _uuid = require("../../utils/uuid");
/**
 * @name placeholder_node
 */

var placeholder = exports.placeholder = (0, _nodeTypes.placeholder)({
  parseDOM: [{
    tag: 'span[data-placeholder]',
    getAttrs: function getAttrs(dom) {
      return {
        text:
        // eslint-disable-next-line @atlaskit/editor/no-as-casting, @typescript-eslint/no-non-null-assertion
        dom.getAttribute('data-placeholder') || placeholder.attrs.text.default
      };
    }
  }],
  toDOM: function toDOM(node) {
    var text = node.attrs.text;
    var attrs = {
      'data-placeholder': text,
      // Needs to be edtiable for mobile to not close keyboard
      contenteditable: 'true'
    };
    return ['span', attrs, text];
  }
});
var placeholderWithLocalId = exports.placeholderWithLocalId = (0, _nodeTypes.placeholder)({
  parseDOM: [{
    tag: 'span[data-placeholder]',
    getAttrs: function getAttrs(dom) {
      return {
        text:
        // eslint-disable-next-line @atlaskit/editor/no-as-casting, @typescript-eslint/no-non-null-assertion
        dom.getAttribute('data-placeholder') || placeholder.attrs.text.default,
        localId: _uuid.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    var text = node.attrs.text;
    var attrs = {
      'data-placeholder': text,
      // Needs to be edtiable for mobile to not close keyboard
      contenteditable: 'true',
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    };
    return ['span', attrs, text];
  }
});
