// @atlaskit/react-ufo/dist/es2019/create-payload/utils/add-performance-measures.js
export const addPerformanceMeasures = (start, measures) => {
  try {
    measures.sort((a, b) => b.revision.localeCompare(a.revision)).slice(0, 2).forEach(entry => {
      var _entry$vcDetails, _entry$vcDetails$;
      if (((_entry$vcDetails = entry.vcDetails) === null || _entry$vcDetails === void 0 ? void 0 : (_entry$vcDetails$ = _entry$vcDetails['90']) === null || _entry$vcDetails$ === void 0 ? void 0 : _entry$vcDetails$.t) !== undefined) {
        var _entry$vcDetails2, _entry$vcDetails2$;
        performance.mark(`VC90 (${entry.revision})`, {
          startTime: start + ((_entry$vcDetails2 = entry.vcDetails) === null || _entry$vcDetails2 === void 0 ? void 0 : (_entry$vcDetails2$ = _entry$vcDetails2['90']) === null || _entry$vcDetails2$ === void 0 ? void 0 : _entry$vcDetails2$.t),
          detail: {
            devtools: {
              dataType: 'marker'
            }
          }
        });
      }
    });
    measures.forEach(entry => {
      var _entry$vcDetails3, _entry$vcDetails3$;
      if (!entry || !entry.vcDetails) {
        return;
      }
      const VCParts = Object.keys(entry.vcDetails);
      performance.measure(`VC90 (${entry.revision})`, {
        start,
        duration: (_entry$vcDetails3 = entry.vcDetails) === null || _entry$vcDetails3 === void 0 ? void 0 : (_entry$vcDetails3$ = _entry$vcDetails3['90']) === null || _entry$vcDetails3$ === void 0 ? void 0 : _entry$vcDetails3$.t,
        detail: {
          devtools: {
            track: `main metrics`,
            trackGroup: '🛸 reactUFO metrics',
            color: 'tertiary'
          }
        }
      });
      VCParts.forEach(key => {
        var _entry$vcDetails4;
        const duration = (_entry$vcDetails4 = entry.vcDetails) === null || _entry$vcDetails4 === void 0 ? void 0 : _entry$vcDetails4[key].t;
        if (typeof duration !== 'number') {
          return;
        }
        performance.measure(`VC${key}`, {
          start,
          duration,
          detail: {
            devtools: {
              track: `VC ${entry.revision}`,
              trackGroup: '🛸 reactUFO metrics',
              color: key === '90' ? 'tertiary' : 'primary-light'
            }
          }
        });
      });
    });
  } catch {}
};
