// @atlaskit/tag/dist/esm/tag-new/avatar-tag.js
/* avatar-tag.tsx generated by @compiled/babel-plugin v0.38.1 */
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["text", "type", "avatar", "href", "linkComponent", "isRemovable", "removeButtonLabel", "onBeforeRemoveAction", "onAfterRemoveAction", "testId", "isVerified", "maxWidth"];
import "./avatar-tag.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef, memo } from 'react';
import StatusVerifiedIcon from '@atlaskit/icon/core/status-verified';
import { LinkWrapper, RemovableWrapper, useButtonInteraction, useLink, useRemoveButton, useTagRemoval } from './shared';

/**
 * The type values that AvatarTag accepts.
 */

/**
 * Props controlled by AvatarTag that will be passed to the avatar component.
 * These props are managed internally by AvatarTag.
 */

/**
 * Common props shared across all AvatarTag types.
 */

/**
 * Props for user avatar tags with circular avatars.
 */

/**
 * Props for other/teams avatar tags with square avatars.
 */

/**
 * Props for agent avatar tags with hexagonal avatars.
 */

/**
 * Props for AvatarTag component. Uses discriminated union based on `type`.
 */

var styles = {
  baseStyles: "_zulpv77o _2rko1rr0 _11c8wadc _1dqonqa1 _189ee4h9 _1mou1b66 _195g1b66 _1rjcze3t _1e0c116y _vchhusvi _1ul9idpf _p12ff6dh _4t3iv47k _kqswh2mm _4cvr1h6o _s7n4nkob _bfhksm61 _syazi7uo _80om73ad _y4tii2wt _bozgze3t",
  otherBaseStyles: "_2rko12b0 _y4ti1b66",
  agentBaseStyles: "_2rko12b0 _y4ti1b66",
  removableStyles: "_y4ti1l7b",
  userRemovableStyles: "_y4ti1l7b",
  avatarStyles: "_4bfu1r31 _1hmsglyw _ajmmnqa1 _2rko1rr0 _1reo15vq _18m915vq _1e0c116y _4cvr1h6o _1o9zidpf _ahbq1n1a",
  otherAvatarStyles: "_2rko12b0",
  agentAvatarStyles: "_2rkoidpf",
  textStyles: "_1reo15vq _18m915vq _1bto1l2s _o5721q9c _16jlkb7n _1ul9idpf _syazi7uo",
  afterStyles: "_1e0c1txw _4cvr1h6o _1o9zidpf _lcxv1wug _ahbqv77o _kqswh2mm",
  verifiedIconStyles: "_1e0c116y _4cvr1h6o _1o9zidpf _syaz6txe _ahbqt94y",
  focusRingStyles: "_1ygbdfik _1ah3v77o",
  childFocusRingStyles: "_1a37dfik _1cwgv77o",
  interactiveBaseStyles: "_80omtlke _7nm0v77o _59381r31 _yha2glyw _1nmznqa1 _1ufy15vq _turg15vq _jr6g1snc _1ymragmp _1y1s10yn _1hw7116y _6wji1h6o _1k6fidpf _43dsjdj5 _jwdiagmp _1cayjdj5 _102s1snc _yi8hagmp _h79h10yn _9m7f1r31 _w7swglyw _15a9nqa1 _jxlhagmp",
  interactiveHoverStyles: "_fs7a1r31 _1uxk8stv _12u8nqa1 _irr31dpa _oxk4jdj5 _1di6fcek"
};

/* eslint-disable @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors, @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-important-styles */
// Border color style - gray only for AvatarTag
var borderColorStyles = {
  root: "_119b8yml"
};

// Border filter styles - base state
var borderFilterStyles = {
  root: "_1h6dpryq _1k671038 _13muu67f _oe6mmko7 _u8p61h4e _hg5mr3uz _1easrqqc"
};

// Border filter styles - interactive states (hover/pressed)
var borderInteractiveFilterStyles = {
  root: "_18dicjiy _hyog26zw _ylgq1r5k _bg3q9by0 _l0666xdd _pclnn8xv _4cvxpc7y _q4zt1ij0 _75gvn2k3 _ui3h1vs5"
};
/* eslint-enable @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors, @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-important-styles */

/**
 * __AvatarTag__
 *
 * AvatarTag is a specialized tag for representing people, teams/other entities, or agents with an avatar and name.
 * The visual appearance (round, square, or hexagonal) is determined by the `type` prop.
 */
var AvatarTagComponent = /*#__PURE__*/forwardRef(function AvatarTag(_ref, ref) {
  var text = _ref.text,
    type = _ref.type,
    AvatarComponent = _ref.avatar,
    href = _ref.href,
    linkComponent = _ref.linkComponent,
    _ref$isRemovable = _ref.isRemovable,
    isRemovable = _ref$isRemovable === void 0 ? true : _ref$isRemovable,
    removeButtonLabel = _ref.removeButtonLabel,
    onBeforeRemoveAction = _ref.onBeforeRemoveAction,
    onAfterRemoveAction = _ref.onAfterRemoveAction,
    testId = _ref.testId,
    isVerified = _ref.isVerified,
    maxWidth = _ref.maxWidth,
    other = _objectWithoutProperties(_ref, _excluded);
  var _useTagRemoval = useTagRemoval(text, onBeforeRemoveAction, onAfterRemoveAction),
    status = _useTagRemoval.status,
    handleRemoveRequest = _useTagRemoval.handleRemoveRequest,
    onKeyPress = _useTagRemoval.onKeyPress,
    removingTag = _useTagRemoval.removingTag,
    showingTag = _useTagRemoval.showingTag;
  var _useLink = useLink(href, linkComponent),
    isLink = _useLink.isLink,
    LinkComponent = _useLink.LinkComponent;
  var _useButtonInteraction = useButtonInteraction(),
    isLinkHovered = _useButtonInteraction.isLinkHovered,
    isOverButton = _useButtonInteraction.isOverButton,
    isButtonFocused = _useButtonInteraction.isButtonFocused,
    isLinkFocused = _useButtonInteraction.isLinkFocused,
    buttonHandlers = _useButtonInteraction.buttonHandlers,
    linkHandlers = _useButtonInteraction.linkHandlers;

  // Determine styles based on type
  var isOtherType = type === 'other';
  var isAgentType = type === 'agent';
  var isUserType = type === 'user';

  // isVerified is only available for 'other' type
  var showVerified = isOtherType && isVerified;

  // Determine avatar appearance based on type
  var avatarAppearance = isUserType ? 'circle' : isAgentType ? 'hexagon' : 'square';

  // Determine remove button shape based on type
  var removeButtonShape = isUserType ? 'circle' : 'default';
  var removeButton = useRemoveButton({
    isRemovable: isRemovable,
    tagText: text,
    removeButtonLabel: removeButtonLabel,
    testId: testId,
    handleRemoveRequest: handleRemoveRequest,
    removingTag: removingTag,
    showingTag: showingTag,
    onKeyPress: onKeyPress,
    shape: removeButtonShape,
    buttonHandlers: buttonHandlers
  });

  // Render the avatar component with controlled props
  // Cast to ComponentType<AvatarRenderProps> to inject controlled props at runtime
  var AvatarWithControlledProps = AvatarComponent;
  var avatarElement = /*#__PURE__*/React.createElement(AvatarWithControlledProps, {
    size: "xsmall",
    appearance: avatarAppearance,
    borderColor: "transparent"
  });
  var tagContent = /*#__PURE__*/React.createElement("span", _extends({}, other, {
    ref: ref,
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: maxWidth !== undefined ? {
      maxWidth: maxWidth
    } : undefined,
    className: ax([styles.baseStyles, isOtherType && styles.otherBaseStyles, isAgentType && styles.agentBaseStyles, borderColorStyles.root, borderFilterStyles.root, isLink && styles.interactiveBaseStyles, isLink && styles.focusRingStyles, isLink && isLinkHovered && !isOverButton && borderInteractiveFilterStyles.root, isLink && isLinkHovered && !isOverButton && styles.interactiveHoverStyles, isRemovable && !isUserType && styles.removableStyles, isRemovable && isUserType && styles.userRemovableStyles, isLinkFocused && !isButtonFocused && styles.childFocusRingStyles])
  }), /*#__PURE__*/React.createElement(LinkWrapper, {
    isLink: isLink,
    href: href,
    LinkComponent: LinkComponent,
    testId: testId,
    linkHandlers: linkHandlers
  }, /*#__PURE__*/React.createElement("span", {
    className: ax([styles.avatarStyles, isOtherType && styles.otherAvatarStyles, isAgentType && styles.agentAvatarStyles])
  }, avatarElement), /*#__PURE__*/React.createElement("span", {
    "data-tag-text": true,
    className: ax([styles.textStyles])
  }, text)), showVerified && /*#__PURE__*/React.createElement("span", {
    className: ax([styles.verifiedIconStyles])
  }, /*#__PURE__*/React.createElement(StatusVerifiedIcon, {
    label: "Verified",
    size: "small"
  })), removeButton && /*#__PURE__*/React.createElement("span", {
    className: ax([styles.afterStyles])
  }, removeButton));
  return /*#__PURE__*/React.createElement(RemovableWrapper, {
    isRemovable: isRemovable,
    status: status
  }, tagContent);
});

/**
 * __Avatar tag__
 */
var AvatarTag = /*#__PURE__*/memo(AvatarTagComponent);
export default AvatarTag;
