// @atlaskit/tag/dist/esm/tag-new/tag-new.js
/* tag-new.tsx generated by @compiled/babel-plugin v0.38.1 */
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["color", "text", "elemBefore", "href", "linkComponent", "isRemovable", "removeButtonLabel", "onBeforeRemoveAction", "onAfterRemoveAction", "testId", "maxWidth"];
import "./tag-new.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef, memo } from 'react';
import { LinkWrapper, RemovableWrapper, useButtonInteraction, useLink, useRemoveButton, useTagRemoval } from './shared';

// CSS variable name for icon color - must be used as literal string in cssMap due to Compiled CSS static analysis
var iconColorVar = '--ds-tag-icon';
// Color mapping from old color names to new color names
export var colorMapping = {
  standard: 'gray',
  grey: 'gray',
  blue: 'blue',
  green: 'green',
  red: 'red',
  yellow: 'yellow',
  purple: 'purple',
  lime: 'lime',
  magenta: 'magenta',
  orange: 'orange',
  teal: 'teal',
  // Light variants map to same as their non-light counterparts
  greyLight: 'gray',
  blueLight: 'blue',
  greenLight: 'green',
  redLight: 'red',
  yellowLight: 'yellow',
  purpleLight: 'purple',
  limeLight: 'lime',
  magentaLight: 'magenta',
  orangeLight: 'orange',
  tealLight: 'teal'
};
var styles = {
  baseStyles: "_zulp1b66 _2rko12b0 _11c8wadc _1dqonqa1 _189ee4h9 _1mou1b66 _195g1b66 _1rjcv77o _18zr1mie _1e0c116y _vchhusvi _1ul9idpf _p12ff6dh _4t3i1k92 _kqswh2mm _4cvr1h6o _80om73ad _bfhksm61",
  removableStyles: "_zulp1b66",
  beforeStyles: "_4bfu1snc _1hmsagmp _ajmmvv82 _1e0c116y _4cvr1h6o _1o9zidpf _syaz125o",
  textStyles: "_1reo15vq _18m915vq _1bto1l2s _o5721q9c _16jlkb7n _1ul9idpf _syazi7uo",
  afterStyles: "_1e0c1txw _4cvr1h6o _1o9zidpf _lcxv1wug _kqswh2mm",
  focusRingStyles: "_1ygbdfik _1ah3v77o",
  childFocusRingStyles: "_1a37dfik _1cwgv77o",
  interactiveBaseStyles: "_80omtlke _1all1b66 _qsdj1r31 _17ncglyw _15hpnqa1 _1rg815vq _1yoo15vq _j8j91snc _1afaagmp _gdbt10yn _h5vp116y _1qxz1h6o _8zx0idpf _1km4jdj5 _wc1fagmp _9408jdj5 _usew1snc _ebtnagmp _uvk910yn _10nj1r31 _1xq0glyw _18x1nqa1 _2mpsagmp",
  interactiveHoverStyles: "_17501r31 _qrsc8stv _1wodnqa1 _irr31dpa _rlyhjdj5 _1di6fcek"
};

// Border and icon color styles - each color sets CSS variables for its token values
var colorStyles = {
  gray: "_119b8yml _1ib21diz",
  blue: "_119b1w55 _1ib26txe",
  green: "_119b1crs _1ib217hp",
  red: "_119b16y8 _1ib2nbxb",
  yellow: "_119bfeiv _1ib21oqg",
  purple: "_119b1hqq _1ib2uq3u",
  lime: "_119bnf89 _1ib28645",
  magenta: "_119buwd0 _1ib2v8hk",
  orange: "_119b1b2s _1ib21j6y",
  teal: "_119b1h3o _1ib28uof"
};

/* eslint-disable @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-important-styles, @atlaskit/ui-styling-standard/no-unsafe-values */
// Border + icon filter styles - base state
var borderIconFilterStyles = {
  root: "_1h6dpryq _1k671038 _13muu67f _oe6mmko7 _m71tr3uz _168qx19p _wi6h1o1e _4uq11yqs _u8p61h4e _hg5mr3uz _1t75u67f _1f28f1ug _189qkb7n _1easrqqc _1knhjbtf"
};

// Border + icon filter styles - interactive states (hover/pressed)
var borderIconInteractiveFilterStyles = {
  root: "_18dicjiy _hyog26zw _ylgq1r5k _bg3q9by0 _fdt01r5k _46v5hwvj _1iel6xdd _111i1h4e _l0666xdd _pclnn8xv _ockq1r5k _aewt1n8h _196p1btq _18mn783i _4cvxpc7y _zqoikzix _q4zt1ij0 _ba351r2o _75gvn2k3 _kgu3swso _ui3h1vs5 _1r84p7hp"
};
/* eslint-enable @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-important-styles, @atlaskit/ui-styling-standard/no-unsafe-values */

/**
 * __TagNew__
 *
 * TagNew is the visual uplift implementation of Tag
 */
var TagNewComponent = /*#__PURE__*/forwardRef(function TagNew(_ref, ref) {
  var _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'gray' : _ref$color,
    text = _ref.text,
    elemBefore = _ref.elemBefore,
    href = _ref.href,
    linkComponent = _ref.linkComponent,
    _ref$isRemovable = _ref.isRemovable,
    isRemovable = _ref$isRemovable === void 0 ? true : _ref$isRemovable,
    removeButtonLabel = _ref.removeButtonLabel,
    onBeforeRemoveAction = _ref.onBeforeRemoveAction,
    onAfterRemoveAction = _ref.onAfterRemoveAction,
    testId = _ref.testId,
    maxWidth = _ref.maxWidth,
    other = _objectWithoutProperties(_ref, _excluded);
  var _useTagRemoval = useTagRemoval(text, onBeforeRemoveAction, onAfterRemoveAction),
    status = _useTagRemoval.status,
    handleRemoveRequest = _useTagRemoval.handleRemoveRequest,
    onKeyPress = _useTagRemoval.onKeyPress,
    removingTag = _useTagRemoval.removingTag,
    showingTag = _useTagRemoval.showingTag;
  var _useLink = useLink(href, linkComponent),
    isLink = _useLink.isLink,
    LinkComponent = _useLink.LinkComponent;
  var _useButtonInteraction = useButtonInteraction(),
    isLinkHovered = _useButtonInteraction.isLinkHovered,
    isOverButton = _useButtonInteraction.isOverButton,
    isButtonFocused = _useButtonInteraction.isButtonFocused,
    isLinkFocused = _useButtonInteraction.isLinkFocused,
    buttonHandlers = _useButtonInteraction.buttonHandlers,
    linkHandlers = _useButtonInteraction.linkHandlers;
  var removeButton = useRemoveButton({
    isRemovable: isRemovable,
    tagText: text,
    removeButtonLabel: removeButtonLabel,
    testId: testId,
    handleRemoveRequest: handleRemoveRequest,
    removingTag: removingTag,
    showingTag: showingTag,
    onKeyPress: onKeyPress,
    buttonHandlers: buttonHandlers
  });
  var tagContent = /*#__PURE__*/React.createElement("span", _extends({}, other, {
    ref: ref,
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: maxWidth !== undefined ? {
      maxWidth: maxWidth
    } : undefined,
    className: ax([styles.baseStyles, colorStyles[color], borderIconFilterStyles.root, isLink && styles.interactiveBaseStyles, isLink && styles.focusRingStyles, isLink && isLinkHovered && !isOverButton && borderIconInteractiveFilterStyles.root, isLink && isLinkHovered && !isOverButton && styles.interactiveHoverStyles, isRemovable && styles.removableStyles, isLinkFocused && !isButtonFocused && styles.childFocusRingStyles])
  }), /*#__PURE__*/React.createElement(LinkWrapper, {
    isLink: isLink,
    href: href,
    LinkComponent: LinkComponent,
    testId: testId,
    linkHandlers: linkHandlers
  }, elemBefore && /*#__PURE__*/React.createElement("span", {
    className: ax([styles.beforeStyles])
  }, elemBefore), /*#__PURE__*/React.createElement("span", {
    "data-tag-text": true,
    className: ax([styles.textStyles])
  }, text)), removeButton && /*#__PURE__*/React.createElement("span", {
    className: ax([styles.afterStyles])
  }, removeButton));
  return /*#__PURE__*/React.createElement(RemovableWrapper, {
    isRemovable: isRemovable,
    status: status
  }, tagContent);
});

/**
 * __Tag new__
 */
var TagNew = /*#__PURE__*/memo(TagNewComponent);
export default TagNew;
