// @atlaskit/react-ufo/dist/cjs/create-interaction-extra-metrics-payload/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _coinflip = _interopRequireDefault(require("../coinflip"));
var _config = require("../config");
var _utils = require("../create-payload/common/utils");
var _getMoreAccuratePageVisibilityUpToTtai = require("../create-payload/utils/get-more-accurate-page-visibility-up-to-ttai");
var _getPageVisibilityUpToTtai = _interopRequireDefault(require("../create-payload/utils/get-page-visibility-up-to-ttai"));
var _getPayloadSize = _interopRequireDefault(require("../create-payload/utils/get-payload-size"));
var _getReactUfoPayloadVersion = require("../create-payload/utils/get-react-ufo-payload-version");
var _getTtai = _interopRequireDefault(require("../create-payload/utils/get-ttai"));
var _getVcMetrics = _interopRequireDefault(require("../create-payload/utils/get-vc-metrics"));
var _optimizeApdex = require("../create-payload/utils/optimize-apdex");
var _optimizeCustomTimings = require("../create-payload/utils/optimize-custom-timings");
var _optimizeHoldInfo = require("../create-payload/utils/optimize-hold-info");
var _optimizeMarks = require("../create-payload/utils/optimize-marks");
var _optimizeReactProfilerTimings = require("../create-payload/utils/optimize-react-profiler-timings");
var _optimizeRequestInfo = require("../create-payload/utils/optimize-request-info");
var _optimizeSpans = require("../create-payload/utils/optimize-spans");
var _interactionMetrics = require("../interaction-metrics");
var _excluded = ["labelStack"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function createInteractionExtraLogPayload(_x, _x2, _x3, _x4) {
  return _createInteractionExtraLogPayload.apply(this, arguments);
}
function _createInteractionExtraLogPayload() {
  _createInteractionExtraLogPayload = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(interactionId, interaction, lastInteractionFinish, lastInteractionFinishVCResult) {
    var _getTTAI, _lastInteractionFinis, _getTTAI2, _window$location;
    var config, end, start, ufoName, rate, type, abortReason, routeName, previousInteractionName, isPreviousInteractionAborted, abortedByInteractionName, knownSegments, minorInteractions, sanitisedUfoName, configRate, pageVisibilityAtTTAI, isPageLoad, calculatePageVisibilityFromTheStartOfPageLoad, moreAccuratePageVisibilityAtTTAI, extraTTAI, newUFOName, finalVCMetrics, vcRevisionPayload, effectiveVCRevisionPayload, normalTTAI, lastInteractionFinishStart, lastInteractionFinishEnd, lastInteractionFinishVC90, lastInteractionFinishVCClean, lastInteractionFinishVCRev, lastInteractionFinishRevision, isThirdParty, filteredData, getDetailedInteractionMetrics, segments3p, segmentTree, payload;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          config = (0, _config.getConfig)();
          if (config) {
            _context.next = 3;
            break;
          }
          throw Error('UFO Configuration not provided');
        case 3:
          end = interaction.end, start = interaction.start, ufoName = interaction.ufoName, rate = interaction.rate, type = interaction.type, abortReason = interaction.abortReason, routeName = interaction.routeName, previousInteractionName = interaction.previousInteractionName, isPreviousInteractionAborted = interaction.isPreviousInteractionAborted, abortedByInteractionName = interaction.abortedByInteractionName, knownSegments = interaction.knownSegments, minorInteractions = interaction.minorInteractions;
          sanitisedUfoName = (0, _utils.sanitizeUfoName)(ufoName);
          configRate = (0, _config.getExtraInteractionRate)(sanitisedUfoName, type);
          if ((0, _coinflip.default)(configRate)) {
            _context.next = 8;
            break;
          }
          return _context.abrupt("return", null);
        case 8:
          pageVisibilityAtTTAI = (0, _getPageVisibilityUpToTtai.default)(interaction);
          isPageLoad = type === 'page_load' || type === 'transition';
          if (!(!isPageLoad || minorInteractions !== undefined && minorInteractions.length > 0)) {
            _context.next = 12;
            break;
          }
          return _context.abrupt("return", null);
        case 12:
          calculatePageVisibilityFromTheStartOfPageLoad = config.enableBetterPageVisibilityApi && isPageLoad;
          moreAccuratePageVisibilityAtTTAI = calculatePageVisibilityFromTheStartOfPageLoad ? (0, _getMoreAccuratePageVisibilityUpToTtai.getMoreAccuratePageVisibilityUpToTTAI)(interaction) : null;
          extraTTAI = (_getTTAI = (0, _getTtai.default)(interaction)) !== null && _getTTAI !== void 0 ? _getTTAI : undefined;
          newUFOName = (0, _utils.sanitizeUfoName)(ufoName);
          _context.next = 18;
          return (0, _getVcMetrics.default)(interaction, true);
        case 18:
          finalVCMetrics = _context.sent;
          // Check if VC is clean and has valid metric
          vcRevisionPayload = finalVCMetrics === null || finalVCMetrics === void 0 ? void 0 : finalVCMetrics['ufo:vc:rev'];
          effectiveVCRevisionPayload = vcRevisionPayload === null || vcRevisionPayload === void 0 ? void 0 : vcRevisionPayload.find(function (_ref) {
            var revision = _ref.revision;
            return revision === _config.DEFAULT_TTVC_REVISION;
          });
          if (!(!(effectiveVCRevisionPayload !== null && effectiveVCRevisionPayload !== void 0 && effectiveVCRevisionPayload.clean) || (effectiveVCRevisionPayload === null || effectiveVCRevisionPayload === void 0 ? void 0 : effectiveVCRevisionPayload['metric:vc90']) === undefined || typeof (effectiveVCRevisionPayload === null || effectiveVCRevisionPayload === void 0 ? void 0 : effectiveVCRevisionPayload['metric:vc90']) !== 'number' || extraTTAI === undefined || typeof extraTTAI !== 'number' || interaction.errors.length > 0)) {
            _context.next = 23;
            break;
          }
          return _context.abrupt("return", null);
        case 23:
          if (!(!lastInteractionFinish || lastInteractionFinish !== null && lastInteractionFinish !== void 0 && lastInteractionFinish.abortReason || lastInteractionFinish !== null && lastInteractionFinish !== void 0 && (_lastInteractionFinis = lastInteractionFinish.errors) !== null && _lastInteractionFinis !== void 0 && _lastInteractionFinis.length)) {
            _context.next = 25;
            break;
          }
          return _context.abrupt("return", null);
        case 25:
          normalTTAI = (_getTTAI2 = (0, _getTtai.default)(lastInteractionFinish)) !== null && _getTTAI2 !== void 0 ? _getTTAI2 : undefined;
          lastInteractionFinishStart = typeof lastInteractionFinish.start === 'number' ? Math.round(lastInteractionFinish.start) : undefined;
          lastInteractionFinishEnd = typeof lastInteractionFinish.end === 'number' ? Math.round(lastInteractionFinish.end) : undefined;
          lastInteractionFinishVC90 = null;
          lastInteractionFinishVCClean = false;
          if (!lastInteractionFinishVCResult) {
            _context.next = 41;
            break;
          }
          lastInteractionFinishVCRev = lastInteractionFinishVCResult['ufo:vc:rev'];
          lastInteractionFinishRevision = lastInteractionFinishVCRev === null || lastInteractionFinishVCRev === void 0 ? void 0 : lastInteractionFinishVCRev.find(function (_ref2) {
            var revision = _ref2.revision;
            return revision === _config.DEFAULT_TTVC_REVISION;
          });
          if (!(lastInteractionFinishRevision !== null && lastInteractionFinishRevision !== void 0 && lastInteractionFinishRevision.clean)) {
            _context.next = 38;
            break;
          }
          lastInteractionFinishVCClean = true;
          lastInteractionFinishVC90 = lastInteractionFinishRevision['metric:vc90'];
          _context.next = 39;
          break;
        case 38:
          return _context.abrupt("return", null);
        case 39:
          _context.next = 42;
          break;
        case 41:
          if (normalTTAI !== undefined && typeof normalTTAI === 'number' && normalTTAI === extraTTAI) {
            // Because TTAI is equal between with and without 3p, we can assume VC90 is also equal
            lastInteractionFinishVC90 = effectiveVCRevisionPayload === null || effectiveVCRevisionPayload === void 0 ? void 0 : effectiveVCRevisionPayload['metric:vc90'];
            lastInteractionFinishVCClean = effectiveVCRevisionPayload === null || effectiveVCRevisionPayload === void 0 ? void 0 : effectiveVCRevisionPayload.clean;
          }
        case 42:
          // Helper function to check if labelStack contains third-party type
          isThirdParty = function isThirdParty(labelStack) {
            var _labelStack$some;
            return (_labelStack$some = labelStack === null || labelStack === void 0 ? void 0 : labelStack.some(function (entry) {
              return 'type' in entry && entry.type === 'third-party';
            })) !== null && _labelStack$some !== void 0 ? _labelStack$some : false;
          }; // Pre-filter 3p data
          filteredData = {
            errors: interaction.errors.filter(function (error) {
              return isThirdParty(error.labelStack);
            }),
            spans: [].concat((0, _toConsumableArray2.default)(interaction.spans), (0, _toConsumableArray2.default)(_interactionMetrics.interactionSpans)).filter(function (span) {
              return isThirdParty(span.labelStack);
            }),
            requestInfo: interaction.requestInfo.filter(function (req) {
              return isThirdParty(req.labelStack);
            }),
            customTimings: interaction.customTimings.filter(function (timing) {
              return isThirdParty(timing.labelStack);
            }),
            apdex: interaction.apdex.filter(function (apdex) {
              return isThirdParty(apdex.labelStack);
            }),
            reactProfilerTimings: interaction.reactProfilerTimings.filter(function (timing) {
              return isThirdParty(timing.labelStack);
            }),
            customData: interaction.customData.filter(function (data) {
              return isThirdParty(data.labelStack);
            }),
            segments: knownSegments.filter(function (segment) {
              return isThirdParty(segment.labelStack);
            }),
            marks: interaction.marks.filter(function (mark) {
              return isThirdParty(mark.labelStack);
            })
          }; // Clear atlaskit spans after filtering
          _interactionMetrics.interactionSpans.length = 0;

          // Detailed payload
          getDetailedInteractionMetrics = function getDetailedInteractionMetrics() {
            var _interaction$hold3pIn;
            return {
              errors: filteredData.errors.map(function (_ref3) {
                var labelStack = _ref3.labelStack,
                  others = (0, _objectWithoutProperties2.default)(_ref3, _excluded);
                return _objectSpread(_objectSpread({}, others), {}, {
                  labelStack: labelStack && (0, _utils.optimizeLabelStack)(labelStack, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type))
                });
              }),
              holdActive: interaction.hold3pActive ? (0, _toConsumableArray2.default)(interaction.hold3pActive.values()) : [],
              holdInfo: (0, _optimizeHoldInfo.optimizeHoldInfo)((_interaction$hold3pIn = interaction.hold3pInfo) !== null && _interaction$hold3pIn !== void 0 ? _interaction$hold3pIn : [], start, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)),
              spans: (0, _optimizeSpans.optimizeSpans)(filteredData.spans, start, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)),
              requestInfo: (0, _optimizeRequestInfo.optimizeRequestInfo)(filteredData.requestInfo, start, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)),
              customTimings: (0, _optimizeCustomTimings.optimizeCustomTimings)(filteredData.customTimings, start)
            };
          };
          segments3p = !(0, _platformFeatureFlags.fg)('platform_ufo_remove_deprecated_config_fields') && config.killswitchNestedSegments ? [] : filteredData.segments;
          segmentTree = (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type) === '2.0.0' ? (0, _utils.buildSegmentTree)(segments3p.map(function (segment) {
            return segment.labelStack;
          })) : {};
          payload = {
            actionSubject: 'experience',
            action: 'measured',
            eventType: 'operational',
            source: 'measured',
            tags: ['observability'],
            attributes: {
              properties: {
                // basic
                'event:hostname': ((_window$location = window.location) === null || _window$location === void 0 ? void 0 : _window$location.hostname) || 'unknown',
                'event:product': config.product,
                'event:schema': '1.0.0',
                'event:sizeInKb': 0,
                'event:source': {
                  name: 'react-ufo/web',
                  version: (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)
                },
                'event:region': config.region || 'unknown',
                'experience:key': 'custom.interaction-extra-metrics',
                'experience:name': newUFOName,
                interactionMetrics: _objectSpread(_objectSpread({
                  namePrefix: config.namePrefix || '',
                  segmentPrefix: config.segmentPrefix || '',
                  interactionId: interactionId,
                  pageVisibilityAtTTAI: pageVisibilityAtTTAI,
                  experimental__pageVisibilityAtTTAI: moreAccuratePageVisibilityAtTTAI,
                  // raw interaction metrics
                  rate: rate,
                  routeName: routeName,
                  type: type,
                  abortReason: abortReason,
                  previousInteractionName: previousInteractionName,
                  isPreviousInteractionAborted: isPreviousInteractionAborted,
                  abortedByInteractionName: abortedByInteractionName,
                  // performance
                  end: Math.round(end),
                  start: Math.round(start),
                  'metric:ttai:3p': extraTTAI
                }, finalVCMetrics), {}, {
                  segments: (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type) === '2.0.0' ? segmentTree : (0, _utils.getOldSegmentsLabelStack)(segments3p, interaction.type),
                  marks: (0, _optimizeMarks.optimizeMarks)(filteredData.marks, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)),
                  apdex: (0, _optimizeApdex.optimizeApdex)(filteredData.apdex, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)),
                  reactProfilerTimings: (0, _optimizeReactProfilerTimings.optimizeReactProfilerTimings)(filteredData.reactProfilerTimings, start, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type)),
                  customData: filteredData.customData
                }, getDetailedInteractionMetrics()),
                'vc:effective:revision': _config.DEFAULT_TTVC_REVISION,
                lastInteractionFinish: {
                  start: lastInteractionFinishStart,
                  end: lastInteractionFinishEnd,
                  ttai: normalTTAI,
                  vc90: lastInteractionFinishVC90,
                  vcClean: lastInteractionFinishVCClean
                }
              }
            }
          };
          payload.attributes.properties['event:sizeInKb'] = (0, _getPayloadSize.default)(payload.attributes.properties);
          return _context.abrupt("return", payload);
        case 51:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return _createInteractionExtraLogPayload.apply(this, arguments);
}
var _default = exports.default = createInteractionExtraLogPayload;
