// @atlaskit/app-provider/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ThemeProvider", {
  enumerable: true,
  get: function get() {
    return _themeProvider.default;
  }
});
Object.defineProperty(exports, "UNSAFE_useColorModeForMigration", {
  enumerable: true,
  get: function get() {
    return _useColorModeForMigration.UNSAFE_useColorModeForMigration;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _appProvider.default;
  }
});
Object.defineProperty(exports, "useColorMode", {
  enumerable: true,
  get: function get() {
    return _useColorMode.useColorMode;
  }
});
Object.defineProperty(exports, "useRouterLink", {
  enumerable: true,
  get: function get() {
    return _useRouterLink.default;
  }
});
Object.defineProperty(exports, "useSetColorMode", {
  enumerable: true,
  get: function get() {
    return _useSetColorMode.useSetColorMode;
  }
});
Object.defineProperty(exports, "useSetTheme", {
  enumerable: true,
  get: function get() {
    return _useSetTheme.useSetTheme;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function get() {
    return _useTheme.useTheme;
  }
});
var _appProvider = _interopRequireDefault(require("./app-provider"));
var _useColorModeForMigration = require("./theme-provider/hooks/use-color-mode-for-migration");
var _useColorMode = require("./theme-provider/hooks/use-color-mode");
var _useSetColorMode = require("./theme-provider/hooks/use-set-color-mode");
var _useSetTheme = require("./theme-provider/hooks/use-set-theme");
var _useTheme = require("./theme-provider/hooks/use-theme");
var _themeProvider = _interopRequireDefault(require("./theme-provider"));
var _useRouterLink = _interopRequireDefault(require("./router-link-provider/hooks/use-router-link"));
