// @atlaskit/object/dist/cjs/components/object-tile/object-tile-base.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ObjectTileBase;
var _react = _interopRequireDefault(require("react"));
var _tinyInvariant = _interopRequireDefault(require("tiny-invariant"));
var _tile = _interopRequireDefault(require("@atlaskit/tile"));
/**
 * __Object tile base__
 *
 * An object tile represents an Atlassian-specific content type.
 *
 */
function ObjectTileBase(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? '' : _ref$label,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    testId = _ref.testId,
    color = _ref.color,
    backgroundColor = _ref.backgroundColor,
    Icon = _ref.icon;
  // Validate that size is a valid `ObjectTileSize` (and won't allow Tile's `xxsmall` size to be passed)
  var validSizes = ['xsmall', 'small', 'medium', 'large', 'xlarge'];
  (0, _tinyInvariant.default)(validSizes.includes(size), "ObjectTile: size \"".concat(size, "\" is not valid. Valid sizes are: ").concat(validSizes.join(', ')));
  return /*#__PURE__*/_react.default.createElement(_tile.default, {
    label: "",
    backgroundColor: backgroundColor,
    size: size
  }, /*#__PURE__*/_react.default.createElement(Icon, {
    label: label,
    color: color,
    testId: testId
    // @ts-expect-error - Internal, undocumented prop
    ,
    shouldScale: true
  }));
}
