// @atlaskit/layering/dist/cjs/components/open-layer-observer/use-open-layer-observer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useOpenLayerObserver = useOpenLayerObserver;
var _react = require("react");
var _openLayerObserverContext = require("./open-layer-observer-context");
/**
 * A hook for use within an `OpenLayerObserver` component.  It will return `null` if there is no
 * `OpenLayerObserver` in the component tree.
 *
 * It provides access to:
 *
 * - `getCount`: a function that returns the current count of open layers under the observer.
 * - `onChange`: a function that allows you to subscribe to changes in the layer count. It returns a
 * cleanup function to unsubscribe, which you should call when the component unmounts.
 * - `closeLayers`: a function that closes all open layers.
 *
 * Example usage:
 * ```tsx
 * const openLayerObserver = useOpenLayerObserver();
 * useEffect(() => {
 *   return openLayerObserver?.onChange(function onChange({ count }) {
 *     // react to changes in the layer count
 *   });
 * }, [openLayerObserver]);
 * ```
 */
function useOpenLayerObserver() {
  var context = (0, _react.useContext)(_openLayerObserverContext.OpenLayerObserverContext);
  var publicAPI = (0, _react.useMemo)(function () {
    if (context === null) {
      return null;
    }
    return {
      getCount: context.getCount,
      onChange: context.onChange,
      closeLayers: context.closeLayers
    };
  }, [context]);
  return publicAPI;
}
