// @atlaskit/adf-schema-generator/dist/es2019/transforms/adfToPm/codeGenHelpers.js
import { capitalize } from '../../utils/capitalize';
export const convertTypeToTypeName = (type, suffix = 'Definition') => {
  return `${type.split('_').map(capitalize).join('')}${suffix}`;
};
export const _codeBlock = (...blocks) => {
  return blocks.filter(b => b !== '').join('\n\n');
};
function escape(str) {
  // handle special characters: \n
  return str.replace(/\n/gu, '\\n');
}
const buildADFAttributesType = (attrKey, attrValue) => {
  const key = `${attrKey}${attrValue.optional ? '?' : ''}`;
  switch (attrValue.type) {
    case 'string':
    case 'number':
    case 'boolean':
      return [`${key}: ${attrValue.type}`];
    case 'enum':
      return [`${key}: ${attrValue.values.map(v => `"${escape(v)}"`).join('|')}`];
    case 'array':
      return [`${key}: Array<${attrValue.items.type}>`];
    case 'object':
      return [`${key}: Record<string, unknown>`];
    default:
      throw new Error(`Unknown type for "${attrKey}". ADF Attributes must have a type defined in the DSL.`);
  }
};
export const buildADFAttributesTypes = attrs => {
  return Object.entries(attrs).flatMap(([attrKey, attrValue]) => buildADFAttributesType(attrKey, attrValue));
};
export const _interface = (name, body) => {
  return `export interface ${name} {${body}}`;
};
export const _type = (name, body) => {
  return `export type ${name} = ${body}`;
};
export const _functionCallToVariable = (variableName, functionName, args) => {
  return `export const ${variableName} = ${functionName}(${args.join(',')})`;
};
export const _namedImport = (pkg, ...names) => {
  return `import { ${names.join(', ')} } from '${pkg}'`;
};
export const _namedTypeImport = (pkg, ...names) => {
  return `import type { ${names.join(', ')} } from '${pkg}'`;
};

/**
 * This helper function is used to keep undefined in the generated code,
 * because JSON.stringify will strip out undefined key-value pair by default.
 * @param obj object
 * @returns
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function stringifyWithUndefined(obj) {
  return JSON.stringify(obj, (_key, value) => {
    return value !== undefined ? value : 'undefined';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }).replace(/"undefined"/gu, undefined);
}
