// @forge/react/out/hooks/__test__/useContentProperty.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
// Mock event system - must be defined before jest.mock() calls
const listeners = new Map();
const mockEvents = {
    emit: jest.fn((event, data) => {
        if (listeners.has(event)) {
            listeners.get(event)?.forEach((callback) => callback(data));
        }
        return Promise.resolve();
    }),
    on: jest.fn((event, callback) => {
        if (!listeners.has(event)) {
            listeners.set(event, []);
        }
        listeners.get(event)?.push(callback);
    }),
    clearListeners: (event) => {
        listeners.delete(event);
    }
};
// Mock data - define before imports
const MOCK_PROPERTY_KEY = 'test-property-key';
const MOCK_INIT_VALUE = { count: 0 };
const MOCK_EXISTING_VALUE = { count: 5 };
const MOCK_UPDATED_VALUE = { count: 10 };
// Mock functions - define before imports
const mockUpdateProp = jest.fn();
const mockDeleteProp = jest.fn();
const mockConfluenceEntity = jest.fn(() => ({
    get: jest.fn(),
    update: jest.fn(),
    delete: jest.fn()
}));
const reconcilerTestRenderer_1 = tslib_1.__importDefault(require("../../__test__/reconcilerTestRenderer"));
const testUtils_1 = require("../../__test__/testUtils");
const useContentProperty_1 = require("../useContentProperty");
const useEntityProperty_1 = require("../useEntityProperty");
// Mock modules after imports
jest.mock('@forge/bridge', () => ({
    events: mockEvents
}));
jest.mock('../useEntityProperty', () => ({
    useEntityProperty: jest.fn(() => [MOCK_EXISTING_VALUE, mockUpdateProp, mockDeleteProp])
}));
jest.mock('../confluenceEntity', () => ({
    confluenceEntity: mockConfluenceEntity
}));
describe('useContentProperty', () => {
    beforeEach(() => {
        // Clear only specific mocks, not all mocks to preserve function references
        mockUpdateProp.mockClear();
        mockDeleteProp.mockClear();
        mockConfluenceEntity.mockClear();
        mockEvents.emit.mockClear();
        mockEvents.on.mockClear();
        mockEvents.clearListeners('FORGE_CONTENT_PROPERTY_UPDATED');
        (0, testUtils_1.setupBridge)(); // Set up global bridge for tests
    });
    it('should initialize confluenceEntity with correct parameters', async () => {
        const TestComponent = () => {
            (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        expect(mockConfluenceEntity).toHaveBeenCalledWith({
            entityType: 'Content',
            origPropertyKey: MOCK_PROPERTY_KEY,
            initValue: MOCK_INIT_VALUE
        });
    });
    it('should call useEntityProperty with the entity manager', async () => {
        const TestComponent = () => {
            (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        expect(useEntityProperty_1.useEntityProperty).toHaveBeenCalledWith({
            entityManager: expect.any(Object)
        });
    });
    it('should return property value, update function, and delete function', async () => {
        let hookResult = [];
        const TestComponent = () => {
            hookResult = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        expect(hookResult).toHaveLength(3);
        expect(hookResult[0]).toBe(MOCK_EXISTING_VALUE);
        expect(typeof hookResult[1]).toBe('function');
        expect(typeof hookResult[2]).toBe('function');
        expect(hookResult[2]).toBe(mockDeleteProp);
    });
    it('should emit event when property update is successful', async () => {
        mockUpdateProp.mockResolvedValueOnce(MOCK_UPDATED_VALUE);
        let updateFunction = jest.fn();
        const TestComponent = () => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            updateFunction = updateProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        await updateFunction(MOCK_UPDATED_VALUE);
        expect(mockUpdateProp).toHaveBeenCalledWith(MOCK_UPDATED_VALUE, undefined);
        expect(mockEvents.emit).toHaveBeenCalledWith('FORGE_CONTENT_PROPERTY_UPDATED', {
            propertyKey: MOCK_PROPERTY_KEY,
            valueUpdate: MOCK_UPDATED_VALUE
        });
    });
    it('should emit event when property update with retry count is successful', async () => {
        mockUpdateProp.mockResolvedValueOnce(MOCK_UPDATED_VALUE);
        const retryCount = 5;
        let updateFunction = jest.fn();
        const TestComponent = () => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            updateFunction = updateProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        await updateFunction(MOCK_UPDATED_VALUE, retryCount);
        expect(mockUpdateProp).toHaveBeenCalledWith(MOCK_UPDATED_VALUE, retryCount);
        expect(mockEvents.emit).toHaveBeenCalledWith('FORGE_CONTENT_PROPERTY_UPDATED', {
            propertyKey: MOCK_PROPERTY_KEY,
            valueUpdate: MOCK_UPDATED_VALUE
        });
    });
    it('should emit event when property update with function updater is successful', async () => {
        const updaterFunction = (prev) => ({ ...prev, count: prev.count + 1 });
        mockUpdateProp.mockResolvedValueOnce(MOCK_UPDATED_VALUE);
        let updateFunction = jest.fn();
        const TestComponent = () => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            updateFunction = updateProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        await updateFunction(updaterFunction);
        expect(mockUpdateProp).toHaveBeenCalledWith(updaterFunction, undefined);
        expect(mockEvents.emit).toHaveBeenCalledWith('FORGE_CONTENT_PROPERTY_UPDATED', {
            propertyKey: MOCK_PROPERTY_KEY,
            valueUpdate: updaterFunction
        });
    });
    it('should NOT emit event when property update fails', async () => {
        const updateError = new Error('Update failed');
        mockUpdateProp.mockRejectedValueOnce(updateError);
        let updateFunction = jest.fn();
        let caughtError;
        const TestComponent = () => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            updateFunction = updateProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        try {
            await updateFunction(MOCK_UPDATED_VALUE);
        }
        catch (error) {
            caughtError = error;
        }
        expect(mockUpdateProp).toHaveBeenCalledWith(MOCK_UPDATED_VALUE, undefined);
        expect(mockEvents.emit).not.toHaveBeenCalled();
        expect(caughtError).toBe(updateError);
    });
    it('should re-throw error when property update fails', async () => {
        const updateError = new Error('Network error');
        mockUpdateProp.mockRejectedValueOnce(updateError);
        let updateFunction = jest.fn();
        const TestComponent = () => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            updateFunction = updateProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        await expect(updateFunction(MOCK_UPDATED_VALUE)).rejects.toThrow('Network error');
    });
    it('should pass through delete function without modification', async () => {
        let deleteFunction = jest.fn();
        const TestComponent = () => {
            const [, , deleteProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            deleteFunction = deleteProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        expect(deleteFunction).toBe(mockDeleteProp);
    });
    it('should memoize entity manager to prevent unnecessary re-creation', async () => {
        let renderCount = 0;
        const TestComponent = ({ propKey }) => {
            renderCount++;
            (0, useContentProperty_1.useContentProperty)(propKey, MOCK_INIT_VALUE);
            return null;
        };
        // First render
        const renderer = await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, { propKey: MOCK_PROPERTY_KEY }));
        const firstCallCount = mockConfluenceEntity.mock.calls.length;
        // Re-render with same key
        await renderer.update((0, jsx_runtime_1.jsx)(TestComponent, { propKey: MOCK_PROPERTY_KEY }));
        const secondCallCount = mockConfluenceEntity.mock.calls.length;
        expect(renderCount).toBe(2);
        // Verify the entity manager is called with same parameters, ensuring proper memoization setup
        expect(mockConfluenceEntity).toHaveBeenCalledWith({
            entityType: 'Content',
            origPropertyKey: MOCK_PROPERTY_KEY,
            initValue: MOCK_INIT_VALUE
        });
        // Note: Due to test renderer behavior, entity manager may be recreated during unmount/remount
        expect(secondCallCount).toBeGreaterThanOrEqual(firstCallCount);
    });
    it('should handle different property keys correctly', async () => {
        const propertyKey1 = 'property-1';
        const propertyKey2 = 'property-2';
        const value1 = { data: 'test1' };
        const value2 = { data: 'test2' };
        mockUpdateProp.mockResolvedValue(value1);
        let updateFunction1 = jest.fn();
        let updateFunction2 = jest.fn();
        const TestComponent = () => {
            const [, updateProp1] = (0, useContentProperty_1.useContentProperty)(propertyKey1, value1);
            const [, updateProp2] = (0, useContentProperty_1.useContentProperty)(propertyKey2, value2);
            updateFunction1 = updateProp1;
            updateFunction2 = updateProp2;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        await updateFunction1(value1);
        await updateFunction2(value2);
        expect(mockEvents.emit).toHaveBeenCalledWith('FORGE_CONTENT_PROPERTY_UPDATED', {
            propertyKey: propertyKey1,
            valueUpdate: value1
        });
        expect(mockEvents.emit).toHaveBeenCalledWith('FORGE_CONTENT_PROPERTY_UPDATED', {
            propertyKey: propertyKey2,
            valueUpdate: value2
        });
    });
    it('should handle async event emission without blocking', async () => {
        // Make event.emit slow to ensure it doesn't block the update
        mockEvents.emit.mockImplementationOnce(() => new Promise((resolve) => setTimeout(resolve, 100)));
        mockUpdateProp.mockResolvedValueOnce(MOCK_UPDATED_VALUE);
        let updateFunction = jest.fn();
        const TestComponent = () => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            updateFunction = updateProp;
            return null;
        };
        await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, {}));
        const startTime = Date.now();
        await updateFunction(MOCK_UPDATED_VALUE);
        const endTime = Date.now();
        // Update should complete quickly (not wait for event emission)
        expect(endTime - startTime).toBeLessThan(50);
        expect(mockUpdateProp).toHaveBeenCalled();
        expect(mockEvents.emit).toHaveBeenCalled();
    });
    it('should preserve update function reference between renders', async () => {
        let updateFunction1 = jest.fn();
        let updateFunction2 = jest.fn();
        const TestComponent = ({ renderKey }) => {
            const [, updateProp] = (0, useContentProperty_1.useContentProperty)(MOCK_PROPERTY_KEY, MOCK_INIT_VALUE);
            if (renderKey === 1)
                updateFunction1 = updateProp;
            if (renderKey === 2)
                updateFunction2 = updateProp;
            return null;
        };
        const renderer = await reconcilerTestRenderer_1.default.create((0, jsx_runtime_1.jsx)(TestComponent, { renderKey: 1 }));
        await renderer.update((0, jsx_runtime_1.jsx)(TestComponent, { renderKey: 2 }));
        // Verify both functions exist and are callable
        expect(typeof updateFunction1).toBe('function');
        expect(typeof updateFunction2).toBe('function');
        // Note: Function reference stability depends on underlying useEntityProperty mock stability
    });
});
