// @forge/react/out/hooks/__test__/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sleep = exports.simpleBridgeEvents = void 0;
const listeners = new Map();
exports.simpleBridgeEvents = {
    on: (event, callback) => {
        if (!listeners.has(event)) {
            listeners.set(event, []);
        }
        listeners.get(event)?.push(callback);
        return Promise.resolve({
            unsubscribe: () => {
                const eventListeners = listeners.get(event);
                if (eventListeners) {
                    const index = eventListeners.indexOf(callback);
                    if (index > -1) {
                        eventListeners.splice(index, 1);
                    }
                }
            }
        });
    },
    emit: (event, data) => {
        if (listeners.has(event)) {
            listeners.get(event)?.forEach((callback) => callback(data));
        }
    },
    clearListeners: (event) => {
        listeners.delete(event);
    }
};
// Give a tick for state to update
const sleep = () => new Promise((resolve) => setTimeout(resolve, 0));
exports.sleep = sleep;
