// @atlaskit/tmp-editor-statsig/dist/cjs/editor-experiments-test-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.eeTest = eeTest;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _setup = require("./setup");
var _experimentsConfig = require("./experiments-config");
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /* eslint-disable @atlaskit/editor/no-re-export */ // Entry file in package.json
// This is loosely based on the `ffTest` util from `@atlassian/feature-flags-test-utils` package.
/**
 * This is a utility function for testing editor experiments.
 *
 * @example Boolean experiment
 * ```ts
 * eeTest('example-boolean', {
 *   true: () => {
 *     expect(editorExperiment('example-boolean', true)).toBe(true);
 *     expect(editorExperiment('example-boolean', false)).toBe(false);
 *   },
 *   false: () => {
 *     expect(editorExperiment('example-boolean', false)).toBe(true);
 *     expect(editorExperiment('example-boolean', true)).toBe(false);
 *   },
 * })
 * ```
 *
 * @example Multivariate experiment
 * ```ts
 * eeTest('example-multivariate', {
 *   one: () => {
 *     expect(editorExperiment('example-multivariate', 'one')).toBe(true);
 *     expect(editorExperiment('example-multivariate', 'two')).toBe(false);
 *   },
 *   two: () => {
 *     expect(editorExperiment('example-multivariate', 'two')).toBe(true);
 *     expect(editorExperiment('example-multivariate', 'one')).toBe(false);
 *   },
 *   three: () => {
 *     expect(editorExperiment('example-multivariate', 'three')).toBe(true);
 *     expect(editorExperiment('example-multivariate', 'one')).toBe(false);
 *   },
 * })
 * ```
 *
 * API based on Legacy ffTest API
 * - https://hello.atlassian.net/wiki/spaces/AF/pages/2569505829/Task+Testing+your+feature+flag+in+platform+and+product#Legacy-API-lEGACY
 */
function eeTest(experimentName, cases, otherExperiments) {
  (0, _setup.setupEditorExperiments)('test', {});
  describe("eeTest: ".concat(experimentName), function () {
    afterEach(function () {
      (0, _setup.setupEditorExperiments)('test', {});
    });
    var isBooleanExperiment = typeof _experimentsConfig.editorExperimentsConfig[experimentName].defaultValue === 'boolean';
    if (isBooleanExperiment && Object.keys(cases).length !== 2) {
      throw new Error("Expected exactly 2 cases for boolean experiment ".concat(experimentName, ", got ").concat(Object.keys(cases).length));
    }
    test.each(Object.keys(cases))("".concat(experimentName, ": %s"), /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(value) {
        var testCaseKey, convertedValue, testCase;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              testCaseKey = value; // For boolean experiments, we need to convert the 'on' and 'off' cases to boolean `true` and `false` values.
              convertedValue = isBooleanExperiment ? testCaseKey === 'true' ? true : false : testCaseKey;
              (0, _setup.setupEditorExperiments)('test', _objectSpread((0, _defineProperty2.default)({}, experimentName, convertedValue), otherExperiments));
              testCase = cases[testCaseKey]; // @ts-ignore
              _context.next = 6;
              return Promise.resolve(testCase());
            case 6:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
  });
}

// Namespace declaration to add describe property to eeTest function type
// TypeScript automatically merges function and namespace declarations with the same name
// eslint-disable-next-line @typescript-eslint/no-namespace

/**
 * eeTest.describe() Wrapper utility for describe() that runs a test with a editor experiment overides.
 *
 * @example
 * ```ts
 * eeTest.describe('exp-name', 'description of test').variant(true, () => {
 *   it('should do the thing', () => {
 *     expect(editorExperiment('example-boolean', true)).toBe(true);
 *   });
 * })
 * ```
 *
 * @example
 * ```ts
 * eeTest.describe('exp-name', 'description of test').each(() => {
 *   it('should do the thing', () => {
 *     expect(editorExperiment('example-boolean', true)).toBe(true);
 *   });
 * })
 * ```
 *
 * API based on next gen ffTest API
 * - https://hello.atlassian.net/wiki/spaces/AF/pages/2569505829/Task+Testing+your+feature+flag+in+platform+and+product#Next-Generation-API-%E2%9C%A8
 */
eeTest.describe = function eeTestDescribe(experimentName, describeName) {
  function eeTest(value) {
    var describeBody = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : function () {};
    describe("".concat(describeName, " [").concat(value, "]"), function () {
      beforeEach(function () {
        (0, _setup.setupEditorExperiments)('test', (0, _defineProperty2.default)({}, experimentName, value));
      });
      afterEach(function () {
        (0, _setup.setupEditorExperiments)('test', {});
      });
      describeBody();
    });
    (0, _setup.setupEditorExperiments)('test', {});
  }
  function variant(value, describeBody) {
    eeTest(value, describeBody);
  }
  function each(describeBody) {
    var possibleValues =
    // @ts-ignore
    _experimentsConfig.editorExperimentsConfig[experimentName].typeGuard.values;
    if (possibleValues) {
      var _iterator = _createForOfIteratorHelper(possibleValues),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var value = _step.value;
          eeTest(value, describeBody);
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    } else {
      // if there are no possible values, it's a boolean experiment
      eeTest(true, describeBody);
      eeTest(false, describeBody);
    }
  }
  return {
    variant: variant,
    each: each
  };
};
