// @atlaskit/adf-schema/dist/es2019/schema/nodes/embed-card.js
import { embedCard as embedCardFactory } from '../../next-schema/generated/nodeTypes';
import { uuid } from '../../utils/uuid';

/**
 * @name embedCard_node
 */

export const embedCard = embedCardFactory({
  parseDOM: [{
    tag: 'div[data-embed-card]',
    getAttrs: dom => ({
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      url: dom.getAttribute('data-card-url'),
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      layout: dom.getAttribute('data-layout') || 'center',
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      width: Number(dom.getAttribute('data-width')) || null,
      originalWidth:
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      Number(dom.getAttribute('data-card-original-width')) || null,
      originalHeight:
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      Number(dom.getAttribute('data-card-original-height')) || null
    })
  }],
  toDOM(node) {
    const {
      url,
      layout,
      width,
      originalWidth,
      originalHeight
    } = node.attrs;
    const attrs = {
      'data-embed-card': '',
      'data-card-url': url,
      'data-layout': layout,
      'data-width': width,
      'data-original-width': originalWidth,
      'data-original-height': originalHeight
    };
    return ['div', attrs];
  }
});
export const embedCardWithLocalId = embedCardFactory({
  parseDOM: [{
    tag: 'div[data-embed-card]',
    getAttrs: dom => ({
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      url: dom.getAttribute('data-card-url'),
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      layout: dom.getAttribute('data-layout') || 'center',
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      width: Number(dom.getAttribute('data-width')) || null,
      originalWidth:
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      Number(dom.getAttribute('data-card-original-width')) || null,
      originalHeight:
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      Number(dom.getAttribute('data-card-original-height')) || null,
      localId: uuid.generate()
    })
  }],
  toDOM(node) {
    var _node$attrs;
    const {
      url,
      layout,
      width,
      originalWidth,
      originalHeight
    } = node.attrs;
    const attrs = {
      'data-embed-card': '',
      'data-card-url': url,
      'data-layout': layout,
      'data-width': width,
      'data-original-width': originalWidth,
      'data-original-height': originalHeight,
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    };
    return ['div', attrs];
  }
});
