// @atlaskit/navigation-system/dist/cjs/ui/page-layout/main/main.js
/* main.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Main = Main;
require("./main.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _skipLinksContext = require("../../../context/skip-links/skip-links-context");
var _constants = require("../constants");
var _idUtils = require("../id-utils");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var mainElementStyles = {
  root: "_nd5l1gzg _1reo1wug _18m91wug _19121cl4 _152timx3 _qwfh1wug _165teqxy _13wn1if8"
};

/**
 * Use the Main area for the main page content. It has a fluid width and will expand to fill available space.
 */
function Main(_ref) {
  var children = _ref.children,
    xcss = _ref.xcss,
    _ref$skipLinkLabel = _ref.skipLinkLabel,
    skipLinkLabel = _ref$skipLinkLabel === void 0 ? 'Main Content' : _ref$skipLinkLabel,
    testId = _ref.testId,
    providedId = _ref.id;
  var id = (0, _idUtils.useLayoutId)({
    providedId: providedId
  });
  (0, _skipLinksContext.useSkipLink)(id, skipLinkLabel);
  return /*#__PURE__*/React.createElement(_react.Fragment, null, /*#__PURE__*/React.createElement("div", {
    id: id,
    "data-layout-slot": true,
    className: (0, _runtime.ax)([mainElementStyles.root, xcss]),
    role: "main",
    "data-testid": testId
  }, children));
}
