// @atlaskit/analytics-next/dist/cjs/components/AnalyticsDecorator/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof3 = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ContextTypes = exports.AnalyticsDecorator = void 0;
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _matchEvent = _interopRequireDefault(require("../matchEvent"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof3(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var ContextTypes = exports.ContextTypes = {
  onAnalyticsEvent: _propTypes.default.func,
  getParentAnalyticsData: _propTypes.default.func
};

/**
 * The Decorator component extends analytics event data
 * for any events fired by its descendents,
 * then passes the event up the hierarchy
 */
// eslint-disable-next-line @repo/internal/react/no-class-components
var AnalyticsDecorator = exports.AnalyticsDecorator = /*#__PURE__*/function (_Component) {
  function AnalyticsDecorator() {
    var _this;
    (0, _classCallCheck2.default)(this, AnalyticsDecorator);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, AnalyticsDecorator, [].concat(args));
    (0, _defineProperty2.default)(_this, "getDecoratedAnalyticsData", function (name, srcData, isPrivate) {
      var _this$props = _this.props,
        data = _this$props.data,
        getData = _this$props.getData,
        match = _this$props.match,
        matchPrivate = _this$props.matchPrivate;
      var decoratedData = _objectSpread({}, srcData);
      if (matchPrivate === isPrivate && (0, _matchEvent.default)(match, name)) {
        if ((0, _typeof2.default)(data) === 'object') {
          Object.assign(decoratedData, data);
        }
        if (typeof getData === 'function') {
          Object.assign(decoratedData, getData(name, decoratedData));
        }
      }
      return decoratedData;
    });
    (0, _defineProperty2.default)(_this, "onAnalyticsEvent", function (name, srcData, isPrivate) {
      var _ref = _this.context,
        onAnalyticsEvent = _ref.onAnalyticsEvent;
      if (typeof onAnalyticsEvent !== 'function') {
        return;
      }
      var decoratedData = _this.getDecoratedAnalyticsData(name, srcData, isPrivate);
      onAnalyticsEvent(name, decoratedData, isPrivate);
    });
    (0, _defineProperty2.default)(_this, "getParentAnalyticsData", function (name, isPrivate) {
      var parentData = _this.getDecoratedAnalyticsData(name, {}, isPrivate);
      var _ref2 = _this.context,
        getParentAnalyticsData = _ref2.getParentAnalyticsData;
      if (typeof getParentAnalyticsData === 'function') {
        Object.assign(parentData, getParentAnalyticsData(name, isPrivate));
      }
      return parentData;
    });
    return _this;
  }
  (0, _inherits2.default)(AnalyticsDecorator, _Component);
  return (0, _createClass2.default)(AnalyticsDecorator, [{
    key: "getChildContext",
    value: function getChildContext() {
      return {
        onAnalyticsEvent: this.onAnalyticsEvent,
        getParentAnalyticsData: this.getParentAnalyticsData
      };
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return _react.default.Children.only(children);
    }
  }]);
}(_react.Component);
(0, _defineProperty2.default)(AnalyticsDecorator, "defaultProps", {
  match: '*',
  matchPrivate: false
});
(0, _defineProperty2.default)(AnalyticsDecorator, "contextTypes", ContextTypes);
(0, _defineProperty2.default)(AnalyticsDecorator, "childContextTypes", ContextTypes);
var _default = exports.default = AnalyticsDecorator;
