// @atlaskit/react-ufo/dist/cjs/interaction-metrics/interaction-extra-metrics.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _vc = require("../vc");
var _constants = require("./common/constants");
var _index = require("./index");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var InteractionExtraMetrics = exports.default = /*#__PURE__*/function () {
  function InteractionExtraMetrics() {
    (0, _classCallCheck2.default)(this, InteractionExtraMetrics);
    // Store the finished interaction (as non-3p interaction)
    (0, _defineProperty2.default)(this, "finishedInteraction", null);
    // independent VC observer, that observes until `custom.post-interaction-logs` event is sent
    (0, _defineProperty2.default)(this, "vcObserver", null);
    (0, _defineProperty2.default)(this, "sinkHandlerFn", function () {});
  }
  return (0, _createClass2.default)(InteractionExtraMetrics, [{
    key: "initializeVCObserver",
    value: function initializeVCObserver(options) {
      this.vcObserver = new _vc.VCObserverWrapper(_objectSpread(_objectSpread({}, options), {}, {
        isPostInteraction: true
      }));
    }
  }, {
    key: "startVCObserver",
    value: function startVCObserver(_ref, interactionId) {
      var startTime = _ref.startTime;
      if (this.eligibleToMeasure(interactionId)) {
        var _this$vcObserver;
        (_this$vcObserver = this.vcObserver) === null || _this$vcObserver === void 0 || _this$vcObserver.start({
          startTime: startTime
        });
      }
    }
  }, {
    key: "stopVCObserver",
    value: function stopVCObserver() {
      var _this$vcObserver2;
      (_this$vcObserver2 = this.vcObserver) === null || _this$vcObserver2 === void 0 || _this$vcObserver2.stop();
    }

    // Check if the current interaction is eligible for measurement
  }, {
    key: "eligibleToMeasure",
    value: function eligibleToMeasure(interactionId) {
      var interaction = _constants.interactions.get(interactionId);
      return (interaction === null || interaction === void 0 ? void 0 : interaction.type) === 'page_load' || (interaction === null || interaction === void 0 ? void 0 : interaction.type) === 'transition';
    }
  }, {
    key: "updateFinishedInteraction",
    value: function updateFinishedInteraction(interaction) {
      if ((interaction === null || interaction === void 0 ? void 0 : interaction.type) === 'page_load' || (interaction === null || interaction === void 0 ? void 0 : interaction.type) === 'transition') {
        this.finishedInteraction = interaction;
      }
    }
  }, {
    key: "setLastInteractionFinishVCResult",
    value: function setLastInteractionFinishVCResult(result) {
      this.lastInteractionFinishVCResult = result;
    }
  }, {
    key: "sinkHandler",
    value: function sinkHandler(fn) {
      this.sinkHandlerFn = fn;
    }
  }, {
    key: "onInteractionComplete",
    value: function onInteractionComplete(id, data) {
      if (data.ufoName) {
        var _this$vcObserver3;
        var updatedData = _objectSpread(_objectSpread({}, data), {}, {
          vcObserver: (_this$vcObserver3 = this.vcObserver) !== null && _this$vcObserver3 !== void 0 ? _this$vcObserver3 : undefined
        });
        this.sinkHandlerFn(id, updatedData, this.finishedInteraction, this.lastInteractionFinishVCResult);
      }
      this.stopVCObserver();
      (0, _index.remove)(id);
      this.reset();
    }
  }, {
    key: "reset",
    value: function reset() {
      var _this$finishedInterac;
      this.stopVCObserver();
      if ((_this$finishedInterac = this.finishedInteraction) !== null && _this$finishedInterac !== void 0 && _this$finishedInterac.id) {
        (0, _index.remove)(this.finishedInteraction.id);
      }
      this.finishedInteraction = null;
      this.lastInteractionFinishVCResult = undefined;
    }
  }, {
    key: "stopAll",
    value: function stopAll(id) {
      (0, _index.remove)(id);
      this.reset();
    }
  }]);
}();
