// @atlaskit/select/dist/esm/checkbox-select.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["components"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import React, { useMemo } from 'react';
import { CheckboxOption } from './components/input-options';
import Select from './select';
var CheckboxSelectInternal = function CheckboxSelectInternal(_ref) {
  var components = _ref.components,
    props = _objectWithoutProperties(_ref, _excluded);
  var mergedComponents = useMemo(function () {
    return _objectSpread(_objectSpread({}, components), {}, {
      Option: CheckboxOption
    });
  }, [components]);
  return /*#__PURE__*/React.createElement(Select, _extends({
    closeMenuOnSelect: false,
    hideSelectedOptions: false,
    isMulti: true
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
  }, props, {
    components: mergedComponents
  }));
};
var CheckboxSelect = /*#__PURE__*/React.memo(CheckboxSelectInternal
// Type casting as `React.memo` does not forward generic types
// Reference: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/37087#issuecomment-656596623
);

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export default CheckboxSelect;
