// @atlaskit/adf-schema/dist/cjs/steps/link-meta-step.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stepType = exports.invertStepType = exports.LinkMetaStep = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _model = require("@atlaskit/editor-prosemirror/model");
var _transform = require("@atlaskit/editor-prosemirror/transform");
var _excluded = ["sourceEvent"];
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var stepType = exports.stepType = 'editor-linking-meta';
var invertStepType = exports.invertStepType = 'editor-linking-meta-invert';
/**
 * Custom Prosemirror Step to attach metadata about user interactions with links
 * Using a Step means that it will work with prosemirror-history and we get utilise when
 * firing events on history change
 */
var LinkMetaStep = exports.LinkMetaStep = /*#__PURE__*/function (_Step) {
  function LinkMetaStep(pos, metadata) {
    var _this;
    var isInverted = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
    (0, _classCallCheck2.default)(this, LinkMetaStep);
    _this = _callSuper(this, LinkMetaStep);
    _this.pos = pos;
    _this.metadata = metadata;
    _this.isInverted = isInverted;
    return _this;
  }
  (0, _inherits2.default)(LinkMetaStep, _Step);
  return (0, _createClass2.default)(LinkMetaStep, [{
    key: "getMetadata",
    value: function getMetadata() {
      return this.metadata;
    }

    /**
     * Generate new undo/redo analytics event when step is inverted
     */
  }, {
    key: "invert",
    value: function invert() {
      /**
       * Omit sourceEvent in history
       */
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      var _this$metadata = this.metadata,
        sourceEvent = _this$metadata.sourceEvent,
        metadata = (0, _objectWithoutProperties2.default)(_this$metadata, _excluded);
      return new LinkMetaStep(this.pos, metadata, true);
    }

    // Should make no modifications to the doc
  }, {
    key: "apply",
    value: function apply(doc) {
      return _transform.StepResult.ok(doc);
    }
  }, {
    key: "map",
    value: function map(mapping) {
      var newPos = this.pos;
      if (typeof newPos === 'number') {
        newPos = mapping.map(newPos);
      }
      // Return the same events, this step will never be removed
      return new LinkMetaStep(newPos, this.metadata, this.isInverted);
    }
  }, {
    key: "getMap",
    value: function getMap() {
      return new _transform.StepMap([this.pos || 0, 0, 0]);
    }

    // Return null to avoid merging events
  }, {
    key: "merge",
    value: function merge() {
      return null;
    }
  }, {
    key: "toJSON",
    value: function toJSON() {
      // When serialized we should create a noop Replace step
      return {
        stepType: 'replace',
        from: 0,
        to: 0
      };
    }
  }], [{
    key: "fromJSON",
    value: function fromJSON() {
      // This is a "local custom step" once serialized
      // we need to transform it in a no-operation action
      return new _transform.ReplaceStep(0, 0, _model.Slice.empty);
    }
  }]);
}(_transform.Step);
/** Register this step with Prosemirror */
_transform.Step.jsonID(stepType, LinkMetaStep);
