// @atlaskit/navigation-system/dist/esm/ui/page-layout/main/main.js
/* main.tsx generated by @compiled/babel-plugin v0.38.1 */
import "./main.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Fragment } from 'react';
import { useSkipLink } from '../../../context/skip-links/skip-links-context';
import { contentHeightWhenFixed, contentInsetBlockStart } from '../constants';
import { useLayoutId } from '../id-utils';
var mainElementStyles = {
  root: "_nd5l1gzg _1reo1wug _18m91wug _19121cl4 _152timx3 _qwfh1wug _165teqxy _13wn1if8"
};

/**
 * Use the Main area for the main page content. It has a fluid width and will expand to fill available space.
 */
export function Main(_ref) {
  var children = _ref.children,
    xcss = _ref.xcss,
    _ref$skipLinkLabel = _ref.skipLinkLabel,
    skipLinkLabel = _ref$skipLinkLabel === void 0 ? 'Main Content' : _ref$skipLinkLabel,
    testId = _ref.testId,
    providedId = _ref.id;
  var id = useLayoutId({
    providedId: providedId
  });
  useSkipLink(id, skipLinkLabel);
  return /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement("div", {
    id: id,
    "data-layout-slot": true,
    className: ax([mainElementStyles.root, xcss]),
    role: "main",
    "data-testid": testId
  }, children));
}
