// @atlaskit/progress-tracker/dist/cjs/internal/marker.js
/* marker.tsx generated by @compiled/babel-plugin v0.38.1 */
/* eslint-disable @atlaskit/design-system/no-nested-styles */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./marker.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var progressMarkerStyles = null;
var markerColor = {
  unvisited: "_bfhk1aqn",
  current: "_bfhk1856",
  visited: "_bfhk1856",
  disabled: "_bfhkby5v"
};

/**
 * __Progress marker__
 *
 * Similar to `@atlaskit/progress-indicator`, a small visual circle marker
 */
var ProgressMarker = function ProgressMarker(_ref) {
  var testId = _ref.testId,
    _ref$status = _ref.status,
    status = _ref$status === void 0 ? 'unvisited' : _ref$status;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId,
    className: (0, _runtime.ax)(["_2rkou2gc _v564vc7x _1bsbu2gc _4t3iu2gc _kqswstnw _1e021ssb _t9ec157a", markerColor[status]])
  });
};
var _default = exports.default = ProgressMarker;
