// @atlaskit/navigation-system/dist/esm/ui/top-nav-items/help.js
import _extends from "@babel/runtime/helpers/extends";
/**
 * @jsxRuntime classic
 */
import React, { forwardRef } from 'react';
import HelpIcon from '@atlaskit/icon/core/question-circle';
import { BadgeContainer } from '../../components/badge-container';
import { EndItem } from './end-item';
var HELP_NOTIFICATION_BADGE_ID = 'atlassian-navigation-help-notification-count';
/**
 * __Help__
 *
 * The trigger button for the help menu in the top navigation bar.
 */
export var Help = /*#__PURE__*/forwardRef(function Help(_ref, ref) {
  var label = _ref.label,
    onClick = _ref.onClick,
    onMouseEnter = _ref.onMouseEnter,
    isSelected = _ref.isSelected,
    testId = _ref.testId,
    interactionName = _ref.interactionName,
    isListItem = _ref.isListItem,
    badge = _ref.badge,
    shortcut = _ref.shortcut;
  var sharedProps = {
    icon: HelpIcon,
    label: label,
    onClick: onClick,
    onMouseEnter: onMouseEnter,
    isSelected: isSelected,
    testId: testId,
    interactionName: interactionName,
    ref: ref,
    shortcut: shortcut
  };
  if (!badge) {
    return /*#__PURE__*/React.createElement(EndItem, _extends({}, sharedProps, {
      isListItem: isListItem
    }));
  }
  return /*#__PURE__*/React.createElement(BadgeContainer, {
    id: HELP_NOTIFICATION_BADGE_ID,
    badge: badge
    // We only need to set the list item role on the top level element (BadgeContainer)
    ,
    isListItem: isListItem
  }, /*#__PURE__*/React.createElement(EndItem, _extends({}, sharedProps, {
    // We explicitly set the EndItem to not be a list item,
    // because the BadgeContainer already has a list item role (if `isListItem` is true)
    isListItem: false
  })));
});
