// @atlaskit/navigation-system/dist/esm/ui/top-nav-items/log-in.js
import React, { forwardRef } from 'react';
import LogInIcon from '@atlaskit/icon/core/log-in';
import { ListItem } from '../../components/list-item';
import { LinkIconButton } from './themed/migration';
/**
 * __Log in__
 *
 * The Log in button for the top navigation.
 */
export var LogIn = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var onClick = _ref.onClick,
    onMouseEnter = _ref.onMouseEnter,
    testId = _ref.testId,
    interactionName = _ref.interactionName,
    href = _ref.href,
    label = _ref.label;
  return (
    /*#__PURE__*/
    /**
     * This component doesn't use `EndItem` internally because it renders a
     * `LinkIconButton` instead of an `IconButton`.
     *
     * We could make another abstraction, but this is the only top navigation item
     * that renders a link.
     */
    React.createElement(ListItem, null, /*#__PURE__*/React.createElement(LinkIconButton, {
      ref: ref,
      label: label,
      href: href,
      onClick: onClick,
      onMouseEnter: onMouseEnter,
      testId: testId,
      interactionName: interactionName,
      icon: LogInIcon,
      appearance: "subtle",
      isTooltipDisabled: false
    }))
  );
});
