// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/caption.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.caption = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _date = require("./date");
var _emoji = require("./emoji");
var _hardBreak = require("./hardBreak");
var _inlineCard = require("./inlineCard");
var _mention = require("./mention");
var _placeholder = require("./placeholder");
var _status = require("./status");
var _text = require("./text");
var _unsupportedInline = require("./unsupportedInline");
var caption = exports.caption = (0, _adfSchemaGenerator.adfNode)('caption').define({
  isolating: true,
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  allowAnyChildMark: true,
  content: [(0, _adfSchemaGenerator.$zeroPlus)((0, _adfSchemaGenerator.$or)(_hardBreak.hardBreak, _mention.mention, _emoji.emoji, _date.date, _placeholder.placeholder, _inlineCard.inlineCard, _status.status, _text.text.use('formatted'), _text.text.use('code_inline'), _unsupportedInline.unsupportedInline))]
});
