// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/expand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expand = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _breakout = require("../marks/breakout");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _nestedExpand = require("./nestedExpand");
var _nonNestableBlockContentGroup = require("../groups/nonNestableBlockContentGroup");
var expand = exports.expand = (0, _adfSchemaGenerator.adfNode)('expand').define({
  isolating: true,
  selectable: true,
  noMarks: true,
  attrs: {
    title: {
      type: 'string',
      default: '',
      optional: true
    },
    __expanded: {
      type: 'boolean',
      default: true,
      optional: true
    }
  },
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(_nonNestableBlockContentGroup.nonNestableBlockContentGroup, _nestedExpand.nestedExpand.use('with_no_marks'), _nestedExpand.nestedExpand.use('without_non_bodied_macros')))]
}).variant('root_only', {
  marks: [_breakout.breakout, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  noMarks: false,
  noExtend: true
});
