// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/inlineExtension.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineExtension = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _dataConsumer = require("../marks/dataConsumer");
var _fragment = require("../marks/fragment");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var inlineExtension = exports.inlineExtension = (0, _adfSchemaGenerator.adfNode)('inlineExtension').define({
  inline: true,
  selectable: true,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  hasEmptyMarks: true,
  attrs: {
    extensionKey: {
      minLength: 1,
      type: 'string',
      default: ''
    },
    extensionType: {
      minLength: 1,
      type: 'string',
      default: ''
    },
    parameters: {
      type: 'object',
      optional: true,
      default: null
    },
    text: {
      type: 'string',
      optional: true,
      default: null
    },
    localId: {
      minLength: 1,
      type: 'string',
      optional: true,
      default: null
    }
  }
}).variant('with_marks', {
  marks: [_dataConsumer.dataConsumer, _fragment.fragment, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute]
});
