// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/layoutColumn.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.layoutColumn = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _blockContentGroup = require("../groups/blockContentGroup");
var _blockGroup = require("../groups/blockGroup");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _unsupportedBlock = require("../nodes/unsupportedBlock");
var layoutColumn = exports.layoutColumn = (0, _adfSchemaGenerator.adfNode)('layoutColumn').define({
  isolating: true,
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    width: {
      type: 'number',
      minimum: 0,
      maximum: 100,
      default: undefined
    }
  },
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(_blockGroup.blockGroup, _blockContentGroup.blockContentGroup, _unsupportedBlock.unsupportedBlock))]
});
