// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/nestedExpand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.nestedExpand = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _blockquote = require("./blockquote");
var _codeBlock = require("./codeBlock");
var _decisionList = require("./decisionList");
var _extension = require("./extension");
var _heading = require("./heading");
var _list = require("./list");
var _mediaGroup = require("./mediaGroup");
var _mediaSingle = require("./mediaSingle");
var _panel = require("./panel");
var _paragraph = require("./paragraph");
var _rule = require("./rule");
var _task = require("./task");
var _unsupportedBlock = require("./unsupportedBlock");
var nestedExpandContent = [_paragraph.paragraph.use('with_no_marks'), _heading.heading.use('with_no_marks'), _mediaSingle.mediaSingle.use('caption'), _mediaSingle.mediaSingle.use('full'), _mediaGroup.mediaGroup, _codeBlock.codeBlock, _list.bulletList, _list.orderedList, _task.taskList, _decisionList.decisionList, _rule.rule, _panel.panel, _blockquote.blockquote, _unsupportedBlock.unsupportedBlock];
var nestedExpand = exports.nestedExpand = (0, _adfSchemaGenerator.adfNode)('nestedExpand').define({
  isolating: true,
  selectable: true,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    title: {
      type: 'string',
      default: '',
      optional: true
    },
    __expanded: {
      type: 'boolean',
      default: true,
      optional: true
    }
  },
  content: [(0, _adfSchemaGenerator.$onePlus)(_adfSchemaGenerator.$or.apply(void 0, nestedExpandContent.concat([_extension.extension.use('with_marks')])))],
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      'props.attrs.optional': {
        remove: true,
        reason: '@DSLCompatibilityException - mismatch with DSL'
      },
      'props.content': {
        value: 'nestedExpand_content',
        reason: '@DSLCompatibilityException - mismatch with DSL'
      },
      required: {
        reason: '@DSLCompatibilityException - required for nestedExpand validator spec',
        value: ['content']
      }
    }
  }
}).variant('content', {
  ignore: ['pm-spec']
}).variant('with_no_marks', {
  marks: [],
  content: [],
  noMarks: true
}).variant('without_non_bodied_macros', {
  marks: [],
  content: [(0, _adfSchemaGenerator.$onePlus)(_adfSchemaGenerator.$or.apply(void 0, nestedExpandContent))],
  noMarks: true,
  noExtend: true,
  ignore: ['json-schema', 'validator-spec']
});
