// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.panel = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _blockCard = require("./blockCard");
var _codeBlock = require("./codeBlock");
var _decisionList = require("./decisionList");
var _heading = require("./heading");
var _list = require("./list");
var _mediaGroup = require("./mediaGroup");
var _mediaSingle = require("./mediaSingle");
var _paragraph = require("./paragraph");
var _rule = require("./rule");
var _task = require("./task");
var _unsupportedBlock = require("./unsupportedBlock");
var _extension = require("./extension");
var panelContent = [_paragraph.paragraph.use('with_no_marks'), _heading.heading.use('with_no_marks'), _list.bulletList, _list.orderedList, _blockCard.blockCard, _mediaGroup.mediaGroup, _mediaSingle.mediaSingle.use('caption'), _mediaSingle.mediaSingle.use('full'), _codeBlock.codeBlock, _task.taskList, _rule.rule, _decisionList.decisionList, _unsupportedBlock.unsupportedBlock];
var panel = exports.panel = (0, _adfSchemaGenerator.adfNode)('panel').define({
  selectable: true,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    panelType: {
      type: 'enum',
      values: ['info', 'note', 'tip', 'warning', 'error', 'success', 'custom'],
      default: 'info'
    },
    panelIcon: {
      type: 'string',
      default: null,
      optional: true
    },
    panelIconId: {
      type: 'string',
      default: null,
      optional: true
    },
    panelIconText: {
      type: 'string',
      default: null,
      optional: true
    },
    panelColor: {
      type: 'string',
      default: null,
      optional: true
    }
  },
  content: [(0, _adfSchemaGenerator.$onePlus)(_adfSchemaGenerator.$or.apply(void 0, panelContent.concat([_extension.extension.use('with_marks')])))]
}).variant('without_nested_non_bodied_macros', {
  content: [(0, _adfSchemaGenerator.$onePlus)(_adfSchemaGenerator.$or.apply(void 0, panelContent))],
  noExtend: true,
  ignore: ['json-schema', 'validator-spec']
});
