// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/status.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.status = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _annotation = require("../marks/annotation");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var status = exports.status = (0, _adfSchemaGenerator.adfNode)('status').define({
  inline: true,
  selectable: true,
  marks: [_unsupportedNodeAttribute.unsupportedNodeAttribute, _unsupportedMark.unsupportedMark],
  attrs: {
    text: {
      minLength: 1,
      type: 'string',
      default: ''
    },
    color: {
      type: 'enum',
      values: ['neutral', 'purple', 'blue', 'red', 'yellow', 'green'],
      default: ''
    },
    localId: {
      type: 'string',
      optional: true,
      default: ''
    },
    style: {
      type: 'string',
      optional: true,
      default: ''
    }
  },
  stage0: {
    marks: [_unsupportedNodeAttribute.unsupportedNodeAttribute, _unsupportedMark.unsupportedMark, _annotation.annotation]
  }
});
