// @atlaskit/pragmatic-drag-and-drop-react-drop-indicator/dist/cjs/box.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropIndicator = DropIndicator;
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _line = require("./internal/line");
var _presets = require("./presets");
/**
 * __Drop indicator__
 *
 * A drop indicator is used to communicate the intended resting place of the draggable item. The orientation of the drop indicator should always match the direction of the content flow.
 */
function DropIndicator(_ref) {
  var _ref$appearance = _ref.appearance,
    appearance = _ref$appearance === void 0 ? 'default' : _ref$appearance,
    edge = _ref.edge,
    gap = _ref.gap,
    indent = _ref.indent,
    type = _ref.type;
  return /*#__PURE__*/_react.default.createElement(_line.Line, {
    edge: edge,
    gap: gap,
    strokeColor: _presets.presetStrokeColors[appearance],
    type: type,
    indent: indent
  });
}

// This default export is intended for usage with React.lazy
var _default = exports.default = DropIndicator;
