// @atlaskit/comment/dist/cjs/components/author.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _analyticsNext = require("@atlaskit/analytics-next");
var _field = _interopRequireDefault(require("./field"));
var Author = function Author(_ref) {
  var children = _ref.children,
    href = _ref.href,
    providedOnClick = _ref.onClick,
    onFocus = _ref.onFocus,
    onMouseOver = _ref.onMouseOver;
  var onClick = (0, _analyticsNext.usePlatformLeafEventHandler)({
    fn: function fn(event, analyticsEvent) {
      return providedOnClick && providedOnClick(event, analyticsEvent);
    },
    action: 'clicked',
    componentName: 'commentAuthor',
    packageName: "@atlaskit/comment",
    packageVersion: "0.0.0-development"
  });
  return /*#__PURE__*/_react.default.createElement(_field.default, {
    hasAuthor: true,
    href: href,
    onClick: onClick,
    onFocus: onFocus,
    onMouseOver: onMouseOver
  }, children);
};
Author.displayName = 'CommentAuthor';

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = Author;
