// @atlaskit/comment/dist/cjs/components/header.js
/* header.tsx generated by @compiled/babel-plugin v0.38.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./header.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _lockLocked = _interopRequireDefault(require("@atlaskit/icon/core/lock-locked"));
var _lozenge = _interopRequireDefault(require("@atlaskit/lozenge"));
var _compiled = require("@atlaskit/primitives/compiled");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var headingStyles = {
  root: "_11c8fhey"
};
var iconWrapperStyles = {
  root: "_18zrv77o _1e0c1txw"
};
var headerItemStyles = {
  root: "_o5721q9c _1o9zidpf"
};
var authorBoxStyles = {
  root: "_1nmz9jpi _1ul9idpf _p12f1osq"
};

/**
 * __Header items__
 *
 * Comment header items.
 *
 * @internal
 */
var Header = function Header(_ref) {
  var author = _ref.author,
    edited = _ref.edited,
    isError = _ref.isError,
    isSaving = _ref.isSaving,
    restrictedTo = _ref.restrictedTo,
    savingText = _ref.savingText,
    time = _ref.time,
    testId = _ref.testId,
    type = _ref.type,
    _ref$headingLevel = _ref.headingLevel,
    headingLevel = _ref$headingLevel === void 0 ? '3' : _ref$headingLevel,
    shouldHeaderWrap = _ref.shouldHeaderWrap;
  var Heading = "h".concat(headingLevel);
  var shouldRender = author || time || restrictedTo || isSaving && savingText || edited || type;
  return shouldRender ? /*#__PURE__*/React.createElement(Heading, {
    className: (0, _runtime.ax)([headingStyles.root])
  }, shouldHeaderWrap === false ? /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "center",
    testId: testId,
    space: "space.100",
    as: "span",
    shouldWrap: false
  }, author, type && /*#__PURE__*/React.createElement(_lozenge.default, {
    testId: testId && "".concat(testId, "-type")
  }, type), time && !isSaving && !isError && time, edited || null, isSaving ? savingText : null, restrictedTo && /*#__PURE__*/React.createElement(_compiled.Text, {
    color: "color.text.subtlest"
  }, /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "center",
    space: "space.050",
    as: "span"
  }, "\u2022", /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: iconWrapperStyles.root
  }, /*#__PURE__*/React.createElement(_lockLocked.default, {
    color: "currentColor",
    label: "",
    size: "small"
  })), restrictedTo))) : /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "center",
    testId: testId,
    space: "space.100",
    rowSpace: "space.0",
    as: "span",
    shouldWrap: true
  }, author && /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: authorBoxStyles.root
  }, author), type && /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: headerItemStyles.root
  }, /*#__PURE__*/React.createElement(_lozenge.default, {
    testId: testId && "".concat(testId, "-type")
  }, type)), time && !isSaving && !isError && /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: headerItemStyles.root
  }, time), edited && /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: headerItemStyles.root
  }, edited), isSaving && savingText && /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: headerItemStyles.root
  }, savingText), restrictedTo && /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span"
  }, /*#__PURE__*/React.createElement(_compiled.Text, {
    color: "color.text.subtlest"
  }, /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "center",
    space: "space.050",
    as: "span"
  }, /*#__PURE__*/React.createElement(_compiled.Box, {
    as: "span",
    xcss: iconWrapperStyles.root
  }, /*#__PURE__*/React.createElement(_lockLocked.default, {
    color: "currentColor",
    label: "",
    size: "small"
  })), restrictedTo))))) : null;
};
Header.displayName = 'CommentHeader';
var _default = exports.default = Header;
