// @atlaskit/adf-utils/dist/cjs/builders.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "a", {
  enumerable: true,
  get: function get() {
    return _link.link;
  }
});
Object.defineProperty(exports, "alignment", {
  enumerable: true,
  get: function get() {
    return _alignment.alignment;
  }
});
Object.defineProperty(exports, "b", {
  enumerable: true,
  get: function get() {
    return _strong.strong;
  }
});
Object.defineProperty(exports, "backgroundColor", {
  enumerable: true,
  get: function get() {
    return _backgroundColor.backgroundColor;
  }
});
Object.defineProperty(exports, "blockCard", {
  enumerable: true,
  get: function get() {
    return _blockCard.blockCard;
  }
});
Object.defineProperty(exports, "blockQuote", {
  enumerable: true,
  get: function get() {
    return _blockquote.blockQuote;
  }
});
Object.defineProperty(exports, "blockTaskItem", {
  enumerable: true,
  get: function get() {
    return _blockTaskItem.blockTaskItem;
  }
});
Object.defineProperty(exports, "bodiedExtension", {
  enumerable: true,
  get: function get() {
    return _bodiedExtension.bodiedExtension;
  }
});
Object.defineProperty(exports, "bodiedSyncBlock", {
  enumerable: true,
  get: function get() {
    return _bodiedSyncBlock.bodiedSyncBlock;
  }
});
Object.defineProperty(exports, "border", {
  enumerable: true,
  get: function get() {
    return _border.border;
  }
});
Object.defineProperty(exports, "br", {
  enumerable: true,
  get: function get() {
    return _hardBreak.hardBreak;
  }
});
Object.defineProperty(exports, "breakout", {
  enumerable: true,
  get: function get() {
    return _breakout.breakout;
  }
});
Object.defineProperty(exports, "bulletList", {
  enumerable: true,
  get: function get() {
    return _bulletList.bulletList;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function get() {
    return _code.code;
  }
});
Object.defineProperty(exports, "codeBlock", {
  enumerable: true,
  get: function get() {
    return _codeBlock.codeBlock;
  }
});
Object.defineProperty(exports, "dataConsumer", {
  enumerable: true,
  get: function get() {
    return _dataConsumer.dataConsumer;
  }
});
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function get() {
    return _date.date;
  }
});
Object.defineProperty(exports, "decisionItem", {
  enumerable: true,
  get: function get() {
    return _decisionItem.decisionItem;
  }
});
Object.defineProperty(exports, "decisionList", {
  enumerable: true,
  get: function get() {
    return _decisionList.decisionList;
  }
});
Object.defineProperty(exports, "doc", {
  enumerable: true,
  get: function get() {
    return _doc.doc;
  }
});
Object.defineProperty(exports, "em", {
  enumerable: true,
  get: function get() {
    return _em.em;
  }
});
Object.defineProperty(exports, "embedCard", {
  enumerable: true,
  get: function get() {
    return _embedCard.embedCard;
  }
});
Object.defineProperty(exports, "emoji", {
  enumerable: true,
  get: function get() {
    return _emoji.emoji;
  }
});
Object.defineProperty(exports, "expand", {
  enumerable: true,
  get: function get() {
    return _expand.expand;
  }
});
Object.defineProperty(exports, "extension", {
  enumerable: true,
  get: function get() {
    return _extension.extension;
  }
});
Object.defineProperty(exports, "fragment", {
  enumerable: true,
  get: function get() {
    return _fragment.fragment;
  }
});
Object.defineProperty(exports, "hardBreak", {
  enumerable: true,
  get: function get() {
    return _hardBreak.hardBreak;
  }
});
Object.defineProperty(exports, "heading", {
  enumerable: true,
  get: function get() {
    return _heading.heading;
  }
});
Object.defineProperty(exports, "hr", {
  enumerable: true,
  get: function get() {
    return _rule.rule;
  }
});
Object.defineProperty(exports, "indentation", {
  enumerable: true,
  get: function get() {
    return _indentation.indentation;
  }
});
Object.defineProperty(exports, "inlineCard", {
  enumerable: true,
  get: function get() {
    return _inlineCard.inlineCard;
  }
});
Object.defineProperty(exports, "inlineExtension", {
  enumerable: true,
  get: function get() {
    return _inlineExtension.inlineExtension;
  }
});
Object.defineProperty(exports, "layoutColumn", {
  enumerable: true,
  get: function get() {
    return _layoutColumn.layoutColumn;
  }
});
Object.defineProperty(exports, "layoutSection", {
  enumerable: true,
  get: function get() {
    return _layoutSection.layoutSection;
  }
});
Object.defineProperty(exports, "li", {
  enumerable: true,
  get: function get() {
    return _listItem.listItem;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function get() {
    return _link.link;
  }
});
Object.defineProperty(exports, "listItem", {
  enumerable: true,
  get: function get() {
    return _listItem.listItem;
  }
});
Object.defineProperty(exports, "media", {
  enumerable: true,
  get: function get() {
    return _media.media;
  }
});
Object.defineProperty(exports, "mediaGroup", {
  enumerable: true,
  get: function get() {
    return _mediaGroup.mediaGroup;
  }
});
Object.defineProperty(exports, "mediaSingle", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingle;
  }
});
Object.defineProperty(exports, "mention", {
  enumerable: true,
  get: function get() {
    return _mention.mention;
  }
});
Object.defineProperty(exports, "nestedExpand", {
  enumerable: true,
  get: function get() {
    return _nestedExpand.nestedExpand;
  }
});
Object.defineProperty(exports, "ol", {
  enumerable: true,
  get: function get() {
    return _orderedList.orderedList;
  }
});
Object.defineProperty(exports, "orderedList", {
  enumerable: true,
  get: function get() {
    return _orderedList.orderedList;
  }
});
Object.defineProperty(exports, "p", {
  enumerable: true,
  get: function get() {
    return _paragraph.paragraph;
  }
});
Object.defineProperty(exports, "panel", {
  enumerable: true,
  get: function get() {
    return _panel.panel;
  }
});
Object.defineProperty(exports, "paragraph", {
  enumerable: true,
  get: function get() {
    return _paragraph.paragraph;
  }
});
Object.defineProperty(exports, "placeholder", {
  enumerable: true,
  get: function get() {
    return _placeholder.placeholder;
  }
});
Object.defineProperty(exports, "rule", {
  enumerable: true,
  get: function get() {
    return _rule.rule;
  }
});
Object.defineProperty(exports, "status", {
  enumerable: true,
  get: function get() {
    return _status.status;
  }
});
Object.defineProperty(exports, "strike", {
  enumerable: true,
  get: function get() {
    return _strike.strike;
  }
});
Object.defineProperty(exports, "strong", {
  enumerable: true,
  get: function get() {
    return _strong.strong;
  }
});
Object.defineProperty(exports, "subsup", {
  enumerable: true,
  get: function get() {
    return _subsup.subsup;
  }
});
Object.defineProperty(exports, "syncBlock", {
  enumerable: true,
  get: function get() {
    return _syncBlock.syncBlock;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function get() {
    return _table.table;
  }
});
Object.defineProperty(exports, "tableCell", {
  enumerable: true,
  get: function get() {
    return _tableCell.tableCell;
  }
});
Object.defineProperty(exports, "tableHeader", {
  enumerable: true,
  get: function get() {
    return _tableHeader.tableHeader;
  }
});
Object.defineProperty(exports, "tableRow", {
  enumerable: true,
  get: function get() {
    return _tableRow.tableRow;
  }
});
Object.defineProperty(exports, "taskItem", {
  enumerable: true,
  get: function get() {
    return _taskItem.taskItem;
  }
});
Object.defineProperty(exports, "taskList", {
  enumerable: true,
  get: function get() {
    return _taskList.taskList;
  }
});
Object.defineProperty(exports, "td", {
  enumerable: true,
  get: function get() {
    return _tableCell.tableCell;
  }
});
Object.defineProperty(exports, "text", {
  enumerable: true,
  get: function get() {
    return _text.text;
  }
});
Object.defineProperty(exports, "textColor", {
  enumerable: true,
  get: function get() {
    return _textColor.textColor;
  }
});
Object.defineProperty(exports, "th", {
  enumerable: true,
  get: function get() {
    return _tableHeader.tableHeader;
  }
});
Object.defineProperty(exports, "tr", {
  enumerable: true,
  get: function get() {
    return _tableRow.tableRow;
  }
});
Object.defineProperty(exports, "u", {
  enumerable: true,
  get: function get() {
    return _underline.underline;
  }
});
Object.defineProperty(exports, "ul", {
  enumerable: true,
  get: function get() {
    return _bulletList.bulletList;
  }
});
Object.defineProperty(exports, "underline", {
  enumerable: true,
  get: function get() {
    return _underline.underline;
  }
});
var _breakout = require("./builders/marks/breakout");
var _code = require("./builders/marks/code");
var _em = require("./builders/marks/em");
var _link = require("./builders/marks/link");
var _strike = require("./builders/marks/strike");
var _strong = require("./builders/marks/strong");
var _subsup = require("./builders/marks/subsup");
var _textColor = require("./builders/marks/text-color");
var _underline = require("./builders/marks/underline");
var _alignment = require("./builders/marks/alignment");
var _indentation = require("./builders/marks/indentation");
var _dataConsumer = require("./builders/marks/data-consumer");
var _fragment = require("./builders/marks/fragment");
var _border = require("./builders/marks/border");
var _backgroundColor = require("./builders/marks/background-color");
var _blockquote = require("./builders/nodes/blockquote");
var _bodiedExtension = require("./builders/nodes/bodied-extension");
var _blockCard = require("./builders/nodes/block-card");
var _bulletList = require("./builders/nodes/bullet-list");
var _codeBlock = require("./builders/nodes/code-block");
var _date = require("./builders/nodes/date");
var _decisionItem = require("./builders/nodes/decision-item");
var _decisionList = require("./builders/nodes/decision-list");
var _doc = require("./builders/nodes/doc");
var _emoji = require("./builders/nodes/emoji");
var _extension = require("./builders/nodes/extension");
var _expand = require("./builders/nodes/expand");
var _nestedExpand = require("./builders/nodes/nested-expand");
var _hardBreak = require("./builders/nodes/hard-break");
var _heading = require("./builders/nodes/heading");
var _inlineExtension = require("./builders/nodes/inline-extension");
var _inlineCard = require("./builders/nodes/inline-card");
var _layoutColumn = require("./builders/nodes/layout-column");
var _layoutSection = require("./builders/nodes/layout-section");
var _listItem = require("./builders/nodes/list-item");
var _mediaGroup = require("./builders/nodes/media-group");
var _mediaSingle = require("./builders/nodes/media-single");
var _media = require("./builders/nodes/media");
var _mention = require("./builders/nodes/mention");
var _orderedList = require("./builders/nodes/ordered-list");
var _panel = require("./builders/nodes/panel");
var _paragraph = require("./builders/nodes/paragraph");
var _placeholder = require("./builders/nodes/placeholder");
var _rule = require("./builders/nodes/rule");
var _status = require("./builders/nodes/status");
var _tableCell = require("./builders/nodes/table-cell");
var _tableHeader = require("./builders/nodes/table-header");
var _tableRow = require("./builders/nodes/table-row");
var _table = require("./builders/nodes/table");
var _taskItem = require("./builders/nodes/task-item");
var _taskList = require("./builders/nodes/task-list");
var _text = require("./builders/nodes/text");
var _embedCard = require("./builders/nodes/embed-card");
var _blockTaskItem = require("./builders/nodes/block-task-item");
var _syncBlock = require("./builders/nodes/sync-block");
var _bodiedSyncBlock = require("./builders/nodes/bodied-sync-block");
