// @atlaskit/side-nav-items/dist/es2019/ui/menu-item/button-menu-item.js
import React, { forwardRef } from 'react';
import { MenuItemBase } from './menu-item';
import { MenuListItem } from './menu-list-item';
/**
 * __ButtonMenuItem__
 *
 * A menu item button. It should be used within a `ul`, as it renders a list item.
 */
export const ButtonMenuItem = /*#__PURE__*/forwardRef(({
  testId,
  actions,
  children,
  description,
  elemAfter,
  isDisabled,
  isSelected,
  elemBefore,
  actionsOnHover,
  onClick,
  'aria-controls': ariaControls,
  'aria-expanded': ariaExpanded,
  'aria-haspopup': ariaHasPopup,
  interactionName,
  isContentTooltipDisabled,
  visualContentRef,
  listItemRef,
  isDragging,
  hasDragIndicator,
  dropIndicator
}, forwardedRef) => {
  return /*#__PURE__*/React.createElement(MenuListItem, {
    ref: listItemRef
  }, /*#__PURE__*/React.createElement(MenuItemBase, {
    testId: testId,
    description: description,
    elemAfter: elemAfter,
    elemBefore: elemBefore,
    isDisabled: isDisabled
    /**
     * Not passing `actions` and `actionsOnHover` to MenuItemBase when `isDisabled`,
     * so they aren't rendered in the disabled state.
     */,
    actions: isDisabled ? undefined : actions,
    actionsOnHover: isDisabled ? undefined : actionsOnHover,
    onClick: onClick,
    ariaControls: ariaControls,
    ariaExpanded: ariaExpanded,
    ariaHasPopup: ariaHasPopup,
    isSelected: isSelected,
    ref: forwardedRef,
    visualContentRef: visualContentRef,
    interactionName: interactionName,
    isContentTooltipDisabled: isContentTooltipDisabled,
    isDragging: isDragging,
    hasDragIndicator: hasDragIndicator,
    dropIndicator: dropIndicator
  }, children));
});
