// @atlaskit/adf-schema-generator/dist/es2019/transforms/adfToValidatorSpec/buildContent.js
export function buildContent(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
node, content) {
  const isTupleLike = content.length > 1;
  if (isTupleLike) {
    return buildTupleLikeContent(node, content);
  }
  return buildRegularContent(content[0]);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function buildTupleLikeContent(node, content) {
  const nodeSpec = node.getSpec();
  const hasUnsupportedInline = content.some(item => item.hasUnsupportedInline);
  const hasUnsupportedBlock = content.some(item => item.hasUnsupportedBlock);
  const returnContent = {
    type: 'array',
    isTupleLike: true
  };
  const items = [];
  for (const child of content) {
    items.push(buildItems(child.contentTypes));
  }
  returnContent.items = items;
  if ('contentMinItems' in nodeSpec) {
    returnContent.minItems = nodeSpec.contentMinItems;
  }
  if ('contentMaxItems' in nodeSpec) {
    returnContent.maxItems = nodeSpec.contentMaxItems;
  }
  if (hasUnsupportedInline) {
    returnContent.allowUnsupportedInline = true;
  }
  if (hasUnsupportedBlock) {
    returnContent.allowUnsupportedBlock = true;
  }
  return returnContent;
}
function buildRegularContent(content) {
  const returnContent = {
    type: 'array'
  };
  returnContent.items = [buildItems(content.contentTypes)];
  if (content !== null && content !== void 0 && content.minItems) {
    returnContent.minItems = content === null || content === void 0 ? void 0 : content.minItems;
  }
  if (content !== null && content !== void 0 && content.maxItems) {
    returnContent.maxItems = content === null || content === void 0 ? void 0 : content.maxItems;
  }
  if (content !== null && content !== void 0 && content.optional) {
    returnContent.optional = content === null || content === void 0 ? void 0 : content.optional;
  }
  if (content.hasUnsupportedInline) {
    returnContent.allowUnsupportedInline = true;
  }
  if (content.hasUnsupportedBlock) {
    returnContent.allowUnsupportedBlock = true;
  }
  return returnContent;
}

/**
 * If contentTypes only has one item it will be a single array [],
 * however if it has more than one item it will be an array within an array [[]].
 */
function buildItems(contentTypes) {
  if (contentTypes.length === 1) {
    return contentTypes[0];
  } else {
    return contentTypes;
  }
}
