// @atlaskit/adf-schema-generator/dist/es2019/transforms/adfToValidatorSpec/buildMarks.js
import { ValidatorSpecTransformerName } from '../transformerNames';
export function buildMarks(marks, noMarks = false, hasEmptyMarks = false) {
  if (noMarks) {
    return {
      type: 'array',
      maxItems: 0,
      items: [],
      optional: true
    };
  }
  if (hasEmptyMarks) {
    return {
      type: 'array',
      items: [],
      optional: true
    };
  }
  const filteredMarks = filterIgnoredMarks(marks);
  if (!filteredMarks || filteredMarks.length === 0) {
    // @ts-expect-error
    return;
  }
  const returnMarks = {
    type: 'array',
    optional: true,
    items: []
  };
  if (filteredMarks.length > 1) {
    returnMarks.items = [filteredMarks.map(mark => mark.getType())];
  } else {
    returnMarks.items = filteredMarks.map(mark => mark.getType());
  }
  return returnMarks;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function filterIgnoredMarks(marks) {
  return (marks !== null && marks !== void 0 ? marks : []).filter(mark => !mark.isIgnored(ValidatorSpecTransformerName));
}
