// @atlaskit/adf-schema-generator/dist/esm/transforms/adfToJson/adfToJsonVisitor.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
import { JSONSchemaTransformerName } from '../transformerNames';
import { resolveName } from './nameResolver';
import { buildMarks } from './marksBuilder';

// Lots of placeholder code for now

export function isNodeReturnValue(value) {
  return value && 'node' in value;
}
export function isGroupReturnValue(value) {
  return value && 'group' in value;
}
export var buildVisitor = function buildVisitor(res, buildJson, fullSchema) {
  return {
    node: function node(_node, content, cycle) {
      var _node$getSpec$marks;
      if (_node.isStage0Only() && fullSchema) {
        return undefined;
      }
      if (_node.isIgnored(JSONSchemaTransformerName)) {
        return undefined;
      }
      var json = buildJson(_node, content, fullSchema);
      var processedNode = {
        node: _node,
        json: json
      };
      if (cycle) {
        return processedNode;
      }
      res[resolveName(_node.getName())] = {
        json: json
      };
      var marks = buildMarks((_node$getSpec$marks = _node.getSpec().marks) !== null && _node$getSpec$marks !== void 0 ? _node$getSpec$marks : []);
      Object.entries(marks).forEach(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 2),
          key = _ref2[0],
          value = _ref2[1];
        res[key] = value;
      });
      return processedNode;
    },
    group: function group(_group, members) {
      if (_group.isIgnored(JSONSchemaTransformerName)) {
        return undefined;
      }
      res[resolveName(_group.group)] = {
        json: {
          anyOf: _group.members.filter(function (m) {
            return !m.isIgnored(JSONSchemaTransformerName) && !(m.isStage0Only() && fullSchema);
          }).map(function (m) {
            return {
              $ref: "#/definitions/".concat(resolveName(m.getName()))
            };
          })
        }
      };
      return {
        group: _group.group,
        members: members
      };
    },
    $or: function $or(children) {
      var contentTypes = [];
      var _iterator = _createForOfIteratorHelper(children),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var child = _step.value;
          if (isGroupReturnValue(child)) {
            // only include content types if group has members
            // e.g. blockRootOnly don't have members, so won't be included in contentTypes of doc
            if (child.members && child.members.length > 0) {
              contentTypes.push(child.group);
            }
          } else if (isNodeReturnValue(child)) {
            contentTypes.push(child.node.getName());
          }
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
      return {
        type: '$or',
        minItems: 1,
        contentTypes: contentTypes
      };
    },
    $onePlus: function $onePlus(child) {
      return {
        type: '$onePlus',
        minItems: 1,
        contentTypes: child.contentTypes
      };
    },
    $zeroPlus: function $zeroPlus(child) {
      return {
        type: '$zeroPlus',
        minItems: 0,
        contentTypes: child.contentTypes
      };
    },
    $range: function $range(range, child) {
      return {
        type: '$range',
        contentTypes: child.contentTypes,
        range: range
      };
    }
  };
};
