// @atlaskit/adf-schema-generator/dist/esm/transforms/adfToJson/contentBuilder.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import uniqBy from 'lodash/uniqBy';
import { ADFNode } from '../../adfNode';
import { JSONSchemaTransformerName } from '../transformerNames';
import { resolveName } from './nameResolver';
export function buildContent(content, name, nodeSpec, fullSchema) {
  var _nodeSpec$contentMinI, _nodeSpec$contentMaxI;
  if (content.length === 0) {
    return {};
  }
  var adfNodeContent = nodeSpec.content || [];
  var minItems = (_nodeSpec$contentMinI = nodeSpec.contentMinItems) !== null && _nodeSpec$contentMinI !== void 0 ? _nodeSpec$contentMinI : determineMinItems(content);
  var items = determineItems(content, name, adfNodeContent, fullSchema);
  var maxItems = (_nodeSpec$contentMaxI = nodeSpec.contentMaxItems) !== null && _nodeSpec$contentMaxI !== void 0 ? _nodeSpec$contentMaxI : determineMaxItems(content);
  return {
    content: _objectSpread(_objectSpread({
      type: 'array',
      items: items
    }, minItems !== null ? {
      minItems: minItems
    } : {}), maxItems !== null ? {
      maxItems: maxItems
    } : {})
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isADFGroup(value) {
  return value && 'group' in value;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isADFNode(value) {
  return value && value instanceof ADFNode;
}
export function determineItems(content, name, adfNodeContent, fullSchema) {
  // @DSLCompatibilityException Special case for doc_node.
  // It flattens the groups using the nodes themselves instead of the group name in the content.
  if (name === 'doc') {
    var jsonSchema = adfNodeContent.reduce(function (acc, value) {
      // We're expecting a single one+ with a nested $or
      if (value.type === '$one+' && value.content.type === '$or') {
        return [].concat(_toConsumableArray(acc), _toConsumableArray(flattenContent(value.content.content).filter(function (node) {
          return !node.isIgnored(JSONSchemaTransformerName) && !(node.isStage0Only() && fullSchema);
        }).map(function (node) {
          return {
            $ref: "#/definitions/".concat(resolveName(node.getName()))
          };
        })));
      }
      return acc;
    }, []);
    var anyOfItems = uniqBy(jsonSchema, function (v) {
      return v.$ref;
    });
    if (!anyOfItems.length) {
      return;
    }
    return {
      anyOf: anyOfItems
    };
  }
  var processedContentGroups = processContentGroups(content);
  // The JSON schema omits the array if there is only 1 item
  return flattenArray(processedContentGroups);
}
export function processContentTypes(contentTypes) {
  var itemsArray = [];
  contentTypes.forEach(function (piece) {
    itemsArray.push({
      $ref: "#/definitions/".concat(resolveName(piece))
    });
  });
  if (itemsArray.length === 0) {
    return;
  }

  // We flatten an array here as well, but using anyOf to fit the JSON schema
  if (itemsArray.length === 1) {
    return itemsArray[0];
  } else {
    return {
      anyOf: itemsArray
    };
  }
}
export function processContentGroups(content) {
  return content.map(function (item) {
    return processContentTypes(item.contentTypes);
  }).filter(Boolean);
}
export function determineMinItems(content) {
  var _content$find, _content$find2;
  // Despite it being possible for there to be multiple content groups on one node in DSL,
  // the JSON schema has only one minItem value for all content on a node.
  if (!content) {
    return null;
  }
  var _ref = (_content$find = content.find(function (v) {
      return v.minItems !== undefined && !isNaN(v.minItems);
    })) !== null && _content$find !== void 0 ? _content$find : {},
    minItems = _ref.minItems;
  var _ref2 = (_content$find2 = content.find(function (v) {
      return v.range;
    })) !== null && _content$find2 !== void 0 ? _content$find2 : {},
    range = _ref2.range;
  if (minItems) {
    return minItems;
  }

  // If it's a range, take minItems from that
  if (range) {
    return range.min;
  }
  return null;
}

// This function is not comprehensive, it is only defined for certain inputs
export function determineMaxItems(content) {
  // Despite it being possible for there to be multiple content groups on one node in DSL,
  // the JSON schema has only one maxItem value for all content on a node.

  // If there's only one item, we can simply calculate it and return
  if (content.length === 1) {
    // If it's a range, grab maxItems from that
    if (content[0].range) {
      return content[0].range.max;
    }
    return null;
  }
  var maxItems = 0;
  var _iterator = _createForOfIteratorHelper(content),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var item = _step.value;
      // ignoring empty content items, e.g. when all nodes in this particular content items
      // are filtered out by being ignored for json-schema output
      if (!item.contentTypes.length) {
        continue;
      }
      if (item.type === '$or') {
        maxItems += 1;
      } else if (item.type === '$zeroPlus' || item.type === '$onePlus') {
        // with zeroPlus and onePlus it means parent can have infinite number of children
        return null;
      } else if (item.type === '$range' && item.range) {
        maxItems += item.range.max;
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  return maxItems > 0 ? maxItems : null;
}
export function flattenArray(array) {
  if (array.length === 1) {
    return array[0];
  } else {
    return array;
  }
}
/**
 * Flattens ADF groups and nodes into an array of nodes.
 * @param content
 * @returns ADFNode[]
 */
export function flattenContent(content) {
  return content.reduce(function (acc, item) {
    if (isADFGroup(item)) {
      // Expand the group into its member nodes
      return [].concat(_toConsumableArray(acc), _toConsumableArray(flattenContent(item.members)));
    } else if (isADFNode(item) && !item.isIgnored(JSONSchemaTransformerName)) {
      return [].concat(_toConsumableArray(acc), [item]);
    }
    return acc;
  }, []);
}
