// @atlaskit/adf-schema-generator/dist/esm/transforms/adfToJson/marksBuilder.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { JSONSchemaTransformerName } from '../transformerNames';
import { buildAttrs } from './attrBuilder';
import { buildRequired } from './requiredBuilder';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function buildMarks(marks) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  var res = {};
  marks.forEach(function (mark) {
    if (mark.isIgnored(JSONSchemaTransformerName)) {
      return;
    }
    var attrs = buildAttrs(mark.getSpec().attrs);
    var required = buildRequired(mark.getSpec().attrs, false, '');
    res["".concat(mark.getType(), "_mark")] = {
      json: {
        type: 'object',
        properties: _objectSpread({
          type: {
            enum: [mark.getType()]
          }
        }, attrs),
        required: required,
        additionalProperties: false
      }
    };
  });
  return res;
}
