// @atlaskit/atlassian-context/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getATLContextDomain", {
  enumerable: true,
  get: function get() {
    return _domainLookup.getATLContextDomain;
  }
});
Object.defineProperty(exports, "getATLContextUrl", {
  enumerable: true,
  get: function get() {
    return _domainLookup.getATLContextUrl;
  }
});
Object.defineProperty(exports, "getDomainInContext", {
  enumerable: true,
  get: function get() {
    return _generalizedDomainLookup.getDomainInContext;
  }
});
Object.defineProperty(exports, "getUrlForDomainInContext", {
  enumerable: true,
  get: function get() {
    return _generalizedDomainLookup.getUrlForDomainInContext;
  }
});
Object.defineProperty(exports, "isFedRamp", {
  enumerable: true,
  get: function get() {
    return _isFedramp.isFedRamp;
  }
});
Object.defineProperty(exports, "isFedrampModerate", {
  enumerable: true,
  get: function get() {
    return _perimeter.isFedrampModerate;
  }
});
Object.defineProperty(exports, "isIsolatedCloud", {
  enumerable: true,
  get: function get() {
    return _perimeter.isIsolatedCloud;
  }
});
Object.defineProperty(exports, "isolatedCloudDomain", {
  enumerable: true,
  get: function get() {
    return _perimeter.isolatedCloudDomain;
  }
});
Object.defineProperty(exports, "isolationContextId", {
  enumerable: true,
  get: function get() {
    return _perimeter.isolationContextId;
  }
});
var _isFedramp = require("./services/host-based-lookup/is-fedramp");
var _domainLookup = require("./services/host-based-lookup/domain-lookup");
var _perimeter = require("./services/perimeter");
var _generalizedDomainLookup = require("./services/generalized-domain-lookup");
